<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>nico_covariations.Covariations &mdash; NiCo 1.0.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="../../_static/css/theme.css?v=19f00094" />

  
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../_static/jquery.js?v=5d32c60e"></script>
        <script src="../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="../../_static/documentation_options.js?v=8d563738"></script>
        <script src="../../_static/doctools.js?v=888ff710"></script>
        <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../index.html">
            
              <img src="../../_static/nico-logo.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../examples.html">Examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../examples.html#tutorials">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../modules.html">NiCo_Modules</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../api_reference.html">API Reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../faq.html">Frequently Asked Questions (FAQ)</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">NiCo</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../index.html">Module code</a></li>
      <li class="breadcrumb-item active">nico_covariations.Covariations</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for nico_covariations.Covariations</h1><div class="highlight"><pre>
<span></span>
<span class="kn">import</span> <span class="nn">matplotlib</span>
<span class="c1">#matplotlib.rcParams[&#39;pdf.fonttype&#39;] = 42</span>
<span class="c1">#matplotlib.rcParams[&#39;ps.fonttype&#39;] = 42</span>
<span class="c1">#matplotlib.rcParams[&#39;axes.linewidth&#39;] = 0.1 #set the value globally</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="c1">#plt.rc(&#39;font&#39;, family=&#39;Helvetica&#39;)</span>

<span class="kn">from</span> <span class="nn">matplotlib.colors</span> <span class="kn">import</span> <span class="n">LinearSegmentedColormap</span>
<span class="kn">from</span> <span class="nn">matplotlib</span> <span class="kn">import</span> <span class="n">gridspec</span>
<span class="c1">#from scipy.spatial import Voronoi, ConvexHull,voronoi_plot_2d, Delaunay</span>
<span class="kn">from</span> <span class="nn">numpy.linalg</span> <span class="kn">import</span> <span class="n">norm</span>

<span class="kn">from</span> <span class="nn">sklearn.datasets</span> <span class="kn">import</span> <span class="n">make_classification</span>
<span class="kn">from</span> <span class="nn">sklearn.multioutput</span> <span class="kn">import</span> <span class="n">MultiOutputRegressor</span>
<span class="kn">from</span> <span class="nn">sklearn.linear_model</span> <span class="kn">import</span> <span class="n">LogisticRegression</span><span class="p">,</span><span class="n">LogisticRegressionCV</span><span class="p">,</span> <span class="n">Lasso</span><span class="p">,</span><span class="n">Ridge</span><span class="p">,</span> <span class="n">RidgeCV</span><span class="p">,</span><span class="n">LassoCV</span><span class="p">,</span> <span class="n">LinearRegression</span>
<span class="kn">from</span> <span class="nn">sklearn.neighbors</span> <span class="kn">import</span> <span class="n">KNeighborsRegressor</span>
<span class="kn">from</span> <span class="nn">sklearn.tree</span> <span class="kn">import</span> <span class="n">DecisionTreeRegressor</span>
<span class="kn">from</span> <span class="nn">sklearn.model_selection</span> <span class="kn">import</span> <span class="n">RandomizedSearchCV</span><span class="p">,</span><span class="n">GridSearchCV</span><span class="p">,</span><span class="n">cross_val_predict</span><span class="p">,</span> <span class="n">cross_val_score</span><span class="p">,</span><span class="n">RepeatedKFold</span><span class="p">,</span><span class="n">RepeatedStratifiedKFold</span><span class="p">,</span><span class="n">StratifiedShuffleSplit</span>
<span class="c1">#from sklearn.metrics import make_scorer,accuracy_score, f1_score, classification_report,confusion_matrix,roc_curve, roc_auc_score, precision_score, recall_score, precision_recall_curve</span>
<span class="kn">from</span> <span class="nn">sklearn.metrics</span> <span class="kn">import</span> <span class="n">confusion_matrix</span><span class="p">,</span><span class="n">r2_score</span><span class="p">,</span><span class="n">mean_absolute_error</span><span class="p">,</span><span class="n">mean_squared_error</span><span class="p">,</span><span class="n">mean_squared_log_error</span><span class="p">,</span><span class="n">mean_absolute_percentage_error</span><span class="p">,</span><span class="n">median_absolute_error</span><span class="p">,</span> <span class="n">max_error</span><span class="p">,</span> <span class="n">explained_variance_score</span>

<span class="kn">from</span> <span class="nn">sklearn.preprocessing</span> <span class="kn">import</span> <span class="n">PolynomialFeatures</span><span class="p">,</span> <span class="n">StandardScaler</span>
<span class="kn">from</span> <span class="nn">sklearn.pipeline</span> <span class="kn">import</span> <span class="n">Pipeline</span>
<span class="c1">#from sklearn.metrics import precision_recall_fscore_support as score</span>
<span class="c1">#from imblearn.over_sampling import SMOTE, SMOTEN,ADASYN, KMeansSMOTE, SVMSMOTE</span>
<span class="kn">from</span> <span class="nn">sklearn.utils</span> <span class="kn">import</span> <span class="n">class_weight</span>
<span class="kn">from</span> <span class="nn">sklearn.metrics</span> <span class="kn">import</span> <span class="n">roc_curve</span><span class="p">,</span> <span class="n">auc</span><span class="p">,</span><span class="n">consensus_score</span>

<span class="c1">#from sklearn.datasets import make_checkerboard</span>
<span class="kn">from</span> <span class="nn">sklearn.cluster</span> <span class="kn">import</span> <span class="n">SpectralBiclustering</span>
<span class="kn">from</span> <span class="nn">sklearn.decomposition</span> <span class="kn">import</span> <span class="n">NMF</span>
<span class="kn">from</span> <span class="nn">matplotlib.tri</span> <span class="kn">import</span> <span class="n">Triangulation</span>
<span class="kn">from</span> <span class="nn">matplotlib.collections</span> <span class="kn">import</span> <span class="n">PatchCollection</span>
<span class="kn">from</span> <span class="nn">matplotlib.gridspec</span> <span class="kn">import</span> <span class="n">SubplotSpec</span>

<span class="kn">from</span> <span class="nn">sklearn.metrics.pairwise</span> <span class="kn">import</span> <span class="n">cosine_similarity</span>
<span class="c1">#bicluster</span>

<span class="kn">from</span> <span class="nn">gseapy.plot</span> <span class="kn">import</span> <span class="n">gseaplot</span><span class="p">,</span> <span class="n">heatmap</span>
<span class="kn">import</span> <span class="nn">gseapy</span>
<span class="kn">from</span> <span class="nn">sklearn.decomposition</span> <span class="kn">import</span> <span class="n">PCA</span> <span class="k">as</span> <span class="n">skPCA</span>

<span class="kn">from</span> <span class="nn">scipy.spatial</span> <span class="kn">import</span> <span class="n">cKDTree</span>
<span class="kn">from</span> <span class="nn">scipy.spatial.distance</span> <span class="kn">import</span> <span class="n">cosine</span>
<span class="kn">import</span> <span class="nn">scipy.sparse</span> <span class="k">as</span> <span class="nn">scipy_sparse</span>
<span class="kn">from</span> <span class="nn">sklearn.metrics.pairwise</span> <span class="kn">import</span> <span class="n">cosine_similarity</span>

<span class="c1">#Metrics</span>
<span class="kn">from</span> <span class="nn">sklearn.metrics</span> <span class="kn">import</span> <span class="n">cohen_kappa_score</span>
<span class="kn">from</span> <span class="nn">sklearn.metrics</span> <span class="kn">import</span> <span class="n">hamming_loss</span>
<span class="kn">from</span> <span class="nn">sklearn.metrics</span> <span class="kn">import</span> <span class="n">log_loss</span>
<span class="kn">from</span> <span class="nn">sklearn.metrics</span> <span class="kn">import</span> <span class="n">zero_one_loss</span>
<span class="kn">from</span> <span class="nn">sklearn.metrics</span> <span class="kn">import</span> <span class="n">matthews_corrcoef</span>
<span class="kn">from</span> <span class="nn">scipy.stats</span> <span class="kn">import</span> <span class="n">pearsonr</span><span class="p">,</span><span class="n">spearmanr</span>

<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">snn</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">random</span>
<span class="kn">import</span> <span class="nn">warnings</span>
<span class="kn">import</span> <span class="nn">time</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">pickle</span>
<span class="kn">import</span> <span class="nn">xlsxwriter</span>
<span class="kn">from</span> <span class="nn">types</span> <span class="kn">import</span> <span class="n">SimpleNamespace</span>
<span class="kn">import</span> <span class="nn">math</span>
<span class="kn">import</span> <span class="nn">scipy</span>
<span class="kn">from</span> <span class="nn">sklearn.utils.extmath</span> <span class="kn">import</span> <span class="n">svd_flip</span>
<span class="c1">#import statsmodels.api as sm</span>
<span class="kn">import</span> <span class="nn">statsmodels.stats.multitest</span> <span class="k">as</span> <span class="nn">sm</span>
<span class="c1">#import shap</span>
<span class="kn">from</span> <span class="nn">scipy.stats</span> <span class="kn">import</span> <span class="n">entropy</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">from</span> <span class="nn">sklearn.model_selection</span> <span class="kn">import</span> <span class="n">KFold</span>


<span class="n">fpath</span><span class="o">=</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="vm">__file__</span><span class="p">),</span><span class="s1">&#39;utils&#39;</span><span class="p">)</span>
<span class="n">sys</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fpath</span><span class="p">)</span>
<span class="c1">#sys.path.insert(1,&#39;./utilities/&#39;)</span>
<span class="c1">#sys.path.insert(1,&#39;./ionmf/factorization/&#39;)</span>
<span class="c1">#from ionmf.factorization.onmf import onmf</span>
<span class="c1">#from ionmf.factorization.model import iONMF</span>
<span class="kn">from</span> <span class="nn">pyliger_utilities</span> <span class="kn">import</span> <span class="n">nnlsm_blockpivot</span><span class="p">,</span><span class="n">iNMF</span><span class="p">,</span><span class="n">NMF_obj_eval</span>


<div class="viewcode-block" id="gene_covariation_analysis">
<a class="viewcode-back" href="../../nico_covariations.html#nico_covariations.Covariations.gene_covariation_analysis">[docs]</a>
<span class="k">def</span> <span class="nf">gene_covariation_analysis</span><span class="p">(</span><span class="n">Radius</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span><span class="n">output_niche_prediction_dir</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<span class="n">refpath</span><span class="o">=</span><span class="s1">&#39;./inputRef/&#39;</span><span class="p">,</span><span class="n">quepath</span><span class="o">=</span><span class="s1">&#39;./inputQuery/&#39;</span><span class="p">,</span><span class="n">ref_cluster_tag</span><span class="o">=</span><span class="s1">&#39;cluster&#39;</span><span class="p">,</span>
<span class="n">ref_original_counts</span><span class="o">=</span><span class="s1">&#39;Original_counts.h5ad&#39;</span><span class="p">,</span>
<span class="n">LRdbFilename</span><span class="o">=</span><span class="s1">&#39;./utils/NiCoLRdb.txt&#39;</span><span class="p">,</span><span class="n">iNMFmode</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">no_of_factors</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span>
<span class="n">shap_analysis</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">shap_cluster_cutoff</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span>
<span class="n">cutoff_to_count_exp_cell_population</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
<span class="n">seed</span><span class="o">=</span><span class="mi">541</span><span class="p">,</span> <span class="n">spatial_integration_modality</span><span class="o">=</span><span class="s1">&#39;double&#39;</span><span class="p">,</span>
<span class="n">anndata_object_name</span><span class="o">=</span><span class="s1">&#39;nico_celltype_annotation.h5ad&#39;</span><span class="p">,</span>
<span class="n">lambda_c</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">power</span><span class="p">(</span><span class="mf">2.0</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="o">-</span><span class="mi">10</span><span class="p">,</span> <span class="mi">10</span><span class="p">))),</span>
<span class="c1">#lambda_c=list(10 * 0.90 ** np.arange(1,100)),</span>
<span class="c1">#lambda_c=[1],</span>
<span class="n">coeff_cutoff_for_rid_reg</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span><span class="n">logistic_coef_cutoff</span><span class="o">=</span><span class="mi">0</span><span class="p">):</span>

    <span class="c1">####Random seed used in RepeatedStratifiedKFold</span>
    <span class="c1">#####seed=3685134seed=3685134,</span>

<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Perform gene covariation analysis within the niche.</span>

<span class="sd">    This is the primary function called by the user to perform gene covariation analysis within the niche.</span>
<span class="sd">    Before calling this function, the user must call the `spatial_neighborhood_analysis` function from the interaction module.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    Radius : int, optional</span>
<span class="sd">        This radius parameter should be the same as used in spatial neighborhood analysis to find the niche interactions.</span>
<span class="sd">        Default is 0.</span>

<span class="sd">    output_niche_prediction_dir : str, optional</span>
<span class="sd">        The output directory location from the previous niche interaction runs generated by the function `spatial_neighborhood_analysis`.</span>
<span class="sd">        Default is &#39;./nico_out/&#39;.</span>

<span class="sd">    refpath : str, optional</span>
<span class="sd">        Path to the reference scRNAseq count matrix in scTransform-like normalization. The filename must be `sct_singleCell.h5ad`.</span>
<span class="sd">        Default is &#39;./inputRef/&#39;.</span>

<span class="sd">    quepath : str, optional</span>
<span class="sd">        Path to the query spatial count matrix in scTransform-like normalization. The filename must be `sct_spatial.h5ad`.</span>
<span class="sd">        Default is &#39;./inputQuery/&#39;.</span>

<span class="sd">    ref_cluster_tag : str, optional</span>
<span class="sd">        The slot in the reference anndata object file where cell type information is stored.</span>
<span class="sd">        Default is &#39;cluster&#39;.</span>

<span class="sd">    ref_original_counts : str, optional</span>
<span class="sd">        Path to the original count data of scRNAseq in anndata object. Must have the cluster information in `.obs` and the umap information in `.obsm[&#39;X_umap&#39;]`.</span>
<span class="sd">        anndata.raw layer should have count matrix data. It will used to find the Spearman correlation and cosine similarity.</span>
<span class="sd">        Default is &#39;Original_counts.h5ad&#39;.</span>

<span class="sd">    LRdbFilename : str, optional</span>
<span class="sd">        Filename of the ligand-receptor database. The first column should be Ligand, the second column Receptor, and the third column the resource list.</span>
<span class="sd">        Default is &#39;./utils/NiCoLRdb.txt&#39;.</span>

<span class="sd">    iNMFmode : bool, optional</span>
<span class="sd">        If True, uses an integrated NMF approach to learn a gene-by-factor submatrix from both modalities.</span>
<span class="sd">        If False, uses an ordinary NMF approach to learn a gene by factor submatrix only from scRNAseq data and transfers these factors to the spatial modality for learning the gene weights.</span>
<span class="sd">        Default is True.</span>

<span class="sd">    no_of_factors : int, optional</span>
<span class="sd">        Number of factors used in NMF for finding the common gene latent dimension space.</span>
<span class="sd">        Default is 3.</span>

<span class="sd">    lambda_c : list, optional</span>
<span class="sd">        Initial range of regularization parameters used in the ridge regression step to find the optimal parameter.</span>
<span class="sd">        Default is `list(np.power(2.0, np.arange(-10, 10)))`.</span>

<span class="sd">    shap_analysis : bool, optional</span>
<span class="sd">        Flag to perform SHAP analysis.</span>
<span class="sd">        Default is False.</span>

<span class="sd">    shap_cluster_cutoff : float, optional</span>
<span class="sd">        SHAP analysis cutoff parameter.</span>
<span class="sd">        Default is 0.5.</span>

<span class="sd">    coeff_cutoff_for_rid_reg : float, optional</span>
<span class="sd">        Cutoff used to create the list of significant celltype_factor-celltype_factor niche covariations with an absolute regression coefficient greater than this.</span>
<span class="sd">        Default is 0.</span>

<span class="sd">    cutoff_to_count_exp_cell_population : int, optional</span>
<span class="sd">        Parameter to find the percentage of the cell population that express a given gene in a given cell type. Value 0 is acceptable with count data.</span>
<span class="sd">        Default is 0.</span>

<span class="sd">    seed : int, optional</span>
<span class="sd">        Random seed used in RepeatedStratifiedKFold.</span>
<span class="sd">        Default is 541.</span>

<span class="sd">    spatial_integration_modality : str, optional</span>
<span class="sd">        Modality for spatial integration if both scRNAseq and Spatial data is available.</span>
<span class="sd">        Default is &#39;double&#39;.</span>
<span class="sd">        For only spatial data this value must be &#39;single&#39;</span>

<span class="sd">    anndata_object_name : str, optional</span>
<span class="sd">        Name of the spatial anndata object name.</span>
<span class="sd">        Default is &#39;nico_celltype_annotation.h5ad&#39;.</span>

<span class="sd">    logistic_coef_cutoff : float, optional</span>
<span class="sd">        Cutoff to retrieve the positive niche interactions (cell type - cell type). For values &gt;0, cell type pairs are likely to interact.</span>
<span class="sd">        Default is 0.</span>


<span class="sd">    Outputs</span>
<span class="sd">    -------</span>
<span class="sd">    The output is saved in the directory specified by `output_niche_prediction_dir`, with default location being &#39;./nico_out/covariations_R*_F*&#39;.</span>

<span class="sd">    Notes</span>
<span class="sd">    -----</span>
<span class="sd">    - Please provide `Original_counts.h5ad`, `sct_singleCell.h5ad` files from scRNAseq data.</span>
<span class="sd">    - Provide  `sct_spatial.h5ad` files for the spatial transcriptomics data.</span>
<span class="sd">    - `Original_counts.h5ad` object should also have the cluster information in `.obs` and the umap information in `.obsm` and .raw layer has count data.</span>
<span class="sd">    &quot;&quot;&quot;</span>


    <span class="c1">#ref_h5ad=refpath+&#39;sct_singleCell.h5ad&#39;</span>
    <span class="c1">#que_h5ad=quepath+&#39;sct_spatial.h5ad&#39;</span>



    <span class="k">if</span> <span class="n">output_niche_prediction_dir</span><span class="o">==</span><span class="kc">None</span><span class="p">:</span>
        <span class="n">niche_prediction_dir</span><span class="o">=</span><span class="s1">&#39;./nico_out/&#39;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">niche_prediction_dir</span><span class="o">=</span><span class="n">output_niche_prediction_dir</span>


    <span class="k">if</span> <span class="n">spatial_integration_modality</span><span class="o">==</span><span class="s1">&#39;single&#39;</span><span class="p">:</span>
        <span class="n">original_h5ad</span><span class="o">=</span><span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="n">niche_prediction_dir</span><span class="o">+</span><span class="n">anndata_object_name</span><span class="p">)</span>
        <span class="n">spatial_adata</span><span class="o">=</span><span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="n">niche_prediction_dir</span><span class="o">+</span><span class="n">anndata_object_name</span><span class="p">)</span>


    <span class="k">if</span> <span class="n">spatial_integration_modality</span><span class="o">==</span><span class="s1">&#39;double&#39;</span><span class="p">:</span>
        <span class="n">spatial_adata</span><span class="o">=</span><span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="n">quepath</span><span class="o">+</span><span class="s1">&#39;sct_spatial.h5ad&#39;</span><span class="p">)</span>
        <span class="n">ref_adata</span><span class="o">=</span><span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="n">refpath</span><span class="o">+</span><span class="s1">&#39;sct_singleCell.h5ad&#39;</span><span class="p">)</span>
        <span class="n">original_h5ad</span><span class="o">=</span><span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="n">refpath</span><span class="o">+</span><span class="n">ref_original_counts</span><span class="p">)</span>

    <span class="n">df</span><span class="o">=</span><span class="n">original_h5ad</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">ref_cluster_tag</span><span class="p">]</span>
    <span class="n">annotation_singlecell_celltypename</span><span class="o">=</span><span class="n">df</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span>
    <span class="n">cellname</span><span class="o">=</span><span class="n">df</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span>


    <span class="n">sc_ct_name</span><span class="o">=</span><span class="p">[]</span>
    <span class="n">A</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">annotation_singlecell_celltypename</span><span class="p">)))</span>
    <span class="n">d</span><span class="o">=</span><span class="p">{}</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">A</span><span class="p">)):</span>
        <span class="n">sc_ct_name</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">i</span><span class="p">,</span><span class="n">A</span><span class="p">[</span><span class="n">i</span><span class="p">]])</span>
        <span class="n">d</span><span class="p">[</span><span class="n">A</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span><span class="o">=</span><span class="n">i</span>
    <span class="n">sc_ct_name</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">sc_ct_name</span><span class="p">)</span>

    <span class="n">sc_cluster</span><span class="o">=</span><span class="p">[]</span>
    <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">annotation_singlecell_celltypename</span><span class="p">)):</span>
        <span class="n">sc_cluster</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">cellname</span><span class="p">[</span><span class="n">j</span><span class="p">],</span><span class="n">d</span><span class="p">[</span><span class="n">annotation_singlecell_celltypename</span><span class="p">[</span><span class="n">j</span><span class="p">]]])</span>
    <span class="n">sc_cluster</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">sc_cluster</span><span class="p">)</span>
    <span class="n">annotation_singlecell_barcode_id</span><span class="o">=</span><span class="n">sc_cluster</span><span class="p">[:,</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">annotation_singlecell_cluster_id</span><span class="o">=</span><span class="n">sc_cluster</span><span class="p">[:,</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">singlecell_unique_clustername</span><span class="o">=</span><span class="n">sc_ct_name</span><span class="p">[:,</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">singlecell_unique_clusterid</span><span class="o">=</span><span class="n">sc_ct_name</span><span class="p">[:,</span><span class="mi">0</span><span class="p">]</span>


    <span class="n">f</span><span class="o">=</span><span class="nb">open</span><span class="p">(</span><span class="n">LRdbFilename</span><span class="p">,</span><span class="s1">&#39;r&#39;</span><span class="p">)</span>
    <span class="n">LRdb</span><span class="o">=</span><span class="n">f</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
    <span class="n">f</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>


    <span class="c1">#print(&quot;spatial&quot;,np.sum(spatial_adata.X),np.sum(spatial_adata.raw.X))</span>
    <span class="c1">#print(&quot;singlecell&quot;,np.sum(ref_adata.X),np.sum(ref_adata.raw.X))</span>

    <span class="k">if</span> <span class="n">spatial_integration_modality</span><span class="o">==</span><span class="s1">&#39;single&#39;</span><span class="p">:</span>
        <span class="n">sct_ad_sp</span><span class="o">=</span><span class="n">original_h5ad</span><span class="o">.</span><span class="n">raw</span>
        <span class="n">sct_ad_sc</span><span class="o">=</span><span class="n">original_h5ad</span><span class="o">.</span><span class="n">raw</span>
        <span class="n">full_ad_sc</span><span class="o">=</span><span class="n">original_h5ad</span><span class="o">.</span><span class="n">raw</span>

    <span class="k">if</span> <span class="n">spatial_integration_modality</span><span class="o">==</span><span class="s1">&#39;double&#39;</span><span class="p">:</span>
        <span class="n">sct_ad_sp</span><span class="o">=</span><span class="n">spatial_adata</span><span class="o">.</span><span class="n">raw</span>
        <span class="n">sct_ad_sc</span><span class="o">=</span><span class="n">ref_adata</span><span class="o">.</span><span class="n">raw</span>
        <span class="n">full_ad_sc</span><span class="o">=</span><span class="n">original_h5ad</span><span class="o">.</span><span class="n">raw</span>

    <span class="n">covariation_outdir</span><span class="o">=</span><span class="n">niche_prediction_dir</span><span class="o">+</span><span class="s1">&#39;covariations_&#39;</span>

    <span class="n">strategy</span><span class="o">=</span><span class="s1">&#39;niche_prediction_linear/&#39;</span>
    <span class="n">gene_set_names</span><span class="o">=</span><span class="p">[]</span>

    <span class="c1">#print(&#39;sc1 annotation_singlecell_cluster_id&#39;,len(annotation_singlecell_cluster_id))</span>
    <span class="c1">#print(&#39;sc2 annotation_singlecell_barcode_id&#39;,len(annotation_singlecell_barcode_id))</span>
    <span class="c1">#print(&#39;sc3 annotation_singlecell_celltypename&#39;,len(annotation_singlecell_celltypename))</span>
    <span class="c1">#print(&#39;sc4 singlecell_unique_clustername&#39;, len(singlecell_unique_clustername))</span>


    <span class="c1"># load spatial dat</span>
    <span class="n">sp_genename</span><span class="o">=</span><span class="n">sct_ad_sp</span><span class="o">.</span><span class="n">var_names</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span>
    <span class="n">sc_genename</span><span class="o">=</span><span class="n">sct_ad_sc</span><span class="o">.</span><span class="n">var_names</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span>
    <span class="n">index_sp</span><span class="p">,</span><span class="n">index_sc</span><span class="o">=</span><span class="n">find_index</span><span class="p">(</span><span class="n">sp_genename</span><span class="p">,</span><span class="n">sc_genename</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;common genes between sc and sp&#39;</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">index_sp</span><span class="p">),</span><span class="nb">len</span><span class="p">(</span><span class="n">index_sc</span><span class="p">))</span>
    <span class="n">ad_sp_ori</span><span class="o">=</span><span class="n">sct_ad_sp</span><span class="p">[:,</span><span class="n">index_sp</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">ad_sc_ori</span><span class="o">=</span><span class="n">sct_ad_sc</span><span class="p">[:,</span><span class="n">index_sc</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

    <span class="n">sct_ad_sp</span><span class="o">=</span><span class="mi">0</span>
    <span class="n">sct_ad_sc</span><span class="o">=</span><span class="mi">0</span>
    <span class="c1">#spatial_adata=0</span>
    <span class="n">ref_adata</span><span class="o">=</span><span class="mi">0</span>

    <span class="n">inputRadius</span><span class="o">=</span><span class="p">[</span><span class="n">Radius</span><span class="p">]</span>
    <span class="k">for</span> <span class="n">radius</span> <span class="ow">in</span> <span class="n">inputRadius</span><span class="p">:</span>
        <span class="n">celltypeFilename</span><span class="o">=</span><span class="n">niche_prediction_dir</span><span class="o">+</span><span class="s1">&#39;used_CT.txt&#39;</span>
        <span class="n">clusterFilename</span><span class="o">=</span><span class="n">niche_prediction_dir</span><span class="o">+</span><span class="s1">&#39;used_Clusters&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">radius</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;.csv&#39;</span>

        <span class="n">annotation_spatial_celltypename</span><span class="p">,</span><span class="n">annotation_spatial_barcode_id</span><span class="p">,</span><span class="n">annotation_spatial_cluster_id</span><span class="p">,</span><span class="n">spatialcell_unique_clustername</span><span class="p">,</span><span class="n">spatialcell_unique_clusterid</span><span class="o">=</span><span class="n">read_spatial_data</span><span class="p">(</span><span class="n">clusterFilename</span><span class="p">,</span><span class="n">celltypeFilename</span><span class="p">)</span>


        <span class="n">neighbors</span><span class="o">=</span><span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span> <span class="nb">open</span><span class="p">(</span><span class="n">niche_prediction_dir</span><span class="o">+</span><span class="s1">&#39;neighbors_&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">radius</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;.p&#39;</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span> <span class="p">)</span> <span class="p">)</span>
        <span class="n">distances</span><span class="o">=</span><span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span> <span class="nb">open</span><span class="p">(</span><span class="n">niche_prediction_dir</span><span class="o">+</span><span class="s1">&#39;distances_&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">radius</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;.p&#39;</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span> <span class="p">)</span> <span class="p">)</span>

        <span class="n">covariation_dir</span><span class="o">=</span><span class="n">covariation_outdir</span><span class="o">+</span><span class="s1">&#39;R&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">radius</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;_F&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">no_of_factors</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;/&#39;</span>
        <span class="n">create_directory</span><span class="p">(</span><span class="n">covariation_dir</span><span class="p">)</span>
        <span class="n">outputname</span><span class="o">=</span><span class="n">covariation_dir</span><span class="o">+</span><span class="s1">&#39;Principal_component_feature_matrix.npz&#39;</span>
        <span class="n">inputdata</span><span class="o">=</span><span class="p">{}</span>
        <span class="n">inputdata</span><span class="p">[</span><span class="s1">&#39;no_of_pc&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">no_of_factors</span>
        <span class="n">inputdata</span><span class="p">[</span><span class="s1">&#39;outputname&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">outputname</span>
        <span class="n">inputdata</span><span class="p">[</span><span class="s1">&#39;covariation_dir&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">covariation_dir</span>


        <span class="n">fname</span><span class="o">=</span><span class="n">niche_prediction_dir</span><span class="o">+</span><span class="n">strategy</span><span class="o">+</span><span class="s1">&#39;/classifier_matrices_&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">radius</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;.npz&#39;</span>
        <span class="n">data</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">fname</span><span class="p">,</span><span class="n">allow_pickle</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">logistic_coef</span><span class="o">=</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;coef&#39;</span><span class="p">]</span>
        <span class="n">logistic_cmn</span><span class="o">=</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;cmn&#39;</span><span class="p">]</span>
        <span class="n">logistic_cmn_std</span><span class="o">=</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;cmn_std&#39;</span><span class="p">]</span>
        <span class="n">logistic_coef_std</span><span class="o">=</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;coef_std&#39;</span><span class="p">]</span>
        <span class="n">logistic_CTFeatures</span><span class="o">=</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;CTFeatures&#39;</span><span class="p">]</span>
        <span class="c1">#f=open(input_spatial+&#39;BiologicalNameOfCT.dat&#39;)</span>
        <span class="n">f</span><span class="o">=</span><span class="nb">open</span><span class="p">(</span><span class="n">celltypeFilename</span><span class="p">)</span>
        <span class="n">nameOfCellType</span><span class="o">=</span><span class="p">{}</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">l</span><span class="o">=</span><span class="n">line</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
            <span class="n">nameOfCellType</span><span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">l</span><span class="p">[</span><span class="mi">0</span><span class="p">])]</span><span class="o">=</span><span class="n">l</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>

        <span class="n">logistic_predicted_interactions</span><span class="o">=</span><span class="n">find_logistic_regression_interacting_score</span><span class="p">(</span><span class="n">logistic_cmn</span><span class="p">,</span><span class="n">logistic_coef</span><span class="p">,</span><span class="n">logistic_CTFeatures</span><span class="p">,</span><span class="n">nameOfCellType</span><span class="p">,</span><span class="n">logistic_coef_cutoff</span><span class="p">)</span>



        <span class="n">inputdata</span><span class="p">[</span><span class="s1">&#39;ad_sp&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">ad_sp_ori</span> <span class="c1">#sct_ad_sp</span>
        <span class="n">inputdata</span><span class="p">[</span><span class="s1">&#39;ad_sc&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">ad_sc_ori</span><span class="c1">#sct_ad_sc#</span>
        <span class="n">inputdata</span><span class="p">[</span><span class="s1">&#39;umap_plot_sc&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">original_h5ad</span>
        <span class="n">inputdata</span><span class="p">[</span><span class="s1">&#39;umap_plot_sp&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">spatial_adata</span>
        <span class="n">inputdata</span><span class="p">[</span><span class="s1">&#39;annotation_spatial_cluster_id&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">annotation_spatial_cluster_id</span>
        <span class="n">inputdata</span><span class="p">[</span><span class="s1">&#39;annotation_spatial_barcode_id&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">annotation_spatial_barcode_id</span>
        <span class="n">inputdata</span><span class="p">[</span><span class="s1">&#39;annotation_spatial_celltypename&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">annotation_spatial_celltypename</span>
        <span class="n">inputdata</span><span class="p">[</span><span class="s1">&#39;spatialcell_unique_clustername&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">spatialcell_unique_clustername</span>
        <span class="n">inputdata</span><span class="p">[</span><span class="s1">&#39;spatialcell_unique_clusterid&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">spatialcell_unique_clusterid</span>

        <span class="n">inputdata</span><span class="p">[</span><span class="s1">&#39;annotation_singlecell_cluster_id&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">annotation_singlecell_cluster_id</span>
        <span class="n">inputdata</span><span class="p">[</span><span class="s1">&#39;annotation_singlecell_barcode_id&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">annotation_singlecell_barcode_id</span>
        <span class="n">inputdata</span><span class="p">[</span><span class="s1">&#39;annotation_singlecell_celltypename&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">annotation_singlecell_celltypename</span>
        <span class="n">inputdata</span><span class="p">[</span><span class="s1">&#39;singlecell_unique_clustername&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">singlecell_unique_clustername</span>
        <span class="n">inputdata</span><span class="p">[</span><span class="s1">&#39;singlecell_unique_clusterid&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">singlecell_unique_clusterid</span>
        <span class="n">inputdata</span><span class="p">[</span><span class="s1">&#39;neighbors&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">neighbors</span>
        <span class="n">inputdata</span><span class="p">[</span><span class="s1">&#39;neigh_distances&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">distances</span>
        <span class="n">inputdata</span><span class="p">[</span><span class="s1">&#39;nmf_output&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">covariation_dir</span><span class="o">+</span><span class="s1">&#39;NMF_output/&#39;</span>

        <span class="n">regression_outdir</span><span class="o">=</span><span class="n">covariation_dir</span><span class="o">+</span><span class="s1">&#39;Regression_outputs&#39;</span><span class="o">+</span><span class="s1">&#39;/&#39;</span>
        <span class="n">create_directory</span><span class="p">(</span><span class="n">regression_outdir</span><span class="p">)</span>
        <span class="n">inputdata</span><span class="p">[</span><span class="s1">&#39;seed&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">seed</span>
        <span class="n">inputdata</span><span class="p">[</span><span class="s1">&#39;lambda_c&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">lambda_c</span>
        <span class="n">inputdata</span><span class="p">[</span><span class="s1">&#39;iNMFmode&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">iNMFmode</span>
        <span class="n">inputdata</span><span class="p">[</span><span class="s1">&#39;regression_outdir&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">regression_outdir</span>
        <span class="c1">#inputdata[&#39;K_fold&#39;]=K_fold</span>
        <span class="c1">#inputdata[&#39;n_repeats&#39;]=n_repeats</span>
        <span class="c1">#inputdata[&#39;n_jobs&#39;]=n_jobs</span>
        <span class="n">inputdata</span><span class="p">[</span><span class="s1">&#39;shap_analysis&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">shap_analysis</span>
        <span class="n">inputdata</span><span class="p">[</span><span class="s1">&#39;shap_cluster_cutoff&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">shap_cluster_cutoff</span>

        <span class="n">inputdata</span><span class="p">[</span><span class="s1">&#39;logistic_coef_cutoff&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">logistic_coef_cutoff</span>
        <span class="n">inputdata</span><span class="p">[</span><span class="s1">&#39;coeff_cutoff_for_rid_reg&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">coeff_cutoff_for_rid_reg</span>
        <span class="n">inputdata</span><span class="p">[</span><span class="s1">&#39;gene_set_names&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">gene_set_names</span>
        <span class="n">inputdata</span><span class="p">[</span><span class="s1">&#39;spatial_integration_modality&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">spatial_integration_modality</span>
        <span class="c1">#inputdata[&#39;pvalueCutoff&#39;]=pvalueCutoff</span>

        <span class="n">inputdata</span><span class="p">[</span><span class="s1">&#39;cutoff_to_count_exp_cell_population&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">cutoff_to_count_exp_cell_population</span>
        <span class="n">inputdata</span><span class="p">[</span><span class="s1">&#39;LRdb&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">LRdb</span>

        <span class="nb">input</span><span class="o">=</span><span class="n">SimpleNamespace</span><span class="p">(</span><span class="o">**</span><span class="n">inputdata</span><span class="p">)</span>

        <span class="n">flag</span><span class="o">=</span><span class="mi">1</span>
        <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">outputname</span><span class="p">):</span>
            <span class="n">filesize</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">getsize</span><span class="p">(</span><span class="n">outputname</span><span class="p">)</span>
            <span class="c1">#if filesize&gt;0: #If file is already exist and have size greater than 0 then no need to run again. It will save some time if you want to run it again with different parameters</span>
            <span class="c1">#    flag=0</span>

        <span class="k">if</span> <span class="n">flag</span><span class="o">==</span><span class="mi">1</span><span class="p">:</span>
            <span class="n">pc_of_sp_clusterid</span><span class="p">,</span><span class="n">PCA_of_sc_cluster_accordingto_spatial_clusterid</span><span class="p">,</span><span class="n">save_scFactors</span><span class="p">,</span><span class="n">save_spFactors</span><span class="o">=</span><span class="n">compute_PC_space</span><span class="p">(</span><span class="nb">input</span><span class="p">,</span><span class="n">full_ad_sc</span><span class="p">)</span>
            <span class="c1"># full_ad_sc use in only find_PC_of_invidualCluster_in_SC function</span>
            <span class="c1"># ideally it should be sctransform way of normalized matrix equivalent to sct_ad_sc but</span>
            <span class="c1"># if not then need to do perform scaling HVG etc</span>
            <span class="n">pickle</span><span class="o">.</span><span class="n">dump</span><span class="p">((</span><span class="n">PCA_of_sc_cluster_accordingto_spatial_clusterid</span><span class="p">,</span><span class="n">save_scFactors</span><span class="p">,</span><span class="n">save_spFactors</span><span class="p">),</span><span class="nb">open</span><span class="p">(</span><span class="n">covariation_dir</span><span class="o">+</span><span class="s1">&#39;factors_info.p&#39;</span><span class="p">,</span> <span class="s1">&#39;wb&#39;</span><span class="p">))</span>
            <span class="n">inputdata</span><span class="p">[</span><span class="s1">&#39;pc_of_sp_clusterid&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">pc_of_sp_clusterid</span>
            <span class="nb">input</span><span class="o">=</span><span class="n">SimpleNamespace</span><span class="p">(</span><span class="o">**</span><span class="n">inputdata</span><span class="p">)</span>
            <span class="n">makePCneighboorhoodFeatureMatrix</span><span class="p">(</span><span class="nb">input</span><span class="p">)</span>


        <span class="n">save_reg_coef</span><span class="o">=</span><span class="n">model_linear_regression</span><span class="p">(</span><span class="nb">input</span><span class="p">,</span><span class="n">logistic_predicted_interactions</span><span class="p">)</span>
        <span class="n">inputdata</span><span class="p">[</span><span class="s1">&#39;save_reg_coef&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">save_reg_coef</span>
        <span class="nb">input</span><span class="o">=</span><span class="n">SimpleNamespace</span><span class="p">(</span><span class="o">**</span><span class="n">inputdata</span><span class="p">)</span>
    <span class="k">return</span> <span class="nb">input</span></div>



<div class="viewcode-block" id="plot_cosine_and_spearman_correlation_to_factors">
<a class="viewcode-back" href="../../nico_covariations.html#nico_covariations.Covariations.plot_cosine_and_spearman_correlation_to_factors">[docs]</a>
<span class="k">def</span> <span class="nf">plot_cosine_and_spearman_correlation_to_factors</span><span class="p">(</span><span class="nb">input</span><span class="p">,</span><span class="n">choose_celltypes</span><span class="o">=</span><span class="p">[],</span><span class="n">NOG_Fa</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span><span class="n">saveas</span><span class="o">=</span><span class="s1">&#39;pdf&#39;</span><span class="p">,</span><span class="n">transparent_mode</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">showit</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">15</span><span class="p">,</span><span class="mi">10</span><span class="p">)):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Plots cosine and Spearman correlation to factors for given cell types.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    -----------</span>
<span class="sd">    input : object</span>
<span class="sd">        The main input object containing the output from gene_covariation_analysis.</span>

<span class="sd">    choose_celltypes : list, optional</span>
<span class="sd">        The cell types for which you want to inspect the covariation pattern.</span>
<span class="sd">        If the list is empty, the output will be generated for all cell types.</span>
<span class="sd">        Default is [].</span>

<span class="sd">    NOG_Fa : int, optional</span>
<span class="sd">        Number of genes to visualize for each factor.</span>
<span class="sd">        Default is 30.</span>

<span class="sd">    saveas : str, optional</span>
<span class="sd">        Format to save the figures. Options are &#39;pdf&#39; or &#39;png&#39;. If &#39;png&#39; is chosen, dpi is set to 300.</span>
<span class="sd">        Default is &#39;pdf&#39;.</span>

<span class="sd">    transparent_mode : bool, optional</span>
<span class="sd">        Whether to save the figure with a transparent background.</span>
<span class="sd">        Default is False.</span>

<span class="sd">    showit : bool, optional</span>
<span class="sd">        Whether to display the plot interactively.</span>
<span class="sd">        Default is True.</span>

<span class="sd">    figsize : tuple, optional</span>
<span class="sd">        Dimensions of the figure size.</span>
<span class="sd">        Default is (15, 10).</span>

<span class="sd">    Outputs:</span>
<span class="sd">    --------</span>
<span class="sd">    The output NMF plots are saved in ./&lt;output_nico_dir&gt;/covariations_R*_F*/NMF_output.</span>
<span class="sd">    &quot;&quot;&quot;</span>


    <span class="n">create_directory</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">nmf_output</span><span class="p">)</span>

    <span class="n">n</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clusterid</span><span class="p">)</span>

    <span class="n">perform</span><span class="o">=</span><span class="p">[]</span>
    <span class="n">found</span><span class="o">=</span><span class="p">[]</span>
    <span class="k">for</span> <span class="n">fi</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="p">):</span>
          <span class="n">CC_celltype_name</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clustername</span><span class="p">[</span><span class="n">fi</span><span class="p">]</span>
          <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">choose_celltypes</span><span class="p">)</span><span class="o">==</span><span class="mi">0</span><span class="p">:</span>
              <span class="n">perform</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fi</span><span class="p">)</span>
          <span class="k">else</span><span class="p">:</span>
              <span class="k">if</span> <span class="n">CC_celltype_name</span> <span class="ow">in</span> <span class="n">choose_celltypes</span><span class="p">:</span>
                  <span class="n">perform</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fi</span><span class="p">)</span>
                  <span class="n">found</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">CC_celltype_name</span><span class="p">)</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">choose_celltypes</span><span class="p">)</span><span class="o">!=</span><span class="mi">0</span><span class="p">:</span>
          <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;cell types found &quot;</span><span class="p">,</span><span class="n">found</span><span class="p">)</span>


    <span class="n">xlabels</span><span class="o">=</span><span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">no_of_pc</span><span class="p">):</span>
        <span class="n">xlabels</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;NMF&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">))</span>

    <span class="n">PCA_of_sc_cluster_accordingto_spatial_clusterid</span><span class="p">,</span><span class="n">save_scFactors</span><span class="p">,</span><span class="n">save_spFactors</span><span class="o">=</span><span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">covariation_dir</span><span class="o">+</span><span class="s1">&#39;factors_info.p&#39;</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">))</span>
    <span class="n">n</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clustername</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">fi</span> <span class="ow">in</span> <span class="n">perform</span><span class="p">:</span>
        <span class="n">clid</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clusterid</span><span class="p">[</span><span class="n">fi</span><span class="p">]</span>
        <span class="n">spearman_factors</span><span class="p">,</span><span class="n">CC_PCA</span><span class="p">,</span><span class="n">CC_gene</span><span class="p">,</span><span class="n">CC_meanExpression</span><span class="p">,</span><span class="n">CC_popExpression</span><span class="p">,</span><span class="n">cosine_factors</span><span class="p">,</span><span class="n">alpha</span><span class="o">=</span><span class="n">PCA_of_sc_cluster_accordingto_spatial_clusterid</span><span class="p">[</span><span class="n">clid</span><span class="p">]</span>
        <span class="n">CC_celltype_name</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clustername</span><span class="p">[</span><span class="n">fi</span><span class="p">]</span>

        <span class="n">genename_full</span><span class="o">=</span><span class="n">CC_gene</span>
        <span class="c1">#sc_cosine=find_correlation_bw_genes_and_PC_component_in_singlecell_cosine(H.T,CbyG)</span>
        <span class="n">gname2b</span><span class="p">,</span><span class="n">geneNMF2b</span><span class="o">=</span><span class="n">top_genes_in_correlation_list_without</span><span class="p">(</span><span class="n">genename_full</span><span class="p">,</span><span class="n">cosine_factors</span><span class="p">,</span><span class="n">NOG_Fa</span><span class="p">)</span>
        <span class="c1">#sc_spearman=find_correlation_bw_genes_and_PC_component_in_singlecell(H.T,CbyG)</span>
        <span class="n">gname3b</span><span class="p">,</span><span class="n">geneNMF3b</span><span class="o">=</span><span class="n">top_genes_in_correlation_list_without</span><span class="p">(</span><span class="n">genename_full</span><span class="p">,</span><span class="n">spearman_factors</span><span class="p">,</span><span class="n">NOG_Fa</span><span class="p">)</span>

        <span class="n">selectedGenesAvgExp_cosine</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">gname2b</span><span class="p">),</span><span class="mi">1</span><span class="p">)</span> <span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">gname2b</span><span class="p">)):</span>
            <span class="n">ind</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">genename_full</span><span class="o">==</span><span class="n">gname2b</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
            <span class="n">selectedGenesAvgExp_cosine</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="mi">0</span><span class="p">]</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">CC_meanExpression</span><span class="p">[</span><span class="n">ind</span><span class="p">[</span><span class="mi">0</span><span class="p">]])</span>

        <span class="n">selectedGenesAvgExp</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">gname3b</span><span class="p">),</span><span class="mi">1</span><span class="p">)</span> <span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">gname3b</span><span class="p">)):</span>
            <span class="n">ind</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">genename_full</span><span class="o">==</span><span class="n">gname3b</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
            <span class="n">selectedGenesAvgExp</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="mi">0</span><span class="p">]</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">CC_meanExpression</span><span class="p">[</span><span class="n">ind</span><span class="p">[</span><span class="mi">0</span><span class="p">]])</span>


        <span class="n">fig</span><span class="o">=</span><span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">)</span>
        <span class="n">gs</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_gridspec</span><span class="p">(</span><span class="n">ncols</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span> <span class="n">nrows</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">wspace</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span><span class="n">width_ratios</span><span class="o">=</span><span class="p">[</span><span class="mf">0.5</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mf">0.5</span><span class="p">])</span>
        <span class="n">ax0</span><span class="o">=</span><span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="n">ax1</span><span class="o">=</span><span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="n">ax2</span><span class="o">=</span><span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
        <span class="n">ax3</span><span class="o">=</span><span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>
        <span class="n">b</span><span class="o">=</span><span class="n">snn</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">selectedGenesAvgExp_cosine</span><span class="p">,</span><span class="n">yticklabels</span><span class="o">=</span><span class="n">gname2b</span><span class="p">,</span><span class="n">ax</span><span class="o">=</span><span class="n">ax0</span><span class="p">)</span><span class="c1">#componentlabel,ax=ax</span>
        <span class="n">b</span><span class="o">.</span><span class="n">set_yticklabels</span><span class="p">(</span><span class="n">b</span><span class="o">.</span><span class="n">get_ymajorticklabels</span><span class="p">(),</span> <span class="n">fontsize</span> <span class="o">=</span> <span class="mi">6</span><span class="p">)</span>
        <span class="n">b</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;log(avg exp)&#39;</span><span class="p">)</span>


        <span class="n">b</span><span class="o">=</span><span class="n">snn</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">geneNMF2b</span><span class="p">,</span><span class="n">yticklabels</span><span class="o">=</span><span class="n">gname2b</span><span class="p">,</span><span class="n">ax</span><span class="o">=</span><span class="n">ax1</span><span class="p">)</span><span class="c1">#componentlabel,ax=ax</span>
        <span class="n">b</span><span class="o">.</span><span class="n">set_xticklabels</span><span class="p">(</span><span class="n">xlabels</span><span class="p">,</span><span class="n">size</span> <span class="o">=</span> <span class="mi">8</span><span class="p">,</span><span class="n">rotation</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span>
        <span class="n">b</span><span class="o">.</span><span class="n">set_yticklabels</span><span class="p">(</span><span class="n">b</span><span class="o">.</span><span class="n">get_ymajorticklabels</span><span class="p">(),</span> <span class="n">fontsize</span> <span class="o">=</span> <span class="mi">6</span><span class="p">)</span>
        <span class="n">b</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;cosine&#39;</span><span class="p">)</span>
        <span class="c1">#b.set_title(&#39;spatial&#39;+entropy_SH)</span>
        <span class="n">b</span><span class="o">=</span><span class="n">snn</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">geneNMF3b</span><span class="p">,</span><span class="n">yticklabels</span><span class="o">=</span><span class="n">gname3b</span><span class="p">,</span><span class="n">ax</span><span class="o">=</span><span class="n">ax2</span><span class="p">)</span><span class="c1">#componentlabel,ax=ax</span>
        <span class="n">b</span><span class="o">.</span><span class="n">set_xticklabels</span><span class="p">(</span><span class="n">xlabels</span><span class="p">,</span><span class="n">size</span> <span class="o">=</span> <span class="mi">8</span><span class="p">,</span><span class="n">rotation</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span>
        <span class="n">b</span><span class="o">.</span><span class="n">set_yticklabels</span><span class="p">(</span><span class="n">b</span><span class="o">.</span><span class="n">get_ymajorticklabels</span><span class="p">(),</span> <span class="n">fontsize</span> <span class="o">=</span> <span class="mi">6</span><span class="p">)</span>
        <span class="n">b</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;spearman corr &#39;</span><span class="o">+</span><span class="n">CC_celltype_name</span><span class="o">+</span><span class="s1">&#39;, lambda = &#39;</span><span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">alpha</span><span class="p">))</span>

        <span class="n">b</span><span class="o">=</span><span class="n">snn</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">selectedGenesAvgExp</span><span class="p">,</span><span class="n">yticklabels</span><span class="o">=</span><span class="n">gname3b</span><span class="p">,</span><span class="n">ax</span><span class="o">=</span><span class="n">ax3</span><span class="p">)</span><span class="c1">#componentlabel,ax=ax</span>
        <span class="c1">#b.set_xticklabels(&#39;exp&#39;,size = 8,rotation=90)</span>
        <span class="n">b</span><span class="o">.</span><span class="n">set_yticklabels</span><span class="p">(</span><span class="n">b</span><span class="o">.</span><span class="n">get_ymajorticklabels</span><span class="p">(),</span> <span class="n">fontsize</span> <span class="o">=</span> <span class="mi">6</span><span class="p">)</span>
        <span class="n">b</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;log(avg exp)&#39;</span><span class="p">)</span>
        <span class="c1">#plt.tight_layout()</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;The figures are saved: &quot;</span><span class="p">,</span> <span class="nb">input</span><span class="o">.</span><span class="n">nmf_output</span><span class="o">+</span><span class="n">remove_extra_character_from_name</span><span class="p">(</span><span class="n">CC_celltype_name</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;.&#39;</span><span class="o">+</span><span class="n">saveas</span><span class="p">)</span>
        <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">nmf_output</span><span class="o">+</span><span class="n">remove_extra_character_from_name</span><span class="p">(</span><span class="n">CC_celltype_name</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;.&#39;</span><span class="o">+</span><span class="n">saveas</span><span class="p">,</span><span class="n">bbox_inches</span><span class="o">=</span><span class="s1">&#39;tight&#39;</span><span class="p">,</span><span class="n">transparent</span><span class="o">=</span><span class="n">transparent_mode</span><span class="p">,</span><span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">showit</span><span class="p">:</span>
            <span class="k">pass</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="s1">&#39;all&#39;</span><span class="p">)</span></div>


<div class="viewcode-block" id="plot_feature_matrices">
<a class="viewcode-back" href="../../nico_covariations.html#nico_covariations.Covariations.plot_feature_matrices">[docs]</a>
<span class="k">def</span> <span class="nf">plot_feature_matrices</span><span class="p">(</span><span class="nb">input</span><span class="p">,</span><span class="n">showit</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">saveas</span><span class="o">=</span><span class="s1">&#39;pdf&#39;</span><span class="p">,</span><span class="n">transparent_mode</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span><span class="mi">10</span><span class="p">)):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Plots feature vectors of the spatial factors from the regression step.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1">#maindir1=outputdir+&#39;covariations_&#39;</span>
    <span class="c1">#maindir=maindir1+str(radius)+&#39;/&#39;</span>
    <span class="c1">#outputname=maindir+&#39;Principal_component_feature_matrix&#39;+str(no_of_factors)+&#39;.npz&#39;</span>
    <span class="n">ylabelname</span><span class="o">=</span><span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clustername</span><span class="p">)):</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">no_of_pc</span><span class="p">):</span>
            <span class="n">ylabelname</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clustername</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">+</span><span class="s1">&#39;_&#39;</span><span class="o">+</span><span class="s1">&#39;Fa&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">j</span><span class="o">+</span><span class="mi">1</span><span class="p">))</span>

    <span class="n">data1</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">outputname</span><span class="p">,</span><span class="n">allow_pickle</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">data</span><span class="o">=</span><span class="n">data1</span><span class="p">[</span><span class="s1">&#39;weighted_neighborhood_of_factors_in_niche&#39;</span><span class="p">]</span>
    <span class="n">fig</span><span class="p">,</span><span class="n">axs</span><span class="o">=</span><span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">)</span>
    <span class="c1">#data=np.genfromtxt(open(name, &quot;rb&quot;), delimiter=&#39;,&#39;, skip_header=0)</span>
    <span class="n">Feature</span><span class="o">=</span><span class="n">data</span><span class="p">[:,(</span><span class="mi">1</span><span class="o">+</span><span class="nb">input</span><span class="o">.</span><span class="n">no_of_pc</span><span class="p">):</span><span class="n">data</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span>
    <span class="n">index</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">annotation_spatial_cluster_id</span><span class="p">)</span>
    <span class="n">snn</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">Feature</span><span class="p">[</span><span class="n">index</span><span class="p">,:]),</span><span class="n">xticklabels</span><span class="o">=</span><span class="n">ylabelname</span><span class="p">)</span>
    <span class="c1">#fig.tight_layout()</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;The figures are saved: &quot;</span><span class="p">,</span> <span class="nb">input</span><span class="o">.</span><span class="n">covariation_dir</span><span class="o">+</span><span class="s1">&#39;Feature_matrix_PC&#39;</span><span class="o">+</span><span class="s1">&#39;.&#39;</span><span class="o">+</span><span class="n">saveas</span><span class="p">)</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">covariation_dir</span><span class="o">+</span><span class="s1">&#39;Feature_matrix_PC&#39;</span><span class="o">+</span><span class="s1">&#39;.&#39;</span><span class="o">+</span><span class="n">saveas</span><span class="p">,</span><span class="n">bbox_inches</span><span class="o">=</span><span class="s1">&#39;tight&#39;</span><span class="p">,</span><span class="n">transparent</span><span class="o">=</span><span class="n">transparent_mode</span><span class="p">,</span><span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">showit</span><span class="p">:</span>
        <span class="k">pass</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="s1">&#39;all&#39;</span><span class="p">)</span></div>



<div class="viewcode-block" id="plot_significant_regression_covariations_as_circleplot">
<a class="viewcode-back" href="../../nico_covariations.html#nico_covariations.Covariations.plot_significant_regression_covariations_as_circleplot">[docs]</a>
<span class="k">def</span> <span class="nf">plot_significant_regression_covariations_as_circleplot</span><span class="p">(</span><span class="nb">input</span><span class="p">,</span><span class="n">choose_celltypes</span><span class="o">=</span><span class="p">[],</span><span class="n">saveas</span><span class="o">=</span><span class="s1">&#39;pdf&#39;</span><span class="p">,</span><span class="n">pvalue_cutoff</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span><span class="n">mention_pvalue</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="n">transparent_mode</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">showit</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span><span class="mf">1.25</span><span class="p">)):</span>

<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Plot significant regression covariations as a circle plot.</span>

<span class="sd">    This function visualizes the significant regression covariations identified in the gene covariation analysis.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    input : object</span>
<span class="sd">        The main input is the output from gene_covariation_analysis.</span>

<span class="sd">    choose_celltypes : list, optional</span>
<span class="sd">        The cell type(s) for which you want to inspect the covariation pattern.</span>
<span class="sd">        If the list is empty, the output will be generated for all cell types.</span>
<span class="sd">        Default is [].</span>

<span class="sd">    saveas : str, optional</span>
<span class="sd">        Format to save the figures in, either &#39;pdf&#39; or &#39;png&#39; (dpi for PNG format is 300).</span>
<span class="sd">        Default is &#39;pdf&#39;.</span>

<span class="sd">    pvalue_cutoff : float, optional</span>
<span class="sd">        The p-value cutoff used to print the -log10(pvalue) on top of the circle.</span>
<span class="sd">        Default is 0.05.</span>

<span class="sd">    mention_pvalue : bool, optional</span>
<span class="sd">        Whether to highlight the p-value on the circle plot. If False, it will not be shown.</span>
<span class="sd">        Default is True.</span>

<span class="sd">    transparent_mode : bool, optional</span>
<span class="sd">        Background color in the figures. If True, the background will be transparent.</span>
<span class="sd">        Default is False.</span>

<span class="sd">    showit : bool, optional</span>
<span class="sd">        Whether to display the plot. If False, the plot will be saved but not shown.</span>
<span class="sd">        Default is True.</span>

<span class="sd">    figsize : tuple, optional</span>
<span class="sd">        Dimension of the figure size.</span>
<span class="sd">        Default is (6, 1.25).</span>

<span class="sd">    Outputs</span>
<span class="sd">    -------</span>
<span class="sd">    The regression figures are saved in &#39;./nico_out/covariations_R*_F*/Regression_outputs/&#39;.</span>

<span class="sd">    Notes</span>
<span class="sd">    -----</span>
<span class="sd">    - The main input is the output from gene_covariation_analysis.</span>
<span class="sd">    - The output directory for saving the figures should exist prior to running this function.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">n</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clusterid</span><span class="p">)</span>

    <span class="n">perform</span><span class="o">=</span><span class="p">[]</span>
    <span class="n">found</span><span class="o">=</span><span class="p">[]</span>
    <span class="k">for</span> <span class="n">fi</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="p">):</span>
          <span class="n">CC_celltype_name</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clustername</span><span class="p">[</span><span class="n">fi</span><span class="p">]</span>
          <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">choose_celltypes</span><span class="p">)</span><span class="o">==</span><span class="mi">0</span><span class="p">:</span>
              <span class="n">perform</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fi</span><span class="p">)</span>
          <span class="k">else</span><span class="p">:</span>
              <span class="k">if</span> <span class="n">CC_celltype_name</span> <span class="ow">in</span> <span class="n">choose_celltypes</span><span class="p">:</span>
                  <span class="n">perform</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fi</span><span class="p">)</span>
                  <span class="n">found</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">CC_celltype_name</span><span class="p">)</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">choose_celltypes</span><span class="p">)</span><span class="o">!=</span><span class="mi">0</span><span class="p">:</span>
          <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;cell types found &quot;</span><span class="p">,</span><span class="n">found</span><span class="p">)</span>

    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;The regression figures as pvalue circle plots are saved in following path &quot;</span><span class="p">,</span> <span class="nb">input</span><span class="o">.</span><span class="n">regression_outdir</span><span class="o">+</span><span class="s1">&#39;pvalue_coeff_circleplot_*&#39;</span><span class="p">)</span>


    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">perform</span><span class="p">:</span>
        <span class="n">filename</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clustername</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
        <span class="n">temp</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clusterid</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">==</span><span class="nb">input</span><span class="o">.</span><span class="n">annotation_spatial_cluster_id</span><span class="p">)</span>
        <span class="n">index</span><span class="o">=</span><span class="n">temp</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

        <span class="n">data</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">save_reg_coef</span><span class="p">[</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clusterid</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span>
        <span class="n">coef_mu</span><span class="p">,</span><span class="n">intercept</span><span class="p">,</span><span class="n">alpha</span><span class="p">,</span><span class="n">xlabel</span><span class="p">,</span><span class="n">score</span><span class="p">,</span><span class="n">target</span><span class="p">,</span><span class="n">neighborhoodClass</span><span class="p">,</span><span class="n">pvalue</span><span class="p">,</span><span class="n">pve</span><span class="p">,</span><span class="n">rve</span><span class="o">=</span><span class="n">data</span>

<span class="w">        </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">        savedata=input.regression_outdir+&#39;coef&#39;+str(input.spatialcell_unique_clusterid[i])+&#39;.npz&#39;</span>
<span class="sd">        data=np.load(savedata,allow_pickle=True)</span>
<span class="sd">        coef_mu=data[&#39;coef_mu&#39;]</span>
<span class="sd">        intercept=data[&#39;intercept&#39;]</span>
<span class="sd">        pve=data[&#39;pve&#39;] # percentage variance explanined</span>
<span class="sd">        rve=data[&#39;rve&#39;] # residual variance explained</span>
<span class="sd">        pvalue=data[&#39;pvalue&#39;]</span>
<span class="sd">        #coef_std=data[&#39;coef_std&#39;]</span>
<span class="sd">        #comp_score=data[&#39;comp_score&#39;]</span>
<span class="sd">        #comp_score_std=data[&#39;comp_score_std&#39;]</span>
<span class="sd">        alpha=data[&#39;alpha&#39;]</span>
<span class="sd">        xlabel=data[&#39;xlabel&#39;]</span>
<span class="sd">        score=data[&#39;score&#39;]</span>
<span class="sd">        &#39;&#39;&#39;</span>

        <span class="n">componentlabel</span><span class="o">=</span><span class="p">[]</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">no_of_pc</span><span class="p">):</span>
            <span class="n">componentlabel</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;Fa&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">j</span><span class="o">+</span><span class="mi">1</span><span class="p">))</span>

        <span class="n">percentVE</span><span class="o">=</span><span class="s1">&#39;&#39;</span>
        <span class="n">percentRE</span><span class="o">=</span><span class="s1">&#39;&#39;</span>

        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">pve</span><span class="p">)):</span>
            <span class="k">if</span> <span class="n">j</span><span class="o">!=</span><span class="mi">0</span><span class="p">:</span>
                <span class="n">percentVE</span><span class="o">+=</span><span class="s1">&#39;, &#39;</span>
                <span class="n">percentRE</span><span class="o">+=</span><span class="s1">&#39;, &#39;</span>
            <span class="n">percentVE</span><span class="o">+=</span><span class="s1">&#39;</span><span class="si">%0.3f</span><span class="s1">&#39;</span><span class="o">%</span><span class="n">pve</span><span class="p">[</span><span class="n">j</span><span class="p">]</span>
            <span class="c1">#percentRE+=&#39;%0.1f&#39;%rve[j]</span>

        <span class="n">ylabelname</span><span class="o">=</span><span class="p">[]</span>
        <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">xlabel</span><span class="p">)):</span>
            <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">no_of_pc</span><span class="p">):</span>
                <span class="n">ylabelname</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">xlabel</span><span class="p">[</span><span class="n">k</span><span class="p">]</span><span class="o">+</span><span class="s1">&#39;_s&#39;</span><span class="o">+</span><span class="s1">&#39;</span><span class="si">%0.3f</span><span class="s1">&#39;</span><span class="o">%</span><span class="n">score</span><span class="p">[</span><span class="n">k</span><span class="p">]</span><span class="o">+</span><span class="s1">&#39;_Fa&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">j</span><span class="o">+</span><span class="mi">1</span><span class="p">))</span>


        <span class="c1">#tempG=pvalue&lt;0.1</span>
        <span class="c1">#m1,m2=tempG.nonzero()</span>
        <span class="c1">#coef_mu[m1,m2]=0</span>
        <span class="c1">#coef_mu=tempG.astype(int)</span>

        <span class="c1">#pvalue=pvalue&lt;0.05</span>
        <span class="n">pvalue</span><span class="p">[</span><span class="n">pvalue</span><span class="o">&lt;</span><span class="mi">10</span><span class="o">**-</span><span class="mi">10</span><span class="p">]</span><span class="o">=</span><span class="mi">10</span><span class="o">**-</span><span class="mi">10</span>
        <span class="n">pvalue</span><span class="o">=-</span><span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">pvalue</span><span class="p">)</span>
        <span class="n">pvalue</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">nan_to_num</span><span class="p">(</span><span class="n">pvalue</span><span class="p">)</span>
        <span class="n">pvcut</span><span class="o">=-</span><span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">pvalue_cutoff</span><span class="p">)</span>

        <span class="n">factor</span><span class="o">=</span><span class="mf">0.5</span>
        <span class="n">newfigsize</span><span class="o">=</span><span class="p">(</span><span class="n">factor</span><span class="o">*</span><span class="nb">len</span><span class="p">(</span><span class="n">ylabelname</span><span class="p">),</span><span class="n">figsize</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>

        <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="n">figsize</span><span class="o">=</span><span class="n">newfigsize</span><span class="p">)</span>
        <span class="n">M</span><span class="o">=</span><span class="n">pvalue</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">N</span><span class="o">=</span><span class="n">pvalue</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">c</span><span class="o">=</span><span class="n">coef_mu</span>
        <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">meshgrid</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">M</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">N</span><span class="p">))</span>
        <span class="n">R</span> <span class="o">=</span> <span class="n">pvalue</span><span class="o">/</span><span class="mf">10.0</span><span class="o">/</span><span class="mi">2</span>
        <span class="n">maxp</span><span class="o">=</span><span class="n">pvalue</span><span class="o">.</span><span class="n">max</span><span class="p">()</span>
        <span class="n">circles</span> <span class="o">=</span> <span class="p">[</span><span class="n">plt</span><span class="o">.</span><span class="n">Circle</span><span class="p">((</span><span class="n">j</span><span class="p">,</span><span class="n">i</span><span class="p">),</span> <span class="n">radius</span><span class="o">=</span><span class="n">r</span><span class="p">)</span> <span class="k">for</span> <span class="n">r</span><span class="p">,</span> <span class="n">j</span><span class="p">,</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">R</span><span class="o">.</span><span class="n">flat</span><span class="p">,</span> <span class="n">x</span><span class="o">.</span><span class="n">flat</span><span class="p">,</span> <span class="n">y</span><span class="o">.</span><span class="n">flat</span><span class="p">)]</span>
        <span class="n">col</span> <span class="o">=</span> <span class="n">PatchCollection</span><span class="p">(</span><span class="n">circles</span><span class="p">,</span> <span class="n">array</span><span class="o">=</span><span class="n">c</span><span class="o">.</span><span class="n">flatten</span><span class="p">(),</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;jet&#39;</span><span class="p">)</span><span class="c1">#cmap=&quot;RdYlGn&quot;)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">add_collection</span><span class="p">(</span><span class="n">col</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">xticks</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">M</span><span class="p">),</span> <span class="n">yticks</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">N</span><span class="p">),</span>
               <span class="n">xticklabels</span><span class="o">=</span><span class="n">ylabelname</span><span class="p">,</span> <span class="n">yticklabels</span><span class="o">=</span><span class="n">componentlabel</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">M</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span><span class="o">-</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">minor</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">N</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span><span class="o">-</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">minor</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xticklabels</span><span class="p">(</span><span class="n">ylabelname</span><span class="p">,</span><span class="n">size</span> <span class="o">=</span> <span class="mi">8</span><span class="p">,</span><span class="n">rotation</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">filename</span><span class="o">+</span><span class="sa">r</span><span class="s1">&#39;,$\alpha$=&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">alpha</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;,EVS=&#39;</span><span class="o">+</span><span class="n">percentVE</span><span class="p">,</span><span class="n">fontsize</span><span class="o">=</span><span class="mi">6</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="n">which</span><span class="o">=</span><span class="s1">&#39;minor&#39;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">mention_pvalue</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">e</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">M</span><span class="p">):</span>
                <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">N</span><span class="p">):</span>
                    <span class="k">if</span> <span class="n">pvalue</span><span class="p">[</span><span class="n">f</span><span class="p">,</span><span class="n">e</span><span class="p">]</span><span class="o">&gt;</span><span class="n">pvcut</span><span class="p">:</span>
                        <span class="n">ax</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">e</span><span class="p">,</span><span class="n">f</span><span class="p">,</span><span class="s1">&#39;</span><span class="si">%0.2f</span><span class="s1">&#39;</span><span class="o">%</span><span class="n">pvalue</span><span class="p">[</span><span class="n">f</span><span class="p">,</span><span class="n">e</span><span class="p">],</span><span class="n">fontsize</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>
        <span class="n">fig</span><span class="o">.</span><span class="n">colorbar</span><span class="p">(</span><span class="n">col</span><span class="p">)</span>
        <span class="c1">#fig.tight_layout()</span>
        <span class="n">savefname</span><span class="o">=</span><span class="n">remove_extra_character_from_name</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clusterid</span><span class="p">[</span><span class="n">i</span><span class="p">])</span><span class="o">+</span><span class="s1">&#39;_&#39;</span><span class="o">+</span><span class="n">filename</span><span class="p">)</span>
        <span class="c1">#print(&#39;\n\n\n&#39;,input.regression_outdir+&#39;pvalue_coeff_circleplot_&#39;+savefname+&#39;.&#39;+saveas)</span>
        <span class="c1">#print(&quot;The figures are saved: &quot;, input.regression_outdir+&#39;pvalue_significance_coeff_matrix_&#39;+savefname+&#39;.&#39;+saveas)</span>
        <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">regression_outdir</span><span class="o">+</span><span class="s1">&#39;pvalue_coeff_circleplot_&#39;</span><span class="o">+</span><span class="n">savefname</span><span class="o">+</span><span class="s1">&#39;.&#39;</span><span class="o">+</span><span class="n">saveas</span><span class="p">,</span><span class="n">bbox_inches</span><span class="o">=</span><span class="s1">&#39;tight&#39;</span><span class="p">,</span><span class="n">transparent</span><span class="o">=</span><span class="n">transparent_mode</span><span class="p">,</span><span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">showit</span><span class="p">:</span>
            <span class="k">pass</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="s1">&#39;all&#39;</span><span class="p">)</span></div>



<div class="viewcode-block" id="plot_significant_regression_covariations_as_heatmap">
<a class="viewcode-back" href="../../nico_covariations.html#nico_covariations.Covariations.plot_significant_regression_covariations_as_heatmap">[docs]</a>
<span class="k">def</span> <span class="nf">plot_significant_regression_covariations_as_heatmap</span><span class="p">(</span><span class="nb">input</span><span class="p">,</span><span class="n">choose_celltypes</span><span class="o">=</span><span class="p">[],</span><span class="n">saveas</span><span class="o">=</span><span class="s1">&#39;pdf&#39;</span><span class="p">,</span><span class="n">transparent_mode</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">showit</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span><span class="mi">10</span><span class="p">)):</span>

<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Plot significant regression covariations as a heatmap.</span>

<span class="sd">    This function visualizes the significant regression covariations from the gene covariation analysis as a heatmap.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    input : object</span>
<span class="sd">        The main input is the output from the gene_covariation_analysis.</span>

<span class="sd">    choose_celltypes : list, optional</span>
<span class="sd">        The cell types for which you want to inspect the covariation regression pattern.</span>
<span class="sd">        If the list is empty, the output will be generated for all cell types.</span>
<span class="sd">        Default is an empty list [].</span>

<span class="sd">    saveas : str, optional</span>
<span class="sd">        Format to save the figures in, either &#39;pdf&#39; or &#39;png&#39; (dpi for PNG format is 300).</span>
<span class="sd">        Default is &#39;pdf&#39;.</span>

<span class="sd">    transparent_mode : bool, optional</span>
<span class="sd">        Background color in the figures. If True, the background will be transparent.</span>
<span class="sd">        Default is False.</span>

<span class="sd">    showit : bool, optional</span>
<span class="sd">        Whether to display the plot. If False, the plot will be saved but not shown.</span>
<span class="sd">        Default is True.</span>

<span class="sd">    figsize : tuple, optional</span>
<span class="sd">        Dimension of the figure size.</span>
<span class="sd">        Default is (6, 10).</span>

<span class="sd">    Outputs</span>
<span class="sd">    -------</span>
<span class="sd">    The regression heatmap figures are saved in the specified format and location as defined in the function implementation.</span>
<span class="sd">    Default save location is ./nico_out/covariations_R*_F*/Regression_outputs/</span>

<span class="sd">    Notes</span>
<span class="sd">    -----</span>
<span class="sd">    - Ensure the input contains the necessary data from the gene covariation analysis.</span>
<span class="sd">    - Ensure the output directory exists and is writable before running this function.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">n</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clusterid</span><span class="p">)</span>

    <span class="n">perform</span><span class="o">=</span><span class="p">[]</span>
    <span class="n">found</span><span class="o">=</span><span class="p">[]</span>
    <span class="k">for</span> <span class="n">fi</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="p">):</span>
          <span class="n">CC_celltype_name</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clustername</span><span class="p">[</span><span class="n">fi</span><span class="p">]</span>
          <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">choose_celltypes</span><span class="p">)</span><span class="o">==</span><span class="mi">0</span><span class="p">:</span>
              <span class="n">perform</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fi</span><span class="p">)</span>
          <span class="k">else</span><span class="p">:</span>
              <span class="k">if</span> <span class="n">CC_celltype_name</span> <span class="ow">in</span> <span class="n">choose_celltypes</span><span class="p">:</span>
                  <span class="n">perform</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fi</span><span class="p">)</span>
                  <span class="n">found</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">CC_celltype_name</span><span class="p">)</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">choose_celltypes</span><span class="p">)</span><span class="o">!=</span><span class="mi">0</span><span class="p">:</span>
          <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;cell types found &quot;</span><span class="p">,</span><span class="n">found</span><span class="p">)</span>

    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;The regression figures as pvalue heatmap plots are saved in following path &quot;</span><span class="p">,</span> <span class="nb">input</span><span class="o">.</span><span class="n">regression_outdir</span><span class="o">+</span><span class="s1">&#39;pvalue_coeff_heatmap_*&#39;</span><span class="p">)</span>


    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">perform</span><span class="p">:</span>
        <span class="n">filename</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clustername</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
        <span class="n">temp</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clusterid</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">==</span><span class="nb">input</span><span class="o">.</span><span class="n">annotation_spatial_cluster_id</span><span class="p">)</span>
        <span class="n">index</span><span class="o">=</span><span class="n">temp</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

        <span class="n">data</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">save_reg_coef</span><span class="p">[</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clusterid</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span>
        <span class="n">coef_mu</span><span class="p">,</span><span class="n">intercept</span><span class="p">,</span><span class="n">alpha</span><span class="p">,</span><span class="n">xlabel</span><span class="p">,</span><span class="n">score</span><span class="p">,</span><span class="n">target</span><span class="p">,</span><span class="n">neighborhoodClass</span><span class="p">,</span><span class="n">pvalue</span><span class="p">,</span><span class="n">pve</span><span class="p">,</span><span class="n">rve</span><span class="o">=</span><span class="n">data</span>

        <span class="n">componentlabel</span><span class="o">=</span><span class="p">[]</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">no_of_pc</span><span class="p">):</span>
            <span class="n">componentlabel</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;Fa&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">j</span><span class="o">+</span><span class="mi">1</span><span class="p">))</span>

        <span class="n">percentVE</span><span class="o">=</span><span class="s1">&#39;&#39;</span>
        <span class="n">percentRE</span><span class="o">=</span><span class="s1">&#39;&#39;</span>

        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">pve</span><span class="p">)):</span>
            <span class="k">if</span> <span class="n">j</span><span class="o">!=</span><span class="mi">0</span><span class="p">:</span>
                <span class="n">percentVE</span><span class="o">+=</span><span class="s1">&#39;, &#39;</span>
                <span class="n">percentRE</span><span class="o">+=</span><span class="s1">&#39;, &#39;</span>
            <span class="n">percentVE</span><span class="o">+=</span><span class="s1">&#39;</span><span class="si">%0.3f</span><span class="s1">&#39;</span><span class="o">%</span><span class="n">pve</span><span class="p">[</span><span class="n">j</span><span class="p">]</span>
            <span class="c1">#percentRE+=&#39;%0.1f&#39;%rve[j]</span>

        <span class="n">xlabelname</span><span class="o">=</span><span class="p">[]</span>
        <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">xlabel</span><span class="p">)):</span>
            <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">no_of_pc</span><span class="p">):</span>
                <span class="n">xlabelname</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">xlabel</span><span class="p">[</span><span class="n">k</span><span class="p">]</span><span class="o">+</span><span class="s1">&#39;_s&#39;</span><span class="o">+</span><span class="s1">&#39;</span><span class="si">%0.3f</span><span class="s1">&#39;</span><span class="o">%</span><span class="n">score</span><span class="p">[</span><span class="n">k</span><span class="p">]</span><span class="o">+</span><span class="s1">&#39;_Fa&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">j</span><span class="o">+</span><span class="mi">1</span><span class="p">))</span>

        <span class="n">pvalue</span><span class="p">[</span><span class="n">pvalue</span><span class="o">&lt;</span><span class="mi">10</span><span class="o">**-</span><span class="mi">10</span><span class="p">]</span><span class="o">=</span><span class="mi">10</span><span class="o">**-</span><span class="mi">10</span>
        <span class="n">pvalue</span><span class="o">=-</span><span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">pvalue</span><span class="p">)</span>
        <span class="n">pvalue</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">nan_to_num</span><span class="p">(</span><span class="n">pvalue</span><span class="p">)</span>

        <span class="n">factor</span><span class="o">=</span><span class="mf">0.5</span>
        <span class="n">newfigsize</span><span class="o">=</span><span class="p">(</span><span class="n">factor</span><span class="o">*</span><span class="nb">len</span><span class="p">(</span><span class="n">xlabelname</span><span class="p">),</span><span class="n">figsize</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>

        <span class="n">fig</span><span class="o">=</span><span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="n">newfigsize</span><span class="p">)</span>
        <span class="n">gs</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_gridspec</span><span class="p">(</span><span class="n">ncols</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">nrows</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">height_ratios</span><span class="o">=</span><span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
        <span class="n">ax0</span><span class="o">=</span><span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="n">ax1</span><span class="o">=</span><span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>


        <span class="n">a</span><span class="o">=</span><span class="n">snn</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">coef_mu</span><span class="p">,</span><span class="n">xticklabels</span><span class="o">=</span><span class="n">xlabelname</span><span class="p">,</span><span class="n">yticklabels</span><span class="o">=</span><span class="n">componentlabel</span><span class="p">,</span><span class="n">ax</span><span class="o">=</span><span class="n">ax0</span><span class="p">,</span><span class="n">cbar_kws</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;shrink&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">})</span>
        <span class="c1">#snn.axes_style(xtick.top=True)</span>
        <span class="n">xlabels</span><span class="o">=</span> <span class="n">a</span><span class="o">.</span><span class="n">get_xticklabels</span><span class="p">()</span>
        <span class="n">a</span><span class="o">.</span><span class="n">set_xticklabels</span><span class="p">([])</span>
        <span class="c1">#a.set_xticklabels(xlabels,size = 8,rotation=90)</span>
        <span class="n">a</span><span class="o">.</span><span class="n">set_yticklabels</span><span class="p">(</span><span class="n">componentlabel</span><span class="p">,</span><span class="n">rotation</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span><span class="n">size</span><span class="o">=</span><span class="mi">6</span><span class="p">)</span>
        <span class="c1">#a.set_ylabel(&#39;Principal components&#39;)</span>
        <span class="n">a</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">filename</span><span class="o">+</span><span class="sa">r</span><span class="s1">&#39;,$\alpha$=&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">alpha</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;,EVS=&#39;</span><span class="o">+</span><span class="n">percentVE</span><span class="p">,</span><span class="n">fontsize</span><span class="o">=</span><span class="mi">6</span><span class="p">)</span>


        <span class="n">b</span><span class="o">=</span><span class="n">snn</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">pvalue</span><span class="p">,</span><span class="n">annot</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">fmt</span><span class="o">=</span><span class="s1">&#39;.2f&#39;</span><span class="p">,</span><span class="n">cmap</span><span class="o">=</span><span class="n">snn</span><span class="o">.</span><span class="n">cm</span><span class="o">.</span><span class="n">rocket_r</span><span class="p">,</span><span class="n">annot_kws</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;size&quot;</span><span class="p">:</span> <span class="mi">3</span><span class="p">},</span><span class="n">xticklabels</span><span class="o">=</span><span class="n">xlabelname</span><span class="p">,</span><span class="n">yticklabels</span><span class="o">=</span><span class="n">componentlabel</span><span class="p">,</span><span class="n">ax</span><span class="o">=</span><span class="n">ax1</span><span class="p">,</span><span class="n">cbar_kws</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;shrink&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">})</span>
        <span class="c1">#b.xaxis.tick_top()</span>
        <span class="n">xlabels</span><span class="o">=</span> <span class="n">b</span><span class="o">.</span><span class="n">get_xticks</span><span class="p">()</span>
        <span class="n">xlabels</span><span class="o">=</span> <span class="n">b</span><span class="o">.</span><span class="n">get_xticklabels</span><span class="p">()</span>

        <span class="c1">#b.set_xticklabels([])#xlabels,size = 0)</span>
        <span class="n">b</span><span class="o">.</span><span class="n">set_xticklabels</span><span class="p">(</span><span class="n">xlabels</span><span class="p">,</span><span class="n">size</span> <span class="o">=</span> <span class="mi">6</span><span class="p">,</span><span class="n">rotation</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span>
        <span class="n">b</span><span class="o">.</span><span class="n">set_yticklabels</span><span class="p">(</span><span class="n">componentlabel</span><span class="p">,</span><span class="n">rotation</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span><span class="n">size</span><span class="o">=</span><span class="mi">6</span><span class="p">)</span>




        <span class="c1">#fig.tight_layout()</span>
        <span class="n">savefname</span><span class="o">=</span><span class="n">remove_extra_character_from_name</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clusterid</span><span class="p">[</span><span class="n">i</span><span class="p">])</span><span class="o">+</span><span class="s1">&#39;_&#39;</span><span class="o">+</span><span class="n">filename</span><span class="p">)</span>
        <span class="c1">#print(&quot;The figures are saved: &quot;, input.regression_outdir+&#39;pvalue_significance_coeff_matrix_&#39;+savefname+&#39;.&#39;+saveas)</span>
        <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">regression_outdir</span><span class="o">+</span><span class="s1">&#39;pvalue_coeff_heatmap_&#39;</span><span class="o">+</span><span class="n">savefname</span><span class="o">+</span><span class="s1">&#39;.&#39;</span><span class="o">+</span><span class="n">saveas</span><span class="p">,</span><span class="n">bbox_inches</span><span class="o">=</span><span class="s1">&#39;tight&#39;</span><span class="p">,</span><span class="n">transparent</span><span class="o">=</span><span class="n">transparent_mode</span><span class="p">,</span><span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">showit</span><span class="p">:</span>
            <span class="k">pass</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="s1">&#39;all&#39;</span><span class="p">)</span></div>



<div class="viewcode-block" id="save_LR_interactions_in_excelsheet_and_regression_summary_in_textfile_for_interacting_cell_types">
<a class="viewcode-back" href="../../nico_covariations.html#nico_covariations.Covariations.save_LR_interactions_in_excelsheet_and_regression_summary_in_textfile_for_interacting_cell_types">[docs]</a>
<span class="k">def</span> <span class="nf">save_LR_interactions_in_excelsheet_and_regression_summary_in_textfile_for_interacting_cell_types</span><span class="p">(</span><span class="nb">input</span><span class="p">,</span><span class="n">pvalueCutoff</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span> <span class="n">correlation_with_spearman</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="n">LR_plot_NMF_Fa_thres</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span> <span class="n">LR_plot_Exp_thres</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span><span class="n">number_of_top_genes_to_print</span><span class="o">=</span><span class="mi">20</span><span class="p">):</span>

<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Save ligand-receptor (LR) interactions in an Excel sheet and regression summary in a text file for interacting cell types.</span>

<span class="sd">    This function processes the output from gene_covariation_analysis to identify significant LR interactions and saves the results in an Excel sheet and a text file.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    -----------</span>
<span class="sd">    input : object</span>
<span class="sd">        The main input is the output from gene_covariation_analysis.</span>

<span class="sd">    pvalueCutoff : float, optional</span>
<span class="sd">        The cutoff used to select the significant central cell type factor and niche cell type factor covariations.</span>
<span class="sd">        Default is 0.05.</span>

<span class="sd">    correlation_with_spearman : bool, optional</span>
<span class="sd">        If True, genes factor correlations are computed as Spearman correlation coefficient; otherwise, cosine similarities are computed.</span>
<span class="sd">        Default is True.</span>

<span class="sd">    LR_plot_NMF_Fa_thres : float, optional</span>
<span class="sd">        Only ligands or receptors are retained that exhibit a correlation to the respective factors higher than this cutoff.</span>
<span class="sd">        Default is 0.2.</span>

<span class="sd">    LR_plot_Exp_thres : float, optional</span>
<span class="sd">        Only ligands or receptors are retained that are expressed in a fraction of cells of the respective cell types exceeding this cutoff.</span>
<span class="sd">        Default is 0.2.</span>

<span class="sd">    number_of_top_genes_to_print : int, optional</span>
<span class="sd">        The number of top correlating genes to print in the regression summary text file.</span>
<span class="sd">        Default is 20.</span>

<span class="sd">    Outputs:</span>
<span class="sd">    --------</span>

<span class="sd">     - An Excel sheet with ligand-receptor interaction information for easy access.</span>
<span class="sd">       The columns are structured as follows in the sheets:</span>

<span class="sd">        - A. ID of the cell type-cell type interaction</span>
<span class="sd">        - BC. Interacting cell types A and B</span>
<span class="sd">        - D. Normalized interaction scores from the logistic regression classifier</span>
<span class="sd">        - EF. NMF factor IDs (metagenes) in cell types A and B</span>
<span class="sd">        - G. Ridge regression coefficient indicating the factors covariation</span>
<span class="sd">        - H. Ligand in cell type A</span>
<span class="sd">        - I. Receptor in cell type B</span>
<span class="sd">        - JK. Pearson correlation of ligand and receptor genes in cell types A and B with the corresponding factors</span>
<span class="sd">        - LM. Average expression of ligands and receptors in cell types A and B</span>
<span class="sd">        - NO. Fraction of cells expressing these genes with counts greater than zero in cell types A and B</span>

<span class="sd">     - A regression summary text file with the following structure:</span>
<span class="sd">     - First row: CC-Fa(i), CC (cell type), niche_score (from classifier), NC-Fa*, NC (cell type), RegCoeff (covariation score), p-value on normal scale, p-value on -log10 scale</span>
<span class="sd">     - Second row: Top 20 (number_of_top_genes_to_print) genes correlated to Fa(i) of central cell type, with genes and their factor ID indicated in the pair</span>
<span class="sd">     - Third row: Top 20 (number_of_top_genes_to_print) genes correlated to Fa(j) of niche cell type, with genes and their factor ID indicated in the pair</span>

<span class="sd">    Notes</span>
<span class="sd">    -----</span>
<span class="sd">    - Our analysis accounts for bidirectional cellular crosstalk interactions of ligands and receptors in cell types A and B.</span>
<span class="sd">    - The ligand can be expressed on cell type A and signal to the receptor detected on cell type B, or vice versa.</span>
<span class="sd">    - Both ligand-receptor plots and Excel sheets profile bidirectional cellular crosstalk of ligand and receptor in cell types A and B.</span>
<span class="sd">    - Each central cell type is represented in a separate Excel sheet, while the LR enrichment sheet aggregates all interactions across central cell types.</span>

<span class="sd">    &quot;&quot;&quot;</span>


    <span class="n">totalLRpairs</span><span class="p">,</span><span class="n">ligand</span><span class="p">,</span><span class="n">receptor</span><span class="p">,</span><span class="n">either</span><span class="o">=</span><span class="n">read_LigRecDb</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">LRdb</span><span class="p">)</span>
    <span class="n">coeff_cutoff_for_log_reg</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">logistic_coef_cutoff</span>
    <span class="n">coeff_cutoff_for_rid_reg</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">coeff_cutoff_for_rid_reg</span>
    <span class="n">gene_set_names</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">gene_set_names</span>
    <span class="n">LRcutoff</span><span class="o">=</span><span class="n">LR_plot_NMF_Fa_thres</span> <span class="c1">#Used in excel sheet to show the enrichment of ligand receptor intera</span>

    <span class="n">PCA_of_sc_cluster_accordingto_spatial_clusterid</span><span class="p">,</span><span class="n">save_scFactors</span><span class="p">,</span><span class="n">save_spFactors</span><span class="o">=</span><span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">covariation_dir</span><span class="o">+</span><span class="s1">&#39;factors_info.p&#39;</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">))</span>
    <span class="n">n</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clustername</span><span class="p">)</span>

    <span class="n">workbook</span> <span class="o">=</span> <span class="n">xlsxwriter</span><span class="o">.</span><span class="n">Workbook</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">covariation_dir</span><span class="o">+</span><span class="s1">&#39;Lig_and_Rec_enrichment_in_interacting_celltypes.xlsx&#39;</span><span class="p">)</span>
    <span class="n">fout</span><span class="o">=</span><span class="nb">open</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">covariation_dir</span><span class="o">+</span><span class="s1">&#39;Regression_summary.txt&#39;</span><span class="p">,</span><span class="s1">&#39;w&#39;</span><span class="p">)</span>
    <span class="n">worksheet</span> <span class="o">=</span> <span class="n">workbook</span><span class="o">.</span><span class="n">add_worksheet</span><span class="p">(</span><span class="s1">&#39;LR enrichment&#39;</span><span class="p">)</span>
    <span class="n">worksheetrow</span><span class="o">=</span><span class="mi">0</span>
    <span class="n">main_header</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Id&#39;</span><span class="p">,</span><span class="s1">&#39;A&#39;</span><span class="p">,</span><span class="s1">&#39;B&#39;</span><span class="p">,</span><span class="s1">&#39;localized score&#39;</span><span class="p">,</span><span class="s1">&#39;Fa(A)&#39;</span><span class="p">,</span><span class="s1">&#39;Fa(B)&#39;</span><span class="p">,</span> <span class="s1">&#39;Coeff&#39;</span> <span class="p">,</span><span class="s1">&#39;Ligand(A)&#39;</span><span class="p">,</span><span class="s1">&#39;Receptor(B)&#39;</span><span class="p">,</span><span class="s1">&#39;GeneCor(Lig)&#39;</span><span class="p">,</span><span class="s1">&#39;GeneCor(Rec)&#39;</span><span class="p">,</span><span class="s1">&#39;AvgExp(A)&#39;</span><span class="p">,</span><span class="s1">&#39;AvgExp(B)&#39;</span><span class="p">,</span><span class="s1">&#39;PopExp(A)&#39;</span><span class="p">,</span><span class="s1">&#39;PopExp(B)&#39;</span><span class="p">]</span>
    <span class="k">for</span> <span class="n">ri</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">main_header</span><span class="p">)):</span>
        <span class="n">worksheet</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">worksheetrow</span><span class="p">,</span><span class="n">ri</span><span class="p">,</span><span class="n">main_header</span><span class="p">[</span><span class="n">ri</span><span class="p">])</span>
    <span class="n">worksheetrow</span><span class="o">+=</span><span class="mi">1</span>

    <span class="n">d</span><span class="o">=</span><span class="p">{}</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="p">):</span>
        <span class="n">clid</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clusterid</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
        <span class="n">clname</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clustername</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
        <span class="n">d</span><span class="p">[</span><span class="n">clname</span><span class="p">]</span><span class="o">=</span><span class="n">clid</span>

    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;The Excel sheet is saved: &quot;</span><span class="p">,</span><span class="nb">input</span><span class="o">.</span><span class="n">covariation_dir</span><span class="o">+</span><span class="s1">&#39;Lig_and_Rec_enrichment_in_interacting_celltypes.xlsx&#39;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;The text file is saved:&quot;</span><span class="p">,</span><span class="nb">input</span><span class="o">.</span><span class="n">covariation_dir</span><span class="o">+</span><span class="s1">&#39;Regression_summary.txt&#39;</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="p">):</span>
        <span class="n">clid</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clusterid</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
        <span class="n">CC_corr_spearman</span><span class="p">,</span><span class="n">CC_PCA</span><span class="p">,</span><span class="n">CC_gene</span><span class="p">,</span><span class="n">CC_meanExpression</span><span class="p">,</span><span class="n">CC_popExpression</span><span class="p">,</span><span class="n">CC_corr_cosine</span><span class="p">,</span><span class="n">alpha</span><span class="o">=</span><span class="n">PCA_of_sc_cluster_accordingto_spatial_clusterid</span><span class="p">[</span><span class="n">clid</span><span class="p">]</span>
        <span class="n">CC_celltype_name</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clustername</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
        <span class="c1">#temp=np.where(input.spatialcell_unique_clusterid[i]==input.annotation_spatial_cluster_id)</span>
        <span class="c1">#index=temp[0]</span>

        <span class="n">data</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">save_reg_coef</span><span class="p">[</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clusterid</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span>
        <span class="n">coef_mu</span><span class="p">,</span><span class="n">intercept</span><span class="p">,</span><span class="n">alpha</span><span class="p">,</span><span class="n">xlabel</span><span class="p">,</span><span class="n">score</span><span class="p">,</span><span class="n">target</span><span class="p">,</span><span class="n">neighborhoodClass</span><span class="p">,</span><span class="n">pvalue</span><span class="p">,</span><span class="n">pve</span><span class="p">,</span><span class="n">rve</span><span class="o">=</span><span class="n">data</span>
        <span class="n">NC_celltype_name</span><span class="o">=</span><span class="n">xlabel</span>
        <span class="n">largest</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="nb">abs</span><span class="p">(</span><span class="n">coef_mu</span><span class="p">))</span>
        <span class="n">normalized_ridge_coef</span><span class="o">=</span><span class="n">coef_mu</span><span class="o">/</span><span class="n">largest</span>

        <span class="n">ylabelname</span><span class="o">=</span><span class="p">[]</span>
        <span class="n">componentlabel</span><span class="o">=</span><span class="p">[]</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">no_of_pc</span><span class="p">):</span>
            <span class="n">ylabelname</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;CC_&#39;</span><span class="o">+</span><span class="n">CC_celltype_name</span><span class="o">+</span><span class="s1">&#39;_Fa&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">j</span><span class="o">+</span><span class="mi">1</span><span class="p">))</span>
            <span class="n">componentlabel</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;Fa&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">j</span><span class="o">+</span><span class="mi">1</span><span class="p">))</span>

        <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">NC_celltype_name</span><span class="p">)):</span>
            <span class="k">if</span> <span class="n">score</span><span class="p">[</span><span class="n">k</span><span class="p">]</span><span class="o">&gt;</span><span class="n">coeff_cutoff_for_log_reg</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">CC_celltype_name</span><span class="o">!=</span><span class="n">NC_celltype_name</span><span class="p">[</span><span class="n">k</span><span class="p">]:</span>
                    <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">no_of_pc</span><span class="p">):</span>
                        <span class="n">ylabelname</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;NC_&#39;</span><span class="o">+</span><span class="n">NC_celltype_name</span><span class="p">[</span><span class="n">k</span><span class="p">]</span><span class="o">+</span><span class="s1">&#39;_s&#39;</span><span class="o">+</span><span class="s1">&#39;</span><span class="si">%0.3f</span><span class="s1">&#39;</span><span class="o">%</span><span class="n">score</span><span class="p">[</span><span class="n">k</span><span class="p">]</span><span class="o">+</span><span class="s1">&#39;_Fa&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">j</span><span class="o">+</span><span class="mi">1</span><span class="p">))</span>

        <span class="n">pc_index_nc</span><span class="o">=</span><span class="p">[]</span>
        <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">NC_celltype_name</span><span class="p">)):</span>
            <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">no_of_pc</span><span class="p">):</span>
                <span class="n">pc_index_nc</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">j</span><span class="p">)</span>

        <span class="n">CC_celltype_sheetname</span><span class="o">=</span><span class="n">remove_extra_character_from_name</span><span class="p">(</span><span class="n">CC_celltype_name</span><span class="p">)</span>
        <span class="n">worksheet_local</span> <span class="o">=</span> <span class="n">workbook</span><span class="o">.</span><span class="n">add_worksheet</span><span class="p">(</span><span class="n">CC_celltype_sheetname</span><span class="p">)</span>
        <span class="n">worksheetrow_local</span><span class="o">=</span><span class="mi">0</span>
        <span class="k">for</span> <span class="n">ri</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">main_header</span><span class="p">)):</span>
            <span class="n">worksheet_local</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">worksheetrow_local</span><span class="p">,</span><span class="n">ri</span><span class="p">,</span><span class="n">main_header</span><span class="p">[</span><span class="n">ri</span><span class="p">])</span>
        <span class="n">worksheetrow_local</span><span class="o">+=</span><span class="mi">1</span>

        <span class="n">interaction_id</span><span class="o">=</span><span class="mi">0</span>
        <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">normalized_ridge_coef</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]):</span>
            <span class="c1">#k is PC of central cell type</span>
            <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">normalized_ridge_coef</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]):</span>
                <span class="n">interaction_id</span><span class="o">+=</span><span class="mi">1</span>
                <span class="n">index</span><span class="o">=</span><span class="n">math</span><span class="o">.</span><span class="n">floor</span><span class="p">(</span><span class="n">j</span><span class="o">/</span><span class="nb">input</span><span class="o">.</span><span class="n">no_of_pc</span><span class="p">)</span>
                <span class="k">if</span> <span class="p">(</span><span class="n">pvalue</span><span class="p">[</span><span class="n">k</span><span class="p">,</span><span class="n">j</span><span class="p">]</span><span class="o">&lt;</span><span class="n">pvalueCutoff</span><span class="p">)</span><span class="o">&amp;</span><span class="p">(</span><span class="nb">abs</span><span class="p">(</span><span class="n">normalized_ridge_coef</span><span class="p">[</span><span class="n">k</span><span class="p">,</span><span class="n">j</span><span class="p">])</span><span class="o">&gt;</span><span class="n">coeff_cutoff_for_rid_reg</span><span class="p">):</span>
                <span class="c1">#if True:</span>
                    <span class="k">if</span> <span class="n">score</span><span class="p">[</span><span class="n">index</span><span class="p">]</span><span class="o">&gt;</span><span class="n">coeff_cutoff_for_log_reg</span><span class="p">:</span>
                        <span class="n">NC_corr_spearman</span><span class="p">,</span><span class="n">NC_PCA</span><span class="p">,</span><span class="n">NC_gene</span><span class="p">,</span><span class="n">NC_meanExpression</span><span class="p">,</span><span class="n">NC_popExpression</span><span class="p">,</span><span class="n">NC_corr_cosine</span><span class="p">,</span><span class="n">alpha</span><span class="o">=</span><span class="n">PCA_of_sc_cluster_accordingto_spatial_clusterid</span><span class="p">[</span><span class="n">d</span><span class="p">[</span><span class="n">NC_celltype_name</span><span class="p">[</span><span class="n">index</span><span class="p">]]]</span>
                        <span class="k">if</span> <span class="n">correlation_with_spearman</span><span class="p">:</span>
                            <span class="n">top_genes_in_CC</span><span class="p">,</span><span class="n">top_genes_in_NC</span><span class="p">,</span><span class="n">genesWithUP</span><span class="p">,</span><span class="n">genesWithDown</span><span class="p">,</span><span class="n">Found1</span><span class="p">,</span><span class="n">Found2</span><span class="o">=</span><span class="n">find_fold_change</span><span class="p">(</span><span class="n">CC_corr_spearman</span><span class="p">,</span><span class="n">NC_corr_spearman</span><span class="p">,</span><span class="n">CC_gene</span><span class="p">,</span><span class="n">k</span><span class="p">,</span><span class="n">pc_index_nc</span><span class="p">[</span><span class="n">j</span><span class="p">],</span><span class="n">totalLRpairs</span><span class="p">,</span><span class="n">LRcutoff</span><span class="p">,</span><span class="n">CC_meanExpression</span><span class="p">,</span><span class="n">NC_meanExpression</span><span class="p">,</span><span class="n">CC_popExpression</span><span class="p">,</span><span class="n">NC_popExpression</span><span class="p">,</span><span class="n">number_of_top_genes_to_print</span><span class="p">)</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="n">top_genes_in_CC</span><span class="p">,</span><span class="n">top_genes_in_NC</span><span class="p">,</span><span class="n">genesWithUP</span><span class="p">,</span><span class="n">genesWithDown</span><span class="p">,</span><span class="n">Found1</span><span class="p">,</span><span class="n">Found2</span><span class="o">=</span><span class="n">find_fold_change</span><span class="p">(</span><span class="n">CC_corr_cosine</span><span class="p">,</span><span class="n">NC_corr_cosine</span><span class="p">,</span><span class="n">CC_gene</span><span class="p">,</span><span class="n">k</span><span class="p">,</span><span class="n">pc_index_nc</span><span class="p">[</span><span class="n">j</span><span class="p">],</span><span class="n">totalLRpairs</span><span class="p">,</span><span class="n">LRcutoff</span><span class="p">,</span><span class="n">CC_meanExpression</span><span class="p">,</span><span class="n">NC_meanExpression</span><span class="p">,</span><span class="n">CC_popExpression</span><span class="p">,</span><span class="n">NC_popExpression</span><span class="p">,</span><span class="n">number_of_top_genes_to_print</span><span class="p">)</span>

                        <span class="n">common_genes</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">top_genes_in_CC</span><span class="p">)</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">top_genes_in_NC</span><span class="p">)))</span>


                        <span class="k">if</span> <span class="n">CC_celltype_name</span><span class="o">!=</span><span class="n">NC_celltype_name</span><span class="p">[</span><span class="n">index</span><span class="p">]:</span>
                                <span class="k">for</span> <span class="n">ele</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">Found1</span><span class="p">)):</span>
                                    <span class="n">header</span><span class="o">=</span><span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;-&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">interaction_id</span><span class="p">),</span><span class="n">CC_celltype_name</span><span class="o">+</span><span class="s1">&#39;(cc)&#39;</span><span class="p">,</span><span class="n">NC_celltype_name</span><span class="p">[</span><span class="n">index</span><span class="p">]</span><span class="o">+</span><span class="s1">&#39;(nc)&#39;</span><span class="p">,</span><span class="n">score</span><span class="p">[</span><span class="n">index</span><span class="p">],</span><span class="n">k</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="o">+</span><span class="n">pc_index_nc</span><span class="p">[</span><span class="n">j</span><span class="p">],</span><span class="n">normalized_ridge_coef</span><span class="p">[</span><span class="n">k</span><span class="p">,</span><span class="n">j</span><span class="p">]</span> <span class="p">,</span><span class="s1">&#39;Ligand(A)&#39;</span><span class="p">,</span><span class="s1">&#39;Receptor(B)&#39;</span><span class="p">,</span><span class="s1">&#39;GeneCor(Lig)&#39;</span><span class="p">,</span><span class="s1">&#39;GeneCor(Rec)&#39;</span><span class="p">,</span><span class="s1">&#39;Receptor(A)&#39;</span><span class="p">,</span><span class="s1">&#39;Ligand(B)&#39;</span><span class="p">,</span><span class="s1">&#39;GeneCor(Rec)&#39;</span><span class="p">,</span><span class="s1">&#39;GeneCor(Lig)&#39;</span><span class="p">]</span>
                                    <span class="n">header</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span><span class="o">=</span><span class="n">Found1</span><span class="p">[</span><span class="n">ele</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                                    <span class="n">header</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span><span class="o">=</span><span class="n">Found1</span><span class="p">[</span><span class="n">ele</span><span class="p">][</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                                    <span class="n">header</span><span class="p">[</span><span class="mi">9</span><span class="p">]</span><span class="o">=</span><span class="n">Found1</span><span class="p">[</span><span class="n">ele</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span>
                                    <span class="n">header</span><span class="p">[</span><span class="mi">10</span><span class="p">]</span><span class="o">=</span><span class="n">Found1</span><span class="p">[</span><span class="n">ele</span><span class="p">][</span><span class="mi">1</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span>
                                    <span class="n">header</span><span class="p">[</span><span class="mi">11</span><span class="p">]</span><span class="o">=</span><span class="n">Found1</span><span class="p">[</span><span class="n">ele</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="mi">2</span><span class="p">]</span>
                                    <span class="n">header</span><span class="p">[</span><span class="mi">12</span><span class="p">]</span><span class="o">=</span><span class="n">Found1</span><span class="p">[</span><span class="n">ele</span><span class="p">][</span><span class="mi">1</span><span class="p">][</span><span class="mi">2</span><span class="p">]</span>
                                    <span class="n">header</span><span class="p">[</span><span class="mi">13</span><span class="p">]</span><span class="o">=</span><span class="n">Found1</span><span class="p">[</span><span class="n">ele</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="mi">3</span><span class="p">]</span>
                                    <span class="n">header</span><span class="p">[</span><span class="mi">14</span><span class="p">]</span><span class="o">=</span><span class="n">Found1</span><span class="p">[</span><span class="n">ele</span><span class="p">][</span><span class="mi">1</span><span class="p">][</span><span class="mi">3</span><span class="p">]</span>
                                    <span class="k">for</span> <span class="n">ri</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">15</span><span class="p">):</span>
                                        <span class="n">worksheet</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">worksheetrow</span><span class="p">,</span><span class="n">ri</span><span class="p">,</span><span class="n">header</span><span class="p">[</span><span class="n">ri</span><span class="p">])</span>
                                        <span class="n">worksheet_local</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">worksheetrow_local</span><span class="p">,</span><span class="n">ri</span><span class="p">,</span><span class="n">header</span><span class="p">[</span><span class="n">ri</span><span class="p">])</span>
                                    <span class="n">worksheetrow</span><span class="o">+=</span><span class="mi">1</span>
                                    <span class="n">worksheetrow_local</span><span class="o">+=</span><span class="mi">1</span>


                                <span class="k">for</span> <span class="n">ele</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">Found2</span><span class="p">)):</span>
                                    <span class="n">header</span><span class="o">=</span><span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;-&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">interaction_id</span><span class="p">),</span><span class="n">NC_celltype_name</span><span class="p">[</span><span class="n">index</span><span class="p">]</span><span class="o">+</span><span class="s1">&#39;(nc)&#39;</span><span class="p">,</span><span class="n">CC_celltype_name</span><span class="o">+</span><span class="s1">&#39;(cc)&#39;</span><span class="p">,</span><span class="n">score</span><span class="p">[</span><span class="n">index</span><span class="p">],</span><span class="mi">1</span><span class="o">+</span><span class="n">pc_index_nc</span><span class="p">[</span><span class="n">j</span><span class="p">],</span><span class="n">k</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span><span class="n">normalized_ridge_coef</span><span class="p">[</span><span class="n">k</span><span class="p">,</span><span class="n">j</span><span class="p">]</span> <span class="p">,</span><span class="s1">&#39;Ligand(A)&#39;</span><span class="p">,</span><span class="s1">&#39;Receptor(B)&#39;</span><span class="p">,</span><span class="s1">&#39;GeneCor(Lig)&#39;</span><span class="p">,</span><span class="s1">&#39;GeneCor(Rec)&#39;</span><span class="p">,</span><span class="s1">&#39;Receptor(A)&#39;</span><span class="p">,</span><span class="s1">&#39;Ligand(B)&#39;</span><span class="p">,</span><span class="s1">&#39;GeneCor(Rec)&#39;</span><span class="p">,</span><span class="s1">&#39;GeneCor(Lig)&#39;</span><span class="p">]</span>
                                    <span class="n">header</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span><span class="o">=</span><span class="n">Found2</span><span class="p">[</span><span class="n">ele</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                                    <span class="n">header</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span><span class="o">=</span><span class="n">Found2</span><span class="p">[</span><span class="n">ele</span><span class="p">][</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                                    <span class="n">header</span><span class="p">[</span><span class="mi">9</span><span class="p">]</span><span class="o">=</span><span class="n">Found2</span><span class="p">[</span><span class="n">ele</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span>
                                    <span class="n">header</span><span class="p">[</span><span class="mi">10</span><span class="p">]</span><span class="o">=</span><span class="n">Found2</span><span class="p">[</span><span class="n">ele</span><span class="p">][</span><span class="mi">1</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span>
                                    <span class="n">header</span><span class="p">[</span><span class="mi">11</span><span class="p">]</span><span class="o">=</span><span class="n">Found2</span><span class="p">[</span><span class="n">ele</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="mi">2</span><span class="p">]</span>
                                    <span class="n">header</span><span class="p">[</span><span class="mi">12</span><span class="p">]</span><span class="o">=</span><span class="n">Found2</span><span class="p">[</span><span class="n">ele</span><span class="p">][</span><span class="mi">1</span><span class="p">][</span><span class="mi">2</span><span class="p">]</span>
                                    <span class="n">header</span><span class="p">[</span><span class="mi">13</span><span class="p">]</span><span class="o">=</span><span class="n">Found2</span><span class="p">[</span><span class="n">ele</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="mi">3</span><span class="p">]</span>
                                    <span class="n">header</span><span class="p">[</span><span class="mi">14</span><span class="p">]</span><span class="o">=</span><span class="n">Found2</span><span class="p">[</span><span class="n">ele</span><span class="p">][</span><span class="mi">1</span><span class="p">][</span><span class="mi">3</span><span class="p">]</span>
                                    <span class="k">for</span> <span class="n">ri</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">15</span><span class="p">):</span>
                                        <span class="n">worksheet</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">worksheetrow</span><span class="p">,</span><span class="n">ri</span><span class="p">,</span><span class="n">header</span><span class="p">[</span><span class="n">ri</span><span class="p">])</span>
                                        <span class="n">worksheet_local</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">worksheetrow_local</span><span class="p">,</span><span class="n">ri</span><span class="p">,</span><span class="n">header</span><span class="p">[</span><span class="n">ri</span><span class="p">])</span>
                                    <span class="n">worksheetrow</span><span class="o">+=</span><span class="mi">1</span>
                                    <span class="n">worksheetrow_local</span><span class="o">+=</span><span class="mi">1</span>

                        <span class="n">fout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;CC-Fa&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">k</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="o">+</span><span class="n">CC_celltype_name</span><span class="o">+</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="o">+</span><span class="s1">&#39;</span><span class="si">%0.3f</span><span class="s1">&#39;</span><span class="o">%</span><span class="p">(</span><span class="n">score</span><span class="p">[</span><span class="n">index</span><span class="p">])</span><span class="o">+</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">NC-Fa&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="mi">1</span><span class="o">+</span><span class="n">pc_index_nc</span><span class="p">[</span><span class="n">j</span><span class="p">])</span><span class="o">+</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="o">+</span><span class="n">NC_celltype_name</span><span class="p">[</span><span class="n">index</span><span class="p">]</span><span class="o">+</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">RegCoeff=</span><span class="si">%0.3f</span><span class="s1">&#39;</span><span class="o">%</span><span class="p">(</span><span class="n">normalized_ridge_coef</span><span class="p">[</span><span class="n">k</span><span class="p">,</span><span class="n">j</span><span class="p">])</span><span class="o">+</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="o">+</span><span class="s1">&#39;pvalue=</span><span class="si">%0.2e</span><span class="s1">&#39;</span><span class="o">%</span><span class="n">pvalue</span><span class="p">[</span><span class="n">k</span><span class="p">,</span><span class="n">j</span><span class="p">]</span><span class="o">+</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">-log10(pvalue)=</span><span class="si">%0.2f</span><span class="s1">&#39;</span><span class="o">%</span><span class="p">(</span><span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">pvalue</span><span class="p">[</span><span class="n">k</span><span class="p">,</span><span class="n">j</span><span class="p">])))</span><span class="c1">#str(interaction_id)</span>
                        <span class="n">fout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>

                        <span class="n">fout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;CC&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">genesWithUP</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
                        <span class="n">fout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;NC&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">genesWithDown</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
                        <span class="n">fout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="n">fout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n\n</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="n">workbook</span><span class="o">.</span><span class="n">close</span><span class="p">()</span></div>






<div class="viewcode-block" id="find_LR_interactions_in_interacting_cell_types">
<a class="viewcode-back" href="../../nico_covariations.html#nico_covariations.Covariations.find_LR_interactions_in_interacting_cell_types">[docs]</a>
<span class="k">def</span> <span class="nf">find_LR_interactions_in_interacting_cell_types</span><span class="p">(</span><span class="nb">input</span><span class="p">,</span><span class="n">choose_interacting_celltype_pair</span><span class="o">=</span><span class="p">[],</span><span class="n">choose_factors_id</span><span class="o">=</span><span class="p">[],</span><span class="n">pvalueCutoff</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span>
<span class="n">correlation_with_spearman</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">LR_plot_NMF_Fa_thres</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span> <span class="n">LR_plot_Exp_thres</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span><span class="n">saveas</span><span class="o">=</span><span class="s1">&#39;pdf&#39;</span><span class="p">,</span><span class="n">transparent_mode</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">showit</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span><span class="mi">10</span><span class="p">)):</span>

<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Find ligand-receptor (LR) interactions in interacting cell types and visualize them.</span>

<span class="sd">    This function processes the output from gene_covariation_analysis to identify significant LR interactions between specified cell type pairs and visualizes the results.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    input : object</span>
<span class="sd">        The main input is the output from gene_covariation_analysis.</span>

<span class="sd">    choose_interacting_celltype_pair : list, optional</span>
<span class="sd">        Define the cell type pairs for which information on LR communication should be returned. The first element of the list is the central cell type (CC), and the second element is the niche cell type (NC).</span>
<span class="sd">        If the list is empty, LR interactions will be returned for all significant interacting cell types.</span>
<span class="sd">        Default is [].</span>

<span class="sd">    choose_factors_id : list, optional</span>
<span class="sd">        Define factor IDs for which LR interactions are visualized. The first element of the list is the factor ID of the central cell type, and the second element is the factor ID of the niche cell type.</span>
<span class="sd">        If the list is empty, LR plots will be saved for all significant niche cell type factor interactions.</span>
<span class="sd">        Default is [].</span>

<span class="sd">    pvalueCutoff : float, optional</span>
<span class="sd">        The p-value cutoff used to find the significant central cell type factor and niche cell type factor interactions.</span>
<span class="sd">        Default is 0.05.</span>

<span class="sd">    correlation_with_spearman : bool, optional</span>
<span class="sd">        If True, compute gene-factor correlation as Spearman correlation coefficient; otherwise, compute as cosine similarity.</span>
<span class="sd">        Default is True.</span>

<span class="sd">    LR_plot_NMF_Fa_thres : float, optional</span>
<span class="sd">        Only ligands or receptors that exhibit a correlation to the respective factors higher than this cutoff are retained.</span>
<span class="sd">        Default is 0.2.</span>

<span class="sd">    LR_plot_Exp_thres : float, optional</span>
<span class="sd">        Only ligands or receptors that are expressed in a fraction of cells of the respective cell types exceeding this cutoff are retained.</span>
<span class="sd">        Default is 0.2.</span>

<span class="sd">    saveas : str, optional</span>
<span class="sd">        Save the figures in PDF or PNG format (dpi for PNG format is 300).</span>
<span class="sd">        Default is &#39;pdf&#39;.</span>

<span class="sd">    transparent_mode : bool, optional</span>
<span class="sd">        Background color in the figures.</span>
<span class="sd">        Default is False.</span>

<span class="sd">    showit : bool, optional</span>
<span class="sd">        If True, the figures are shown interactively.</span>
<span class="sd">        Default is True.</span>

<span class="sd">    figsize : tuple, optional</span>
<span class="sd">        Dimension of the figure size. The figure size on the X-axis direction is the (number of genes) multiplied by factor 12/34.</span>
<span class="sd">        The figure size on the Y-axis direction is the (number of genes) multiplied by factor 10/44.</span>
<span class="sd">        All generated figure size are scaled according to the above factors.</span>
<span class="sd">        Initital figure size is (12, 10).</span>


<span class="sd">    Outputs</span>
<span class="sd">    -------</span>
<span class="sd">    - The LR interaction figures are saved in &quot;./nico_out/covariations_R*_F*/Plot_ligand_receptor_in_niche*&quot;.</span>

<span class="sd">    Notes</span>
<span class="sd">    -----</span>
<span class="sd">    - Our analysis accounts for bidirectional cellular crosstalk interactions of ligands and receptors in cell types A and B.</span>
<span class="sd">    - The ligand can be expressed on cell type A and signal to the receptor detected on cell type B, or vice versa.</span>
<span class="sd">    - Both ligand-receptor plots and Excel sheets profile bidirectional cellular crosstalk of ligand and receptor in cell types A and B.</span>
<span class="sd">    &quot;&quot;&quot;</span>





    <span class="n">totalLRpairs</span><span class="p">,</span><span class="n">ligand</span><span class="p">,</span><span class="n">receptor</span><span class="p">,</span><span class="n">either</span><span class="o">=</span><span class="n">read_LigRecDb</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">LRdb</span><span class="p">)</span>
    <span class="n">coeff_cutoff_for_log_reg</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">logistic_coef_cutoff</span>
    <span class="n">coeff_cutoff_for_rid_reg</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">coeff_cutoff_for_rid_reg</span>
    <span class="n">gene_set_names</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">gene_set_names</span>
    <span class="n">LRcutoff</span><span class="o">=</span><span class="n">LR_plot_NMF_Fa_thres</span> <span class="c1">#Used in excel sheet to show the enrichment of ligand receptor intera</span>

    <span class="n">PCA_of_sc_cluster_accordingto_spatial_clusterid</span><span class="p">,</span><span class="n">save_scFactors</span><span class="p">,</span><span class="n">save_spFactors</span><span class="o">=</span><span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">covariation_dir</span><span class="o">+</span><span class="s1">&#39;factors_info.p&#39;</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">))</span>
    <span class="n">n</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clustername</span><span class="p">)</span>




    <span class="n">saveLRplots</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">covariation_dir</span><span class="o">+</span><span class="s1">&#39;Plot_ligand_receptor_in_niche/&#39;</span>
    <span class="n">create_directory</span><span class="p">(</span><span class="n">saveLRplots</span><span class="p">)</span>
    <span class="n">saveLRplotsFirst</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">covariation_dir</span><span class="o">+</span><span class="s1">&#39;Plot_ligand_receptor_in_niche_cc_vs_nc/&#39;</span>
    <span class="n">create_directory</span><span class="p">(</span><span class="n">saveLRplotsFirst</span><span class="p">)</span>
    <span class="n">saveLRplotsSecond</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">covariation_dir</span><span class="o">+</span><span class="s1">&#39;Plot_ligand_receptor_in_niche_nc_vs_cc/&#39;</span>
    <span class="n">create_directory</span><span class="p">(</span><span class="n">saveLRplotsSecond</span><span class="p">)</span>

    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;LR figures for both ways are saved in following path &quot;</span><span class="p">,</span> <span class="n">saveLRplots</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;LR figures for CC to NC are saved in following path &quot;</span><span class="p">,</span> <span class="n">saveLRplotsFirst</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;LR figures for NC to CC are saved in following path &quot;</span><span class="p">,</span> <span class="n">saveLRplotsSecond</span><span class="p">)</span>


    <span class="n">d</span><span class="o">=</span><span class="p">{}</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="p">):</span>
        <span class="n">clid</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clusterid</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
        <span class="n">clname</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clustername</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
        <span class="n">d</span><span class="p">[</span><span class="n">clname</span><span class="p">]</span><span class="o">=</span><span class="n">clid</span>


    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">choose_interacting_celltype_pair</span><span class="p">)</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">:</span>
        <span class="n">choose_CC_celltypes</span><span class="o">=</span><span class="p">[</span><span class="n">choose_interacting_celltype_pair</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">choose_CC_celltypes</span><span class="o">=</span><span class="p">[]</span>

    <span class="n">perform</span><span class="o">=</span><span class="p">[]</span>
    <span class="k">for</span> <span class="n">fi</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="p">):</span>
        <span class="n">CC_celltype_name</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clustername</span><span class="p">[</span><span class="n">fi</span><span class="p">]</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">choose_CC_celltypes</span><span class="p">)</span><span class="o">==</span><span class="mi">0</span><span class="p">:</span>
            <span class="n">perform</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fi</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">CC_celltype_name</span> <span class="ow">in</span> <span class="n">choose_CC_celltypes</span><span class="p">:</span>
                <span class="n">perform</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fi</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">perform</span><span class="p">:</span>
        <span class="n">clid</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clusterid</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
        <span class="n">CC_corr_spearman</span><span class="p">,</span><span class="n">CC_PCA</span><span class="p">,</span><span class="n">CC_gene</span><span class="p">,</span><span class="n">CC_meanExpression</span><span class="p">,</span><span class="n">CC_popExpression</span><span class="p">,</span><span class="n">CC_corr_cosine</span><span class="p">,</span><span class="n">alpha</span><span class="o">=</span><span class="n">PCA_of_sc_cluster_accordingto_spatial_clusterid</span><span class="p">[</span><span class="n">clid</span><span class="p">]</span>
        <span class="n">CC_celltype_name</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clustername</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
        <span class="c1">#temp=np.where(input.spatialcell_unique_clusterid[i]==input.annotation_spatial_cluster_id)</span>
        <span class="c1">#index=temp[0]</span>

        <span class="n">data</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">save_reg_coef</span><span class="p">[</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clusterid</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span>
        <span class="n">coef_mu</span><span class="p">,</span><span class="n">intercept</span><span class="p">,</span><span class="n">alpha</span><span class="p">,</span><span class="n">xlabel</span><span class="p">,</span><span class="n">score</span><span class="p">,</span><span class="n">target</span><span class="p">,</span><span class="n">neighborhoodClass</span><span class="p">,</span><span class="n">pvalue</span><span class="p">,</span><span class="n">pve</span><span class="p">,</span><span class="n">rve</span><span class="o">=</span><span class="n">data</span>

        <span class="n">NC_celltype_name</span><span class="o">=</span><span class="n">xlabel</span>
        <span class="n">largest</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="nb">abs</span><span class="p">(</span><span class="n">coef_mu</span><span class="p">))</span>
        <span class="n">normalized_ridge_coef</span><span class="o">=</span><span class="n">coef_mu</span><span class="o">/</span><span class="n">largest</span>

        <span class="n">ylabelname</span><span class="o">=</span><span class="p">[]</span>
        <span class="n">componentlabel</span><span class="o">=</span><span class="p">[]</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">no_of_pc</span><span class="p">):</span>
            <span class="n">ylabelname</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;CC_&#39;</span><span class="o">+</span><span class="n">CC_celltype_name</span><span class="o">+</span><span class="s1">&#39;_Fa&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">j</span><span class="o">+</span><span class="mi">1</span><span class="p">))</span>
            <span class="n">componentlabel</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;Fa&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">j</span><span class="o">+</span><span class="mi">1</span><span class="p">))</span>

        <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">NC_celltype_name</span><span class="p">)):</span>
            <span class="k">if</span> <span class="n">score</span><span class="p">[</span><span class="n">k</span><span class="p">]</span><span class="o">&gt;</span><span class="n">coeff_cutoff_for_log_reg</span><span class="p">:</span>
                <span class="c1">#in ylabelname first (# of pc) is the central cell type</span>
                <span class="c1">#and remaining are (# of pc) from the negihborhood cell type</span>
                <span class="k">if</span> <span class="n">CC_celltype_name</span><span class="o">!=</span><span class="n">NC_celltype_name</span><span class="p">[</span><span class="n">k</span><span class="p">]:</span>
                    <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">no_of_pc</span><span class="p">):</span>
                        <span class="n">ylabelname</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;NC_&#39;</span><span class="o">+</span><span class="n">NC_celltype_name</span><span class="p">[</span><span class="n">k</span><span class="p">]</span><span class="o">+</span><span class="s1">&#39;_s&#39;</span><span class="o">+</span><span class="s1">&#39;</span><span class="si">%0.3f</span><span class="s1">&#39;</span><span class="o">%</span><span class="n">score</span><span class="p">[</span><span class="n">k</span><span class="p">]</span><span class="o">+</span><span class="s1">&#39;_Fa&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">j</span><span class="o">+</span><span class="mi">1</span><span class="p">))</span>

        <span class="n">pc_index_nc</span><span class="o">=</span><span class="p">[]</span>
        <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">NC_celltype_name</span><span class="p">)):</span>
            <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">no_of_pc</span><span class="p">):</span>
                <span class="n">pc_index_nc</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">j</span><span class="p">)</span>

        <span class="c1">#normalized_ridge_coef  noofPC x (noofPC x +ve coff in log reg)</span>


        <span class="n">interaction_id</span><span class="o">=</span><span class="mi">0</span>
        <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">normalized_ridge_coef</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]):</span>
            <span class="c1">#k is PC of central cell type</span>
            <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">normalized_ridge_coef</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]):</span>
                <span class="n">interaction_id</span><span class="o">+=</span><span class="mi">1</span>
                <span class="n">index</span><span class="o">=</span><span class="n">math</span><span class="o">.</span><span class="n">floor</span><span class="p">(</span><span class="n">j</span><span class="o">/</span><span class="nb">input</span><span class="o">.</span><span class="n">no_of_pc</span><span class="p">)</span>

                <span class="c1">#index is the id neighboring cell type</span>
                <span class="c1">#if abs(normalized_ridge_coef[k,j])&gt;coeff_cutoff_for_rid_reg:</span>
                <span class="c1">#pvalueCutoff=1</span>
                <span class="k">if</span> <span class="p">(</span><span class="n">pvalue</span><span class="p">[</span><span class="n">k</span><span class="p">,</span><span class="n">j</span><span class="p">]</span><span class="o">&lt;</span><span class="n">pvalueCutoff</span><span class="p">)</span><span class="o">&amp;</span><span class="p">(</span><span class="nb">abs</span><span class="p">(</span><span class="n">normalized_ridge_coef</span><span class="p">[</span><span class="n">k</span><span class="p">,</span><span class="n">j</span><span class="p">])</span><span class="o">&gt;</span><span class="n">coeff_cutoff_for_rid_reg</span><span class="p">):</span>
                <span class="c1">#if True:</span>
                    <span class="k">if</span> <span class="n">score</span><span class="p">[</span><span class="n">index</span><span class="p">]</span><span class="o">&gt;</span><span class="n">coeff_cutoff_for_log_reg</span><span class="p">:</span>
                        <span class="n">NC_corr_spearman</span><span class="p">,</span><span class="n">NC_PCA</span><span class="p">,</span><span class="n">NC_gene</span><span class="p">,</span><span class="n">NC_meanExpression</span><span class="p">,</span><span class="n">NC_popExpression</span><span class="p">,</span><span class="n">NC_corr_cosine</span><span class="p">,</span><span class="n">alpha</span><span class="o">=</span><span class="n">PCA_of_sc_cluster_accordingto_spatial_clusterid</span><span class="p">[</span><span class="n">d</span><span class="p">[</span><span class="n">NC_celltype_name</span><span class="p">[</span><span class="n">index</span><span class="p">]]]</span>
                        <span class="k">if</span> <span class="n">correlation_with_spearman</span><span class="p">:</span>
                            <span class="n">top_genes_in_CC</span><span class="p">,</span><span class="n">top_genes_in_NC</span><span class="p">,</span><span class="n">genesWithUP</span><span class="p">,</span><span class="n">genesWithDown</span><span class="p">,</span><span class="n">Found1</span><span class="p">,</span><span class="n">Found2</span><span class="o">=</span><span class="n">find_fold_change</span><span class="p">(</span><span class="n">CC_corr_spearman</span><span class="p">,</span><span class="n">NC_corr_spearman</span><span class="p">,</span><span class="n">CC_gene</span><span class="p">,</span><span class="n">k</span><span class="p">,</span><span class="n">pc_index_nc</span><span class="p">[</span><span class="n">j</span><span class="p">],</span><span class="n">totalLRpairs</span><span class="p">,</span><span class="n">LRcutoff</span><span class="p">,</span><span class="n">CC_meanExpression</span><span class="p">,</span><span class="n">NC_meanExpression</span><span class="p">,</span><span class="n">CC_popExpression</span><span class="p">,</span><span class="n">NC_popExpression</span><span class="p">,</span><span class="mi">1</span><span class="p">)</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="n">top_genes_in_CC</span><span class="p">,</span><span class="n">top_genes_in_NC</span><span class="p">,</span><span class="n">genesWithUP</span><span class="p">,</span><span class="n">genesWithDown</span><span class="p">,</span><span class="n">Found1</span><span class="p">,</span><span class="n">Found2</span><span class="o">=</span><span class="n">find_fold_change</span><span class="p">(</span><span class="n">CC_corr_cosine</span><span class="p">,</span><span class="n">NC_corr_cosine</span><span class="p">,</span><span class="n">CC_gene</span><span class="p">,</span><span class="n">k</span><span class="p">,</span><span class="n">pc_index_nc</span><span class="p">[</span><span class="n">j</span><span class="p">],</span><span class="n">totalLRpairs</span><span class="p">,</span><span class="n">LRcutoff</span><span class="p">,</span><span class="n">CC_meanExpression</span><span class="p">,</span><span class="n">NC_meanExpression</span><span class="p">,</span><span class="n">CC_popExpression</span><span class="p">,</span><span class="n">NC_popExpression</span><span class="p">,</span><span class="mi">1</span><span class="p">)</span>

                        <span class="n">common_genes</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">top_genes_in_CC</span><span class="p">)</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">top_genes_in_NC</span><span class="p">)))</span>

                        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">choose_interacting_celltype_pair</span><span class="p">)</span><span class="o">&gt;</span><span class="mi">1</span><span class="p">:</span>
                            <span class="k">if</span> <span class="n">NC_celltype_name</span><span class="p">[</span><span class="n">index</span><span class="p">]</span><span class="o">==</span><span class="n">choose_interacting_celltype_pair</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span>
                                <span class="n">ncflag</span><span class="o">=</span><span class="mi">1</span>
                            <span class="k">else</span><span class="p">:</span>
                                <span class="n">ncflag</span><span class="o">=</span><span class="mi">0</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="n">ncflag</span><span class="o">=</span><span class="mi">1</span>
                        <span class="k">if</span> <span class="n">ncflag</span><span class="o">==</span><span class="mi">1</span><span class="p">:</span>
                            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">choose_factors_id</span><span class="p">)</span><span class="o">==</span><span class="mi">2</span><span class="p">:</span>
                                <span class="n">flag</span><span class="o">=</span><span class="mi">0</span>
                                <span class="n">CC_factor</span><span class="o">=</span><span class="n">choose_factors_id</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">-</span><span class="mi">1</span>
                                <span class="n">NC_factor</span><span class="o">=</span><span class="n">choose_factors_id</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">-</span><span class="mi">1</span>
                                <span class="k">if</span> <span class="p">((</span><span class="n">CC_factor</span><span class="o">==</span><span class="n">k</span><span class="p">)</span><span class="o">&amp;</span> <span class="p">(</span><span class="n">NC_factor</span><span class="o">==</span><span class="n">pc_index_nc</span><span class="p">[</span><span class="n">j</span><span class="p">])):</span>
                                    <span class="n">flag</span><span class="o">=</span><span class="mi">1</span>
                            <span class="k">else</span><span class="p">:</span>
                                <span class="n">flag</span><span class="o">=</span><span class="mi">1</span>
                            <span class="k">if</span> <span class="n">flag</span><span class="o">==</span><span class="mi">1</span><span class="p">:</span>
                                <span class="n">plot_ligand_receptor_in_interacting_celltypes</span><span class="p">(</span><span class="n">CC_celltype_name</span><span class="p">,</span><span class="n">NC_celltype_name</span><span class="p">[</span><span class="n">index</span><span class="p">],</span><span class="n">score</span><span class="p">[</span><span class="n">index</span><span class="p">],</span><span class="n">k</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="o">+</span><span class="n">pc_index_nc</span><span class="p">[</span><span class="n">j</span><span class="p">],</span><span class="n">normalized_ridge_coef</span><span class="p">[</span><span class="n">k</span><span class="p">,</span><span class="n">j</span><span class="p">],</span><span class="n">pvalue</span><span class="p">[</span><span class="n">k</span><span class="p">,</span><span class="n">j</span><span class="p">],</span><span class="n">Found1</span><span class="p">,</span><span class="n">Found2</span><span class="p">,</span><span class="n">saveLRplots</span><span class="p">,</span><span class="n">LR_plot_Exp_thres</span><span class="p">,</span><span class="n">saveas</span><span class="p">,</span><span class="n">transparent_mode</span><span class="p">,</span><span class="n">showit</span><span class="p">,</span><span class="n">figsize</span><span class="p">,</span><span class="s1">&#39;Both&#39;</span><span class="p">)</span>
                                <span class="n">plot_ligand_receptor_in_interacting_celltypes</span><span class="p">(</span><span class="n">CC_celltype_name</span><span class="p">,</span><span class="n">NC_celltype_name</span><span class="p">[</span><span class="n">index</span><span class="p">],</span><span class="n">score</span><span class="p">[</span><span class="n">index</span><span class="p">],</span><span class="n">k</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="o">+</span><span class="n">pc_index_nc</span><span class="p">[</span><span class="n">j</span><span class="p">],</span><span class="n">normalized_ridge_coef</span><span class="p">[</span><span class="n">k</span><span class="p">,</span><span class="n">j</span><span class="p">],</span><span class="n">pvalue</span><span class="p">[</span><span class="n">k</span><span class="p">,</span><span class="n">j</span><span class="p">],</span><span class="n">Found1</span><span class="p">,</span><span class="n">Found2</span><span class="p">,</span><span class="n">saveLRplotsFirst</span><span class="p">,</span><span class="n">LR_plot_Exp_thres</span><span class="p">,</span><span class="n">saveas</span><span class="p">,</span><span class="n">transparent_mode</span><span class="p">,</span><span class="n">showit</span><span class="p">,</span><span class="n">figsize</span><span class="p">,</span><span class="s1">&#39;First&#39;</span><span class="p">)</span>
                                <span class="n">plot_ligand_receptor_in_interacting_celltypes</span><span class="p">(</span><span class="n">CC_celltype_name</span><span class="p">,</span><span class="n">NC_celltype_name</span><span class="p">[</span><span class="n">index</span><span class="p">],</span><span class="n">score</span><span class="p">[</span><span class="n">index</span><span class="p">],</span><span class="n">k</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="o">+</span><span class="n">pc_index_nc</span><span class="p">[</span><span class="n">j</span><span class="p">],</span><span class="n">normalized_ridge_coef</span><span class="p">[</span><span class="n">k</span><span class="p">,</span><span class="n">j</span><span class="p">],</span><span class="n">pvalue</span><span class="p">[</span><span class="n">k</span><span class="p">,</span><span class="n">j</span><span class="p">],</span><span class="n">Found1</span><span class="p">,</span><span class="n">Found2</span><span class="p">,</span><span class="n">saveLRplotsSecond</span><span class="p">,</span><span class="n">LR_plot_Exp_thres</span><span class="p">,</span><span class="n">saveas</span><span class="p">,</span><span class="n">transparent_mode</span><span class="p">,</span><span class="n">showit</span><span class="p">,</span><span class="n">figsize</span><span class="p">,</span><span class="s1">&#39;Second&#39;</span><span class="p">)</span>

    <span class="k">return</span> <span class="mi">0</span></div>



<div class="viewcode-block" id="make_excel_sheet_for_gene_correlation">
<a class="viewcode-back" href="../../nico_covariations.html#nico_covariations.Covariations.make_excel_sheet_for_gene_correlation">[docs]</a>
<span class="k">def</span> <span class="nf">make_excel_sheet_for_gene_correlation</span><span class="p">(</span><span class="nb">input</span><span class="p">):</span>

<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create an Excel sheet compiling gene correlations with factors across different cell types.</span>

<span class="sd">    This function generates an Excel sheet that provides a structured and accessible representation of gene factors associated with cell types. It includes various types of information, such as average gene expression, Spearman correlation values, and cosine similarity values for both scRNASeq and spatial data.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    input : object</span>
<span class="sd">        The main input is the output from gene_covariation_analysis.</span>

<span class="sd">    Outputs</span>
<span class="sd">    -------</span>
<span class="sd">    - Excel sheets categorized into different types of information:</span>
<span class="sd">        - &#39;avg gene exp&#39;: Average gene expression.</span>
<span class="sd">        - &#39;spearman scRNAseq Fa(i)&#39;: Spearman correlation values for different factors within scRNASeq data.</span>
<span class="sd">        - &#39;cosine scRNAseq Fa(i)&#39;: Cosine similarity values within scRNASeq data.</span>
<span class="sd">        - &#39;spearman spatial Fa(i)&#39;: Spearman correlation values for common genes in the spatial data.</span>
<span class="sd">        - &#39;cosine spatial Fa(i)&#39;: Cosine similarity values for common genes in the spatial data.</span>

<span class="sd">    Notes</span>
<span class="sd">    -----</span>
<span class="sd">    - In the sheet names, i corresponds to the factor ID.</span>
<span class="sd">    - Columns include factors representing all cell types.</span>
<span class="sd">    - For each factor, genes are sorted based on their association with the factor ID corresponding to the respective sheet.</span>
<span class="sd">    - A color-coding scheme is used to distinguish genes:</span>
<span class="sd">        - Ligands are depicted in blue.</span>
<span class="sd">        - Receptors are depicted in red.</span>
<span class="sd">        - Genes with both ligand and receptor functions are depicted in magenta.</span>
<span class="sd">    &quot;&quot;&quot;</span>



    <span class="n">totalLRpairs</span><span class="p">,</span><span class="n">ligand</span><span class="p">,</span><span class="n">receptor</span><span class="p">,</span><span class="n">either</span><span class="o">=</span><span class="n">read_LigRecDb</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">LRdb</span><span class="p">)</span>

    <span class="n">workbook</span> <span class="o">=</span> <span class="n">xlsxwriter</span><span class="o">.</span><span class="n">Workbook</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">covariation_dir</span><span class="o">+</span><span class="s1">&#39;gene_correlation.xlsx&#39;</span><span class="p">)</span>

    <span class="n">worksheetAvgGeneExp</span><span class="o">=</span> <span class="n">workbook</span><span class="o">.</span><span class="n">add_worksheet</span><span class="p">(</span><span class="s1">&#39;avg gene exp&#39;</span><span class="p">)</span>
    <span class="n">worksheetFullGene_spearman</span><span class="o">=</span><span class="p">[]</span>
    <span class="n">worksheetFullGene_cosine</span><span class="o">=</span><span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">no_of_pc</span><span class="p">):</span>
        <span class="n">worksheetFullGene_spearman</span><span class="o">.</span><span class="n">append</span><span class="p">(</span> <span class="n">workbook</span><span class="o">.</span><span class="n">add_worksheet</span><span class="p">(</span><span class="s1">&#39;spearman scRNAseq Fa&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">)))</span>
        <span class="n">worksheetFullGene_cosine</span><span class="o">.</span><span class="n">append</span><span class="p">(</span> <span class="n">workbook</span><span class="o">.</span><span class="n">add_worksheet</span><span class="p">(</span><span class="s1">&#39;cosine scRNAseq Fa&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">)))</span>

    <span class="n">worksheetSpatialGene_spearman</span><span class="o">=</span><span class="p">[]</span>
    <span class="n">worksheetSpatialGene_cosine</span><span class="o">=</span><span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">no_of_pc</span><span class="p">):</span>
        <span class="n">worksheetSpatialGene_spearman</span><span class="o">.</span><span class="n">append</span><span class="p">(</span> <span class="n">workbook</span><span class="o">.</span><span class="n">add_worksheet</span><span class="p">(</span><span class="s1">&#39;spearman spatial Fa&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">)))</span>
        <span class="n">worksheetSpatialGene_cosine</span><span class="o">.</span><span class="n">append</span><span class="p">(</span> <span class="n">workbook</span><span class="o">.</span><span class="n">add_worksheet</span><span class="p">(</span><span class="s1">&#39;cosine spatial Fa&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">)))</span>


    <span class="n">PCA_of_sc_cluster_accordingto_spatial_clusterid</span><span class="p">,</span><span class="n">save_scFactors</span><span class="p">,</span><span class="n">save_spFactors</span><span class="o">=</span><span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">covariation_dir</span><span class="o">+</span><span class="s1">&#39;factors_info.p&#39;</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">))</span>


    <span class="c1">#outputFolder=maindir+&#39;geneCorr&#39;+str(input.no_of_pc)+&#39;/&#39;</span>
    <span class="c1">#create_directory(outputFolder)</span>


    <span class="n">genenames</span><span class="o">=</span><span class="nb">sorted</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">ad_sp</span><span class="o">.</span><span class="n">var_names</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()))</span>
    <span class="n">n</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clustername</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="p">):</span>
        <span class="n">clid</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clusterid</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
        <span class="n">CC_corr_spearman</span><span class="p">,</span><span class="n">CC_PCA</span><span class="p">,</span><span class="n">gene</span><span class="p">,</span><span class="n">CC_meanExpression</span><span class="p">,</span><span class="n">CC_popExpression</span><span class="p">,</span><span class="n">CC_corr_cosine</span><span class="p">,</span><span class="n">alpha</span><span class="o">=</span><span class="n">PCA_of_sc_cluster_accordingto_spatial_clusterid</span><span class="p">[</span><span class="n">clid</span><span class="p">]</span>
        <span class="n">worksheetrow</span><span class="o">=</span><span class="mi">0</span>
        <span class="n">worksheetAvgGeneExp</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">worksheetrow</span><span class="p">,</span><span class="mi">3</span><span class="o">*</span><span class="n">i</span><span class="p">,</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clustername</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">no_of_pc</span><span class="p">):</span>
            <span class="n">worksheetFullGene_spearman</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">worksheetrow</span><span class="p">,(</span><span class="nb">input</span><span class="o">.</span><span class="n">no_of_pc</span><span class="o">+</span><span class="mi">2</span><span class="p">)</span><span class="o">*</span><span class="n">i</span><span class="p">,</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clustername</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
            <span class="n">worksheetSpatialGene_spearman</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">worksheetrow</span><span class="p">,(</span><span class="nb">input</span><span class="o">.</span><span class="n">no_of_pc</span><span class="o">+</span><span class="mi">2</span><span class="p">)</span><span class="o">*</span><span class="n">i</span><span class="p">,</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clustername</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
            <span class="n">worksheetFullGene_cosine</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">worksheetrow</span><span class="p">,(</span><span class="nb">input</span><span class="o">.</span><span class="n">no_of_pc</span><span class="o">+</span><span class="mi">2</span><span class="p">)</span><span class="o">*</span><span class="n">i</span><span class="p">,</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clustername</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
            <span class="n">worksheetSpatialGene_cosine</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">worksheetrow</span><span class="p">,(</span><span class="nb">input</span><span class="o">.</span><span class="n">no_of_pc</span><span class="o">+</span><span class="mi">2</span><span class="p">)</span><span class="o">*</span><span class="n">i</span><span class="p">,</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clustername</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
        <span class="n">worksheetrow</span><span class="o">+=</span><span class="mi">1</span>
        <span class="n">fixvalue</span><span class="o">=</span><span class="n">worksheetrow</span>


        <span class="n">index</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="o">-</span><span class="n">CC_meanExpression</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">index</span><span class="p">)):</span>
            <span class="n">worksheetAvgGeneExp</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">j</span><span class="o">+</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="o">*</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span><span class="n">CC_meanExpression</span><span class="p">[</span><span class="n">index</span><span class="p">[</span><span class="n">j</span><span class="p">]])</span>
            <span class="n">worksheetAvgGeneExp</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">j</span><span class="o">+</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="o">*</span><span class="n">i</span><span class="p">,</span><span class="n">gene</span><span class="p">[</span><span class="n">index</span><span class="p">[</span><span class="n">j</span><span class="p">]])</span>


        <span class="n">red</span> <span class="o">=</span> <span class="n">workbook</span><span class="o">.</span><span class="n">add_format</span><span class="p">({</span><span class="s1">&#39;bold&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span><span class="s1">&#39;color&#39;</span><span class="p">:</span> <span class="s1">&#39;red&#39;</span><span class="p">})</span>
        <span class="n">green</span> <span class="o">=</span> <span class="n">workbook</span><span class="o">.</span><span class="n">add_format</span><span class="p">({</span><span class="s1">&#39;bold&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span><span class="s1">&#39;color&#39;</span><span class="p">:</span> <span class="s1">&#39;green&#39;</span><span class="p">})</span>
        <span class="n">blue</span> <span class="o">=</span> <span class="n">workbook</span><span class="o">.</span><span class="n">add_format</span><span class="p">({</span><span class="s1">&#39;bold&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span><span class="s1">&#39;color&#39;</span><span class="p">:</span> <span class="s1">&#39;blue&#39;</span><span class="p">})</span>
        <span class="n">magenta</span> <span class="o">=</span> <span class="n">workbook</span><span class="o">.</span><span class="n">add_format</span><span class="p">({</span><span class="s1">&#39;bold&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span><span class="s1">&#39;color&#39;</span><span class="p">:</span> <span class="s1">&#39;magenta&#39;</span><span class="p">})</span>


<span class="w">        </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">        fig,(ax)=plt.subplots(1,1,figsize=(8,6))</span>
<span class="sd">        ax.plot(CC_corr[:,0],CC_corr[:,1],&#39;.&#39;,markersize=1)</span>
<span class="sd">        ax.set_xlabel(&#39;PC1&#39;)</span>
<span class="sd">        ax.set_ylabel(&#39;PC2&#39;)</span>
<span class="sd">        ax.set_title(input.spatialcell_unique_clustername[i])</span>
<span class="sd">        #fig.tight_layout()</span>
<span class="sd">        fig.savefig(outputFolder+&#39;correlation_&#39;+input.spatialcell_unique_clustername[i]+&#39;.png&#39;,bbox_inches=&#39;tight&#39;,transparent=True,dpi=300)</span>
<span class="sd">        plt.close(&#39;all&#39;)</span>
<span class="sd">        &#39;&#39;&#39;</span>

        <span class="n">headersave_full</span><span class="p">,</span><span class="n">headersave_common</span><span class="p">,</span><span class="n">sort_full</span><span class="p">,</span><span class="n">sort_common</span><span class="o">=</span><span class="n">sorting_of_factors_for_showing_the_value_in_excelsheet</span><span class="p">(</span><span class="n">CC_corr_spearman</span><span class="p">,</span><span class="nb">input</span><span class="o">.</span><span class="n">no_of_pc</span><span class="p">,</span><span class="n">gene</span><span class="p">,</span><span class="n">genenames</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">no_of_pc</span><span class="p">):</span>
            <span class="n">worksheetrow</span><span class="o">=</span><span class="n">fixvalue</span>
            <span class="n">indsort</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">sort_full</span><span class="p">[</span><span class="n">k</span><span class="p">]))</span>
            <span class="k">for</span> <span class="n">rj</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">indsort</span><span class="p">)):</span>
                <span class="n">header</span><span class="o">=</span><span class="n">headersave_full</span><span class="p">[</span><span class="n">indsort</span><span class="p">[</span><span class="n">rj</span><span class="p">]]</span>
                <span class="n">mygene</span><span class="o">=</span><span class="n">header</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">genecolor</span><span class="o">=</span><span class="s1">&#39;&#39;</span>
                <span class="k">if</span> <span class="n">mygene</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="ow">in</span> <span class="n">ligand</span><span class="p">:</span>
                    <span class="n">genecolor</span><span class="o">=</span><span class="n">blue</span>
                <span class="k">elif</span> <span class="n">mygene</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="ow">in</span> <span class="n">receptor</span><span class="p">:</span>
                    <span class="n">genecolor</span><span class="o">=</span><span class="n">red</span>
                <span class="k">elif</span> <span class="n">mygene</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="ow">in</span> <span class="n">either</span><span class="p">:</span>
                    <span class="n">genecolor</span><span class="o">=</span><span class="n">magenta</span>

                <span class="k">for</span> <span class="n">ri</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">header</span><span class="p">)):</span>
                    <span class="n">worksheetFullGene_spearman</span><span class="p">[</span><span class="n">k</span><span class="p">]</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">worksheetrow</span><span class="p">,(</span><span class="nb">input</span><span class="o">.</span><span class="n">no_of_pc</span><span class="o">+</span><span class="mi">2</span><span class="p">)</span><span class="o">*</span><span class="n">i</span><span class="o">+</span><span class="n">ri</span><span class="p">,</span><span class="n">header</span><span class="p">[</span><span class="n">ri</span><span class="p">],</span><span class="n">genecolor</span><span class="p">)</span>
                <span class="n">worksheetrow</span><span class="o">+=</span><span class="mi">1</span>

            <span class="n">worksheetrow</span><span class="o">=</span><span class="n">fixvalue</span>
            <span class="n">indsort</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">sort_common</span><span class="p">[</span><span class="n">k</span><span class="p">]))</span>
            <span class="k">for</span> <span class="n">rj</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">indsort</span><span class="p">)):</span>
                <span class="n">header</span><span class="o">=</span><span class="n">headersave_common</span><span class="p">[</span><span class="n">indsort</span><span class="p">[</span><span class="n">rj</span><span class="p">]]</span>
                <span class="k">for</span> <span class="n">ri</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">header</span><span class="p">)):</span>
                    <span class="n">worksheetSpatialGene_spearman</span><span class="p">[</span><span class="n">k</span><span class="p">]</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">worksheetrow</span><span class="p">,(</span><span class="nb">input</span><span class="o">.</span><span class="n">no_of_pc</span><span class="o">+</span><span class="mi">2</span><span class="p">)</span><span class="o">*</span><span class="n">i</span><span class="o">+</span><span class="n">ri</span><span class="p">,</span><span class="n">header</span><span class="p">[</span><span class="n">ri</span><span class="p">])</span>
                <span class="n">worksheetrow</span><span class="o">+=</span><span class="mi">1</span>



        <span class="n">headersave_full</span><span class="p">,</span><span class="n">headersave_common</span><span class="p">,</span><span class="n">sort_full</span><span class="p">,</span><span class="n">sort_common</span><span class="o">=</span><span class="n">sorting_of_factors_for_showing_the_value_in_excelsheet</span><span class="p">(</span><span class="n">CC_corr_cosine</span><span class="p">,</span><span class="nb">input</span><span class="o">.</span><span class="n">no_of_pc</span><span class="p">,</span><span class="n">gene</span><span class="p">,</span><span class="n">genenames</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">no_of_pc</span><span class="p">):</span>
            <span class="n">worksheetrow</span><span class="o">=</span><span class="n">fixvalue</span>
            <span class="n">indsort</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">sort_full</span><span class="p">[</span><span class="n">k</span><span class="p">]))</span>
            <span class="k">for</span> <span class="n">rj</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">indsort</span><span class="p">)):</span>
                <span class="n">header</span><span class="o">=</span><span class="n">headersave_full</span><span class="p">[</span><span class="n">indsort</span><span class="p">[</span><span class="n">rj</span><span class="p">]]</span>
                <span class="n">mygene</span><span class="o">=</span><span class="n">header</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">genecolor</span><span class="o">=</span><span class="s1">&#39;&#39;</span>
                <span class="k">if</span> <span class="n">mygene</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="ow">in</span> <span class="n">ligand</span><span class="p">:</span>
                    <span class="n">genecolor</span><span class="o">=</span><span class="n">blue</span>
                <span class="k">elif</span> <span class="n">mygene</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="ow">in</span> <span class="n">receptor</span><span class="p">:</span>
                    <span class="n">genecolor</span><span class="o">=</span><span class="n">red</span>
                <span class="k">elif</span> <span class="n">mygene</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="ow">in</span> <span class="n">either</span><span class="p">:</span>
                    <span class="n">genecolor</span><span class="o">=</span><span class="n">magenta</span>

                <span class="k">for</span> <span class="n">ri</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">header</span><span class="p">)):</span>
                    <span class="n">worksheetFullGene_cosine</span><span class="p">[</span><span class="n">k</span><span class="p">]</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">worksheetrow</span><span class="p">,(</span><span class="nb">input</span><span class="o">.</span><span class="n">no_of_pc</span><span class="o">+</span><span class="mi">2</span><span class="p">)</span><span class="o">*</span><span class="n">i</span><span class="o">+</span><span class="n">ri</span><span class="p">,</span><span class="n">header</span><span class="p">[</span><span class="n">ri</span><span class="p">],</span><span class="n">genecolor</span><span class="p">)</span>
                <span class="n">worksheetrow</span><span class="o">+=</span><span class="mi">1</span>

            <span class="n">worksheetrow</span><span class="o">=</span><span class="n">fixvalue</span>
            <span class="n">indsort</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">sort_common</span><span class="p">[</span><span class="n">k</span><span class="p">]))</span>
            <span class="k">for</span> <span class="n">rj</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">indsort</span><span class="p">)):</span>
                <span class="n">header</span><span class="o">=</span><span class="n">headersave_common</span><span class="p">[</span><span class="n">indsort</span><span class="p">[</span><span class="n">rj</span><span class="p">]]</span>
                <span class="k">for</span> <span class="n">ri</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">header</span><span class="p">)):</span>
                    <span class="n">worksheetSpatialGene_cosine</span><span class="p">[</span><span class="n">k</span><span class="p">]</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">worksheetrow</span><span class="p">,(</span><span class="nb">input</span><span class="o">.</span><span class="n">no_of_pc</span><span class="o">+</span><span class="mi">2</span><span class="p">)</span><span class="o">*</span><span class="n">i</span><span class="o">+</span><span class="n">ri</span><span class="p">,</span><span class="n">header</span><span class="p">[</span><span class="n">ri</span><span class="p">])</span>
                <span class="n">worksheetrow</span><span class="o">+=</span><span class="mi">1</span>

    <span class="n">workbook</span><span class="o">.</span><span class="n">close</span><span class="p">()</span></div>



<div class="viewcode-block" id="pathway_analysis">
<a class="viewcode-back" href="../../nico_covariations.html#nico_covariations.Covariations.pathway_analysis">[docs]</a>
<span class="k">def</span> <span class="nf">pathway_analysis</span><span class="p">(</span><span class="nb">input</span><span class="p">,</span><span class="n">NOG_pathway</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">choose_factors_id</span><span class="o">=</span><span class="p">[],</span><span class="n">correlation_with_spearman</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">saveas</span><span class="o">=</span><span class="s1">&#39;pdf&#39;</span><span class="p">,</span><span class="n">savefigure</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
 <span class="n">positively_correlated</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">rps_rpl_mt_genes_included</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">choose_celltypes</span><span class="o">=</span><span class="p">[],</span><span class="n">circlesize</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span><span class="n">pathwayCutoff</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span>
 <span class="n">pathwayorganism</span><span class="o">=</span><span class="s1">&#39;Mouse&#39;</span><span class="p">,</span><span class="n">database</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;GO_Biological_Process_2021&#39;</span><span class="p">,</span><span class="s1">&#39;BioPlanet_2019&#39;</span><span class="p">,</span><span class="s1">&#39;Reactome_2016&#39;</span><span class="p">]):</span><span class="c1">#background_geneName,background_expression</span>

<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Perform pathway analysis for gene covariations within niches.</span>

<span class="sd">    This function utilizes the gene covariation data from the gene_covariation_analysis to perform pathway enrichment analysis using the GSEApy package. It identifies and visualizes the pathways that are significantly enriched among the top genes associated with NMF factors.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    input : object</span>
<span class="sd">        The main input is the output from gene_covariation_analysis.</span>

<span class="sd">    NOG_pathway : int, optional</span>
<span class="sd">        Number of top genes associated with NMF factors to search in the pathway database. If no pathways are observed, increase the number of genes or try different databases.</span>
<span class="sd">        (default is 50)</span>

<span class="sd">    choose_factors_id : list, optional</span>
<span class="sd">        Define the factor IDs for which pathway enrichments should be visualized. If empty, pathways will be saved for all factors.</span>
<span class="sd">        (default is [])</span>

<span class="sd">    correlation_with_spearman : bool, optional</span>
<span class="sd">        If True, visualize gene-factor correlations obtained by Spearman correlation coefficient; otherwise, use cosine similarities.</span>
<span class="sd">        (default is True)</span>

<span class="sd">    positively_correlated : bool, optional</span>
<span class="sd">        If True and correlation_with_spearman is selected, select positively correlated genes; otherwise, select negatively correlated genes.</span>
<span class="sd">        (default is True)</span>

<span class="sd">    rps_rpl_mt_genes_included : bool, optional</span>
<span class="sd">        If True, include rps, rpl, and mt genes in the pathway analysis; if False, filter these genes out.</span>
<span class="sd">        (default is True)</span>

<span class="sd">    pathwayCutoff : float, optional</span>
<span class="sd">        The cutoff parameter for finding pathway-enriched libraries from the top genes for each factor of a given cell type using GSEApy.</span>
<span class="sd">        (default is 0.5)</span>

<span class="sd">    pathwayorganism : str, optional</span>
<span class="sd">        The organism used in the GSEApy package.</span>
<span class="sd">        (default is &#39;Mouse&#39;)</span>

<span class="sd">    database : list, optional</span>
<span class="sd">        The databases used in the GSEApy package for pathway analysis. The default includes &#39;GO_Biological_Process_2021&#39;, &#39;BioPlanet_2019&#39;, and &#39;Reactome_2016&#39;.</span>
<span class="sd">        (default is [&#39;GO_Biological_Process_2021&#39;, &#39;BioPlanet_2019&#39;, &#39;Reactome_2016&#39;])</span>

<span class="sd">    choose_celltypes : list, optional</span>
<span class="sd">        Define the cell types for which pathway enrichments should be returned from the Enrichr library. If empty, the output will be generated for all cell types.</span>
<span class="sd">        (default is [])</span>

<span class="sd">    saveas : str, optional</span>
<span class="sd">        Save the figures in PDF or PNG format (dpi for PNG format is 300).</span>
<span class="sd">        (default is &#39;pdf&#39;)</span>

<span class="sd">    circlesize : int, optional</span>
<span class="sd">        The size of the points in the pathway figure. Increase this value if the dots are too small.</span>
<span class="sd">        (default is 12)</span>

<span class="sd">    savefigure : bool, optional</span>
<span class="sd">        If True, the generated figures will be saved.</span>
<span class="sd">        (default is False)</span>

<span class="sd">    showit : bool, optional</span>
<span class="sd">        If True, the generated figures will be displayed.</span>
<span class="sd">        (default is True)</span>

<span class="sd">    figsize : tuple, optional</span>
<span class="sd">        The dimension of the figure size.</span>
<span class="sd">        (default is (12, 10))</span>

<span class="sd">    Outputs</span>
<span class="sd">    -------</span>
<span class="sd">    The pathway figures are saved in &quot;./nico_out/covariations_R*_F*/Pathway_figures/&quot;.</span>

<span class="sd">    Notes</span>
<span class="sd">    -----</span>
<span class="sd">    - In the sheet names, i corresponds to the factor ID.</span>
<span class="sd">    - Columns include factors representing all cell types.</span>
<span class="sd">    - For each factor, genes are sorted based on their association with the factor ID corresponding to the respective sheet.</span>
<span class="sd">    - A color-coding scheme is used to distinguish genes:</span>
<span class="sd">        - Ligands are depicted in blue.</span>
<span class="sd">        - Receptors are depicted in red.</span>
<span class="sd">        - Genes with both ligand and receptor functions are depicted in magenta.</span>

<span class="sd">    Example</span>
<span class="sd">    -------</span>
<span class="sd">    For other available databases, check for species Human, Mouse, Yeast, Fly, Fish, and Worm in the following way:</span>

<span class="sd">        &gt;&gt;&gt; import gseapy as gp</span>
<span class="sd">        &gt;&gt;&gt; mouse = gp.get_library_name(organism=&#39;Mouse&#39;)</span>
<span class="sd">        &gt;&gt;&gt; human = gp.get_library_name(organism=&#39;Human&#39;)</span>
<span class="sd">    &quot;&quot;&quot;</span>


    <span class="n">savename</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">covariation_dir</span><span class="o">+</span><span class="s1">&#39;Pathway_figures/&#39;</span>
    <span class="n">create_directory</span><span class="p">(</span><span class="n">savename</span><span class="p">)</span>

    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;The pathway figures are saved in &quot;</span><span class="p">,</span> <span class="n">savename</span><span class="p">)</span>


    <span class="c1">#coeff_cutoff_for_log_reg=input.logistic_coef_cutoff</span>
    <span class="c1">#coeff_cutoff_for_rid_reg=input.coeff_cutoff_for_rid_reg</span>
    <span class="c1">#gene_set_names=input.gene_set_names</span>
    <span class="n">nog</span><span class="o">=</span><span class="n">NOG_pathway</span>
    <span class="n">PCA_of_sc_cluster_accordingto_spatial_clusterid</span><span class="p">,</span><span class="n">save_scFactors</span><span class="p">,</span><span class="n">save_spFactors</span><span class="o">=</span><span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">covariation_dir</span><span class="o">+</span><span class="s1">&#39;factors_info.p&#39;</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">))</span>
    <span class="n">n</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clustername</span><span class="p">)</span>

    <span class="n">perform</span><span class="o">=</span><span class="p">[]</span>
    <span class="n">found</span><span class="o">=</span><span class="p">[]</span>
    <span class="k">for</span> <span class="n">fi</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="p">):</span>
        <span class="n">CC_celltype_name</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clustername</span><span class="p">[</span><span class="n">fi</span><span class="p">]</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">choose_celltypes</span><span class="p">)</span><span class="o">==</span><span class="mi">0</span><span class="p">:</span>
            <span class="n">perform</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fi</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">CC_celltype_name</span> <span class="ow">in</span> <span class="n">choose_celltypes</span><span class="p">:</span>
                <span class="n">perform</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fi</span><span class="p">)</span>
                <span class="n">found</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">CC_celltype_name</span><span class="p">)</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">choose_celltypes</span><span class="p">)</span><span class="o">!=</span><span class="mi">0</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;cell types found &quot;</span><span class="p">,</span><span class="n">found</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">fi</span> <span class="ow">in</span> <span class="n">perform</span><span class="p">:</span>
        <span class="n">clid</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clusterid</span><span class="p">[</span><span class="n">fi</span><span class="p">]</span>
        <span class="n">spearman_factors</span><span class="p">,</span><span class="n">CC_PCA</span><span class="p">,</span><span class="n">CC_gene</span><span class="p">,</span><span class="n">CC_meanExpression</span><span class="p">,</span><span class="n">CC_popExpression</span><span class="p">,</span><span class="n">cosine_factors</span><span class="p">,</span><span class="n">alpha</span><span class="o">=</span><span class="n">PCA_of_sc_cluster_accordingto_spatial_clusterid</span><span class="p">[</span><span class="n">clid</span><span class="p">]</span>
        <span class="n">CC_celltype_name</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clustername</span><span class="p">[</span><span class="n">fi</span><span class="p">]</span>

        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">no_of_pc</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">correlation_with_spearman</span><span class="p">:</span>
                <span class="n">source</span><span class="o">=</span><span class="n">spearman_factors</span><span class="p">[:,</span><span class="n">j</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">source</span><span class="o">=</span><span class="n">cosine_factors</span><span class="p">[:,</span><span class="n">j</span><span class="p">]</span>
            <span class="n">ind</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="o">-</span><span class="n">source</span><span class="p">)</span>
            <span class="n">interestofGene</span><span class="o">=</span><span class="p">[]</span>
            <span class="n">value</span><span class="o">=</span><span class="p">[]</span>
            <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">source</span><span class="p">)):</span>
                <span class="n">temp</span><span class="o">=</span><span class="n">CC_gene</span><span class="p">[</span><span class="n">ind</span><span class="p">[</span><span class="n">k</span><span class="p">]]</span>
                <span class="k">if</span> <span class="n">rps_rpl_mt_genes_included</span><span class="p">:</span>
                    <span class="n">flag</span><span class="o">=</span><span class="mi">1</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">flag</span><span class="o">=</span><span class="mi">1</span>
                    <span class="k">if</span> <span class="n">temp</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">3</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;Rps&#39;</span><span class="p">:</span>
                        <span class="n">flag</span><span class="o">=</span><span class="mi">0</span>
                    <span class="k">if</span> <span class="n">temp</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">3</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;Rpl&#39;</span><span class="p">:</span>
                        <span class="n">flag</span><span class="o">=</span><span class="mi">0</span>
                    <span class="k">if</span> <span class="n">temp</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">3</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;mt-&#39;</span><span class="p">:</span>
                        <span class="n">flag</span><span class="o">=</span><span class="mi">0</span>
                <span class="k">if</span> <span class="n">flag</span><span class="o">==</span><span class="mi">1</span><span class="p">:</span>
                    <span class="n">interestofGene</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">CC_gene</span><span class="p">[</span><span class="n">ind</span><span class="p">[</span><span class="n">k</span><span class="p">]])</span>
                    <span class="n">value</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">source</span><span class="p">[</span><span class="n">ind</span><span class="p">[</span><span class="n">k</span><span class="p">]])</span>

            <span class="n">value</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">value</span><span class="p">)</span>
            <span class="n">interestofGene</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">interestofGene</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">positively_correlated</span><span class="p">:</span>
                <span class="n">index</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="o">-</span><span class="n">value</span><span class="p">)</span>
                <span class="n">tname</span><span class="o">=</span><span class="s1">&#39;pos&#39;</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">index</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">value</span><span class="p">)</span>
                <span class="n">tname</span><span class="o">=</span><span class="s1">&#39;neg&#39;</span>

            <span class="n">value</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">value</span><span class="p">[</span><span class="n">index</span><span class="p">])</span>
            <span class="n">interestofGene</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">interestofGene</span><span class="p">[</span><span class="n">index</span><span class="p">])</span>

            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">interestofGene</span><span class="p">)</span><span class="o">&gt;</span><span class="n">nog</span><span class="p">:</span>
                <span class="n">va1</span><span class="o">=</span><span class="n">value</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="n">nog</span><span class="p">]</span>
                <span class="n">ga1</span><span class="o">=</span><span class="n">interestofGene</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="n">nog</span><span class="p">]</span>
                <span class="n">cutoff</span><span class="o">=</span><span class="n">va1</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">ga1</span><span class="o">=</span><span class="n">interestofGene</span>
                <span class="n">va1</span><span class="o">=</span><span class="n">value</span>

            <span class="n">ccname</span><span class="o">=</span><span class="n">remove_extra_character_from_name</span><span class="p">(</span><span class="n">CC_celltype_name</span><span class="p">)</span>
            <span class="n">titlename</span><span class="o">=</span><span class="n">tname</span><span class="o">+</span><span class="s1">&#39; Fa&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">j</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39; &#39;</span><span class="o">+</span><span class="n">CC_celltype_name</span><span class="o">+</span><span class="s1">&#39; c&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="mi">100</span><span class="o">*</span><span class="n">cutoff</span><span class="p">))</span>
            <span class="n">sname1</span><span class="o">=</span><span class="n">tname</span><span class="o">+</span><span class="s1">&#39;Fa&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">j</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;_&#39;</span><span class="o">+</span><span class="n">ccname</span><span class="o">+</span><span class="s1">&#39;_c&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="mi">100</span><span class="o">*</span><span class="n">cutoff</span><span class="p">))</span>

            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">choose_factors_id</span><span class="p">)</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">:</span>
                <span class="k">if</span> <span class="p">(</span><span class="n">j</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span> <span class="ow">in</span> <span class="n">choose_factors_id</span><span class="p">:</span>
                    <span class="n">flag</span><span class="o">=</span><span class="mi">1</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">flag</span><span class="o">=</span><span class="mi">0</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">flag</span><span class="o">=</span><span class="mi">1</span>

            <span class="k">if</span> <span class="n">flag</span><span class="o">==</span><span class="mi">1</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">database</span><span class="p">)):</span>
                    <span class="n">titlename1</span><span class="o">=</span><span class="n">titlename</span><span class="o">+</span><span class="s1">&#39;[&#39;</span><span class="o">+</span><span class="n">database</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">+</span><span class="s1">&#39;]&#39;</span><span class="o">+</span><span class="s1">&#39; #G=&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">ga1</span><span class="p">))</span>
                    <span class="n">sname2</span><span class="o">=</span><span class="n">sname1</span><span class="o">+</span><span class="s1">&#39;_&#39;</span><span class="o">+</span><span class="n">database</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
                    <span class="n">finalsavename</span><span class="o">=</span><span class="n">savename</span><span class="o">+</span><span class="n">sname2</span><span class="o">+</span><span class="s1">&#39;.&#39;</span><span class="o">+</span><span class="n">saveas</span>

                    <span class="n">enr_res1</span> <span class="o">=</span> <span class="n">gseapy</span><span class="o">.</span><span class="n">enrichr</span><span class="p">(</span><span class="n">gene_list</span><span class="o">=</span><span class="n">ga1</span><span class="p">,</span><span class="n">organism</span><span class="o">=</span><span class="n">pathwayorganism</span><span class="p">,</span><span class="n">gene_sets</span><span class="o">=</span><span class="n">database</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">cutoff</span> <span class="o">=</span> <span class="n">pathwayCutoff</span><span class="p">)</span>
                    <span class="c1">#enr_res1 = gseapy.enrichr(gene_list=g1,organism=&#39;Mouse&#39;,gene_sets=background_model,description=&#39;pathway&#39;,cutoff = 0.5)</span>
                    <span class="n">finalsavename</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">,</span><span class="s1">&#39;_&#39;</span><span class="p">)</span>
                    <span class="k">try</span><span class="p">:</span>
                        <span class="c1">#gseapy.barplot(enr_res1.res2d,title=titlename1,ofname=finalsavename,fontsize=12)#database[i]+titlename</span>
                        <span class="k">if</span> <span class="n">savefigure</span><span class="p">:</span>
                            <span class="n">gseapy</span><span class="o">.</span><span class="n">dotplot</span><span class="p">(</span><span class="n">enr_res1</span><span class="o">.</span><span class="n">res2d</span><span class="p">,</span><span class="n">title</span><span class="o">=</span><span class="n">titlename1</span><span class="p">,</span><span class="n">ofname</span><span class="o">=</span><span class="n">finalsavename</span><span class="p">,</span><span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span><span class="n">size</span><span class="o">=</span><span class="n">circlesize</span><span class="p">,</span><span class="n">cmap</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">cm</span><span class="o">.</span><span class="n">autumn_r</span><span class="p">)</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="n">gseapy</span><span class="o">.</span><span class="n">dotplot</span><span class="p">(</span><span class="n">enr_res1</span><span class="o">.</span><span class="n">res2d</span><span class="p">,</span><span class="n">title</span><span class="o">=</span><span class="n">titlename1</span><span class="p">,</span><span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span><span class="n">size</span><span class="o">=</span><span class="n">circlesize</span><span class="p">,</span><span class="n">cmap</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">cm</span><span class="o">.</span><span class="n">autumn_r</span><span class="p">)</span>
                    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span> <span class="c1">#Exception: Error getting the Enrichr libraries</span>
                        <span class="k">pass</span></div>



<div class="viewcode-block" id="extract_and_plot_top_genes_from_chosen_factor_in_celltype">
<a class="viewcode-back" href="../../nico_covariations.html#nico_covariations.Covariations.extract_and_plot_top_genes_from_chosen_factor_in_celltype">[docs]</a>
<span class="k">def</span> <span class="nf">extract_and_plot_top_genes_from_chosen_factor_in_celltype</span><span class="p">(</span><span class="nb">input</span><span class="p">,</span><span class="n">choose_celltype</span><span class="p">,</span><span class="n">choose_factor_id</span><span class="p">,</span><span class="n">top_NOG</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span><span class="n">rps_rpl_mt_genes_included</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="n">correlation_with_spearman</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">positively_correlated</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">saveas</span><span class="o">=</span><span class="s1">&#39;pdf&#39;</span><span class="p">,</span><span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;RdBu_r&#39;</span><span class="p">,</span><span class="n">transparent_mode</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">showit</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">6</span><span class="p">)):</span>


<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Extract and plot top genes associated with a chosen factor in a specified cell type.</span>

<span class="sd">    This function uses the output from gene_covariation_analysis to identify and visualize the top genes associated with a chosen factor in a specified cell type. The genes can be filtered and visualized based on their correlation with the factor, with options to include or exclude specific gene types.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    input : object</span>
<span class="sd">        The main input is the output from gene_covariation_analysis.</span>

<span class="sd">    choose_celltype : str</span>
<span class="sd">        Define the cell type to include in the analysis.</span>

<span class="sd">    choose_factor_id : int</span>
<span class="sd">        Define the factor ID of the cell type to be analyzed.</span>

<span class="sd">    top_NOG : int, optional</span>
<span class="sd">        Number of top genes to visualize.</span>
<span class="sd">        (default is 30)</span>

<span class="sd">    rps_rpl_mt_genes_included : bool, optional</span>
<span class="sd">        Decide whether to include rps, rpl, and mt genes in the pathway analysis. If True, they are included.</span>
<span class="sd">        (default is True)</span>

<span class="sd">    correlation_with_spearman : bool, optional</span>
<span class="sd">        If True, visualize gene-factor association using the Spearman correlation coefficient; otherwise, use cosine similarity.</span>
<span class="sd">        (default is True)</span>

<span class="sd">    positively_correlated : bool, optional</span>
<span class="sd">        If the gene-factor association is selected as Spearman correlation, choose whether the associated genes should be positively correlated (True) or negatively correlated (False).</span>
<span class="sd">        (default is True)</span>

<span class="sd">    saveas : str, optional</span>
<span class="sd">        Save the figures in PDF or PNG format (dpi for PNG format is 300).</span>
<span class="sd">        (default is &#39;pdf&#39;)</span>

<span class="sd">    cmap : str, optional</span>
<span class="sd">        Define the colormap for visualizing factors.</span>
<span class="sd">        (default is &#39;RdBu_r&#39;)</span>

<span class="sd">    transparent_mode : bool, optional</span>
<span class="sd">        Define the background color of the figures. If True, figures have a transparent background.</span>
<span class="sd">        (default is False)</span>

<span class="sd">    showit : bool, optional</span>
<span class="sd">        If True, the generated figures will be displayed.</span>
<span class="sd">        (default is True)</span>

<span class="sd">    figsize : tuple, optional</span>
<span class="sd">        Dimension of the figure size.</span>
<span class="sd">        (default is (5, 6))</span>

<span class="sd">    Outputs</span>
<span class="sd">    -------</span>
<span class="sd">    pd.DataFrame</span>
<span class="sd">        Returns a DataFrame containing the gene, factor, average expression, and proportion of the population expressing that gene.</span>

<span class="sd">    Notes</span>
<span class="sd">    -----</span>
<span class="sd">    - The function saves the figures in the directory &quot;nico_out/covariations_R*_F*/dotplots/Factors*&quot;.</span>
<span class="sd">    - The DataFrame returned includes detailed information about the top genes associated with the chosen factor.</span>

<span class="sd">    Example</span>
<span class="sd">    -------</span>
<span class="sd">    &gt;&gt;&gt; extract_and_plot_top_genes_from_chosen_factor_in_celltype(input_data, &#39;CellTypeA&#39;, 1, top_NOG=50, saveas=&#39;png&#39;, figsize=(10, 8))</span>
<span class="sd">    &quot;&quot;&quot;</span>




    <span class="n">savefigdir</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">covariation_dir</span><span class="o">+</span> <span class="s1">&#39;dotplots/&#39;</span>
    <span class="n">create_directory</span><span class="p">(</span><span class="n">savefigdir</span><span class="p">)</span>

    <span class="n">PCA_of_sc_cluster_accordingto_spatial_clusterid</span><span class="p">,</span><span class="n">save_scFactors</span><span class="p">,</span><span class="n">save_spFactors</span><span class="o">=</span><span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">covariation_dir</span><span class="o">+</span><span class="s1">&#39;factors_info.p&#39;</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">))</span>
    <span class="n">n</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clustername</span><span class="p">)</span>
    <span class="n">perform</span><span class="o">=</span><span class="p">[]</span>
    <span class="k">for</span> <span class="n">fi</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="p">):</span>
        <span class="n">CC_celltype_name</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clustername</span><span class="p">[</span><span class="n">fi</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">CC_celltype_name</span><span class="o">==</span><span class="n">choose_celltype</span><span class="p">:</span>
                <span class="n">perform</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fi</span><span class="p">)</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">perform</span><span class="p">)</span><span class="o">==</span><span class="mi">0</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Cell type name do not match&quot;</span><span class="p">)</span>
        <span class="n">flag_correct</span><span class="o">=</span><span class="mi">0</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">flag_correct</span><span class="o">=</span><span class="mi">1</span>

    <span class="k">if</span> <span class="mi">1</span><span class="o">&lt;=</span><span class="n">choose_factor_id</span><span class="o">&lt;=</span><span class="nb">input</span><span class="o">.</span><span class="n">no_of_pc</span><span class="p">:</span>
        <span class="n">flag_correct</span><span class="o">=</span><span class="mi">1</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Factor ID is wrong&quot;</span><span class="p">)</span>
        <span class="n">flag_correct</span><span class="o">=</span><span class="mi">0</span>

    <span class="n">df</span><span class="o">=</span><span class="mi">0</span>
    <span class="k">if</span> <span class="n">flag_correct</span><span class="o">==</span><span class="mi">1</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">fi</span> <span class="ow">in</span> <span class="n">perform</span><span class="p">:</span>
            <span class="n">clid</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clusterid</span><span class="p">[</span><span class="n">fi</span><span class="p">]</span>
            <span class="n">spearman_factors</span><span class="p">,</span><span class="n">CC_PCA</span><span class="p">,</span><span class="n">CC_gene</span><span class="p">,</span><span class="n">CC_meanExpression</span><span class="p">,</span><span class="n">CC_popExpression</span><span class="p">,</span><span class="n">cosine_factors</span><span class="p">,</span><span class="n">alpha</span><span class="o">=</span><span class="n">PCA_of_sc_cluster_accordingto_spatial_clusterid</span><span class="p">[</span><span class="n">clid</span><span class="p">]</span>
            <span class="n">CC_celltype_name</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clustername</span><span class="p">[</span><span class="n">fi</span><span class="p">]</span>
            <span class="n">mu</span><span class="o">=</span><span class="n">CC_meanExpression</span>
            <span class="n">pop</span><span class="o">=</span><span class="n">CC_popExpression</span>

            <span class="c1">#for j in range(input.no_of_pc):</span>
            <span class="k">if</span> <span class="kc">True</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">correlation_with_spearman</span><span class="p">:</span>
                    <span class="n">source</span><span class="o">=</span><span class="n">spearman_factors</span><span class="p">[:,</span><span class="n">choose_factor_id</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">source</span><span class="o">=</span><span class="n">cosine_factors</span><span class="p">[:,</span><span class="n">choose_factor_id</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
                <span class="n">ind</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="o">-</span><span class="n">source</span><span class="p">)</span>
                <span class="n">interestofGene</span><span class="o">=</span><span class="p">[]</span>
                <span class="n">value_fact</span><span class="o">=</span><span class="p">[]</span>
                <span class="n">value_pop</span><span class="o">=</span><span class="p">[]</span>
                <span class="n">value_avgexp</span><span class="o">=</span><span class="p">[]</span>
                <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">source</span><span class="p">)):</span>
                    <span class="n">temp</span><span class="o">=</span><span class="n">CC_gene</span><span class="p">[</span><span class="n">ind</span><span class="p">[</span><span class="n">k</span><span class="p">]]</span>
                    <span class="k">if</span> <span class="n">rps_rpl_mt_genes_included</span><span class="p">:</span>
                        <span class="n">flag</span><span class="o">=</span><span class="mi">1</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">flag</span><span class="o">=</span><span class="mi">1</span>
                        <span class="k">if</span> <span class="n">temp</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">3</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;Rps&#39;</span><span class="p">:</span>
                            <span class="n">flag</span><span class="o">=</span><span class="mi">0</span>
                        <span class="k">if</span> <span class="n">temp</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">3</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;Rpl&#39;</span><span class="p">:</span>
                            <span class="n">flag</span><span class="o">=</span><span class="mi">0</span>
                        <span class="k">if</span> <span class="n">temp</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">3</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;mt-&#39;</span><span class="p">:</span>
                            <span class="n">flag</span><span class="o">=</span><span class="mi">0</span>
                    <span class="k">if</span> <span class="n">flag</span><span class="o">==</span><span class="mi">1</span><span class="p">:</span>
                        <span class="n">interestofGene</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">CC_gene</span><span class="p">[</span><span class="n">ind</span><span class="p">[</span><span class="n">k</span><span class="p">]])</span>
                        <span class="n">value_fact</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">source</span><span class="p">[</span><span class="n">ind</span><span class="p">[</span><span class="n">k</span><span class="p">]])</span>
                        <span class="n">value_pop</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">pop</span><span class="p">[</span><span class="n">ind</span><span class="p">[</span><span class="n">k</span><span class="p">]])</span>
                        <span class="n">value_avgexp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mu</span><span class="p">[</span><span class="n">ind</span><span class="p">[</span><span class="n">k</span><span class="p">]])</span>

                <span class="n">value_fact</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">value_fact</span><span class="p">)</span>
                <span class="n">value_pop</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">value_pop</span><span class="p">)</span>
                <span class="n">value_avgexp</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">value_avgexp</span><span class="p">)</span>

                <span class="n">interestofGene</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">interestofGene</span><span class="p">)</span>
                <span class="n">index_pos</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="o">-</span><span class="n">value_fact</span><span class="p">)</span>
                <span class="n">index_neg</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">value_fact</span><span class="p">)</span>

                <span class="n">gp1</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">interestofGene</span><span class="p">[</span><span class="n">index_pos</span><span class="p">])</span>
                <span class="n">gn1</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">interestofGene</span><span class="p">[</span><span class="n">index_neg</span><span class="p">])</span>

                <span class="n">gex</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">top_NOG</span><span class="p">,</span><span class="mi">1</span><span class="p">),</span><span class="n">dtype</span><span class="o">=</span><span class="nb">float</span><span class="p">)</span>

                <span class="n">vp1</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">value_fact</span><span class="p">[</span><span class="n">index_pos</span><span class="p">])[</span><span class="mi">0</span><span class="p">:</span><span class="n">top_NOG</span><span class="p">]</span>
                <span class="n">vn1</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">value_fact</span><span class="p">[</span><span class="n">index_neg</span><span class="p">])[</span><span class="mi">0</span><span class="p">:</span><span class="n">top_NOG</span><span class="p">]</span>
                <span class="n">pos_pop1</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">value_pop</span><span class="p">[</span><span class="n">index_pos</span><span class="p">])[</span><span class="mi">0</span><span class="p">:</span><span class="n">top_NOG</span><span class="p">]</span>
                <span class="n">neg_pop1</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">value_pop</span><span class="p">[</span><span class="n">index_neg</span><span class="p">])[</span><span class="mi">0</span><span class="p">:</span><span class="n">top_NOG</span><span class="p">]</span>
                <span class="n">pos_avg1</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">value_avgexp</span><span class="p">[</span><span class="n">index_pos</span><span class="p">])[</span><span class="mi">0</span><span class="p">:</span><span class="n">top_NOG</span><span class="p">]</span>
                <span class="n">neg_avg1</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">value_avgexp</span><span class="p">[</span><span class="n">index_neg</span><span class="p">])[</span><span class="mi">0</span><span class="p">:</span><span class="n">top_NOG</span><span class="p">]</span>

                <span class="k">if</span> <span class="n">positively_correlated</span><span class="p">:</span>
                    <span class="n">nvr1</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">hstack</span><span class="p">((</span><span class="n">np</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">vp1</span><span class="p">,(</span><span class="nb">len</span><span class="p">(</span><span class="n">vp1</span><span class="p">),</span><span class="mi">1</span><span class="p">)),</span><span class="n">gex</span><span class="p">))</span>
                    <span class="n">nvr2</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">hstack</span><span class="p">((</span><span class="n">np</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">pos_pop1</span><span class="p">,(</span><span class="nb">len</span><span class="p">(</span><span class="n">pos_pop1</span><span class="p">),</span><span class="mi">1</span><span class="p">)),</span><span class="n">gex</span><span class="p">))</span>
                    <span class="n">nvr3</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">hstack</span><span class="p">((</span><span class="n">np</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">pos_avg1</span><span class="p">,(</span><span class="nb">len</span><span class="p">(</span><span class="n">pos_avg1</span><span class="p">),</span><span class="mi">1</span><span class="p">)),</span><span class="n">gex</span><span class="p">))</span>
                    <span class="n">comgene</span><span class="o">=</span><span class="n">gp1</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="n">top_NOG</span><span class="p">]</span>
                    <span class="n">title</span><span class="o">=</span><span class="s1">&#39;Pos Fa&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">choose_factor_id</span><span class="p">)</span>

                    <span class="n">d</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;Gene&#39;</span><span class="p">:</span> <span class="n">comgene</span><span class="p">,</span> <span class="s1">&#39;Fa&#39;</span><span class="p">:</span> <span class="n">vp1</span><span class="p">,</span><span class="s1">&#39;mean_expression&#39;</span><span class="p">:</span><span class="n">pos_avg1</span><span class="p">,</span><span class="s1">&#39;proportion_of_population_expressed&#39;</span><span class="p">:</span><span class="n">pos_pop1</span><span class="p">}</span>
                    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">d</span><span class="p">)</span>
                    <span class="n">df</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;Gene&#39;</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">nvr1</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">hstack</span><span class="p">((</span><span class="n">np</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">vn1</span><span class="p">,(</span><span class="nb">len</span><span class="p">(</span><span class="n">vn1</span><span class="p">),</span><span class="mi">1</span><span class="p">)),</span><span class="n">gex</span><span class="p">))</span>
                    <span class="n">nvr2</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">hstack</span><span class="p">((</span><span class="n">np</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">neg_pop1</span><span class="p">,(</span><span class="nb">len</span><span class="p">(</span><span class="n">neg_pop1</span><span class="p">),</span><span class="mi">1</span><span class="p">)),</span><span class="n">gex</span><span class="p">))</span>
                    <span class="n">nvr3</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">hstack</span><span class="p">((</span><span class="n">np</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">neg_avg1</span><span class="p">,(</span><span class="nb">len</span><span class="p">(</span><span class="n">neg_avg1</span><span class="p">),</span><span class="mi">1</span><span class="p">)),</span><span class="n">gex</span><span class="p">))</span>
                    <span class="n">comgene</span><span class="o">=</span><span class="n">gn1</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="n">top_NOG</span><span class="p">]</span>
                    <span class="n">title</span><span class="o">=</span><span class="s1">&#39;Neg Fa&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">choose_factor_id</span><span class="p">)</span>
                    <span class="n">d</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;Gene&#39;</span><span class="p">:</span> <span class="n">comgene</span><span class="p">,</span> <span class="s1">&#39;Fa&#39;</span><span class="p">:</span> <span class="n">vn1</span><span class="p">,</span><span class="s1">&#39;mean_expression&#39;</span><span class="p">:</span><span class="n">neg_avg1</span><span class="p">,</span><span class="s1">&#39;proportion_of_population_expressed&#39;</span><span class="p">:</span><span class="n">neg_pop1</span><span class="p">}</span>
                    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">d</span><span class="p">)</span>
                    <span class="n">df</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;Gene&#39;</span><span class="p">)</span>


            <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">)</span>
            <span class="n">x</span><span class="p">,</span><span class="n">y</span><span class="p">,</span><span class="n">z</span><span class="p">,</span><span class="n">bigs</span><span class="o">=</span><span class="n">findXYZC</span><span class="p">(</span><span class="n">nvr1</span><span class="p">,</span><span class="n">nvr2</span><span class="p">)</span>
            <span class="n">p0</span><span class="o">=</span><span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">x</span><span class="p">,</span><span class="n">y</span><span class="p">,</span><span class="n">s</span><span class="o">=</span><span class="n">bigs</span><span class="p">,</span><span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">,</span><span class="n">c</span><span class="o">=</span><span class="n">z</span><span class="p">,</span><span class="n">cmap</span><span class="o">=</span><span class="n">cmap</span><span class="p">)</span> <span class="c1">#&#39;cm.cmap_name</span>
            <span class="n">x</span><span class="p">,</span><span class="n">y</span><span class="p">,</span><span class="n">z</span><span class="p">,</span><span class="n">bigs</span><span class="o">=</span><span class="n">findXYZC</span><span class="p">(</span><span class="n">nvr3</span><span class="p">,</span><span class="n">nvr2</span><span class="p">)</span>
            <span class="n">p1</span><span class="o">=</span><span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">x</span><span class="p">,</span><span class="n">y</span><span class="p">,</span><span class="n">s</span><span class="o">=</span><span class="n">bigs</span><span class="p">,</span><span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">,</span><span class="n">c</span><span class="o">=</span><span class="n">z</span><span class="p">,</span><span class="n">cmap</span><span class="o">=</span><span class="n">cmap</span><span class="p">)</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">colorbar</span><span class="p">(</span><span class="n">p0</span><span class="p">,</span><span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="n">shrink</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">colorbar</span><span class="p">(</span><span class="n">p1</span><span class="p">,</span><span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span><span class="n">shrink</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
            <span class="n">kw</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">prop</span><span class="o">=</span><span class="s2">&quot;sizes&quot;</span><span class="p">,</span> <span class="n">num</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.6</span><span class="p">,</span> <span class="n">fmt</span><span class="o">=</span><span class="s2">&quot;% </span><span class="si">{x:.0f}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">legend2</span> <span class="o">=</span> <span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="o">*</span><span class="n">p0</span><span class="o">.</span><span class="n">legend_elements</span><span class="p">(</span><span class="o">**</span><span class="n">kw</span><span class="p">),</span><span class="n">loc</span><span class="o">=</span><span class="s2">&quot;lower center&quot;</span><span class="p">,</span> <span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mf">0.25</span><span class="p">,</span> <span class="o">-</span><span class="mf">0.25</span><span class="p">),</span><span class="n">title</span><span class="o">=</span><span class="s2">&quot;Fraction of cells expressed&quot;</span><span class="p">,</span><span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="c1">#</span>
            <span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">title</span><span class="p">)</span>
            <span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Avg expression&#39;</span><span class="p">)</span>

            <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">2</span><span class="p">):</span>
                <span class="n">ax</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">nvr1</span><span class="p">)))</span>
                <span class="n">ax</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticklabels</span><span class="p">(</span><span class="n">comgene</span><span class="p">,</span><span class="n">style</span><span class="o">=</span><span class="s1">&#39;italic&#39;</span><span class="p">)</span>
                <span class="n">ax</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span><span class="c1">#range(1))</span>
                <span class="n">ax</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">set_xticklabels</span><span class="p">([])</span><span class="c1">#xlabels[i],rotation=30)</span>
                <span class="n">ax</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">([</span><span class="o">-</span><span class="mf">0.5</span><span class="p">,</span><span class="mf">0.5</span><span class="p">])</span>
                <span class="n">ax</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">([</span><span class="o">-</span><span class="mf">0.5</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">nvr1</span><span class="p">)</span><span class="o">+</span><span class="mf">0.5</span><span class="p">])</span>


            <span class="c1">#create_subtitle(fig, grid[0, ::], CC_celltype_name+&#39; Spearman correlation&#39;)</span>
            <span class="c1">#create_subtitle(fig, grid[1, ::],  CC_celltype_name+&#39; log(avg expression)&#39;)</span>
            <span class="n">fig</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;The figures are saved: &quot;</span><span class="p">,</span> <span class="n">savefigdir</span><span class="o">+</span><span class="s1">&#39;Factors_&#39;</span><span class="o">+</span><span class="n">remove_extra_character_from_name</span><span class="p">(</span><span class="n">CC_celltype_name</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;.&#39;</span><span class="o">+</span><span class="n">saveas</span><span class="p">)</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">savefigdir</span><span class="o">+</span><span class="s1">&#39;Factors_&#39;</span><span class="o">+</span><span class="n">remove_extra_character_from_name</span><span class="p">(</span><span class="n">CC_celltype_name</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;.&#39;</span><span class="o">+</span><span class="n">saveas</span><span class="p">,</span><span class="n">bbox_inches</span><span class="o">=</span><span class="s1">&#39;tight&#39;</span><span class="p">,</span><span class="n">transparent</span><span class="o">=</span><span class="n">transparent_mode</span><span class="p">,</span><span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">showit</span><span class="p">:</span>
                <span class="k">pass</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="s1">&#39;all&#39;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">df</span></div>




<div class="viewcode-block" id="create_directory">
<a class="viewcode-back" href="../../nico_covariations.html#nico_covariations.Covariations.create_directory">[docs]</a>
<span class="k">def</span> <span class="nf">create_directory</span><span class="p">(</span><span class="n">outputFolder</span><span class="p">):</span>


<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create an empty directory.</span>

<span class="sd">    This function checks if a specified directory exists, and if not, it creates the directory.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    outputFolder : str</span>
<span class="sd">        The path of the directory to be created.</span>

<span class="sd">    Raises</span>
<span class="sd">    ------</span>
<span class="sd">    OSError</span>
<span class="sd">        If the directory cannot be created due to permission issues or other OS-related errors.</span>

<span class="sd">    Notes</span>
<span class="sd">    -----</span>
<span class="sd">    - If the directory already exists, no action is taken.</span>
<span class="sd">    - This function ensures that the directory path is available for subsequent file operations.</span>

<span class="sd">    Example</span>
<span class="sd">    -------</span>
<span class="sd">    &gt;&gt;&gt; create_directory(&#39;./new_out/&#39;)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">answer</span><span class="o">=</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isdir</span><span class="p">(</span><span class="n">outputFolder</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">answer</span><span class="o">==</span><span class="kc">True</span><span class="p">:</span>
        <span class="k">pass</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">os</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">outputFolder</span><span class="p">)</span></div>



<div class="viewcode-block" id="find_index">
<a class="viewcode-back" href="../../nico_covariations.html#nico_covariations.Covariations.find_index">[docs]</a>
<span class="k">def</span> <span class="nf">find_index</span><span class="p">(</span><span class="n">sp_genename</span><span class="p">,</span><span class="n">sc_genename</span><span class="p">):</span>


<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Find the common gene space submatrix between two modalities.</span>

<span class="sd">    This helper function is used within the `gene_covariation_analysis` function to identify the indices of common genes</span>
<span class="sd">    between two lists of gene names corresponding to spatial and scRNAseq modalities.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    sp_genename : list</span>
<span class="sd">        A list of gene names from the spatial modality.</span>

<span class="sd">    sc_genename : list</span>
<span class="sd">        A list of gene names from the scRNAseq modality.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    list</span>
<span class="sd">        A list of indices corresponding to the common genes found in both `sp_genename` and `sc_genename`.</span>

<span class="sd">    Example</span>
<span class="sd">    -------</span>
<span class="sd">    &gt;&gt;&gt; sp_genes = [&#39;gene1&#39;, &#39;gene2&#39;, &#39;gene3&#39;, &#39;gene4&#39;]</span>
<span class="sd">    &gt;&gt;&gt; sc_genes = [&#39;gene3&#39;, &#39;gene4&#39;, &#39;gene5&#39;, &#39;gene6&#39;]</span>
<span class="sd">    &gt;&gt;&gt; index_sp,index_sc = find_index(sp_genes, sc_genes)</span>
<span class="sd">    &gt;&gt;&gt; print(index_sp)</span>
<span class="sd">    [2, 3]</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">index_sc</span><span class="o">=</span><span class="p">[]</span>
    <span class="n">index_sp</span><span class="o">=</span><span class="p">[]</span>
    <span class="n">d</span><span class="o">=</span><span class="p">{}</span>
    <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">sc_genename</span><span class="p">)):</span>
        <span class="n">name</span><span class="o">=</span><span class="n">sc_genename</span><span class="p">[</span><span class="n">j</span><span class="p">]</span>
        <span class="n">d</span><span class="p">[</span><span class="n">name</span><span class="p">]</span><span class="o">=</span><span class="n">j</span>

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">sp_genename</span><span class="p">)):</span>
        <span class="n">name</span><span class="o">=</span><span class="n">sp_genename</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">d</span><span class="p">[</span><span class="n">name</span><span class="p">]</span>
            <span class="n">flag</span><span class="o">=</span><span class="mi">1</span>
        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="n">flag</span><span class="o">=</span><span class="mi">0</span>
        <span class="k">if</span> <span class="n">flag</span><span class="o">==</span><span class="mi">1</span><span class="p">:</span>
            <span class="n">index_sc</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">d</span><span class="p">[</span><span class="n">name</span><span class="p">])</span>
            <span class="n">index_sp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">index_sp</span><span class="p">,</span><span class="n">index_sc</span></div>


<div class="viewcode-block" id="read_spatial_data">
<a class="viewcode-back" href="../../nico_covariations.html#nico_covariations.Covariations.read_spatial_data">[docs]</a>
<span class="k">def</span> <span class="nf">read_spatial_data</span><span class="p">(</span><span class="n">clusterFilename</span><span class="p">,</span><span class="n">celltypeFilename</span><span class="p">):</span>


<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Read the cluster information for spatial data.</span>

<span class="sd">    This helper function is used within the `gene_covariation_analysis` function to read the cluster</span>
<span class="sd">    and cell type information from the specified files.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    clusterFilename : str</span>
<span class="sd">        The file path of the cluster information file.</span>

<span class="sd">    celltypeFilename : str</span>
<span class="sd">        The file path of the cell type information file.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">df</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">celltypeFilename</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span><span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
    <span class="n">data</span><span class="o">=</span><span class="n">df</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span>
    <span class="n">spatialcell_unique_clustername</span><span class="o">=</span><span class="n">data</span><span class="p">[:,</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">spatialcell_unique_clusterid</span><span class="o">=</span><span class="n">data</span><span class="p">[:,</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">CTname</span><span class="o">=</span><span class="n">spatialcell_unique_clustername</span>
    <span class="n">CTid</span><span class="o">=</span><span class="n">spatialcell_unique_clusterid</span>

    <span class="n">df</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">clusterFilename</span><span class="p">)</span>
    <span class="n">louvainFull</span><span class="o">=</span><span class="n">df</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span>


    <span class="n">celltype</span><span class="o">=</span><span class="p">{}</span>
    <span class="n">cellsinCT</span><span class="o">=</span><span class="p">{}</span>
    <span class="n">index</span><span class="o">=</span><span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">louvainFull</span><span class="p">)):</span>
        <span class="n">clu_id</span><span class="o">=</span><span class="n">louvainFull</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">cel_id</span><span class="o">=</span><span class="n">louvainFull</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">clu_id</span> <span class="ow">in</span> <span class="n">CTid</span><span class="p">:</span>
            <span class="n">index</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
            <span class="c1">#celltype[cel_id]=clu_id</span>
            <span class="k">if</span> <span class="n">clu_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">cellsinCT</span><span class="p">:</span>
                <span class="n">cellsinCT</span><span class="p">[</span><span class="n">clu_id</span><span class="p">]</span><span class="o">=</span><span class="p">[</span><span class="n">cel_id</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">cellsinCT</span><span class="p">[</span><span class="n">clu_id</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cel_id</span><span class="p">)</span>

    <span class="n">louvain</span><span class="o">=</span><span class="n">louvainFull</span><span class="p">[</span><span class="n">index</span><span class="p">,:]</span>
    <span class="n">annotation_spatial_barcode_id</span><span class="o">=</span> <span class="n">louvain</span><span class="p">[:,</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">annotation_spatial_cluster_id</span><span class="o">=</span> <span class="n">louvain</span><span class="p">[:,</span><span class="mi">1</span><span class="p">]</span>

    <span class="n">d</span><span class="o">=</span><span class="p">{}</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">spatialcell_unique_clustername</span><span class="p">)):</span>
        <span class="n">d</span><span class="p">[</span><span class="n">spatialcell_unique_clusterid</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span><span class="o">=</span><span class="n">spatialcell_unique_clustername</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
    <span class="n">annotation_spatial_celltypename</span><span class="o">=</span><span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">annotation_spatial_cluster_id</span><span class="p">)):</span>
        <span class="n">annotation_spatial_celltypename</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">d</span><span class="p">[</span><span class="n">annotation_spatial_cluster_id</span><span class="p">[</span><span class="n">i</span><span class="p">]])</span>
    <span class="n">annotation_spatial_celltypename</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">annotation_spatial_celltypename</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">annotation_spatial_celltypename</span><span class="p">,</span><span class="n">annotation_spatial_barcode_id</span><span class="p">,</span><span class="n">annotation_spatial_cluster_id</span><span class="p">,</span><span class="n">spatialcell_unique_clustername</span><span class="p">,</span><span class="n">spatialcell_unique_clusterid</span></div>



<div class="viewcode-block" id="find_correlation_bw_genes_and_PC_component_in_singlecell">
<a class="viewcode-back" href="../../nico_covariations.html#nico_covariations.Covariations.find_correlation_bw_genes_and_PC_component_in_singlecell">[docs]</a>
<span class="k">def</span> <span class="nf">find_correlation_bw_genes_and_PC_component_in_singlecell</span><span class="p">(</span><span class="n">KcomponentCluster</span><span class="p">,</span><span class="n">clusterExpression</span><span class="p">):</span>


<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Calculate Spearman correlation between genes and principal components in single-cell data.</span>

<span class="sd">    This helper function is used within the `find_PC_of_invidualCluster_in_SC` function to determine</span>
<span class="sd">    the Spearman correlation between common gene scRNAseq factors (principal components) and scRNAseq</span>
<span class="sd">    gene expression data.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    KcomponentCluster : numpy.ndarray or pandas.DataFrame</span>
<span class="sd">        The matrix representing the principal components (factors) from scRNAseq data. Each column</span>
<span class="sd">        corresponds to a principal component.</span>

<span class="sd">    clusterExpression : numpy.ndarray or pandas.DataFrame</span>
<span class="sd">        The matrix representing the gene expression data from scRNAseq. Each row corresponds to a gene</span>
<span class="sd">        and each column corresponds to a cell.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">mat</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">clusterExpression</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span><span class="n">KcomponentCluster</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]),</span><span class="n">dtype</span><span class="o">=</span><span class="nb">float</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">clusterExpression</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]):</span>
        <span class="n">v1</span><span class="o">=</span><span class="n">clusterExpression</span><span class="p">[:,</span><span class="n">i</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">KcomponentCluster</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]):</span>
            <span class="n">v2</span><span class="o">=</span><span class="n">KcomponentCluster</span><span class="p">[:,</span><span class="n">j</span><span class="p">]</span>
            <span class="c1">#corr,_ = pearsonr(v1,v2)</span>
            <span class="n">corr</span><span class="p">,</span><span class="n">_</span> <span class="o">=</span><span class="n">spearmanr</span><span class="p">(</span><span class="n">v1</span><span class="p">,</span><span class="n">v2</span><span class="p">)</span>
            <span class="c1">#corr=np.corrcoef(v1,v2)</span>
            <span class="n">mat</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="n">j</span><span class="p">]</span><span class="o">=</span><span class="n">corr</span>

    <span class="c1"># mat shape is (# of genes x # of pc) it is a correlation between (PC and genes) of the single cell cluster</span>
    <span class="c1"># KcomponentCluster shape is (# of single cell in a single cell cluster x # of pc)</span>
    <span class="c1"># clusterExpression shape is (# of single cell in a single cell cluster x # of genes)</span>
    <span class="n">mat</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">nan_to_num</span><span class="p">(</span><span class="n">mat</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">mat</span></div>


<div class="viewcode-block" id="find_correlation_bw_genes_and_PC_component_in_singlecell_cosine">
<a class="viewcode-back" href="../../nico_covariations.html#nico_covariations.Covariations.find_correlation_bw_genes_and_PC_component_in_singlecell_cosine">[docs]</a>
<span class="k">def</span> <span class="nf">find_correlation_bw_genes_and_PC_component_in_singlecell_cosine</span><span class="p">(</span><span class="n">KcomponentCluster</span><span class="p">,</span><span class="n">clusterExpression</span><span class="p">):</span>


<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Calculate cosine similarity between common gene scRNAseq factors and scRNAseq count data.</span>

<span class="sd">    This helper function is used within the `find_PC_of_invidualCluster_in_SC` function to determine</span>
<span class="sd">    the cosine similarity between common gene scRNAseq factors (principal components) and scRNAseq</span>
<span class="sd">    gene expression data.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    KcomponentCluster : numpy.ndarray or pandas.DataFrame</span>
<span class="sd">        The matrix representing the principal components (factors) from scRNAseq data. Each column</span>
<span class="sd">        corresponds to a principal component.</span>

<span class="sd">    clusterExpression : numpy.ndarray or pandas.DataFrame</span>
<span class="sd">        The matrix representing the gene expression data from scRNAseq. Each row corresponds to a gene</span>
<span class="sd">        and each column corresponds to a cell.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    numpy.ndarray</span>
<span class="sd">        A matrix containing the cosine similarity scores between each gene and each principal component.</span>
<span class="sd">        Each row corresponds to a gene, and each column corresponds to a principal component.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1">#same vector =1 perpendicular vector 0</span>
    <span class="c1">#print(KcomponentCluster.shape,clusterExpression.shape)</span>
    <span class="c1">#mat=np.zeros((clusterExpression.shape[1],KcomponentCluster.shape[1]),dtype=float)</span>
    <span class="n">mat</span><span class="o">=</span><span class="n">cosine_similarity</span><span class="p">(</span><span class="n">clusterExpression</span><span class="o">.</span><span class="n">T</span><span class="p">,</span><span class="n">KcomponentCluster</span><span class="o">.</span><span class="n">T</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">mat</span></div>




<div class="viewcode-block" id="top_genes_in_correlation_list_without">
<a class="viewcode-back" href="../../nico_covariations.html#nico_covariations.Covariations.top_genes_in_correlation_list_without">[docs]</a>
<span class="k">def</span> <span class="nf">top_genes_in_correlation_list_without</span><span class="p">(</span><span class="n">genename</span><span class="p">,</span><span class="n">corr_NMFfactors_genes</span><span class="p">,</span><span class="n">n_top_words</span><span class="p">):</span>


<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Identify top genes associated with NMF factors, excluding duplicates.</span>

<span class="sd">    This helper function sorts the factor values and selects the top genes</span>
<span class="sd">    associated with each factor. It is used in `plot_cosine_and_spearman_correlation_to_factors`.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    genename : numpy.ndarray or pandas.Series</span>
<span class="sd">        Array or Series containing gene names.</span>

<span class="sd">    corr_NMFfactors_genes : numpy.ndarray or pandas.DataFrame</span>
<span class="sd">        The matrix representing the correlation values between genes and NMF factors.</span>
<span class="sd">        Each row corresponds to a gene, and each column corresponds to an NMF factor.</span>

<span class="sd">    n_top_words : int</span>
<span class="sd">        The number of top genes to retrieve for each NMF factor.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    gname : numpy.ndarray</span>
<span class="sd">        Array containing the names of the top genes associated with the NMF factors.</span>

<span class="sd">    mat : numpy.ndarray</span>
<span class="sd">        Matrix containing the correlation values of the top genes associated with the NMF factors.</span>
<span class="sd">        Each row corresponds to a selected top gene, and each column corresponds to an NMF factor.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">top_genes_assoc_factors</span><span class="o">=</span><span class="p">[]</span>
    <span class="k">for</span> <span class="n">topic_idx</span><span class="p">,</span> <span class="n">topic</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">corr_NMFfactors_genes</span><span class="o">.</span><span class="n">T</span><span class="p">):</span>
        <span class="n">top_features_ind</span> <span class="o">=</span> <span class="n">topic</span><span class="o">.</span><span class="n">argsort</span><span class="p">()[:</span> <span class="o">-</span><span class="n">n_top_words</span> <span class="o">-</span> <span class="mi">1</span> <span class="p">:</span> <span class="o">-</span><span class="mi">1</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">top_features_ind</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">i</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">top_genes_assoc_factors</span><span class="p">:</span>
                <span class="n">top_genes_assoc_factors</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
    <span class="n">gname</span><span class="o">=</span><span class="n">genename</span><span class="p">[</span><span class="n">top_genes_assoc_factors</span><span class="p">]</span>
    <span class="n">mat</span><span class="o">=</span><span class="n">corr_NMFfactors_genes</span><span class="p">[</span><span class="n">top_genes_assoc_factors</span><span class="p">,:]</span>

    <span class="k">return</span> <span class="n">gname</span><span class="p">,</span><span class="n">mat</span></div>





<div class="viewcode-block" id="alignment_score">
<a class="viewcode-back" href="../../nico_covariations.html#nico_covariations.Covariations.alignment_score">[docs]</a>
<span class="k">def</span> <span class="nf">alignment_score</span><span class="p">(</span><span class="n">H</span><span class="p">,</span><span class="n">spH</span><span class="p">,</span><span class="n">ind_H</span><span class="p">,</span><span class="n">ind_spH</span><span class="p">):</span>


<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Calculate the alignment score between factors from two different modalities during integrated NMF.</span>

<span class="sd">    This helper function is used in `find_PC_of_invidualCluster_in_SC` to evaluate the alignment</span>
<span class="sd">    score between factors from scRNAseq data and spatial data.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    H : numpy.ndarray</span>
<span class="sd">        The matrix representing the factors from the scRNAseq data.</span>
<span class="sd">        Each row corresponds to a sample, and each column corresponds to a factor.</span>

<span class="sd">    spH : numpy.ndarray</span>
<span class="sd">        The matrix representing the factors from the spatial data.</span>
<span class="sd">        Each row corresponds to a sample, and each column corresponds to a factor.</span>

<span class="sd">    ind_H : numpy.ndarray or list</span>
<span class="sd">        Indices of the common genes in the scRNAseq data.</span>

<span class="sd">    ind_spH : numpy.ndarray or list</span>
<span class="sd">        Indices of the common genes in the spatial data.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    float</span>
<span class="sd">        The alignment score between the factors from the scRNAseq and spatial data.</span>

<span class="sd">    Notes</span>
<span class="sd">    -----</span>
<span class="sd">    The alignment score is calculated by computing the cosine similarity between the factors of</span>
<span class="sd">    the common genes in the scRNAseq and spatial data. A higher score indicates better alignment</span>
<span class="sd">    between the factors from the two modalities.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1">#print(H.shape,spH.shape,len(ind_H),len(ind_spH))</span>
    <span class="n">r1</span><span class="o">=</span><span class="n">H</span><span class="p">[:,</span><span class="n">ind_H</span><span class="p">]</span>
    <span class="n">r2</span><span class="o">=</span><span class="n">spH</span><span class="p">[:,</span><span class="n">ind_spH</span><span class="p">]</span>
    <span class="n">comb</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">hstack</span><span class="p">((</span><span class="n">r1</span><span class="p">,</span><span class="n">r2</span><span class="p">))</span><span class="o">.</span><span class="n">T</span>
    <span class="n">n</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">ind_H</span><span class="p">)</span>
    <span class="n">knn</span><span class="o">=</span><span class="nb">max</span><span class="p">([</span><span class="mi">2</span><span class="p">,</span><span class="n">np</span><span class="o">.</span><span class="n">ceil</span><span class="p">(</span><span class="mf">0.01</span><span class="o">*</span><span class="n">n</span><span class="p">)</span> <span class="p">])</span>
    <span class="n">n_jobs</span><span class="o">=-</span><span class="mi">1</span>
    <span class="n">k_d</span><span class="p">,</span><span class="n">k_ind</span> <span class="o">=</span> <span class="n">cKDTree</span><span class="p">(</span><span class="n">comb</span><span class="p">)</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">comb</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="n">knn</span><span class="p">,</span> <span class="n">workers</span><span class="o">=</span><span class="n">n_jobs</span><span class="p">)</span>

    <span class="n">avgc1</span><span class="o">=</span><span class="mi">0</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="p">):</span>
        <span class="n">neigh</span><span class="o">=</span><span class="n">k_ind</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
        <span class="n">c1</span><span class="o">=</span><span class="mi">0</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">neigh</span><span class="p">)):</span>
            <span class="k">if</span> <span class="n">neigh</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">&lt;</span><span class="n">n</span><span class="p">:</span>
                <span class="n">c1</span><span class="o">=</span><span class="n">c1</span><span class="o">+</span><span class="mi">1</span>
        <span class="n">avgc1</span><span class="o">=</span><span class="n">avgc1</span><span class="o">+</span><span class="n">c1</span>
    <span class="n">avgc1</span><span class="o">=</span><span class="n">avgc1</span><span class="o">/</span><span class="n">n</span>
    <span class="c1">#doi:10.1038/nbt.4096</span>
    <span class="n">score</span><span class="o">=</span><span class="mi">1</span> <span class="o">-</span> <span class="p">((</span><span class="n">avgc1</span> <span class="o">-</span> <span class="p">(</span><span class="n">knn</span><span class="o">/</span><span class="n">n</span><span class="p">)</span> <span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="n">knn</span> <span class="o">-</span> <span class="p">(</span><span class="n">knn</span><span class="o">/</span><span class="n">n</span><span class="p">)</span> <span class="p">))</span>

    <span class="k">return</span> <span class="n">score</span></div>


<div class="viewcode-block" id="multiplicative_method">
<a class="viewcode-back" href="../../nico_covariations.html#nico_covariations.Covariations.multiplicative_method">[docs]</a>
<span class="k">def</span> <span class="nf">multiplicative_method</span><span class="p">(</span><span class="n">W</span><span class="p">,</span><span class="n">H</span><span class="p">,</span><span class="n">A</span><span class="p">,</span><span class="n">max_iter</span><span class="p">):</span>

<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Perform conventional Non-negative Matrix Factorization (NMF) using a multiplicative update rule.</span>

<span class="sd">    This helper function is used in `find_PC_of_invidualCluster_in_SC` to decompose matrix `A` into</span>
<span class="sd">    two non-negative matrices `W` and `H` such that `A  W @ H`.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    W : ndarray</span>
<span class="sd">        Initial matrix representing the basis vectors. Shape(n_samples, n_components).</span>
<span class="sd">    H : ndarray</span>
<span class="sd">        Initial matrix representing the coefficients. Shape(n_components, n_features).</span>
<span class="sd">    A : ndarray</span>
<span class="sd">        The input data matrix to be factorized. Shape(n_samples, n_features).</span>
<span class="sd">    max_iter : int</span>
<span class="sd">        The maximum number of iterations for the multiplicative update algorithm.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    W : ndarray</span>
<span class="sd">        Updated basis matrix after NMF. Shape: (n_samples, n_components).</span>
<span class="sd">    H : ndarray</span>
<span class="sd">        Updated coefficient matrix after NMF. Shape: (n_components, n_features).</span>
<span class="sd">    norms : list</span>
<span class="sd">        List of Frobenius norms of the difference between `A` and `W @ H` for each iteration.</span>

<span class="sd">    Notes</span>
<span class="sd">    -----</span>
<span class="sd">    The update rules for `W` and `H` are based on minimizing the Frobenius norm of the difference between `A` and `W @ H`. The update for `H` is performed as:</span>

<span class="sd">    .. math::</span>
<span class="sd">        H_{ij} = H_{ij}  \frac{(W^T A)_{ij}}{(W^T W H)_{ij} + \epsilon}</span>

<span class="sd">    where `` is a small constant to prevent division by zero.</span>

<span class="sd">    The update for `W` has been commented out but follows a similar form. Uncomment the lines under &quot;Update W&quot; to perform updates for `W` as well.</span>

<span class="sd">    .. math::</span>
<span class="sd">        W_{ij} = W_{ij}  \frac{(A H^T)_{ij}}{(W H H^T)_{ij} + \epsilon}</span>

<span class="sd">    This method is sensitive to initializations of `W` and `H`, and the results may vary across runs.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">norms</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">e</span> <span class="o">=</span> <span class="mf">1.0e-10</span>
    <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">max_iter</span><span class="p">):</span>
        <span class="c1"># Update H</span>
        <span class="n">W_TA</span> <span class="o">=</span> <span class="n">W</span><span class="o">.</span><span class="n">T</span><span class="nd">@A</span>
        <span class="n">W_TWH</span> <span class="o">=</span> <span class="n">W</span><span class="o">.</span><span class="n">T</span><span class="nd">@W@H</span><span class="o">+</span><span class="n">e</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">size</span><span class="p">(</span><span class="n">H</span><span class="p">,</span> <span class="mi">0</span><span class="p">)):</span>
            <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">size</span><span class="p">(</span><span class="n">H</span><span class="p">,</span> <span class="mi">1</span><span class="p">)):</span>
                <span class="n">H</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="n">H</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">*</span> <span class="n">W_TA</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">/</span> <span class="n">W_TWH</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span>
        <span class="c1"># Update W</span>
        <span class="c1">#AH_T = A@H.T</span>
        <span class="c1">#WHH_T =  W@H@H.T+ e</span>
        <span class="c1">#for i in range(np.size(W, 0)):</span>
        <span class="c1">#    for j in range(np.size(W, 1)):</span>
        <span class="c1">#        W[i, j] = W[i, j] * AH_T[i, j] / WHH_T[i, j]</span>

        <span class="n">norm</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linalg</span><span class="o">.</span><span class="n">norm</span><span class="p">(</span><span class="n">A</span> <span class="o">-</span> <span class="n">W</span><span class="nd">@H</span><span class="p">,</span> <span class="s1">&#39;fro&#39;</span><span class="p">)</span>
        <span class="n">norms</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">norm</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">W</span> <span class="p">,</span><span class="n">H</span> <span class="p">,</span><span class="n">norms</span></div>



<div class="viewcode-block" id="remove_extra_character_from_name">
<a class="viewcode-back" href="../../nico_covariations.html#nico_covariations.Covariations.remove_extra_character_from_name">[docs]</a>
<span class="k">def</span> <span class="nf">remove_extra_character_from_name</span><span class="p">(</span><span class="n">name</span><span class="p">):</span>

<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Remove special characters from cell type names to avoid errors while saving figures.</span>

<span class="sd">    This function replaces certain special characters in the input `name` with</span>
<span class="sd">    underscores or other appropriate characters to ensure the name is safe for use</span>
<span class="sd">    as a filename.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    name : str</span>
<span class="sd">        The original cell type name that may contain special characters.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    str</span>
<span class="sd">        The modified cell type name with special characters removed or replaced.</span>

<span class="sd">    Example</span>
<span class="sd">    -------</span>
<span class="sd">    &gt;&gt;&gt; name = &#39;T-cell (CD4+)/CD8+&#39;</span>
<span class="sd">    &gt;&gt;&gt; clean_name = remove_extra_character_from_name(name)</span>
<span class="sd">    &gt;&gt;&gt; print(clean_name)</span>
<span class="sd">    &#39;T-cell_CD4p_CD8p&#39;</span>

<span class="sd">    Notes</span>
<span class="sd">    -----</span>
<span class="sd">    The following replacements are made:</span>

<span class="sd">        - &#39;/&#39; is replaced with &#39;_&#39;</span>
<span class="sd">        - &#39; &#39; (space) is replaced with &#39;_&#39;</span>
<span class="sd">        - &#39;&quot;&#39; (double quote) is removed</span>
<span class="sd">        - &quot;&#39;&quot; (single quote) is removed</span>
<span class="sd">        - &#39;)&#39; is removed</span>
<span class="sd">        - &#39;(&#39; is removed</span>
<span class="sd">        - &#39;+&#39; is replaced with &#39;p&#39;</span>
<span class="sd">        - &#39;-&#39; is replaced with &#39;n&#39;</span>
<span class="sd">        - &#39;.&#39; (dot) is removed</span>

<span class="sd">    These substitutions help in creating filenames that do not contain characters</span>
<span class="sd">    that might be problematic for file systems or software.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">name</span><span class="o">=</span><span class="n">name</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;/&#39;</span><span class="p">,</span><span class="s1">&#39;_&#39;</span><span class="p">)</span>
    <span class="n">name</span><span class="o">=</span><span class="n">name</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">,</span><span class="s1">&#39;_&#39;</span><span class="p">)</span>
    <span class="n">name</span><span class="o">=</span><span class="n">name</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;&quot;&#39;</span><span class="p">,</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
    <span class="n">name</span><span class="o">=</span><span class="n">name</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;&#39;&quot;</span><span class="p">,</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
    <span class="n">name</span><span class="o">=</span><span class="n">name</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;)&#39;</span><span class="p">,</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
    <span class="n">name</span><span class="o">=</span><span class="n">name</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;(&#39;</span><span class="p">,</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
    <span class="n">name</span><span class="o">=</span><span class="n">name</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;+&#39;</span><span class="p">,</span><span class="s1">&#39;p&#39;</span><span class="p">)</span>
    <span class="n">name</span><span class="o">=</span><span class="n">name</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">,</span><span class="s1">&#39;n&#39;</span><span class="p">)</span>
    <span class="n">name</span><span class="o">=</span><span class="n">name</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">,</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">name</span></div>



<div class="viewcode-block" id="find_PC_of_invidualCluster_in_SC">
<a class="viewcode-back" href="../../nico_covariations.html#nico_covariations.Covariations.find_PC_of_invidualCluster_in_SC">[docs]</a>
<span class="k">def</span> <span class="nf">find_PC_of_invidualCluster_in_SC</span><span class="p">(</span><span class="n">seed</span><span class="p">,</span><span class="n">spatial_integration_modality</span><span class="p">,</span><span class="n">scbarcode</span><span class="p">,</span><span class="n">iNMFmode</span><span class="p">,</span><span class="n">scadata</span><span class="p">,</span><span class="n">no_of_pc</span><span class="p">,</span><span class="n">spbarcode</span><span class="p">,</span><span class="n">spadata</span><span class="p">,</span> <span class="n">sct_ad_sc_full</span><span class="p">,</span><span class="n">celltype_name</span><span class="p">,</span><span class="n">cutoff_to_count_exp_cell_population</span><span class="p">):</span>

<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Helper function used in compute_PC_space to find principal components (PCs) for individual clusters in single-cell RNA sequencing (scRNA-seq) data and spatial transcriptomics data.</span>

<span class="sd">    This function integrates scRNA-seq and spatial transcriptomics data using non-negative matrix factorization (NMF) or integrative NMF (iNMF), and computes the alignment score, correlation, and other metrics for the identified principal components.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    seed : int</span>
<span class="sd">        Random seed for reproducibility.</span>
<span class="sd">    spatial_integration_modality : str</span>
<span class="sd">        Modality for spatial integration, either &#39;single&#39; or &#39;double&#39;.</span>
<span class="sd">    scbarcode : list</span>
<span class="sd">        List of single-cell barcodes.</span>
<span class="sd">    iNMFmode : bool</span>
<span class="sd">        Flag indicating whether to use iNMF (True) or not (False).</span>
<span class="sd">    scadata : AnnData</span>
<span class="sd">        Single-cell RNA-seq data in AnnData format.</span>
<span class="sd">    no_of_pc : int</span>
<span class="sd">        Number of principal components to compute.</span>
<span class="sd">    spbarcode : list</span>
<span class="sd">        List of spatial transcriptomics barcodes.</span>
<span class="sd">    spadata : AnnData</span>
<span class="sd">        Spatial transcriptomics data in AnnData format.</span>
<span class="sd">    sct_ad_sc_full : AnnData</span>
<span class="sd">        Full single-cell RNA-seq data in AnnData format.</span>
<span class="sd">    celltype_name : str</span>
<span class="sd">        Name of the cell type being analyzed.</span>
<span class="sd">    cutoff_to_count_exp_cell_population : float</span>
<span class="sd">        Expression cutoff to count the proportion of cell population expressing a gene.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    transfer_sp_com : ndarray</span>
<span class="sd">        Transformed spatial component matrix.</span>
<span class="sd">    transfer_sc_com : list</span>
<span class="sd">        Transformed single-cell component matrix (currently not populated).</span>
<span class="sd">    sc_spearman : ndarray</span>
<span class="sd">        Spearman correlation between genes and principal components in single-cell data.</span>
<span class="sd">    sc_cosine : ndarray</span>
<span class="sd">        Cosine similarity between genes and principal components in single-cell data.</span>
<span class="sd">    sc_genenames : ndarray</span>
<span class="sd">        Array of gene names.</span>
<span class="sd">    H : ndarray</span>
<span class="sd">        Principal component matrix for single-cell data.</span>
<span class="sd">    spH : ndarray</span>
<span class="sd">        Principal component matrix for spatial data.</span>
<span class="sd">    sc_cluster_mean_exp : ndarray</span>
<span class="sd">        Mean expression of genes across single-cell clusters.</span>
<span class="sd">    sc_cluster_exp_more_than_threshold : ndarray</span>
<span class="sd">        Proportion of single-cell clusters expressing genes above the cutoff.</span>
<span class="sd">    alpha : int</span>
<span class="sd">        Optimal alpha value used in iNMF.</span>

<span class="sd">    Notes</span>
<span class="sd">    -----</span>
<span class="sd">    - This function normalizes gene expression data and computes principal components using either NMF or iNMF.</span>
<span class="sd">    - It calculates the alignment score for spatial and single-cell data integration.</span>
<span class="sd">    - Spearman correlation and cosine similarity between genes and PCs are computed.</span>
<span class="sd">    - The results include the transformed spatial component matrix, gene correlations, and other metrics for downstream analysis.</span>

<span class="sd">    &quot;&quot;&quot;</span>


    <span class="n">cellname</span><span class="o">=</span><span class="n">sct_ad_sc_full</span><span class="o">.</span><span class="n">obs_names</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span>
    <span class="n">d</span><span class="o">=</span><span class="p">{}</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">cellname</span><span class="p">)):</span>
        <span class="n">d</span><span class="p">[</span><span class="n">cellname</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span><span class="o">=</span><span class="n">i</span>
    <span class="n">index</span><span class="o">=</span><span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">scbarcode</span><span class="p">)):</span>
        <span class="n">index</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">d</span><span class="p">[</span><span class="n">scbarcode</span><span class="p">[</span><span class="n">i</span><span class="p">]])</span>
    <span class="n">full_genes_sc</span><span class="o">=</span><span class="n">sct_ad_sc_full</span><span class="p">[</span><span class="n">index</span><span class="p">,:]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

    <span class="c1">#common gene single cell</span>
    <span class="n">cellname</span><span class="o">=</span><span class="n">scadata</span><span class="o">.</span><span class="n">obs_names</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span>
    <span class="n">d</span><span class="o">=</span><span class="p">{}</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">cellname</span><span class="p">)):</span>
        <span class="n">d</span><span class="p">[</span><span class="n">cellname</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span><span class="o">=</span><span class="n">i</span>
    <span class="n">index</span><span class="o">=</span><span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">scbarcode</span><span class="p">)):</span>
        <span class="n">index</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">d</span><span class="p">[</span><span class="n">scbarcode</span><span class="p">[</span><span class="n">i</span><span class="p">]])</span>
    <span class="n">sct_ad_sc</span><span class="o">=</span><span class="n">scadata</span><span class="p">[</span><span class="n">index</span><span class="p">,:]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

    <span class="c1">#common gene spatial</span>
    <span class="n">cellname</span><span class="o">=</span><span class="n">spadata</span><span class="o">.</span><span class="n">obs_names</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span>
    <span class="n">d</span><span class="o">=</span><span class="p">{}</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">cellname</span><span class="p">)):</span>
        <span class="n">d</span><span class="p">[</span><span class="n">cellname</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span><span class="o">=</span><span class="n">i</span>
    <span class="n">index</span><span class="o">=</span><span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">spbarcode</span><span class="p">)):</span>
        <span class="n">index</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">d</span><span class="p">[</span><span class="n">spbarcode</span><span class="p">[</span><span class="n">i</span><span class="p">]])</span>
    <span class="n">sct_ad_sp</span><span class="o">=</span><span class="n">spadata</span><span class="p">[</span><span class="n">index</span><span class="p">,:]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

    <span class="k">if</span> <span class="n">scipy_sparse</span><span class="o">.</span><span class="n">issparse</span><span class="p">(</span><span class="n">full_genes_sc</span><span class="o">.</span><span class="n">X</span><span class="p">):</span>
        <span class="n">CbyG</span><span class="o">=</span><span class="n">full_genes_sc</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">CbyG</span><span class="o">=</span><span class="n">full_genes_sc</span><span class="o">.</span><span class="n">X</span>

    <span class="c1">#print(&quot;B2 sp sc&quot;,np.sum(sct_ad_sp.X), np.sum(sct_ad_sc.X))</span>


    <span class="k">if</span> <span class="n">scipy_sparse</span><span class="o">.</span><span class="n">issparse</span><span class="p">(</span><span class="n">sct_ad_sc</span><span class="o">.</span><span class="n">X</span><span class="p">):</span>
        <span class="n">msc</span><span class="o">=</span><span class="n">sct_ad_sc</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">msc</span><span class="o">=</span><span class="n">sct_ad_sc</span><span class="o">.</span><span class="n">X</span>

    <span class="k">if</span> <span class="n">scipy_sparse</span><span class="o">.</span><span class="n">issparse</span><span class="p">(</span><span class="n">sct_ad_sp</span><span class="o">.</span><span class="n">X</span><span class="p">):</span>
        <span class="n">msp</span><span class="o">=</span><span class="n">sct_ad_sp</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">msp</span><span class="o">=</span><span class="n">sct_ad_sp</span><span class="o">.</span><span class="n">X</span>

    <span class="c1">#replace nan to zero</span>
    <span class="c1">#msp=np.nan_to_num(msp)</span>
    <span class="c1">#msc=np.nan_to_num(msc)</span>

    <span class="c1">#msc=msc/np.sum(msc)</span>
    <span class="c1">#msp=msp/np.sum(msp)</span>
    <span class="c1">#CbyG=CbyG/np.sum(CbyG)</span>
    <span class="n">genename_joint</span><span class="o">=</span><span class="n">sct_ad_sc</span><span class="o">.</span><span class="n">var_names</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span>
    <span class="n">genename_spatial</span><span class="o">=</span><span class="n">sct_ad_sp</span><span class="o">.</span><span class="n">var_names</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span>


    <span class="c1">#Gene based normalization</span>
    <span class="c1">#msc=np.log10(1+msc)</span>
    <span class="c1">#msp=np.log10(1+msp)</span>
    <span class="c1">#CbyG=np.log10(1+CbyG)</span>
    <span class="n">std1</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">std</span><span class="p">(</span><span class="n">msc</span><span class="p">,</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">std2</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">std</span><span class="p">(</span><span class="n">msp</span><span class="p">,</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">ind</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">((</span><span class="n">std1</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">)</span><span class="o">&amp;</span><span class="p">(</span><span class="n">std2</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">))</span>
    <span class="n">index</span><span class="o">=</span><span class="n">ind</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">n</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">index</span><span class="p">)</span>
    <span class="n">v1</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">msc</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="n">n</span><span class="p">),</span><span class="n">dtype</span><span class="o">=</span><span class="nb">float</span><span class="p">)</span>
    <span class="n">v2</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">msp</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="n">n</span><span class="p">),</span><span class="n">dtype</span><span class="o">=</span><span class="nb">float</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="p">):</span>
        <span class="n">v1</span><span class="p">[:,</span><span class="n">i</span><span class="p">]</span><span class="o">=</span><span class="n">msc</span><span class="p">[:,</span><span class="n">index</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span><span class="o">/</span><span class="n">std1</span><span class="p">[</span><span class="n">index</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span>
        <span class="n">v2</span><span class="p">[:,</span><span class="n">i</span><span class="p">]</span><span class="o">=</span><span class="n">msp</span><span class="p">[:,</span><span class="n">index</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span><span class="o">/</span><span class="n">std2</span><span class="p">[</span><span class="n">index</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span>
    <span class="c1">#sum1=np.std(v1,axis=0)</span>
    <span class="c1">#sum2=np.std(v2,axis=0)</span>

    <span class="n">datasets</span><span class="o">=</span><span class="p">[</span><span class="n">v1</span><span class="p">,</span><span class="n">v2</span><span class="p">]</span>
    <span class="n">n1</span><span class="o">=</span><span class="n">msc</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">n2</span><span class="o">=</span><span class="n">msp</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">threshold</span><span class="o">=</span><span class="mf">0.001</span>
    <span class="n">old_score</span><span class="o">=</span><span class="mi">1</span>

    <span class="k">if</span> <span class="n">spatial_integration_modality</span><span class="o">==</span><span class="s1">&#39;single&#39;</span><span class="p">:</span>
        <span class="n">alpha</span><span class="o">=</span><span class="mi">0</span>
        <span class="n">seed</span><span class="o">=</span><span class="n">seed</span><span class="o">+</span><span class="mi">1</span>
        <span class="n">model</span> <span class="o">=</span> <span class="n">NMF</span><span class="p">(</span><span class="n">n_components</span><span class="o">=</span><span class="n">no_of_pc</span><span class="p">,</span> <span class="n">init</span> <span class="o">=</span> <span class="s2">&quot;nndsvda&quot;</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="n">seed</span><span class="p">,</span><span class="n">beta_loss</span><span class="o">=</span><span class="s2">&quot;kullback-leibler&quot;</span><span class="p">,</span><span class="n">solver</span><span class="o">=</span><span class="s2">&quot;mu&quot;</span><span class="p">,</span><span class="n">max_iter</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span><span class="n">alpha_W</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span><span class="n">alpha_H</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span><span class="n">l1_ratio</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
        <span class="n">W</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">v1</span><span class="o">.</span><span class="n">T</span><span class="p">)</span>
        <span class="n">H</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">components_</span>
        <span class="n">spW</span><span class="o">=</span><span class="n">W</span>
        <span class="n">spH</span><span class="o">=</span><span class="n">H</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">celltype_name</span><span class="p">,</span><span class="s2">&quot;alpha, H size, W size, spH size:&quot;</span><span class="p">,</span> <span class="n">alpha</span><span class="p">,</span><span class="n">H</span><span class="o">.</span><span class="n">shape</span><span class="p">,</span><span class="n">W</span><span class="o">.</span><span class="n">shape</span><span class="p">,</span><span class="n">spH</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">spatial_integration_modality</span><span class="o">==</span><span class="s1">&#39;double&#39;</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">iNMFmode</span><span class="o">==</span><span class="kc">True</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">alpha</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">51</span><span class="p">,</span><span class="mi">2</span><span class="p">):</span>
                <span class="n">arr1</span><span class="o">=</span><span class="p">[</span><span class="o">*</span><span class="nb">range</span><span class="p">(</span><span class="n">n1</span><span class="p">)]</span>
                <span class="n">arr2</span><span class="o">=</span><span class="p">[</span><span class="o">*</span><span class="nb">range</span><span class="p">(</span><span class="n">n2</span><span class="p">)]</span>
                <span class="k">if</span> <span class="n">n1</span><span class="o">&gt;</span><span class="n">n2</span><span class="p">:</span>
                    <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">shuffle</span><span class="p">(</span><span class="n">arr1</span><span class="p">)</span>
                    <span class="n">arr1</span><span class="o">=</span><span class="n">arr1</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="n">n2</span><span class="p">]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">shuffle</span><span class="p">(</span><span class="n">arr2</span><span class="p">)</span>
                    <span class="n">arr2</span><span class="o">=</span><span class="n">arr2</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="n">n1</span><span class="p">]</span>

                <span class="n">seed</span><span class="o">=</span><span class="n">seed</span><span class="o">+</span><span class="mi">1</span>
                <span class="n">H</span><span class="p">,</span><span class="n">spH</span><span class="p">,</span><span class="n">W</span><span class="p">,</span><span class="n">V</span><span class="p">,</span><span class="n">spV</span> <span class="o">=</span> <span class="n">iNMF</span><span class="p">(</span><span class="n">datasets</span><span class="p">,</span><span class="n">no_of_pc</span><span class="p">,</span><span class="n">value_lambda</span><span class="o">=</span><span class="n">alpha</span><span class="p">,</span><span class="n">rand_seed</span><span class="o">=</span><span class="n">seed</span><span class="p">,</span><span class="n">print_obj</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
                <span class="n">spW</span><span class="o">=</span><span class="n">W</span>
                <span class="n">score</span><span class="o">=</span><span class="n">alignment_score</span><span class="p">(</span><span class="n">H</span><span class="p">,</span><span class="n">spH</span><span class="p">,</span><span class="n">arr1</span><span class="p">,</span><span class="n">arr2</span><span class="p">)</span>
                <span class="k">if</span> <span class="nb">abs</span><span class="p">(</span><span class="n">score</span><span class="o">-</span><span class="n">old_score</span><span class="p">)</span><span class="o">&lt;</span><span class="n">threshold</span><span class="p">:</span>
                    <span class="c1"># size1 is scRNAseq and size 2 is spatial</span>
                    <span class="nb">print</span><span class="p">(</span><span class="n">celltype_name</span><span class="p">,</span><span class="s2">&quot;alpha, H size, W size, spH size:&quot;</span><span class="p">,</span> <span class="n">alpha</span><span class="p">,</span><span class="n">H</span><span class="o">.</span><span class="n">shape</span><span class="p">,</span><span class="n">W</span><span class="o">.</span><span class="n">shape</span><span class="p">,</span><span class="n">spH</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
                    <span class="k">break</span>
                <span class="n">old_score</span><span class="o">=</span><span class="n">score</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">alpha</span><span class="o">=</span><span class="mi">0</span>
            <span class="n">seed</span><span class="o">=</span><span class="n">seed</span><span class="o">+</span><span class="mi">1</span>
            <span class="n">model</span> <span class="o">=</span> <span class="n">NMF</span><span class="p">(</span><span class="n">n_components</span><span class="o">=</span><span class="n">no_of_pc</span><span class="p">,</span> <span class="n">init</span> <span class="o">=</span> <span class="s2">&quot;nndsvda&quot;</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="n">seed</span><span class="p">,</span><span class="n">beta_loss</span><span class="o">=</span><span class="s2">&quot;kullback-leibler&quot;</span><span class="p">,</span><span class="n">solver</span><span class="o">=</span><span class="s2">&quot;mu&quot;</span><span class="p">,</span><span class="n">max_iter</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span><span class="n">alpha_W</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span><span class="n">alpha_H</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span><span class="n">l1_ratio</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
            <span class="n">W</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">v1</span><span class="o">.</span><span class="n">T</span><span class="p">)</span>
            <span class="n">H</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">components_</span>
            <span class="n">spW</span><span class="o">=</span><span class="n">W</span>
            <span class="n">spH</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">((</span><span class="n">no_of_pc</span><span class="p">,</span><span class="n">v2</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]),</span><span class="n">dtype</span><span class="o">=</span><span class="nb">float</span><span class="p">)</span>
            <span class="n">spW</span> <span class="p">,</span><span class="n">spH</span> <span class="p">,</span><span class="n">norms</span><span class="o">=</span><span class="n">multiplicative_method</span><span class="p">(</span><span class="n">spW</span><span class="p">,</span><span class="n">spH</span><span class="p">,</span><span class="n">v2</span><span class="o">.</span><span class="n">T</span><span class="p">,</span><span class="mi">200</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="n">celltype_name</span><span class="p">,</span><span class="s2">&quot;alpha, H size, W size, spH size:&quot;</span><span class="p">,</span> <span class="n">alpha</span><span class="p">,</span><span class="n">H</span><span class="o">.</span><span class="n">shape</span><span class="p">,</span><span class="n">W</span><span class="o">.</span><span class="n">shape</span><span class="p">,</span><span class="n">spH</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>



    <span class="n">entropy_H</span><span class="o">=</span><span class="s1">&#39;&#39;</span>
    <span class="n">entropy_SH</span><span class="o">=</span><span class="s1">&#39;&#39;</span>
    <span class="n">entvalue</span><span class="o">=</span><span class="p">[]</span>

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">no_of_pc</span><span class="p">):</span>
        <span class="n">value</span><span class="o">=</span><span class="n">entropy</span><span class="p">(</span><span class="n">H</span><span class="p">[</span><span class="n">i</span><span class="p">,:],</span><span class="n">base</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>  <span class="o">/</span>  <span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">H</span><span class="p">[</span><span class="n">i</span><span class="p">]))</span>
        <span class="n">entvalue</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">value</span><span class="p">)</span>

    <span class="n">entvalue</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">entvalue</span><span class="p">)</span>
    <span class="n">index</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="o">-</span><span class="n">entvalue</span><span class="p">)</span>

    <span class="n">H</span><span class="o">=</span><span class="n">H</span><span class="p">[</span><span class="n">index</span><span class="p">]</span>
    <span class="n">spH</span><span class="o">=</span><span class="n">spH</span><span class="p">[</span><span class="n">index</span><span class="p">]</span>

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">no_of_pc</span><span class="p">):</span>
        <span class="n">entropy_H</span><span class="o">+=</span><span class="s1">&#39;,</span><span class="si">%0.2f</span><span class="s1">&#39;</span><span class="o">%</span><span class="p">(</span><span class="n">entropy</span><span class="p">(</span><span class="n">H</span><span class="p">[</span><span class="n">i</span><span class="p">,:],</span><span class="n">base</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>  <span class="o">/</span>  <span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">H</span><span class="p">[</span><span class="n">i</span><span class="p">])))</span>
        <span class="n">entropy_SH</span><span class="o">+=</span><span class="s1">&#39;,</span><span class="si">%0.2f</span><span class="s1">&#39;</span><span class="o">%</span><span class="p">(</span><span class="n">entropy</span><span class="p">(</span><span class="n">spH</span><span class="p">[</span><span class="n">i</span><span class="p">,:],</span><span class="n">base</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span> <span class="o">/</span> <span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">spH</span><span class="p">[</span><span class="n">i</span><span class="p">]))</span>  <span class="p">)</span>


    <span class="c1">#value1=np.sqrt(np.sum((v1.T-np.matmul(W+V,H))**2))</span>
    <span class="c1">#value2=np.sqrt(np.sum((v2.T-np.matmul(spW+spV,spH))**2))</span>

    <span class="n">sc_cosine</span><span class="o">=</span><span class="n">find_correlation_bw_genes_and_PC_component_in_singlecell_cosine</span><span class="p">(</span><span class="n">H</span><span class="o">.</span><span class="n">T</span><span class="p">,</span><span class="n">CbyG</span><span class="p">)</span>
    <span class="n">sc_spearman</span><span class="o">=</span><span class="n">find_correlation_bw_genes_and_PC_component_in_singlecell</span><span class="p">(</span><span class="n">H</span><span class="o">.</span><span class="n">T</span><span class="p">,</span><span class="n">CbyG</span><span class="p">)</span>


    <span class="n">sc_cluster_mean_exp</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">CbyG</span><span class="p">,</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">sc_cluster_exp_more_than_threshold</span><span class="o">=</span><span class="n">CbyG</span><span class="o">&gt;</span><span class="n">cutoff_to_count_exp_cell_population</span>
    <span class="n">sc_cluster_exp_more_than_threshold</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">sc_cluster_exp_more_than_threshold</span><span class="p">,</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">sc_cluster_exp_more_than_threshold</span><span class="o">=</span><span class="n">sc_cluster_exp_more_than_threshold</span><span class="o">/</span><span class="n">CbyG</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

    <span class="n">transfer_sp_com</span><span class="o">=</span><span class="n">spH</span><span class="o">.</span><span class="n">T</span>
    <span class="n">transfer_sc_com</span><span class="o">=</span><span class="p">[]</span>


    <span class="c1">#sc_barcode=sct_ad_sc.obs_names.to_numpy()</span>
    <span class="c1">#sp_barcode=sct_ad_sp.obs_names.to_numpy()</span>
    <span class="n">sc_genenames</span><span class="o">=</span><span class="n">full_genes_sc</span><span class="o">.</span><span class="n">var_names</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span>


    <span class="c1">#maximum norm or infinity norm normalization</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">transfer_sp_com</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]):</span>
        <span class="c1">#transfer_sp_com[:,i]=transfer_sp_com[:,i]/max(abs(transfer_sp_com[i:,]))</span>
        <span class="c1">#l2norm=np.linalg.norm(transfer_sp_com[:,i],ord=2)</span>
        <span class="n">l2norm</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">std</span><span class="p">(</span><span class="n">transfer_sp_com</span><span class="p">[:,</span><span class="n">i</span><span class="p">])</span>
        <span class="c1">#l1norm=np.linalg.norm(transfer_sp_com[:,i],ord=1)</span>
        <span class="c1">#transfer_sp_com[:,i]=transfer_sp_com[:,i]/l1norm</span>
        <span class="n">transfer_sp_com</span><span class="p">[:,</span><span class="n">i</span><span class="p">]</span><span class="o">=</span><span class="n">transfer_sp_com</span><span class="p">[:,</span><span class="n">i</span><span class="p">]</span><span class="o">/</span><span class="n">l2norm</span>

    <span class="k">return</span> <span class="n">transfer_sp_com</span><span class="p">,</span> <span class="n">transfer_sc_com</span><span class="p">,</span> <span class="n">sc_spearman</span><span class="p">,</span><span class="n">sc_cosine</span><span class="p">,</span><span class="n">sc_genenames</span><span class="p">,</span> <span class="n">H</span><span class="p">,</span> <span class="n">spH</span><span class="p">,</span><span class="n">sc_cluster_mean_exp</span><span class="p">,</span><span class="n">sc_cluster_exp_more_than_threshold</span><span class="p">,</span><span class="n">alpha</span></div>




<div class="viewcode-block" id="makePCneighboorhoodFeatureMatrix">
<a class="viewcode-back" href="../../nico_covariations.html#nico_covariations.Covariations.makePCneighboorhoodFeatureMatrix">[docs]</a>
<span class="k">def</span> <span class="nf">makePCneighboorhoodFeatureMatrix</span><span class="p">(</span><span class="nb">input</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Helper function in gene_covariation_analysis to find the weighted neighborhood</span>
<span class="sd">    average of cell types from the spatial factors.</span>

<span class="sd">    This function computes a matrix where each row corresponds to a cell and each column</span>
<span class="sd">    corresponds to a weighted average of principal components (PCs) from neighboring cells.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    -----------</span>
<span class="sd">    input : object</span>
<span class="sd">        An object containing various attributes required for computation, such as:</span>

<span class="sd">            - spatialcell_unique_clusterid: Unique cluster IDs for spatial cells.</span>
<span class="sd">            - neighbors: List of neighboring cells for each cell.</span>
<span class="sd">            - neigh_distances: Distances to neighbors.</span>
<span class="sd">            - annotation_spatial_barcode_id: Barcode IDs for spatial annotations.</span>
<span class="sd">            - annotation_spatial_cluster_id: Cluster IDs for spatial annotations.</span>
<span class="sd">            - pc_of_sp_clusterid: Principal components for spatial cluster IDs.</span>
<span class="sd">            - no_of_pc: Number of principal components.</span>
<span class="sd">            - outputname: Name of the output file.</span>

<span class="sd">    Outputs:</span>
<span class="sd">    --------</span>
<span class="sd">    A .npz file containing the matrix of weighted neighborhood principal components.</span>
<span class="sd">    &quot;&quot;&quot;</span>


    <span class="n">n</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clusterid</span><span class="p">)</span>
    <span class="n">M</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">neighbors</span><span class="p">),</span><span class="n">n</span><span class="o">*</span><span class="nb">input</span><span class="o">.</span><span class="n">no_of_pc</span><span class="p">),</span><span class="n">dtype</span><span class="o">=</span><span class="nb">float</span><span class="p">)</span>

    <span class="n">dist_neighbors</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">neigh_distances</span>
    <span class="n">avgdistArray</span><span class="o">=</span><span class="mi">0</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">dist_neighbors</span><span class="p">)):</span>
        <span class="n">avgdistArray</span><span class="o">=</span><span class="n">avgdistArray</span><span class="o">+</span><span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">dist_neighbors</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
    <span class="n">avgdist</span><span class="o">=</span><span class="n">avgdistArray</span><span class="o">/</span><span class="nb">len</span><span class="p">(</span><span class="n">dist_neighbors</span><span class="p">)</span>


    <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">neighbors</span><span class="p">)):</span>
        <span class="n">CC_barcode_id</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">annotation_spatial_barcode_id</span><span class="p">[</span><span class="n">j</span><span class="p">]</span>
        <span class="n">CC_cluster_id</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">annotation_spatial_cluster_id</span><span class="p">[</span><span class="n">j</span><span class="p">]</span>
        <span class="n">PC_component_of_CC</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">pc_of_sp_clusterid</span><span class="p">[</span><span class="n">CC_barcode_id</span><span class="p">]</span>
        <span class="n">PC_component_of_CC</span><span class="o">=</span><span class="n">PC_component_of_CC</span><span class="o">.</span><span class="n">reshape</span><span class="p">((</span><span class="mi">1</span><span class="p">,</span><span class="nb">input</span><span class="o">.</span><span class="n">no_of_pc</span><span class="p">))</span>
        <span class="k">if</span> <span class="n">j</span><span class="o">==</span><span class="mi">0</span><span class="p">:</span>
            <span class="n">target</span><span class="o">=</span><span class="n">PC_component_of_CC</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">target</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">vstack</span><span class="p">((</span><span class="n">target</span><span class="p">,</span><span class="n">PC_component_of_CC</span><span class="p">))</span>

        <span class="n">neigh_dist</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">dist_neighbors</span><span class="p">[</span><span class="n">j</span><span class="p">])</span>
        <span class="c1">#weightdist=weightdist/avgdist</span>
        <span class="n">neigh_dist</span><span class="o">=</span><span class="mi">1</span><span class="o">/</span><span class="n">neigh_dist</span>
        <span class="n">sum_weight_dist</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">neigh_dist</span><span class="p">)</span>
        <span class="n">weighted_avg_dist</span><span class="o">=</span><span class="n">neigh_dist</span><span class="o">/</span><span class="n">sum_weight_dist</span>
        <span class="n">temp</span><span class="o">=</span><span class="p">{}</span>
        <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">neighbors</span><span class="p">[</span><span class="n">j</span><span class="p">])):</span>
            <span class="nb">id</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">neighbors</span><span class="p">[</span><span class="n">j</span><span class="p">][</span><span class="n">k</span><span class="p">]</span>
            <span class="n">NC_barcode_id</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">annotation_spatial_barcode_id</span><span class="p">[</span><span class="nb">id</span><span class="p">]</span>
            <span class="n">NC_cluster_id</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">annotation_spatial_cluster_id</span><span class="p">[</span><span class="nb">id</span><span class="p">]</span>
            <span class="n">PC_component_of_NC</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">pc_of_sp_clusterid</span><span class="p">[</span><span class="n">NC_barcode_id</span><span class="p">]</span>
            <span class="n">PC_component_of_NC</span><span class="o">=</span><span class="n">PC_component_of_NC</span><span class="o">.</span><span class="n">reshape</span><span class="p">((</span><span class="mi">1</span><span class="p">,</span><span class="nb">input</span><span class="o">.</span><span class="n">no_of_pc</span><span class="p">))</span>
            <span class="n">factor</span><span class="o">=</span><span class="n">weighted_avg_dist</span><span class="p">[</span><span class="n">k</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">NC_cluster_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">temp</span><span class="p">:</span>
                <span class="n">temp</span><span class="p">[</span><span class="n">NC_cluster_id</span><span class="p">]</span><span class="o">=</span><span class="n">factor</span><span class="o">*</span><span class="n">PC_component_of_NC</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">temp</span><span class="p">[</span><span class="n">NC_cluster_id</span><span class="p">]</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">((</span><span class="n">temp</span><span class="p">[</span><span class="n">NC_cluster_id</span><span class="p">],</span><span class="n">factor</span><span class="o">*</span><span class="n">PC_component_of_NC</span><span class="p">))</span>

        <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clusterid</span><span class="p">:</span>
            <span class="n">start_index</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">no_of_pc</span><span class="o">*</span><span class="n">key</span>
            <span class="n">end_index</span><span class="o">=</span><span class="n">start_index</span><span class="o">+</span><span class="nb">input</span><span class="o">.</span><span class="n">no_of_pc</span>
            <span class="k">if</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">temp</span><span class="p">:</span>
                <span class="n">M</span><span class="p">[</span><span class="n">j</span><span class="p">,</span><span class="n">start_index</span><span class="p">:</span><span class="n">end_index</span><span class="p">]</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">temp</span><span class="p">[</span><span class="n">key</span><span class="p">],</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>


    <span class="c1">#cluster=input.annotation_spatial_cluster_id</span>
    <span class="c1">#cluster=cluster.reshape((len(cluster),1))</span>
    <span class="c1">#df=pd.DataFrame(np.hstack((cluster,M)))</span>
    <span class="n">data</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">hstack</span><span class="p">((</span><span class="n">target</span><span class="p">,</span><span class="n">M</span><span class="p">))</span>
    <span class="c1">#df=pd.DataFrame(np.hstack((target,M)))</span>
    <span class="c1">#df.to_csv(input.outputname,index=True,header=None)</span>
    <span class="n">np</span><span class="o">.</span><span class="n">savez</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">outputname</span><span class="p">,</span><span class="n">weighted_neighborhood_of_factors_in_niche</span><span class="o">=</span><span class="n">data</span><span class="p">)</span></div>






<div class="viewcode-block" id="compute_PC_space">
<a class="viewcode-back" href="../../nico_covariations.html#nico_covariations.Covariations.compute_PC_space">[docs]</a>
<span class="k">def</span> <span class="nf">compute_PC_space</span><span class="p">(</span><span class="nb">input</span><span class="p">,</span><span class="n">sct_ad_sc_full</span><span class="p">):</span>

<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Helper function in gene_covariation_analysis to find the weighted neighborhood</span>
<span class="sd">    average of cell types from the spatial factors.</span>

<span class="sd">    This function computes the weighted neighborhood average of principal components (PCs) for each cell type from spatial transcriptomics data. The weights are based on the inverse of the distances between neighboring cells.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    input : object</span>
<span class="sd">        An object containing the following attributes:</span>

<span class="sd">        - spatialcell_unique_clusterid: list of unique spatial cell cluster IDs.</span>
<span class="sd">        - neighbors: list of neighbors for each cell.</span>
<span class="sd">        - neigh_distances: list of distances to neighbors for each cell.</span>
<span class="sd">        - annotation_spatial_barcode_id: list of spatial barcode IDs for each cell.</span>
<span class="sd">        - annotation_spatial_cluster_id: list of spatial cluster IDs for each cell.</span>
<span class="sd">        - pc_of_sp_clusterid: matrix of principal components for each spatial cluster ID.</span>
<span class="sd">        - no_of_pc: int, number of principal components.</span>
<span class="sd">        - outputname: str, the name of the output file to save the results.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    None</span>
<span class="sd">        This function saves the weighted neighborhood of factors in a niche to a .npz file specified by input.outputname.</span>

<span class="sd">    Notes</span>
<span class="sd">    -----</span>
<span class="sd">    - This function calculates the weighted average of the principal components (PCs) for each cell&#39;s neighborhood, using the inverse of the distances to its neighbors as weights.</span>
<span class="sd">    - The result is a matrix where each row represents a cell, and each column represents the weighted average PC values for each cluster in the cell&#39;s neighborhood.</span>
<span class="sd">    - The weighted neighborhood feature matrix is saved to a file in .npz format.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">a</span><span class="o">=</span><span class="nb">set</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">singlecell_unique_clustername</span><span class="p">)</span>
    <span class="n">b</span><span class="o">=</span><span class="nb">set</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clustername</span><span class="p">)</span>
    <span class="n">common</span><span class="o">=</span><span class="n">a</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="n">b</span><span class="p">)</span>


    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n\n</span><span class="s2"> Spatial and scRNA-seq number of clusters, respectively &quot;</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">b</span><span class="p">),</span><span class="nb">len</span><span class="p">(</span><span class="n">a</span><span class="p">))</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Common cell types between spatial and scRNA-seq data  &#39;</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">common</span><span class="p">),</span><span class="n">common</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">The spatial cluster name does not match the scRNA-seq cluster name &#39;</span><span class="p">,</span> <span class="n">b</span><span class="o">-</span><span class="n">common</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;If the above answer is Null, then everything is okay. However, if any spatial cell type does not exist in the scRNA-seq data, please correct this manually; otherwise, NiCo will not run. &quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n\n</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="n">seed</span> <span class="o">=</span> <span class="nb">input</span><span class="o">.</span><span class="n">seed</span>

    <span class="n">flag</span><span class="o">=</span><span class="mi">1</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">b</span><span class="o">-</span><span class="n">common</span><span class="p">)</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">:</span>
        <span class="n">flag</span><span class="o">=</span><span class="mi">0</span>

    <span class="k">if</span> <span class="n">flag</span><span class="o">==</span><span class="mi">1</span><span class="p">:</span>
        <span class="n">n</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clustername</span><span class="p">)</span>
        <span class="n">pc_of_sp_clusterid</span><span class="o">=</span><span class="p">{}</span>
        <span class="n">save_scFactors</span><span class="o">=</span><span class="p">{}</span>
        <span class="n">save_spFactors</span><span class="o">=</span><span class="p">{}</span>
        <span class="n">PCA_of_sc_cluster_accordingto_spatial_clusterid</span><span class="o">=</span><span class="p">{}</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="p">):</span>
            <span class="n">clidsp</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clusterid</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
            <span class="n">index</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">annotation_spatial_cluster_id</span><span class="o">==</span><span class="n">clidsp</span><span class="p">)</span>
            <span class="n">spbarcode</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">annotation_spatial_barcode_id</span><span class="p">[</span><span class="n">index</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span>
            <span class="n">scbarcode</span><span class="o">=</span><span class="p">[]</span>
            <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">singlecell_unique_clustername</span><span class="p">)):</span>
                <span class="k">if</span> <span class="nb">input</span><span class="o">.</span><span class="n">singlecell_unique_clustername</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">==</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clustername</span><span class="p">[</span><span class="n">i</span><span class="p">]:</span>
                    <span class="n">clid</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">singlecell_unique_clusterid</span><span class="p">[</span><span class="n">j</span><span class="p">]</span>
                    <span class="n">index</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">annotation_singlecell_cluster_id</span><span class="o">==</span><span class="n">clid</span><span class="p">)</span>
                    <span class="n">scbarcode</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">annotation_singlecell_barcode_id</span><span class="p">[</span><span class="n">index</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span>
                    <span class="k">break</span>

            <span class="c1">#pc_sp,pc_sc,sp_barcode,sc_barcode,sc_spearman,sc_cosine,sc_genenames,H, spH,</span>
            <span class="n">seed</span><span class="o">=</span><span class="n">seed</span><span class="o">+</span><span class="mi">1</span>
            <span class="n">pc_sp</span><span class="p">,</span><span class="n">pc_sc</span><span class="p">,</span><span class="n">sc_spearman</span><span class="p">,</span><span class="n">sc_cosine</span><span class="p">,</span><span class="n">sc_genenames</span><span class="p">,</span><span class="n">H</span><span class="p">,</span> <span class="n">spH</span><span class="p">,</span><span class="n">sc_cluster_mean_exp</span><span class="p">,</span><span class="n">sc_cluster_exp_more_than_threshold</span><span class="p">,</span><span class="n">alpha</span><span class="o">=</span><span class="n">find_PC_of_invidualCluster_in_SC</span><span class="p">(</span><span class="n">seed</span><span class="p">,</span><span class="nb">input</span><span class="o">.</span><span class="n">spatial_integration_modality</span><span class="p">,</span><span class="n">scbarcode</span><span class="p">,</span><span class="nb">input</span><span class="o">.</span><span class="n">iNMFmode</span><span class="p">,</span><span class="nb">input</span><span class="o">.</span><span class="n">ad_sc</span><span class="p">,</span><span class="nb">input</span><span class="o">.</span><span class="n">no_of_pc</span><span class="p">,</span><span class="n">spbarcode</span><span class="p">,</span><span class="nb">input</span><span class="o">.</span><span class="n">ad_sp</span><span class="p">,</span> <span class="n">sct_ad_sc_full</span><span class="p">,</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clustername</span><span class="p">[</span><span class="n">i</span><span class="p">],</span><span class="nb">input</span><span class="o">.</span><span class="n">cutoff_to_count_exp_cell_population</span><span class="p">)</span>

            <span class="n">PCA_of_sc_cluster_accordingto_spatial_clusterid</span><span class="p">[</span><span class="n">clidsp</span><span class="p">]</span><span class="o">=</span><span class="p">[</span><span class="n">sc_spearman</span><span class="p">,</span><span class="n">pc_sp</span><span class="p">,</span><span class="n">sc_genenames</span><span class="p">,</span><span class="n">sc_cluster_mean_exp</span><span class="p">,</span><span class="n">sc_cluster_exp_more_than_threshold</span><span class="p">,</span><span class="n">sc_cosine</span><span class="p">,</span><span class="n">alpha</span><span class="p">]</span>
            <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">spbarcode</span><span class="p">)):</span>
                <span class="n">pc_of_sp_clusterid</span><span class="p">[</span><span class="n">spbarcode</span><span class="p">[</span><span class="n">k</span><span class="p">]]</span><span class="o">=</span><span class="n">pc_sp</span><span class="p">[</span><span class="n">k</span><span class="p">]</span>
                <span class="n">save_spFactors</span><span class="p">[</span><span class="n">spbarcode</span><span class="p">[</span><span class="n">k</span><span class="p">]]</span><span class="o">=</span><span class="n">spH</span><span class="p">[:,</span><span class="n">k</span><span class="p">]</span>

            <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">scbarcode</span><span class="p">)):</span>
                <span class="n">save_scFactors</span><span class="p">[</span><span class="n">scbarcode</span><span class="p">[</span><span class="n">k</span><span class="p">]]</span><span class="o">=</span><span class="n">H</span><span class="p">[:,</span><span class="n">k</span><span class="p">]</span>

    <span class="k">return</span> <span class="n">pc_of_sp_clusterid</span><span class="p">,</span><span class="n">PCA_of_sc_cluster_accordingto_spatial_clusterid</span><span class="p">,</span><span class="n">save_scFactors</span><span class="p">,</span><span class="n">save_spFactors</span></div>




<div class="viewcode-block" id="model_linear_regression">
<a class="viewcode-back" href="../../nico_covariations.html#nico_covariations.Covariations.model_linear_regression">[docs]</a>
<span class="k">def</span> <span class="nf">model_linear_regression</span><span class="p">(</span><span class="nb">input</span><span class="p">,</span><span class="n">logistic_predicted_interactions</span><span class="p">):</span>

<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Helper function for gene_covariation_analysis to prepare data Y (central cell factors)</span>
<span class="sd">    and X (neighborhood average spatial cell factors) for each cell type to perform regression.</span>

<span class="sd">    This function loads the precomputed neighborhood feature matrix and prepares the data for linear</span>
<span class="sd">    regression analysis. It then performs ridge regression for each cell type to find the relationship</span>
<span class="sd">    between the central cell factors (Y) and the neighborhood average spatial cell factors (X).</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    input : object</span>
<span class="sd">        An object containing the following attributes:</span>

<span class="sd">        - shap_cluster_cutoff : float, cutoff value for SHAP clustering.</span>
<span class="sd">        - outputname : str, the name of the input file containing precomputed neighborhood features.</span>
<span class="sd">        - no_of_pc : int, number of principal components.</span>
<span class="sd">        - spatialcell_unique_clusterid : list, unique cluster IDs of spatial cells.</span>
<span class="sd">        - annotation_spatial_cluster_id : list, cluster IDs for each spatial cell.</span>
<span class="sd">        - spatialcell_unique_clustername : list, unique cluster names of spatial cells.</span>
<span class="sd">        - seed : int, seed value for regression.</span>
<span class="sd">        - lambda_c : float, regularization parameter for ridge regression.</span>
<span class="sd">        - K_fold : int, number of folds for cross-validation.</span>
<span class="sd">        - n_repeats : int, number of repeats for cross-validation.</span>

<span class="sd">    logistic_predicted_interactions : dict</span>
<span class="sd">        A dictionary where keys are cell type names and values are lists of tuples. Each tuple contains</span>
<span class="sd">        a cell type name and a score representing the predicted interaction strength with the key cell type.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    save_coef : dict</span>
<span class="sd">        A dictionary where keys are unique cluster IDs and values are lists containing the following elements:</span>

<span class="sd">        - coef : array, coefficients of the ridge regression model.</span>
<span class="sd">        - intercept : array, intercepts of the ridge regression model.</span>
<span class="sd">        - alpha : float, regularization parameter of the ridge regression model.</span>
<span class="sd">        - xlabel : array, names of the features.</span>
<span class="sd">        - score : array, scores of the features.</span>
<span class="sd">        - target : array, target values (central cell factors).</span>
<span class="sd">        - neighborhoodClass : array, neighborhood average spatial cell factors.</span>
<span class="sd">        - pv : array, p-values of the regression coefficients.</span>
<span class="sd">        - percent_variance_explained : array, percentage of variance explained by the model.</span>
<span class="sd">        - residual_variance_explained : array, residual variance explained by the model.</span>

<span class="sd">    Notes</span>
<span class="sd">    -----</span>
<span class="sd">    - The function uses ridge regression to model the relationship between central cell factors and neighborhood factors.</span>
<span class="sd">    - The precomputed neighborhood feature matrix is loaded from a file and NaN values are replaced with zeros.</span>
<span class="sd">    - The function selects relevant features based on logistic_predicted_interactions and performs ridge regression.</span>
<span class="sd">    - The results are stored in a dictionary and returned for further analysis.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">shap_cluster_cutoff</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">shap_cluster_cutoff</span>
    <span class="n">data1</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">outputname</span><span class="p">,</span><span class="n">allow_pickle</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">data1</span><span class="o">=</span><span class="n">data1</span><span class="p">[</span><span class="s1">&#39;weighted_neighborhood_of_factors_in_niche&#39;</span><span class="p">]</span>
    <span class="c1">#print(data1.shape)</span>
    <span class="c1">#print(data1[0:5])</span>
    <span class="c1">#data1 = np.genfromtxt(open(input.outputname, &quot;rb&quot;), delimiter=&#39;,&#39;, skip_header=0)</span>
    <span class="c1">#ind=~np.isnan(data1).any(axis=1)</span>
    <span class="c1">#data=data1[ind,:]</span>
    <span class="n">data</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">nan_to_num</span><span class="p">(</span><span class="n">data1</span><span class="p">)</span>

    <span class="n">featureVector</span><span class="o">=</span><span class="nb">range</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">no_of_pc</span><span class="p">,</span><span class="n">data</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="c1"># #just neighborhood</span>
    <span class="n">AllneighborhoodClass</span><span class="o">=</span> <span class="n">data</span><span class="p">[:,</span><span class="n">featureVector</span><span class="p">]</span>
    <span class="n">Alltarget</span><span class="o">=</span> <span class="n">data</span><span class="p">[:,</span><span class="mi">0</span><span class="p">:</span><span class="nb">input</span><span class="o">.</span><span class="n">no_of_pc</span><span class="p">]</span>

    <span class="n">count</span><span class="o">=</span><span class="mi">0</span>
    <span class="n">save_coef</span><span class="o">=</span><span class="p">{}</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clusterid</span><span class="p">)):</span>
        <span class="n">temp</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clusterid</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">==</span><span class="nb">input</span><span class="o">.</span><span class="n">annotation_spatial_cluster_id</span><span class="p">)</span>
        <span class="n">index</span><span class="o">=</span><span class="n">temp</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">neighborhoodClass</span><span class="o">=</span><span class="n">AllneighborhoodClass</span><span class="p">[</span><span class="n">index</span><span class="p">,:]</span>
        <span class="n">target</span><span class="o">=</span><span class="n">Alltarget</span><span class="p">[</span><span class="n">index</span><span class="p">,:]</span>
        <span class="n">positive_interacted_CT</span><span class="o">=</span> <span class="n">logistic_predicted_interactions</span><span class="p">[</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clustername</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span>
        <span class="n">newindex</span><span class="o">=</span><span class="p">[]</span>
        <span class="n">xlabel</span><span class="o">=</span><span class="p">[]</span>
        <span class="n">score</span><span class="o">=</span><span class="p">[]</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clustername</span><span class="p">)):</span>
            <span class="n">start</span><span class="o">=</span><span class="n">j</span><span class="o">*</span><span class="nb">input</span><span class="o">.</span><span class="n">no_of_pc</span>
            <span class="n">end</span><span class="o">=</span><span class="n">start</span><span class="o">+</span><span class="nb">input</span><span class="o">.</span><span class="n">no_of_pc</span>
            <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">positive_interacted_CT</span><span class="p">)):</span>
                <span class="k">if</span> <span class="n">positive_interacted_CT</span><span class="p">[</span><span class="n">k</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">==</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clustername</span><span class="p">[</span><span class="n">j</span><span class="p">]:</span>
                    <span class="n">xlabel</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">positive_interacted_CT</span><span class="p">[</span><span class="n">k</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span>
                    <span class="n">score</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">positive_interacted_CT</span><span class="p">[</span><span class="n">k</span><span class="p">][</span><span class="mi">1</span><span class="p">])</span>
                    <span class="k">for</span> <span class="n">kk</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">start</span><span class="p">,</span><span class="n">end</span><span class="p">):</span>
                        <span class="n">newindex</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">kk</span><span class="p">)</span>

        <span class="n">neighborhoodClass</span><span class="o">=</span><span class="n">neighborhoodClass</span><span class="p">[:,</span><span class="n">newindex</span><span class="p">]</span>
        <span class="n">xlabel</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">xlabel</span><span class="p">)</span>
        <span class="n">score</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">score</span><span class="p">)</span>

        <span class="n">ylabelname</span><span class="o">=</span><span class="p">[]</span>
        <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">xlabel</span><span class="p">)):</span>
            <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">no_of_pc</span><span class="p">):</span>
                <span class="n">ylabelname</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">xlabel</span><span class="p">[</span><span class="n">k</span><span class="p">]</span><span class="o">+</span><span class="s1">&#39;_s&#39;</span><span class="o">+</span><span class="s1">&#39;</span><span class="si">%0.3f</span><span class="s1">&#39;</span><span class="o">%</span><span class="n">score</span><span class="p">[</span><span class="n">k</span><span class="p">]</span><span class="o">+</span><span class="s1">&#39;_Fa&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">j</span><span class="o">+</span><span class="mi">1</span><span class="p">))</span>

        <span class="n">count</span><span class="o">+=</span><span class="n">neighborhoodClass</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">saveoutname</span><span class="o">=</span><span class="nb">str</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clusterid</span><span class="p">[</span><span class="n">i</span><span class="p">])</span><span class="o">+</span><span class="s1">&#39;_&#39;</span><span class="o">+</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clustername</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
        <span class="n">coef</span><span class="p">,</span><span class="n">intercept</span><span class="p">,</span><span class="n">alpha</span><span class="p">,</span><span class="n">percent_variance_explained</span><span class="p">,</span><span class="n">residual_variance_explained</span><span class="p">,</span><span class="n">pv</span><span class="o">=</span><span class="n">run_ridge_regression</span><span class="p">(</span><span class="nb">input</span><span class="p">,</span><span class="n">saveoutname</span><span class="p">,</span><span class="n">ylabelname</span><span class="p">,</span><span class="n">target</span><span class="p">,</span><span class="n">neighborhoodClass</span><span class="p">,</span><span class="n">shap_cluster_cutoff</span><span class="p">)</span>
        <span class="c1">#coef_mu,comp_score,coef_std,comp_score_std,alpha=run_ridge_regression(input.seed ,input.lambda_c,input.K_fold,input.n_repeats,target,neighborhoodClass)</span>
        <span class="c1">#savedata=savedir+&#39;coef&#39;+str(input.spatialcell_unique_clusterid[i])+&#39;.npz&#39;</span>

        <span class="n">save_coef</span><span class="p">[</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clusterid</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span><span class="o">=</span><span class="p">[</span><span class="n">coef</span><span class="p">,</span><span class="n">intercept</span><span class="p">,</span><span class="n">alpha</span><span class="p">,</span><span class="n">xlabel</span><span class="p">,</span><span class="n">score</span><span class="p">,</span><span class="n">target</span><span class="p">,</span><span class="n">neighborhoodClass</span><span class="p">,</span><span class="n">pv</span><span class="p">,</span><span class="n">percent_variance_explained</span><span class="p">,</span><span class="n">residual_variance_explained</span><span class="p">]</span>
        <span class="c1">#np.savez(savedata,coef_mu=coef,intercept=intercept,alpha=alpha,xlabel=xlabel,score=score,Yreg=target,Xreg=neighborhoodClass,pvalue=pv,pve=percent_variance_explained,rve=residual_variance_explained)</span>
        <span class="c1">#np.savez(savedata,coef_mu=coef_mu,coef_std=coef_std,comp_score=comp_score,comp_score_std=comp_score_std,alpha=alpha,xlabel=xlabel,score=score)</span>

    <span class="c1">#print(count)</span>
    <span class="k">return</span> <span class="n">save_coef</span></div>






<div class="viewcode-block" id="run_ridge_regression">
<a class="viewcode-back" href="../../nico_covariations.html#nico_covariations.Covariations.run_ridge_regression">[docs]</a>
<span class="k">def</span> <span class="nf">run_ridge_regression</span><span class="p">(</span><span class="nb">input</span><span class="p">,</span><span class="n">saveoutname</span><span class="p">,</span><span class="n">ylabelname</span><span class="p">,</span><span class="n">target</span><span class="p">,</span><span class="n">neighborhoodClass</span><span class="p">,</span><span class="n">shap_cluster_cutoff</span><span class="p">):</span>

<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Helper function for model_linear_regression to perform ridge regression per cell type.</span>

<span class="sd">    This function performs ridge regression for each target variable (central cell factors) using</span>
<span class="sd">    the neighborhood average spatial cell factors as predictors. It normalizes the data, fits the</span>
<span class="sd">    regression model, and computes various statistics including p-values and explained variance.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    input : object</span>
<span class="sd">        An object containing the following attributes:</span>
<span class="sd">        - shap_analysis : bool, whether to perform SHAP analysis.</span>
<span class="sd">        - regression_outdir : str, directory to save regression outputs.</span>
<span class="sd">        - lambda_c : list, list of regularization parameters for RidgeCV.</span>
<span class="sd">        - no_of_pc : int, number of principal components.</span>

<span class="sd">    saveoutname : str</span>
<span class="sd">        The name to save the output of the regression results.</span>

<span class="sd">    ylabelname : list</span>
<span class="sd">        List of feature names for the predictors.</span>

<span class="sd">    target : array</span>
<span class="sd">        The target values (central cell factors).</span>

<span class="sd">    neighborhoodClass : array</span>
<span class="sd">        The neighborhood average spatial cell factors.</span>

<span class="sd">    shap_cluster_cutoff : float</span>
<span class="sd">        The cutoff value for clustering in SHAP analysis.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    coef : array</span>
<span class="sd">        Coefficients of the ridge regression model.</span>

<span class="sd">    intercept : array</span>
<span class="sd">        Intercepts of the ridge regression model.</span>

<span class="sd">    lambda_c : list</span>
<span class="sd">        List of regularization parameters used in the ridge regression model.</span>

<span class="sd">    percent_variance_explained : list</span>
<span class="sd">        Percentage of variance explained by the model.</span>

<span class="sd">    residual_variance_explained : float</span>
<span class="sd">        Residual variance explained by the model (currently set to 0).</span>

<span class="sd">    pv : array</span>
<span class="sd">        P-values of the regression coefficients.</span>

<span class="sd">    Notes</span>
<span class="sd">    -----</span>
<span class="sd">    - The function normalizes the predictors and target variables.</span>
<span class="sd">    - It fits a ridge regression model for each target variable and computes various statistics.</span>
<span class="sd">    - If SHAP analysis is enabled, it performs SHAP analysis and saves the plots.</span>

<span class="sd">    &quot;&quot;&quot;</span>


    <span class="n">train_index</span><span class="o">=</span><span class="nb">range</span><span class="p">(</span><span class="n">target</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    <span class="n">test_index</span><span class="o">=</span><span class="p">[]</span>

    <span class="n">x_std</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">std</span><span class="p">(</span><span class="n">neighborhoodClass</span><span class="p">,</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">y_std</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">std</span><span class="p">(</span><span class="n">target</span><span class="p">,</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">neighborhoodClass</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]):</span>
        <span class="k">if</span> <span class="n">x_std</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">==</span><span class="mi">0</span><span class="p">:</span>
            <span class="n">x_std</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">=</span><span class="mi">1</span>
        <span class="n">neighborhoodClass</span><span class="p">[:,</span><span class="n">i</span><span class="p">]</span><span class="o">=</span><span class="n">neighborhoodClass</span><span class="p">[:,</span><span class="n">i</span><span class="p">]</span><span class="o">/</span><span class="n">x_std</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">target</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]):</span>
        <span class="k">if</span> <span class="n">y_std</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">==</span><span class="mi">0</span><span class="p">:</span>
            <span class="n">y_std</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">=</span><span class="mi">1</span>
        <span class="n">target</span><span class="p">[:,</span><span class="n">i</span><span class="p">]</span><span class="o">=</span><span class="n">target</span><span class="p">[:,</span><span class="n">i</span><span class="p">]</span><span class="o">/</span><span class="n">y_std</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>

    <span class="c1">#add=np.hstack((target,neighborhoodClass))</span>
    <span class="c1">#ind1=~np.isnan(add).any(axis=0) #1 means rows and 0 means columns</span>
    <span class="c1">#ind2=~np.isnan(add).any(axis=1) #1 means rows and 0 means columns</span>
    <span class="c1">#data=data1[ind,:]</span>
    <span class="c1">#print(target.shape,neighborhoodClass.shape,len(ind1),len(ind2))</span>


    <span class="c1">#print(neighborhoodClass.shape)</span>
    <span class="c1">#print(target.shape,x_std.shape,y_std.shape)</span>
    <span class="n">x_train</span><span class="p">,</span><span class="n">x_test</span><span class="o">=</span><span class="n">neighborhoodClass</span><span class="p">[</span><span class="n">train_index</span><span class="p">],</span><span class="n">neighborhoodClass</span><span class="p">[</span><span class="n">test_index</span><span class="p">]</span>
    <span class="n">y_train</span><span class="p">,</span><span class="n">y_test</span><span class="o">=</span><span class="n">target</span><span class="p">[</span><span class="n">train_index</span><span class="p">],</span><span class="n">target</span><span class="p">[</span><span class="n">test_index</span><span class="p">]</span>

    <span class="c1">#create_directory(savedir+&#39;plot_Y_and_X/&#39;)</span>
    <span class="k">if</span> <span class="nb">input</span><span class="o">.</span><span class="n">shap_analysis</span><span class="p">:</span>
        <span class="n">dir1</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">regression_outdir</span><span class="o">+</span><span class="s1">&#39;Shapley_Interventional/&#39;</span>
        <span class="n">dir2</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">regression_outdir</span><span class="o">+</span><span class="s1">&#39;Shapley_FullConventional/&#39;</span>
        <span class="n">create_directory</span><span class="p">(</span><span class="n">dir1</span><span class="p">)</span>
        <span class="n">create_directory</span><span class="p">(</span><span class="n">dir2</span><span class="p">)</span>

    <span class="n">LRI</span><span class="o">=</span><span class="p">[]</span>
    <span class="n">LRC</span><span class="o">=</span><span class="p">[]</span>
    <span class="n">yhat</span><span class="o">=</span><span class="p">[]</span>
    <span class="n">lambda_c</span><span class="o">=</span><span class="p">[]</span>
    <span class="n">Xdata</span><span class="o">=</span><span class="n">x_train</span>

    <span class="c1">#kf = KFold(10)</span>
    <span class="c1">#print(kf)</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">y_train</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]):</span>
        <span class="n">linear_model</span> <span class="o">=</span> <span class="n">RidgeCV</span><span class="p">(</span><span class="n">alphas</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">lambda_c</span><span class="p">)</span><span class="c1">#,cv=kf,scoring = &#39;neg_mean_squared_error&#39;)</span>
        <span class="c1">#pipe=Pipeline([ (&#39;StandardScaler&#39;,StandardScaler(with_mean=True)),(&#39;ridge_regression&#39;,linear_model)])</span>
        <span class="n">pipe</span><span class="o">=</span><span class="n">Pipeline</span><span class="p">([(</span><span class="s1">&#39;ridge_regression&#39;</span><span class="p">,</span><span class="n">linear_model</span><span class="p">)])</span>
        <span class="n">pipe</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">Xdata</span><span class="p">,</span><span class="n">y_train</span><span class="p">[:,</span><span class="n">i</span><span class="p">])</span>
        <span class="n">yyhat</span><span class="o">=</span><span class="n">pipe</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">Xdata</span><span class="p">)</span>
        <span class="n">yhat</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">yyhat</span><span class="p">)</span>
        <span class="n">LR</span><span class="o">=</span> <span class="n">pipe</span><span class="o">.</span><span class="n">named_steps</span><span class="p">[</span><span class="s1">&#39;ridge_regression&#39;</span><span class="p">]</span>
        <span class="n">coef</span><span class="o">=</span><span class="n">LR</span><span class="o">.</span><span class="n">coef_</span>
        <span class="n">intercept</span><span class="o">=</span><span class="n">LR</span><span class="o">.</span><span class="n">intercept_</span>
        <span class="n">LRI</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">intercept</span><span class="p">)</span>
        <span class="n">LRC</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">coef</span><span class="p">)</span>
        <span class="n">lambda_c</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">%0.2f</span><span class="s1">&#39;</span><span class="o">%</span><span class="n">LR</span><span class="o">.</span><span class="n">alpha_</span><span class="p">)</span>

    <span class="n">LRI</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">LRI</span><span class="p">)</span>
    <span class="n">yhat</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">yhat</span><span class="p">)</span><span class="o">.</span><span class="n">T</span>
    <span class="n">LRC</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">LRC</span><span class="p">)</span>


    <span class="c1">#mu=np.mean(y_train,axis=0)</span>
    <span class="c1">#total_ss= np.sum((y_train-mu)**2,axis=0)</span>
    <span class="c1">#residual_ss=np.sum((y_train-yhat)**2,axis=0)</span>
    <span class="c1">#explained_ss= np.sum((yhat-mu)**2,axis=0)</span>
    <span class="c1">#percent_variance_explained=100*explained_ss/total_ss</span>
    <span class="c1">#residual_variance_explained=100*residual_ss/total_ss</span>

    <span class="n">pv</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="n">LRC</span><span class="o">.</span><span class="n">shape</span><span class="p">,</span><span class="n">dtype</span><span class="o">=</span><span class="nb">float</span><span class="p">)</span>
    <span class="n">EVS</span><span class="o">=</span><span class="p">[]</span>
    <span class="n">rss</span><span class="o">=</span><span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">y_train</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]):</span>
                <span class="c1">#EVS.append(explained_variance_score(save_y_test[:,i], save_y_pred[:,i]))</span>
                <span class="n">EVS</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">explained_variance_score</span><span class="p">(</span><span class="n">y_train</span><span class="p">[:,</span><span class="n">i</span><span class="p">],</span> <span class="n">yhat</span><span class="p">[:,</span><span class="n">i</span><span class="p">]))</span>
                <span class="n">rss</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">((</span><span class="n">y_train</span><span class="p">[:,</span><span class="n">i</span><span class="p">]</span><span class="o">-</span><span class="n">yhat</span><span class="p">[:,</span><span class="n">i</span><span class="p">])</span><span class="o">**</span><span class="mi">2</span><span class="p">,</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">))</span>
                <span class="n">params</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">LRI</span><span class="p">[</span><span class="n">i</span><span class="p">],</span><span class="n">LRC</span><span class="p">[</span><span class="n">i</span><span class="p">,:])</span>
                <span class="n">newX</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="n">Xdata</span><span class="p">),</span><span class="mi">1</span><span class="p">)),</span> <span class="n">Xdata</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
                <span class="n">MSE</span> <span class="o">=</span> <span class="p">(</span><span class="nb">sum</span><span class="p">((</span><span class="n">y_train</span><span class="p">[:,</span><span class="n">i</span><span class="p">]</span><span class="o">-</span><span class="n">yhat</span><span class="p">[:,</span><span class="n">i</span><span class="p">])</span><span class="o">**</span><span class="mi">2</span><span class="p">))</span><span class="o">/</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">newX</span><span class="p">)</span><span class="o">-</span><span class="nb">len</span><span class="p">(</span><span class="n">newX</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span>

                <span class="n">detM</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">linalg</span><span class="o">.</span><span class="n">det</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">newX</span><span class="o">.</span><span class="n">T</span><span class="p">,</span><span class="n">newX</span><span class="p">))</span>
                <span class="k">if</span> <span class="n">detM</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">:</span>
                    <span class="n">flag</span><span class="o">=</span><span class="mi">0</span>
                    <span class="k">try</span><span class="p">:</span>
                        <span class="n">var_b</span> <span class="o">=</span> <span class="n">MSE</span><span class="o">*</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">linalg</span><span class="o">.</span><span class="n">inv</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">newX</span><span class="o">.</span><span class="n">T</span><span class="p">,</span><span class="n">newX</span><span class="p">))</span><span class="o">.</span><span class="n">diagonal</span><span class="p">())</span>
                    <span class="k">except</span> <span class="n">np</span><span class="o">.</span><span class="n">linalg</span><span class="o">.</span><span class="n">LinAlgError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                        <span class="k">if</span> <span class="s1">&#39;Singular matrix&#39;</span> <span class="ow">in</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">):</span>
                            <span class="n">var_b</span><span class="o">=</span><span class="mi">1</span><span class="c1"># your error handling block</span>
                            <span class="n">flag</span><span class="o">=</span><span class="mi">1</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="k">raise</span>
                    <span class="n">sd_b</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">var_b</span><span class="p">)</span>
                    <span class="n">ts_b</span> <span class="o">=</span> <span class="n">params</span><span class="o">/</span> <span class="n">sd_b</span>
                    <span class="n">df</span> <span class="o">=</span> <span class="n">x_train</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">-</span> <span class="n">x_train</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
                    <span class="n">p_values1</span> <span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="mi">2</span><span class="o">*</span><span class="p">(</span><span class="mi">1</span><span class="o">-</span><span class="n">scipy</span><span class="o">.</span><span class="n">stats</span><span class="o">.</span><span class="n">t</span><span class="o">.</span><span class="n">cdf</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">ii</span><span class="p">),</span><span class="n">df</span><span class="o">-</span><span class="mi">1</span><span class="p">))</span> <span class="k">for</span> <span class="n">ii</span> <span class="ow">in</span> <span class="n">ts_b</span><span class="p">]])</span>
                    <span class="n">pv</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">=</span><span class="n">p_values1</span><span class="p">[:,</span><span class="mi">1</span><span class="p">:]</span>
                    <span class="c1">#if flag==1:</span>
                    <span class="c1">#    print(i,saveoutname,MSE,&quot;var_b&quot;,var_b,&quot;pvalue&quot;,pv[i])</span>



        <span class="c1">#print(&quot;LRC&quot;,LRC.shape,LRI.shape)</span>
        <span class="c1">#x_train2 = sm.add_constant(Xdata)</span>
        <span class="c1">#est1=sm.OLS(y_train[:,0],x_train2).fit()</span>
        <span class="c1">#print(&quot;summary1&quot;,est1.summary())</span>
        <span class="c1">#est2=sm.OLS(y_train[:,1],x_train2).fit()</span>
        <span class="c1">#print(&quot;summary2&quot;,est2.summary())</span>
        <span class="c1">#est3=sm.OLS(y_train[:,2],x_train2).fit()</span>
        <span class="c1">#print(&quot;summary3&quot;,est3.summary())</span>

    <span class="c1">#pvalue correction for mulitple hypothesis</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    flatpv=pv.reshape(pv.shape[0]*pv.shape[1],)</span>
<span class="sd">    #print(&quot;flatpv&quot;,flatpv.shape)</span>
<span class="sd">    _, pvals_corrected_bh, _, _ = sm.multipletests(flatpv, alpha=0.05, method=&#39;fdr_bh&#39;)</span>
<span class="sd">    pv=pvals_corrected_bh.reshape(pv.shape[0],pv.shape[1])</span>

<span class="sd">    for i in range(len(flatpv)):</span>
<span class="sd">        if flatpv[i]&lt;0.05:</span>
<span class="sd">            print(i,&quot;a&quot;,saveoutname,flatpv[i],pvals_corrected_bh[i])</span>
<span class="sd">        if pvals_corrected_bh[i]&lt;0.05:</span>
<span class="sd">            print(i,&quot;b&quot;,flatpv[i],pvals_corrected_bh[i])</span>
<span class="sd">    &#39;&#39;&#39;</span>

    <span class="k">if</span> <span class="nb">input</span><span class="o">.</span><span class="n">shap_analysis</span><span class="p">:</span>
            <span class="c1">#explainer = shap.LinearExplainer(LR, x_train)</span>
            <span class="n">explainer</span> <span class="o">=</span> <span class="n">shap</span><span class="o">.</span><span class="n">explainers</span><span class="o">.</span><span class="n">Linear</span><span class="p">(</span><span class="n">LR</span><span class="p">,</span> <span class="n">x_train</span><span class="p">,</span><span class="n">feature_names</span><span class="o">=</span><span class="n">ylabelname</span><span class="p">,</span><span class="n">feature_perturbation</span><span class="o">=</span><span class="s2">&quot;interventional&quot;</span><span class="p">)</span>
            <span class="c1">#explainer = shap.Explainer(LR, x_train,feature_names=ylabelname)</span>
            <span class="c1">#shap_values = explainer.shap_values(x_train)</span>
            <span class="n">shap_values</span> <span class="o">=</span> <span class="n">explainer</span><span class="p">(</span><span class="n">x_train</span><span class="p">)</span>

            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">y_train</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]):</span>
                <span class="c1">#shap.waterfall_plot(explainer.expected_value, shap_values[sample_ind], X.iloc[sample_ind], max_display=14)</span>
                <span class="n">clust</span> <span class="o">=</span> <span class="n">shap</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">hclust</span><span class="p">(</span><span class="n">x_train</span><span class="p">,</span> <span class="n">y_train</span><span class="p">[:,</span><span class="n">i</span><span class="p">],</span> <span class="n">linkage</span><span class="o">=</span><span class="s2">&quot;single&quot;</span><span class="p">)</span>
                <span class="n">shap</span><span class="o">.</span><span class="n">plots</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="n">shap_values</span><span class="p">,</span> <span class="n">clustering</span><span class="o">=</span><span class="n">clust</span><span class="p">,</span> <span class="n">clustering_cutoff</span><span class="o">=</span><span class="n">shap_cluster_cutoff</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
                <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;True to the model &quot;</span><span class="o">+</span><span class="n">saveoutname</span><span class="o">+</span><span class="s1">&#39;_&#39;</span><span class="o">+</span><span class="s1">&#39;Fa&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span><span class="o">+</span><span class="s2">&quot;, EVS = &quot;</span> <span class="o">+</span><span class="s1">&#39;</span><span class="si">%0.4f</span><span class="s1">&#39;</span><span class="o">%</span><span class="n">EVS</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
                <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">dir1</span><span class="o">+</span><span class="n">saveoutname</span><span class="o">+</span><span class="s1">&#39;_Fa&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;.png&#39;</span><span class="p">,</span><span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span> <span class="n">bbox_inches</span> <span class="o">=</span> <span class="s2">&quot;tight&quot;</span><span class="p">)</span>
                <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="s1">&#39;all&#39;</span><span class="p">)</span>

                <span class="n">explainer</span> <span class="o">=</span> <span class="n">shap</span><span class="o">.</span><span class="n">explainers</span><span class="o">.</span><span class="n">Linear</span><span class="p">(</span><span class="n">LR</span><span class="p">,</span> <span class="n">x_train</span><span class="p">,</span><span class="n">feature_names</span><span class="o">=</span><span class="n">ylabelname</span><span class="p">,</span><span class="n">feature_perturbation</span><span class="o">=</span><span class="s2">&quot;correlation_dependent&quot;</span><span class="p">)</span>
                <span class="n">shap_values</span> <span class="o">=</span> <span class="n">explainer</span><span class="p">(</span><span class="n">x_train</span><span class="p">)</span>
                <span class="n">shap</span><span class="o">.</span><span class="n">plots</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="n">shap_values</span><span class="p">,</span> <span class="n">clustering</span><span class="o">=</span><span class="n">clust</span><span class="p">,</span> <span class="n">clustering_cutoff</span><span class="o">=</span><span class="n">shap_cluster_cutoff</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
                <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;True to the data &quot;</span><span class="o">+</span><span class="n">saveoutname</span><span class="o">+</span><span class="s1">&#39;_&#39;</span><span class="o">+</span><span class="s1">&#39;Fa&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span><span class="o">+</span><span class="s2">&quot;, EVS = &quot;</span> <span class="o">+</span><span class="s1">&#39;</span><span class="si">%0.4f</span><span class="s1">&#39;</span><span class="o">%</span><span class="n">EVS</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
                <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">dir2</span><span class="o">+</span><span class="n">saveoutname</span><span class="o">+</span><span class="s1">&#39;_Fa&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;.png&#39;</span><span class="p">,</span><span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span> <span class="n">bbox_inches</span> <span class="o">=</span> <span class="s2">&quot;tight&quot;</span><span class="p">)</span>
                <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="s1">&#39;all&#39;</span><span class="p">)</span>

    <span class="n">coef</span><span class="o">=</span><span class="n">LRC</span>
    <span class="n">intercept</span><span class="o">=</span><span class="n">LRI</span>
    <span class="n">residual_variance_explained</span><span class="o">=</span><span class="mi">0</span>

    <span class="k">return</span> <span class="n">coef</span><span class="p">,</span><span class="n">intercept</span><span class="p">,</span><span class="n">lambda_c</span><span class="p">,</span><span class="n">EVS</span><span class="p">,</span><span class="n">residual_variance_explained</span><span class="p">,</span><span class="n">pv</span></div>




<div class="viewcode-block" id="find_logistic_regression_interacting_score">
<a class="viewcode-back" href="../../nico_covariations.html#nico_covariations.Covariations.find_logistic_regression_interacting_score">[docs]</a>
<span class="k">def</span> <span class="nf">find_logistic_regression_interacting_score</span><span class="p">(</span><span class="n">cmn</span><span class="p">,</span><span class="n">coef</span><span class="p">,</span><span class="n">CTFeatures</span><span class="p">,</span><span class="n">nameOfCellType</span><span class="p">,</span><span class="n">logistic_coef_cutoff</span><span class="p">):</span>

<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Helper function used in gene_covariation_analysis to find niche interaction scores from logistic regression classifier.</span>

<span class="sd">    This function identifies the interacting cell types by analyzing the coefficients of a logistic regression classifier.</span>
<span class="sd">    It normalizes the coefficients, sorts them, and identifies the significant interactions based on a specified cutoff value.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    cmn : array</span>
<span class="sd">        Confusion matrix or similar matrix representing the performance of the logistic regression classifier.</span>

<span class="sd">    coef : array</span>
<span class="sd">        Coefficients of the logistic regression model.</span>

<span class="sd">    CTFeatures : list</span>
<span class="sd">        List of cell type features used in the logistic regression model.</span>

<span class="sd">    nameOfCellType : list</span>
<span class="sd">        List of names corresponding to cell types.</span>

<span class="sd">    logistic_coef_cutoff : float</span>
<span class="sd">        The cutoff value to consider a coefficient as significant for interaction.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    logistic_predicted_interactions : dict</span>
<span class="sd">        A dictionary where keys are cell types and values are lists of interacting cell types with their interaction scores.</span>

<span class="sd">    Notes</span>
<span class="sd">    -----</span>
<span class="sd">    - The function normalizes the logistic regression coefficients.</span>
<span class="sd">    - It identifies the most significant interactions based on the absolute value of the coefficients.</span>
<span class="sd">    - Interactions with coefficients above the cutoff value are considered significant and are included in the output.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">a</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">diag</span><span class="p">(</span><span class="n">cmn</span><span class="p">)</span>
    <span class="c1">#b=np.diag(input.cmn_std)</span>
    <span class="n">goodPredictedCellType</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="o">-</span><span class="n">a</span><span class="p">)</span>
    <span class="n">largest</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="nb">abs</span><span class="p">(</span><span class="n">coef</span><span class="p">))</span>
    <span class="n">normalized_coef</span><span class="o">=</span><span class="n">coef</span><span class="o">/</span><span class="n">largest</span>
    <span class="n">InteractingCTs</span><span class="o">=</span><span class="p">[]</span>
    <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">a</span><span class="p">)):</span>
            <span class="n">meanCoefficients</span><span class="o">=</span><span class="n">normalized_coef</span><span class="p">[</span><span class="n">goodPredictedCellType</span><span class="p">[</span><span class="n">k</span><span class="p">]]</span>
            <span class="c1">#stdCoefficients=input.coef_std[goodPredictedCellType[k]]</span>
            <span class="n">highestIndex</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="o">-</span><span class="nb">abs</span><span class="p">(</span><span class="n">meanCoefficients</span><span class="p">))</span>
            <span class="n">n</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">highestIndex</span><span class="p">)</span>
            <span class="n">coeff_of_CT</span><span class="o">=</span><span class="p">[]</span>
            <span class="n">name_of_the_coeff</span><span class="o">=</span><span class="p">[]</span>
            <span class="n">std_of_coeff</span><span class="o">=</span><span class="p">[]</span>
            <span class="n">predictedCT</span><span class="o">=</span><span class="n">nameOfCellType</span><span class="p">[</span><span class="n">goodPredictedCellType</span><span class="p">[</span><span class="n">k</span><span class="p">]]</span>
            <span class="n">positiveprediction</span><span class="o">=</span><span class="p">[]</span>
            <span class="n">negativeprediction</span><span class="o">=</span><span class="p">[]</span>
            <span class="n">score</span><span class="o">=</span><span class="p">[]</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="p">):</span>
                <span class="n">l</span><span class="o">=</span><span class="n">CTFeatures</span><span class="p">[</span><span class="n">highestIndex</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
                <span class="n">temp</span><span class="o">=</span><span class="s1">&#39;&#39;</span>
                <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">l</span><span class="p">)):</span>
                    <span class="n">temp</span><span class="o">+=</span><span class="n">nameOfCellType</span><span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">l</span><span class="p">[</span><span class="n">j</span><span class="p">][</span><span class="mi">1</span><span class="p">:])]</span>
                    <span class="k">if</span> <span class="n">j</span><span class="o">!=</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">l</span><span class="p">)</span><span class="o">-</span><span class="mi">1</span><span class="p">):</span>
                        <span class="n">temp</span><span class="o">+=</span><span class="s1">&#39;--&#39;</span>
                <span class="k">if</span> <span class="n">meanCoefficients</span><span class="p">[</span> <span class="n">highestIndex</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span><span class="o">&gt;</span><span class="n">logistic_coef_cutoff</span><span class="p">:</span>
                    <span class="n">positiveprediction</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">temp</span><span class="p">)</span>
                    <span class="n">score</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">meanCoefficients</span><span class="p">[</span> <span class="n">highestIndex</span><span class="p">[</span><span class="n">i</span><span class="p">]])</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">negativeprediction</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">temp</span><span class="p">)</span>
            <span class="n">InteractingCTs</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">predictedCT</span><span class="p">,</span><span class="n">positiveprediction</span><span class="p">,</span> <span class="n">score</span>   <span class="p">])</span>

    <span class="n">logistic_predicted_interactions</span><span class="o">=</span><span class="p">{}</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">InteractingCTs</span><span class="p">)):</span>
        <span class="n">cCT</span><span class="o">=</span><span class="n">InteractingCTs</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">nCT</span><span class="o">=</span><span class="n">InteractingCTs</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">Interacting_score</span><span class="o">=</span><span class="n">InteractingCTs</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">2</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">nCT</span><span class="p">)):</span>
            <span class="k">if</span> <span class="n">cCT</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">logistic_predicted_interactions</span><span class="p">:</span>
                <span class="n">logistic_predicted_interactions</span><span class="p">[</span><span class="n">cCT</span><span class="p">]</span><span class="o">=</span><span class="p">[[</span><span class="n">nCT</span><span class="p">[</span><span class="n">j</span><span class="p">],</span><span class="n">Interacting_score</span><span class="p">[</span><span class="n">j</span><span class="p">]]]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">logistic_predicted_interactions</span><span class="p">[</span><span class="n">cCT</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">nCT</span><span class="p">[</span><span class="n">j</span><span class="p">],</span><span class="n">Interacting_score</span><span class="p">[</span><span class="n">j</span><span class="p">]])</span>

    <span class="k">return</span> <span class="n">logistic_predicted_interactions</span></div>



<div class="viewcode-block" id="findXYZC">
<a class="viewcode-back" href="../../nico_covariations.html#nico_covariations.Covariations.findXYZC">[docs]</a>
<span class="k">def</span> <span class="nf">findXYZC</span><span class="p">(</span><span class="n">c</span><span class="p">,</span><span class="n">s</span><span class="p">):</span>

<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Helper function used in plot_top_selected_genes_as_dotplot.</span>

<span class="sd">    This function extracts and transforms data from the given matrices c and s, creating four lists:</span>
<span class="sd">    x-coordinates, y-coordinates, values (z), and sizes (bigs).</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    c : array-like</span>
<span class="sd">        A 2D array (matrix) where each element represents a value at a specific (i, j) coordinate.</span>

<span class="sd">    s : array-like</span>
<span class="sd">        A 2D array (matrix) of the same shape as c, where each element represents a size multiplier for the corresponding</span>
<span class="sd">        element in c.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    x : list</span>
<span class="sd">        List of x-coordinates for each element in c.</span>

<span class="sd">    y : list</span>
<span class="sd">        List of y-coordinates for each element in c.</span>

<span class="sd">    z : list</span>
<span class="sd">        List of values from c corresponding to each (x, y) coordinate.</span>

<span class="sd">    bigs : list</span>
<span class="sd">        List of sizes, where each size is calculated as 100 times the corresponding element in s.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">x</span><span class="o">=</span><span class="p">[]</span>
    <span class="n">y</span><span class="o">=</span><span class="p">[]</span>
    <span class="n">z</span><span class="o">=</span><span class="p">[]</span>
    <span class="n">bigs</span><span class="o">=</span><span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">c</span><span class="p">)):</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">c</span><span class="p">[</span><span class="n">i</span><span class="p">])):</span>
            <span class="n">x</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">j</span><span class="p">)</span>
            <span class="n">y</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
            <span class="n">z</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">c</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="n">j</span><span class="p">])</span>
            <span class="n">bigs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="mi">100</span><span class="o">*</span><span class="n">s</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="n">j</span><span class="p">])</span>
    <span class="k">return</span> <span class="n">x</span><span class="p">,</span><span class="n">y</span><span class="p">,</span><span class="n">z</span><span class="p">,</span><span class="n">bigs</span></div>




<div class="viewcode-block" id="create_subtitle">
<a class="viewcode-back" href="../../nico_covariations.html#nico_covariations.Covariations.create_subtitle">[docs]</a>
<span class="k">def</span> <span class="nf">create_subtitle</span><span class="p">(</span><span class="n">fig</span><span class="p">:</span> <span class="n">plt</span><span class="o">.</span><span class="n">Figure</span><span class="p">,</span> <span class="n">grid</span><span class="p">:</span> <span class="n">SubplotSpec</span><span class="p">,</span> <span class="n">title</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Add a title to a specific set of subplots within a figure.</span>

<span class="sd">    This helper function is used to create a title for a subset of plots within</span>
<span class="sd">    a Matplotlib figure. The title is added with a specific formatting and the</span>
<span class="sd">    subplot is hidden from view (no axes or frames).</span>

<span class="sd">    Parameters:</span>
<span class="sd">    -----------</span>
<span class="sd">    fig : matplotlib.figure.Figure</span>
<span class="sd">        The figure object to which the subplot belongs.</span>
<span class="sd">    grid : matplotlib.gridspec.SubplotSpec</span>
<span class="sd">        The subplot specification that defines the location and size of the subplot within the figure.</span>
<span class="sd">    title : str</span>
<span class="sd">        The title text to be displayed above the subplot.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">row</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="n">grid</span><span class="p">)</span>
    <span class="c1"># the &#39;\n&#39; is important</span>
    <span class="n">row</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">title</span><span class="si">}</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">fontweight</span><span class="o">=</span><span class="s1">&#39;semibold&#39;</span><span class="p">)</span>
    <span class="c1"># hide subplot</span>
    <span class="n">row</span><span class="o">.</span><span class="n">set_frame_on</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">row</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span></div>






<div class="viewcode-block" id="find_fold_change">
<a class="viewcode-back" href="../../nico_covariations.html#nico_covariations.Covariations.find_fold_change">[docs]</a>
<span class="k">def</span> <span class="nf">find_fold_change</span><span class="p">(</span><span class="n">PCA</span><span class="p">,</span><span class="n">NH_PCA</span><span class="p">,</span><span class="n">gene</span><span class="p">,</span><span class="n">CCPC</span><span class="p">,</span><span class="n">NCPC</span><span class="p">,</span><span class="n">totalLRpairs</span><span class="p">,</span><span class="n">LRcutoff</span><span class="p">,</span><span class="n">CC_meanExpression</span><span class="p">,</span><span class="n">NC_meanExpression</span><span class="p">,</span><span class="n">CC_popExpression</span><span class="p">,</span><span class="n">NC_popExpression</span><span class="p">,</span><span class="n">number_of_top_genes_to_print</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Identify ligand-receptor genes for cell type interaction analysis.</span>

<span class="sd">    This helper function is used in `find_LR_interactions_in_interacting_cell_types` to find ligand-receptor (LR) genes based on principal component analysis (PCA) data. It identifies the top genes and checks for LR interactions between specific cell types.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    -----------</span>
<span class="sd">    PCA : numpy.ndarray</span>
<span class="sd">        PCA data for the cell type of interest.</span>
<span class="sd">    NH_PCA : numpy.ndarray</span>
<span class="sd">        PCA data for non-host cell types.</span>
<span class="sd">    gene : list of str</span>
<span class="sd">        List of gene names corresponding to the PCA data.</span>
<span class="sd">    CCPC : int</span>
<span class="sd">        Principal component index for the cell type of interest.</span>
<span class="sd">    NCPC : int</span>
<span class="sd">        Principal component index for the non-host cell type.</span>
<span class="sd">    totalLRpairs : list of tuples</span>
<span class="sd">        List of tuples representing all possible ligand-receptor pairs.</span>
<span class="sd">    LRcutoff : float</span>
<span class="sd">        Threshold for selecting significant ligand-receptor interactions.</span>
<span class="sd">    CC_meanExpression : numpy.ndarray</span>
<span class="sd">        Mean expression values for the central cell type.</span>
<span class="sd">    NC_meanExpression : numpy.ndarray</span>
<span class="sd">        Mean expression values for the niche cell type.</span>
<span class="sd">    CC_popExpression : numpy.ndarray</span>
<span class="sd">        Population expression values for the central cell type.</span>
<span class="sd">    NC_popExpression : numpy.ndarray</span>
<span class="sd">        Population expression values for the niche cell type.</span>
<span class="sd">    number_of_top_genes_to_print : int</span>
<span class="sd">        Number of top genes to include in the output.</span>

<span class="sd">    Returns:</span>
<span class="sd">    --------</span>
<span class="sd">    cc_genes : list of str</span>
<span class="sd">        List of significant genes for the cell type of interest.</span>
<span class="sd">    nc_genes : list of str</span>
<span class="sd">        List of significant genes for the non-host cell type.</span>
<span class="sd">    cc_genes5 : list of list</span>
<span class="sd">        Top genes for the cell type of interest with their PCA scores.</span>
<span class="sd">    nc_genes5 : list of list</span>
<span class="sd">        Top genes for the non-host cell type with their PCA scores.</span>
<span class="sd">    Found1 : list of list</span>
<span class="sd">        Ligand-receptor pairs with ligands in the cell type of interest and receptors in the non-host cell type.</span>
<span class="sd">    Found2 : list of list</span>
<span class="sd">        Ligand-receptor pairs with ligands in the non-host cell type and receptors in the cell type of interest.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">listofallLR</span><span class="o">=</span><span class="p">{}</span>
    <span class="n">uniqueLRpairs</span><span class="o">=</span><span class="p">{}</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">totalLRpairs</span><span class="p">)):</span>
        <span class="n">l</span><span class="o">=</span><span class="n">totalLRpairs</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">r</span><span class="o">=</span><span class="n">totalLRpairs</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">listofallLR</span><span class="p">[</span><span class="n">l</span><span class="p">]</span><span class="o">=</span><span class="mi">1</span>
        <span class="n">listofallLR</span><span class="p">[</span><span class="n">r</span><span class="p">]</span><span class="o">=</span><span class="mi">1</span>
        <span class="n">name</span><span class="o">=</span><span class="n">l</span><span class="o">+</span><span class="s1">&#39;--&#39;</span><span class="o">+</span><span class="n">r</span>
        <span class="k">if</span> <span class="n">name</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">uniqueLRpairs</span><span class="p">:</span>
            <span class="n">uniqueLRpairs</span><span class="p">[</span><span class="n">name</span><span class="p">]</span><span class="o">=</span><span class="mi">1</span>

    <span class="n">first</span><span class="o">=</span><span class="n">PCA</span><span class="p">[:,</span><span class="n">CCPC</span><span class="p">]</span>
    <span class="n">second</span><span class="o">=</span><span class="n">NH_PCA</span><span class="p">[:,</span><span class="n">NCPC</span><span class="p">]</span>
    <span class="n">ind1</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="o">-</span><span class="nb">abs</span><span class="p">(</span><span class="n">first</span><span class="p">))</span>
    <span class="n">ind2</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="o">-</span><span class="nb">abs</span><span class="p">(</span><span class="n">second</span><span class="p">))</span>

    <span class="n">cc_genes</span><span class="o">=</span><span class="p">[]</span>
    <span class="n">cc_genes2</span><span class="o">=</span><span class="p">[]</span>
    <span class="n">cc_genes5</span><span class="o">=</span><span class="p">[]</span>

    <span class="n">nc_genes</span><span class="o">=</span><span class="p">[]</span>
    <span class="n">nc_genes2</span><span class="o">=</span><span class="p">[]</span>
    <span class="n">nc_genes5</span><span class="o">=</span><span class="p">[]</span>


    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">number_of_top_genes_to_print</span><span class="p">):</span>
            <span class="n">cc_genes5</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">gene</span><span class="p">[</span><span class="n">ind1</span><span class="p">[</span><span class="n">i</span><span class="p">]],</span><span class="s1">&#39;</span><span class="si">%0.2f</span><span class="s1">&#39;</span><span class="o">%</span><span class="n">first</span><span class="p">[</span><span class="n">ind1</span><span class="p">[</span><span class="n">i</span><span class="p">]]])</span>

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">number_of_top_genes_to_print</span><span class="p">):</span>
            <span class="n">nc_genes5</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">gene</span><span class="p">[</span> <span class="n">ind2</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="p">],</span><span class="s1">&#39;</span><span class="si">%0.2f</span><span class="s1">&#39;</span><span class="o">%</span><span class="n">second</span><span class="p">[</span> <span class="n">ind2</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="p">]])</span>


    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">ind1</span><span class="p">)):</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">first</span><span class="p">[</span><span class="n">ind1</span><span class="p">[</span><span class="n">i</span><span class="p">]])</span><span class="o">&gt;</span><span class="n">LRcutoff</span><span class="p">:</span>
        <span class="c1">#if (first[ind1[i]])&lt;-0.4:</span>
            <span class="n">cc_genes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gene</span><span class="p">[</span><span class="n">ind1</span><span class="p">[</span><span class="n">i</span><span class="p">]])</span>
            <span class="k">if</span> <span class="n">gene</span><span class="p">[</span><span class="n">ind1</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="ow">in</span> <span class="n">listofallLR</span><span class="p">:</span>
                <span class="n">cc_genes2</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">gene</span><span class="p">[</span><span class="n">ind1</span><span class="p">[</span><span class="n">i</span><span class="p">]],</span><span class="s1">&#39;</span><span class="si">%0.2f</span><span class="s1">&#39;</span><span class="o">%</span><span class="n">first</span><span class="p">[</span><span class="n">ind1</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span> <span class="p">,</span><span class="n">CC_meanExpression</span><span class="p">[</span><span class="n">ind1</span><span class="p">[</span><span class="n">i</span><span class="p">]],</span><span class="n">CC_popExpression</span><span class="p">[</span><span class="n">ind1</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span>       <span class="p">])</span>



    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">ind2</span><span class="p">)):</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">second</span><span class="p">[</span><span class="n">ind2</span><span class="p">[</span><span class="n">i</span><span class="p">]])</span><span class="o">&gt;</span><span class="n">LRcutoff</span><span class="p">:</span>
        <span class="c1">#if (second[ind2[i]])&lt;-0.4:</span>
            <span class="n">nc_genes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gene</span><span class="p">[</span><span class="n">ind2</span><span class="p">[</span><span class="n">i</span><span class="p">]])</span>
            <span class="k">if</span> <span class="n">gene</span><span class="p">[</span><span class="n">ind2</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="ow">in</span> <span class="n">listofallLR</span><span class="p">:</span>
                <span class="n">nc_genes2</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">gene</span><span class="p">[</span> <span class="n">ind2</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="p">],</span><span class="s1">&#39;</span><span class="si">%0.2f</span><span class="s1">&#39;</span><span class="o">%</span><span class="n">second</span><span class="p">[</span> <span class="n">ind2</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="p">],</span> <span class="n">NC_meanExpression</span><span class="p">[</span><span class="n">ind2</span><span class="p">[</span><span class="n">i</span><span class="p">]],</span><span class="n">NC_popExpression</span><span class="p">[</span><span class="n">ind2</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span>   <span class="p">])</span>

    <span class="n">Found1</span><span class="o">=</span><span class="p">[]</span>
    <span class="n">Found2</span><span class="o">=</span><span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">cc_genes2</span><span class="p">)):</span>
        <span class="n">cc</span><span class="o">=</span><span class="n">cc_genes2</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">nc_genes2</span><span class="p">)):</span>
            <span class="n">nc</span><span class="o">=</span><span class="n">nc_genes2</span><span class="p">[</span><span class="n">j</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
            <span class="n">name1</span><span class="o">=</span><span class="n">cc</span><span class="o">+</span><span class="s1">&#39;--&#39;</span><span class="o">+</span><span class="n">nc</span> <span class="c1"># lig in CC and rec in NC</span>
            <span class="n">name2</span><span class="o">=</span><span class="n">nc</span><span class="o">+</span><span class="s1">&#39;--&#39;</span><span class="o">+</span><span class="n">cc</span> <span class="c1"># lig in NC and rec in CC</span>
            <span class="k">if</span> <span class="n">name1</span> <span class="ow">in</span> <span class="n">uniqueLRpairs</span><span class="p">:</span>
                <span class="n">Found1</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">cc_genes2</span><span class="p">[</span><span class="n">i</span><span class="p">],</span><span class="n">nc_genes2</span><span class="p">[</span><span class="n">j</span><span class="p">]</span> <span class="p">])</span>  <span class="c1"># lig in CC and rec in NC</span>
            <span class="k">if</span> <span class="n">name2</span> <span class="ow">in</span> <span class="n">uniqueLRpairs</span><span class="p">:</span>
                <span class="n">Found2</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">nc_genes2</span><span class="p">[</span><span class="n">j</span><span class="p">],</span><span class="n">cc_genes2</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="p">])</span>  <span class="c1"># lig in NC and rec in CC</span>

    <span class="k">return</span> <span class="n">cc_genes</span><span class="p">,</span> <span class="n">nc_genes</span><span class="p">,</span><span class="n">cc_genes5</span><span class="p">,</span><span class="n">nc_genes5</span><span class="p">,</span><span class="n">Found1</span><span class="p">,</span><span class="n">Found2</span></div>



<div class="viewcode-block" id="sorting_of_factors_for_showing_the_value_in_excelsheet">
<a class="viewcode-back" href="../../nico_covariations.html#nico_covariations.Covariations.sorting_of_factors_for_showing_the_value_in_excelsheet">[docs]</a>
<span class="k">def</span> <span class="nf">sorting_of_factors_for_showing_the_value_in_excelsheet</span><span class="p">(</span><span class="n">CC_corr</span><span class="p">,</span><span class="n">no_of_pc</span><span class="p">,</span><span class="n">gene</span><span class="p">,</span><span class="n">genenames</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Sort factor values for displaying in an Excel sheet.</span>

<span class="sd">    This helper function is used in `make_excel_sheet_for_gene_correlation` to sort the factor values from correlation data. It separates the sorted values into all genes and a subset of common genes.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    -----------</span>
<span class="sd">    CC_corr : numpy.ndarray</span>
<span class="sd">        Array of correlation values, where rows represent genes and columns represent principal components.</span>
<span class="sd">    no_of_pc : int</span>
<span class="sd">        Number of principal components.</span>
<span class="sd">    gene : list of str</span>
<span class="sd">        List of gene names corresponding to the rows in `CC_corr`.</span>
<span class="sd">    genenames : list of str</span>
<span class="sd">        List of gene names to be included in the common subset.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">headersave_full</span><span class="o">=</span><span class="p">[]</span>
    <span class="n">headersave_common</span><span class="o">=</span><span class="p">[]</span>
    <span class="n">sort_full</span><span class="o">=</span><span class="p">[]</span>
    <span class="n">sort_common</span><span class="o">=</span><span class="p">[]</span>
    <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">no_of_pc</span><span class="p">):</span>
            <span class="n">sort_full</span><span class="o">.</span><span class="n">append</span><span class="p">([])</span>
            <span class="n">sort_common</span><span class="o">.</span><span class="n">append</span><span class="p">([])</span>
    <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">CC_corr</span><span class="p">)):</span>
            <span class="n">ind</span><span class="o">=~</span><span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">CC_corr</span><span class="p">[</span><span class="n">j</span><span class="p">])</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">ind</span><span class="o">==</span><span class="kc">True</span><span class="p">:</span>
                <span class="c1">#ax.text(CC_corr[j,0],CC_corr[j,1],gene[j],fontsize=5)</span>
                <span class="n">header</span><span class="o">=</span><span class="p">[</span><span class="n">gene</span><span class="p">[</span><span class="n">j</span><span class="p">]]</span>
                <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">no_of_pc</span><span class="p">):</span>
                    <span class="n">sort_full</span><span class="p">[</span><span class="n">k</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">CC_corr</span><span class="p">[</span><span class="n">j</span><span class="p">,</span><span class="n">k</span><span class="p">])</span> <span class="c1">#without absolute</span>
                    <span class="n">header</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">CC_corr</span><span class="p">[</span><span class="n">j</span><span class="p">,</span><span class="n">k</span><span class="p">])</span>
                <span class="n">headersave_full</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">header</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">gene</span><span class="p">[</span><span class="n">j</span><span class="p">]</span> <span class="ow">in</span> <span class="n">genenames</span><span class="p">:</span>
                    <span class="n">headersave_common</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">header</span><span class="p">)</span>
                    <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">no_of_pc</span><span class="p">):</span>
                        <span class="n">sort_common</span><span class="p">[</span><span class="n">k</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">CC_corr</span><span class="p">[</span><span class="n">j</span><span class="p">,</span><span class="n">k</span><span class="p">])</span> <span class="c1">#without absolute</span>
    <span class="k">return</span> <span class="n">headersave_full</span><span class="p">,</span><span class="n">headersave_common</span><span class="p">,</span><span class="n">sort_full</span><span class="p">,</span><span class="n">sort_common</span></div>





<div class="viewcode-block" id="triangulation_for_triheatmap">
<a class="viewcode-back" href="../../nico_covariations.html#nico_covariations.Covariations.triangulation_for_triheatmap">[docs]</a>
<span class="k">def</span> <span class="nf">triangulation_for_triheatmap</span><span class="p">(</span><span class="n">M</span><span class="p">,</span> <span class="n">N</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create triangulation for plotting a ligand-receptor map.</span>

<span class="sd">    This helper function generates the triangulation needed for plotting a rectangle four faces in</span>
<span class="sd">    the `plot_ligand_receptor_in_interacting_celltypes` function. It constructs the vertices</span>
<span class="sd">    and triangles required for visualizing the ligand-receptor interactions on a heatmap.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    -----------</span>
<span class="sd">    M : int</span>
<span class="sd">        Number of columns in the heatmap.</span>
<span class="sd">    N : int</span>
<span class="sd">        Number of rows in the heatmap.</span>

<span class="sd">    Returns:</span>
<span class="sd">    --------</span>
<span class="sd">    list of matplotlib.tri.Triangulation</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">xv</span><span class="p">,</span> <span class="n">yv</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">meshgrid</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="o">-</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">M</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="o">-</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">N</span><span class="p">))</span>  <span class="c1"># vertices of the little squares</span>
    <span class="n">xc</span><span class="p">,</span> <span class="n">yc</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">meshgrid</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">M</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">N</span><span class="p">))</span>  <span class="c1"># centers of the little squares</span>
    <span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">xv</span><span class="o">.</span><span class="n">ravel</span><span class="p">(),</span> <span class="n">xc</span><span class="o">.</span><span class="n">ravel</span><span class="p">()])</span>
    <span class="n">y</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">yv</span><span class="o">.</span><span class="n">ravel</span><span class="p">(),</span> <span class="n">yc</span><span class="o">.</span><span class="n">ravel</span><span class="p">()])</span>
    <span class="n">cstart</span> <span class="o">=</span> <span class="p">(</span><span class="n">M</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="n">N</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>  <span class="c1"># indices of the centers</span>

    <span class="n">trianglesN</span> <span class="o">=</span> <span class="p">[(</span><span class="n">i</span> <span class="o">+</span> <span class="n">j</span> <span class="o">*</span> <span class="p">(</span><span class="n">M</span> <span class="o">+</span> <span class="mi">1</span><span class="p">),</span> <span class="n">i</span> <span class="o">+</span> <span class="mi">1</span> <span class="o">+</span> <span class="n">j</span> <span class="o">*</span> <span class="p">(</span><span class="n">M</span> <span class="o">+</span> <span class="mi">1</span><span class="p">),</span> <span class="n">cstart</span> <span class="o">+</span> <span class="n">i</span> <span class="o">+</span> <span class="n">j</span> <span class="o">*</span> <span class="n">M</span><span class="p">)</span>
                  <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">N</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">M</span><span class="p">)]</span>
    <span class="n">trianglesE</span> <span class="o">=</span> <span class="p">[(</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span> <span class="o">+</span> <span class="n">j</span> <span class="o">*</span> <span class="p">(</span><span class="n">M</span> <span class="o">+</span> <span class="mi">1</span><span class="p">),</span> <span class="n">i</span> <span class="o">+</span> <span class="mi">1</span> <span class="o">+</span> <span class="p">(</span><span class="n">j</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="n">M</span> <span class="o">+</span> <span class="mi">1</span><span class="p">),</span> <span class="n">cstart</span> <span class="o">+</span> <span class="n">i</span> <span class="o">+</span> <span class="n">j</span> <span class="o">*</span> <span class="n">M</span><span class="p">)</span>
                  <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">N</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">M</span><span class="p">)]</span>
    <span class="n">trianglesS</span> <span class="o">=</span> <span class="p">[(</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span> <span class="o">+</span> <span class="p">(</span><span class="n">j</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="n">M</span> <span class="o">+</span> <span class="mi">1</span><span class="p">),</span> <span class="n">i</span> <span class="o">+</span> <span class="p">(</span><span class="n">j</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="n">M</span> <span class="o">+</span> <span class="mi">1</span><span class="p">),</span> <span class="n">cstart</span> <span class="o">+</span> <span class="n">i</span> <span class="o">+</span> <span class="n">j</span> <span class="o">*</span> <span class="n">M</span><span class="p">)</span>
                  <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">N</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">M</span><span class="p">)]</span>
    <span class="n">trianglesW</span> <span class="o">=</span> <span class="p">[(</span><span class="n">i</span> <span class="o">+</span> <span class="p">(</span><span class="n">j</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="n">M</span> <span class="o">+</span> <span class="mi">1</span><span class="p">),</span> <span class="n">i</span> <span class="o">+</span> <span class="n">j</span> <span class="o">*</span> <span class="p">(</span><span class="n">M</span> <span class="o">+</span> <span class="mi">1</span><span class="p">),</span> <span class="n">cstart</span> <span class="o">+</span> <span class="n">i</span> <span class="o">+</span> <span class="n">j</span> <span class="o">*</span> <span class="n">M</span><span class="p">)</span>
                  <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">N</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">M</span><span class="p">)]</span>
    <span class="k">return</span> <span class="p">[</span><span class="n">Triangulation</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">triangles</span><span class="p">)</span> <span class="k">for</span> <span class="n">triangles</span> <span class="ow">in</span> <span class="p">[</span><span class="n">trianglesN</span><span class="p">,</span> <span class="n">trianglesE</span><span class="p">,</span> <span class="n">trianglesS</span><span class="p">,</span> <span class="n">trianglesW</span><span class="p">]]</span></div>



<div class="viewcode-block" id="plot_ligand_receptor_in_interacting_celltypes">
<a class="viewcode-back" href="../../nico_covariations.html#nico_covariations.Covariations.plot_ligand_receptor_in_interacting_celltypes">[docs]</a>
<span class="k">def</span>  <span class="nf">plot_ligand_receptor_in_interacting_celltypes</span><span class="p">(</span><span class="n">CC_celltype_name</span><span class="p">,</span><span class="n">NC_celltype_name</span><span class="p">,</span><span class="n">logRegScore</span><span class="p">,</span><span class="n">pc1</span><span class="p">,</span><span class="n">pc2</span><span class="p">,</span><span class="n">ridgeRegScore</span><span class="p">,</span><span class="n">pvalue</span><span class="p">,</span><span class="n">Found1</span><span class="p">,</span><span class="n">Found2</span><span class="p">,</span><span class="n">saveLRplots</span><span class="p">,</span><span class="n">LR_plot_Exp_thres</span><span class="p">,</span><span class="n">saveas</span><span class="p">,</span><span class="n">transparent_mode</span><span class="p">,</span><span class="n">showit</span><span class="p">,</span><span class="n">figsize</span><span class="p">,</span><span class="n">flag</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Plot ligand-receptor interactions for interacting cell types.</span>

<span class="sd">    This helper function is used in `find_LR_interactions_in_interacting_cell_types` to plot</span>
<span class="sd">    rectangle p-value figures representing ligand-receptor interactions between two cell types.</span>

<span class="sd">    The Y-axis shows the central cell type factors, and the X-axis shows the colocalized neighborhood cell type factors.</span>
<span class="sd">    The circle size denotes the p-values, and the circle size scales with significance.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    -----------</span>
<span class="sd">    CC_celltype_name : str</span>
<span class="sd">        Name of the central cell type.</span>
<span class="sd">    NC_celltype_name : str</span>
<span class="sd">        Name of the neighborhood cell type.</span>
<span class="sd">    logRegScore : float</span>
<span class="sd">        Logistic regression score.</span>
<span class="sd">    pc1 : int</span>
<span class="sd">        Principal component for the central cell type.</span>
<span class="sd">    pc2 : int</span>
<span class="sd">        Principal component for the neighborhood cell type.</span>
<span class="sd">    ridgeRegScore : float</span>
<span class="sd">        Ridge regression score.</span>
<span class="sd">    pvalue : float</span>
<span class="sd">        P-value for the interaction.</span>
<span class="sd">    Found1 : list</span>
<span class="sd">        List of found ligand-receptor interactions where the ligand is in the central cell type.</span>
<span class="sd">    Found2 : list</span>
<span class="sd">        List of found ligand-receptor interactions where the ligand is in the neighborhood cell type.</span>
<span class="sd">    saveLRplots : str</span>
<span class="sd">        Directory to save the ligand-receptor plots.</span>
<span class="sd">    LR_plot_Exp_thres : float</span>
<span class="sd">        Expression threshold for plotting.</span>
<span class="sd">    saveas : str</span>
<span class="sd">        File format to save the plots (e.g., &#39;png&#39;, &#39;pdf&#39;).</span>
<span class="sd">    transparent_mode : bool</span>
<span class="sd">        Whether to save the plots with a transparent background.</span>
<span class="sd">    showit : bool</span>
<span class="sd">        Whether to display the plots.</span>
<span class="sd">    figsize : tuple</span>
<span class="sd">        Size of the figure.</span>
<span class="sd">    flag : str</span>
<span class="sd">        Flag indicating which interactions to plot (&#39;First&#39;, &#39;Second&#39;, &#39;Both&#39;).</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">xfact</span><span class="o">=</span><span class="n">figsize</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">/</span><span class="mf">34.0</span>
    <span class="n">yfact</span><span class="o">=</span><span class="n">figsize</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">/</span><span class="mf">44.0</span>
    <span class="n">LRFigSize</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">figsize</span><span class="p">)</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
    <span class="n">ligand</span><span class="o">=</span><span class="p">[]</span>
    <span class="n">receptor</span><span class="o">=</span><span class="p">[]</span>
    <span class="n">fact_lig</span><span class="o">=</span><span class="p">[]</span>
    <span class="n">fact_rec</span><span class="o">=</span><span class="p">[]</span>
    <span class="n">popExp_lig</span><span class="o">=</span><span class="p">[]</span>
    <span class="n">popExp_rec</span><span class="o">=</span><span class="p">[]</span>
    <span class="n">A</span><span class="o">=</span><span class="p">[]</span>
    <span class="n">B</span><span class="o">=</span><span class="p">[]</span>
    <span class="k">if</span> <span class="n">flag</span><span class="o">==</span><span class="s1">&#39;First&#39;</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">ele</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">Found1</span><span class="p">)):</span>
            <span class="n">ligExpCellPop</span><span class="o">=</span><span class="n">Found1</span><span class="p">[</span><span class="n">ele</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="mi">3</span><span class="p">]</span>
            <span class="n">recExpCellPop</span><span class="o">=</span><span class="n">Found1</span><span class="p">[</span><span class="n">ele</span><span class="p">][</span><span class="mi">1</span><span class="p">][</span><span class="mi">3</span><span class="p">]</span>
            <span class="k">if</span> <span class="p">((</span><span class="n">ligExpCellPop</span><span class="o">&gt;</span><span class="n">LR_plot_Exp_thres</span><span class="p">)</span><span class="o">&amp;</span><span class="p">(</span><span class="n">recExpCellPop</span><span class="o">&gt;</span><span class="n">LR_plot_Exp_thres</span><span class="p">)):</span>
                <span class="n">ligand</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">Found1</span><span class="p">[</span><span class="n">ele</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span>
                <span class="n">receptor</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">Found1</span><span class="p">[</span><span class="n">ele</span><span class="p">][</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span>
                <span class="n">fact_lig</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">Found1</span><span class="p">[</span><span class="n">ele</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">]))</span>
                <span class="n">fact_rec</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">Found1</span><span class="p">[</span><span class="n">ele</span><span class="p">][</span><span class="mi">1</span><span class="p">][</span><span class="mi">1</span><span class="p">]))</span>
                <span class="n">popExp_lig</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">Found1</span><span class="p">[</span><span class="n">ele</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="mi">3</span><span class="p">])</span>
                <span class="n">popExp_rec</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">Found1</span><span class="p">[</span><span class="n">ele</span><span class="p">][</span><span class="mi">1</span><span class="p">][</span><span class="mi">3</span><span class="p">])</span>
                <span class="n">A</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">CC_celltype_name</span><span class="o">+</span><span class="s1">&#39;(cc)_Fa&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">pc1</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;_&#39;</span><span class="o">+</span><span class="n">Found1</span><span class="p">[</span><span class="n">ele</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span>
                <span class="n">B</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">NC_celltype_name</span><span class="o">+</span><span class="s1">&#39;(nc)_Fa&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">pc2</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;_&#39;</span><span class="o">+</span><span class="n">Found1</span><span class="p">[</span><span class="n">ele</span><span class="p">][</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span>

    <span class="k">if</span> <span class="n">flag</span><span class="o">==</span><span class="s1">&#39;Second&#39;</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">ele</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">Found2</span><span class="p">)):</span>
            <span class="n">ligExpCellPop</span><span class="o">=</span><span class="n">Found2</span><span class="p">[</span><span class="n">ele</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="mi">3</span><span class="p">]</span>
            <span class="n">recExpCellPop</span><span class="o">=</span><span class="n">Found2</span><span class="p">[</span><span class="n">ele</span><span class="p">][</span><span class="mi">1</span><span class="p">][</span><span class="mi">3</span><span class="p">]</span>
            <span class="k">if</span> <span class="p">((</span><span class="n">ligExpCellPop</span><span class="o">&gt;</span><span class="n">LR_plot_Exp_thres</span><span class="p">)</span><span class="o">&amp;</span><span class="p">(</span><span class="n">recExpCellPop</span><span class="o">&gt;</span><span class="n">LR_plot_Exp_thres</span><span class="p">)):</span>
                <span class="n">ligand</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">Found2</span><span class="p">[</span><span class="n">ele</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span>
                <span class="n">receptor</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">Found2</span><span class="p">[</span><span class="n">ele</span><span class="p">][</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span>
                <span class="n">fact_lig</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">Found2</span><span class="p">[</span><span class="n">ele</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">]))</span>
                <span class="n">fact_rec</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">Found2</span><span class="p">[</span><span class="n">ele</span><span class="p">][</span><span class="mi">1</span><span class="p">][</span><span class="mi">1</span><span class="p">]))</span>
                <span class="n">popExp_lig</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">Found2</span><span class="p">[</span><span class="n">ele</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="mi">3</span><span class="p">])</span>
                <span class="n">popExp_rec</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">Found2</span><span class="p">[</span><span class="n">ele</span><span class="p">][</span><span class="mi">1</span><span class="p">][</span><span class="mi">3</span><span class="p">])</span>
                <span class="n">A</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">NC_celltype_name</span><span class="o">+</span><span class="s1">&#39;(nc)_Fa&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">pc2</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;_&#39;</span><span class="o">+</span><span class="n">Found2</span><span class="p">[</span><span class="n">ele</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span>
                <span class="n">B</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">CC_celltype_name</span><span class="o">+</span><span class="s1">&#39;(cc)_Fa&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">pc1</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;_&#39;</span><span class="o">+</span><span class="n">Found2</span><span class="p">[</span><span class="n">ele</span><span class="p">][</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span>
    <span class="k">if</span> <span class="n">flag</span><span class="o">==</span><span class="s1">&#39;Both&#39;</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">ele</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">Found1</span><span class="p">)):</span>
            <span class="c1">#header=[&#39;Ligand(A)&#39;,&#39;Receptor(B)&#39;,&#39;GeneCor(Lig)&#39;,&#39;GeneCor(Rec)&#39;,&#39;Receptor(A)&#39;,&#39;Ligand(B)&#39;,&#39;GeneCor(Rec)&#39;,&#39;GeneCor(Lig)&#39;]</span>
            <span class="c1">#header[11]=Found1[ele][0][2]</span>
            <span class="c1">#header[12]=Found1[ele][1][2]</span>
            <span class="n">ligExpCellPop</span><span class="o">=</span><span class="n">Found1</span><span class="p">[</span><span class="n">ele</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="mi">3</span><span class="p">]</span>
            <span class="n">recExpCellPop</span><span class="o">=</span><span class="n">Found1</span><span class="p">[</span><span class="n">ele</span><span class="p">][</span><span class="mi">1</span><span class="p">][</span><span class="mi">3</span><span class="p">]</span>
            <span class="k">if</span> <span class="p">((</span><span class="n">ligExpCellPop</span><span class="o">&gt;</span><span class="n">LR_plot_Exp_thres</span><span class="p">)</span><span class="o">&amp;</span><span class="p">(</span><span class="n">recExpCellPop</span><span class="o">&gt;</span><span class="n">LR_plot_Exp_thres</span><span class="p">)):</span>
                <span class="n">ligand</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">Found1</span><span class="p">[</span><span class="n">ele</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span>
                <span class="n">receptor</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">Found1</span><span class="p">[</span><span class="n">ele</span><span class="p">][</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span>
                <span class="n">fact_lig</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">Found1</span><span class="p">[</span><span class="n">ele</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">]))</span>
                <span class="n">fact_rec</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">Found1</span><span class="p">[</span><span class="n">ele</span><span class="p">][</span><span class="mi">1</span><span class="p">][</span><span class="mi">1</span><span class="p">]))</span>
                <span class="n">popExp_lig</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">Found1</span><span class="p">[</span><span class="n">ele</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="mi">3</span><span class="p">])</span>
                <span class="n">popExp_rec</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">Found1</span><span class="p">[</span><span class="n">ele</span><span class="p">][</span><span class="mi">1</span><span class="p">][</span><span class="mi">3</span><span class="p">])</span>
                <span class="n">A</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">CC_celltype_name</span><span class="o">+</span><span class="s1">&#39;(cc)_Fa&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">pc1</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;_&#39;</span><span class="o">+</span><span class="n">Found1</span><span class="p">[</span><span class="n">ele</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span>
                <span class="n">B</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">NC_celltype_name</span><span class="o">+</span><span class="s1">&#39;(nc)_Fa&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">pc2</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;_&#39;</span><span class="o">+</span><span class="n">Found1</span><span class="p">[</span><span class="n">ele</span><span class="p">][</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span>

        <span class="k">for</span> <span class="n">ele</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">Found2</span><span class="p">)):</span>
            <span class="n">ligExpCellPop</span><span class="o">=</span><span class="n">Found2</span><span class="p">[</span><span class="n">ele</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="mi">3</span><span class="p">]</span>
            <span class="n">recExpCellPop</span><span class="o">=</span><span class="n">Found2</span><span class="p">[</span><span class="n">ele</span><span class="p">][</span><span class="mi">1</span><span class="p">][</span><span class="mi">3</span><span class="p">]</span>
            <span class="k">if</span> <span class="p">((</span><span class="n">ligExpCellPop</span><span class="o">&gt;</span><span class="n">LR_plot_Exp_thres</span><span class="p">)</span><span class="o">&amp;</span><span class="p">(</span><span class="n">recExpCellPop</span><span class="o">&gt;</span><span class="n">LR_plot_Exp_thres</span><span class="p">)):</span>
                <span class="n">ligand</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">Found2</span><span class="p">[</span><span class="n">ele</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span>
                <span class="n">receptor</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">Found2</span><span class="p">[</span><span class="n">ele</span><span class="p">][</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span>
                <span class="n">fact_lig</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">Found2</span><span class="p">[</span><span class="n">ele</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">]))</span>
                <span class="n">fact_rec</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">Found2</span><span class="p">[</span><span class="n">ele</span><span class="p">][</span><span class="mi">1</span><span class="p">][</span><span class="mi">1</span><span class="p">]))</span>
                <span class="n">popExp_lig</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">Found2</span><span class="p">[</span><span class="n">ele</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="mi">3</span><span class="p">])</span>
                <span class="n">popExp_rec</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">Found2</span><span class="p">[</span><span class="n">ele</span><span class="p">][</span><span class="mi">1</span><span class="p">][</span><span class="mi">3</span><span class="p">])</span>
                <span class="n">A</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">NC_celltype_name</span><span class="o">+</span><span class="s1">&#39;(nc)_Fa&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">pc2</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;_&#39;</span><span class="o">+</span><span class="n">Found2</span><span class="p">[</span><span class="n">ele</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span>
                <span class="n">B</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">CC_celltype_name</span><span class="o">+</span><span class="s1">&#39;(cc)_Fa&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">pc1</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;_&#39;</span><span class="o">+</span><span class="n">Found2</span><span class="p">[</span><span class="n">ele</span><span class="p">][</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span>


    <span class="c1">#print(CC_celltype_name,NC_celltype_name,flag,logRegScore,len(Found1),len(Found2),len(ligand),len(A),len(B),LRFigSize1) #,len(nA),len(nB)</span>
    <span class="k">if</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">A</span><span class="p">)</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">)</span><span class="o">&amp;</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">B</span><span class="p">)</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">):</span>
            <span class="n">nA</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">A</span><span class="p">))</span>
            <span class="n">nB</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">B</span><span class="p">))</span>
            <span class="n">fact_lig</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">fact_lig</span><span class="p">)</span>
            <span class="n">fact_rec</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">fact_rec</span><span class="p">)</span>
            <span class="n">popExp_rec</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">popExp_rec</span><span class="p">)</span>
            <span class="n">popExp_lig</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">popExp_lig</span><span class="p">)</span>

            <span class="n">p1</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">fact_lig</span><span class="p">)</span>
            <span class="n">p2</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">fact_rec</span><span class="p">)</span>
            <span class="n">p3</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">popExp_rec</span><span class="p">)</span>
            <span class="n">p4</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">popExp_lig</span><span class="p">)</span>

            <span class="n">q1</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="n">fact_lig</span><span class="p">)</span>
            <span class="n">q2</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="n">fact_rec</span><span class="p">)</span>
            <span class="n">q3</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="n">popExp_rec</span><span class="p">)</span>
            <span class="n">q4</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="n">popExp_lig</span><span class="p">)</span>

            <span class="n">fmin</span><span class="o">=</span><span class="nb">min</span><span class="p">(</span><span class="n">q1</span><span class="p">,</span><span class="n">q2</span><span class="p">)</span>
            <span class="n">fmax</span><span class="o">=</span><span class="nb">max</span><span class="p">(</span><span class="n">p1</span><span class="p">,</span><span class="n">p2</span><span class="p">)</span>
            <span class="n">pmin</span><span class="o">=</span><span class="nb">min</span><span class="p">(</span><span class="n">q3</span><span class="p">,</span><span class="n">q4</span><span class="p">)</span>
            <span class="n">pmax</span><span class="o">=</span><span class="nb">max</span><span class="p">(</span><span class="n">p3</span><span class="p">,</span><span class="n">p4</span><span class="p">)</span>

            <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;cols&#39;</span><span class="p">:</span> <span class="n">B</span><span class="p">,</span>
                               <span class="s1">&#39;rows&#39;</span><span class="p">:</span> <span class="n">A</span><span class="p">,</span>
                               <span class="s1">&#39;north&#39;</span><span class="p">:</span> <span class="n">fact_lig</span><span class="p">,</span>
                               <span class="s1">&#39;south&#39;</span><span class="p">:</span> <span class="n">fact_rec</span><span class="p">,</span>
                               <span class="s1">&#39;east&#39;</span><span class="p">:</span> <span class="n">popExp_rec</span><span class="p">,</span>
                               <span class="s1">&#39;west&#39;</span><span class="p">:</span>  <span class="n">popExp_lig</span><span class="p">})</span>

            <span class="n">df</span><span class="p">[</span><span class="s1">&#39;rows&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Categorical</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;rows&#39;</span><span class="p">],</span><span class="n">categories</span><span class="o">=</span><span class="n">nA</span><span class="p">)</span>  <span class="c1"># fix an ordering,</span>
            <span class="n">df_piv</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">pivot_table</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="s1">&#39;rows&#39;</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="s1">&#39;cols&#39;</span><span class="p">)</span>
            <span class="n">M</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">df_piv</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span> <span class="o">//</span> <span class="mi">4</span>
            <span class="n">N</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">df_piv</span><span class="p">)</span>

            <span class="k">if</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">M</span><span class="p">))</span><span class="o">&lt;</span><span class="mi">13</span><span class="p">)</span><span class="o">|</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">N</span><span class="p">))</span><span class="o">&lt;</span><span class="mi">13</span><span class="p">):</span>
                <span class="n">LRFigSize</span><span class="o">=</span><span class="n">figsize</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">LRFigSize</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">M</span><span class="p">))</span><span class="o">*</span><span class="n">xfact</span>
                <span class="n">LRFigSize</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">N</span><span class="p">))</span><span class="o">*</span><span class="n">yfact</span>

            <span class="c1">#print(&#39;\n\n&#39;,flag,range(M),range(N),CC_celltype_name,[xfact,yfact],LRFigSize)</span>

            <span class="n">values</span> <span class="o">=</span> <span class="p">[</span><span class="n">df_piv</span><span class="p">[</span><span class="nb">dir</span><span class="p">]</span> <span class="k">for</span> <span class="nb">dir</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;north&#39;</span><span class="p">,</span> <span class="s1">&#39;east&#39;</span><span class="p">,</span> <span class="s1">&#39;south&#39;</span><span class="p">,</span> <span class="s1">&#39;west&#39;</span><span class="p">]]</span>  <span class="c1"># these are the 4 column names in df</span>

            <span class="n">triangul</span> <span class="o">=</span> <span class="n">triangulation_for_triheatmap</span><span class="p">(</span><span class="n">M</span><span class="p">,</span> <span class="n">N</span><span class="p">)</span>
            <span class="c1">#cmaps = [&#39;RdYlBu&#39;] * 4</span>
            <span class="c1">#cmaps =[&#39;cool&#39;,&#39;copper&#39;,&#39;cool&#39;,&#39;copper&#39;]</span>
            <span class="n">cmaps</span> <span class="o">=</span><span class="p">[</span><span class="s1">&#39;Blues&#39;</span><span class="p">,</span><span class="s1">&#39;copper_r&#39;</span><span class="p">,</span><span class="s1">&#39;Blues&#39;</span><span class="p">,</span><span class="s1">&#39;copper_r&#39;</span><span class="p">]</span>
            <span class="c1">#norms = [plt.Normalize(0, 1) for _ in range(4)]</span>
            <span class="n">norms</span> <span class="o">=</span> <span class="p">[</span><span class="n">plt</span><span class="o">.</span><span class="n">Normalize</span><span class="p">(</span><span class="n">fmin</span><span class="p">,</span> <span class="n">fmax</span><span class="p">),</span><span class="n">plt</span><span class="o">.</span><span class="n">Normalize</span><span class="p">(</span><span class="n">pmin</span><span class="p">,</span> <span class="n">pmax</span><span class="p">),</span><span class="n">plt</span><span class="o">.</span><span class="n">Normalize</span><span class="p">(</span><span class="n">fmin</span><span class="p">,</span> <span class="n">fmax</span><span class="p">),</span><span class="n">plt</span><span class="o">.</span><span class="n">Normalize</span><span class="p">(</span><span class="n">pmin</span><span class="p">,</span> <span class="n">pmax</span><span class="p">)]</span>

            <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="n">LRFigSize</span><span class="p">)</span>

            <span class="n">imgs</span> <span class="o">=</span> <span class="p">[</span><span class="n">ax</span><span class="o">.</span><span class="n">tripcolor</span><span class="p">(</span><span class="n">t</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ravel</span><span class="p">(</span><span class="n">val</span><span class="p">),</span> <span class="n">norm</span><span class="o">=</span><span class="n">norm</span><span class="p">,</span><span class="n">cmap</span><span class="o">=</span><span class="n">cmap</span><span class="p">,</span><span class="n">ec</span><span class="o">=</span><span class="s1">&#39;white&#39;</span><span class="p">)</span>  <span class="c1">#norm=[]</span>
                    <span class="k">for</span> <span class="n">t</span><span class="p">,</span> <span class="n">val</span><span class="p">,</span> <span class="n">cmap</span><span class="p">,</span> <span class="n">norm</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">triangul</span><span class="p">,</span> <span class="n">values</span><span class="p">,</span> <span class="n">cmaps</span><span class="p">,</span> <span class="n">norms</span><span class="p">)]</span>

            <span class="c1">#ax.tick_params(length=0)</span>
            <span class="c1">#ax.set_title(&#39;localizationCoef=&#39;+&#39;%0.3f&#39;%np.unique(localized)+&#39;,regressionCoef=&#39;+&#39;%0.3f&#39;%np.unique(regCoff))</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">M</span><span class="p">))</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">set_xticklabels</span><span class="p">(</span><span class="n">df_piv</span><span class="p">[</span><span class="s1">&#39;north&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">columns</span><span class="p">,</span><span class="n">rotation</span><span class="o">=</span><span class="mi">90</span><span class="p">,</span><span class="n">style</span><span class="o">=</span><span class="s1">&#39;italic&#39;</span><span class="p">)</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">N</span><span class="p">))</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">set_yticklabels</span><span class="p">(</span><span class="n">df_piv</span><span class="o">.</span><span class="n">index</span><span class="p">,</span><span class="n">style</span><span class="o">=</span><span class="s1">&#39;italic&#39;</span><span class="p">)</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">invert_yaxis</span><span class="p">()</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">margins</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
            <span class="n">pvalue</span><span class="o">=</span><span class="s1">&#39;</span><span class="si">%0.2f</span><span class="s1">&#39;</span><span class="o">%</span><span class="p">(</span><span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">pvalue</span><span class="p">))</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">CC_celltype_name</span><span class="o">+</span><span class="s1">&#39;_Fa&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">pc1</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;, &#39;</span><span class="o">+</span><span class="n">NC_celltype_name</span><span class="o">+</span><span class="s1">&#39;_Fa&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">pc2</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;, SS=</span><span class="si">%0.3f</span><span class="s1">&#39;</span><span class="o">%</span><span class="n">logRegScore</span><span class="o">+</span><span class="s1">&#39;, RRS=</span><span class="si">%0.2f</span><span class="s1">&#39;</span><span class="o">%</span><span class="n">ridgeRegScore</span> <span class="o">+</span><span class="s1">&#39;, pv=&#39;</span><span class="o">+</span><span class="n">pvalue</span>  <span class="p">)</span>
            <span class="c1">#ax.set_aspect(&#39;equal&#39;, &#39;box&#39;)  # square cells</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">colorbar</span><span class="p">(</span><span class="n">imgs</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span><span class="n">label</span><span class="o">=</span><span class="s1">&#39;correlation with factors&#39;</span><span class="p">)</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">colorbar</span><span class="p">(</span><span class="n">imgs</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span><span class="n">label</span><span class="o">=</span><span class="s1">&#39;fraction of cells expressed&#39;</span><span class="p">)</span>

            <span class="c1">#plt.tight_layout()</span>

            <span class="n">savefname</span><span class="o">=</span><span class="n">remove_extra_character_from_name</span><span class="p">(</span><span class="n">CC_celltype_name</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;_Fa&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">pc1</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;_&#39;</span><span class="o">+</span><span class="n">remove_extra_character_from_name</span><span class="p">(</span><span class="n">NC_celltype_name</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;_Fa&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">pc2</span><span class="p">)</span>

            <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">saveLRplots</span><span class="o">+</span><span class="n">savefname</span><span class="o">+</span><span class="s1">&#39;.&#39;</span><span class="o">+</span><span class="n">saveas</span><span class="p">,</span><span class="n">bbox_inches</span><span class="o">=</span><span class="s1">&#39;tight&#39;</span><span class="p">,</span> <span class="n">transparent</span><span class="o">=</span><span class="n">transparent_mode</span><span class="p">,</span><span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">showit</span><span class="p">:</span>
                <span class="k">pass</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="s1">&#39;all&#39;</span><span class="p">)</span>
    <span class="k">return</span> <span class="mi">0</span></div>




<div class="viewcode-block" id="visualize_factors_in_scRNAseq_umap">
<a class="viewcode-back" href="../../nico_covariations.html#nico_covariations.Covariations.visualize_factors_in_scRNAseq_umap">[docs]</a>
<span class="k">def</span> <span class="nf">visualize_factors_in_scRNAseq_umap</span><span class="p">(</span><span class="nb">input</span><span class="p">,</span>
<span class="n">choose_interacting_celltype_pair</span><span class="p">,</span>
<span class="n">visualize_factors_id</span><span class="p">,</span>
<span class="n">umap_tag</span><span class="o">=</span><span class="s1">&#39;X_umap&#39;</span><span class="p">,</span>
<span class="n">msna</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span><span class="n">ms</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span><span class="n">cmap</span><span class="o">=</span><span class="n">matplotlib</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s2">&quot;image.cmap&quot;</span><span class="p">],</span>
<span class="n">saveas</span><span class="o">=</span><span class="s1">&#39;pdf&#39;</span><span class="p">,</span><span class="n">transparent_mode</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">showit</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span><span class="mf">3.5</span><span class="p">)):</span>

<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Visualize factors in scRNAseq UMAP embedding.</span>

<span class="sd">    This function visualizes the factors in single-cell RNA sequencing (scRNAseq) UMAP embeddings. It highlights the interactions between specified cell type pairs and their corresponding factor IDs.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    -----------</span>
<span class="sd">    input : str</span>
<span class="sd">        The primary input is the output from `gene_covariation_analysis`.</span>
<span class="sd">    choose_interacting_celltype_pair : list</span>
<span class="sd">        List defining the cell type single or in pairs to visualize. At least one cell type need to put by the user.</span>
<span class="sd">    visualize_factors_id : list</span>
<span class="sd">        List defining the factor IDs single or in pairs to visualize in the UMAP. The chosen factors analogous to defined cell types.</span>
<span class="sd">    umap_tag : str, optional</span>
<span class="sd">        The UMAP embedding tag in the .obsm field of the AnnData object (default is &#39;X_umap&#39;).</span>
<span class="sd">    msna : float, optional</span>
<span class="sd">        The marker size for non selected (NA) cell types (default is 0.1).</span>
<span class="sd">    ms : int, optional</span>
<span class="sd">        The marker size for selected cell types (default is 5).</span>
<span class="sd">    cmap : str, optional</span>
<span class="sd">        Colormap for visualizing factors (default is `matplotlib.rcParams[&quot;image.cmap&quot;]`).</span>
<span class="sd">    saveas : str, optional</span>
<span class="sd">        Format to save the figures (&#39;pdf&#39; or &#39;png&#39;) (default is &#39;pdf&#39;).</span>
<span class="sd">    transparent_mode : bool, optional</span>
<span class="sd">        Whether to save the figures with a transparent background (default is False).</span>
<span class="sd">    showit : bool, optional</span>
<span class="sd">        Whether to display the figures (default is True).</span>
<span class="sd">    figsize : tuple, optional</span>
<span class="sd">        Size of the figure (default is (8, 3.5)).</span>

<span class="sd">    Outputs:</span>
<span class="sd">    --------</span>
<span class="sd">    The factor visualization in scRNAseq embedding is saved in &quot;./nico_out/covariations_R*_F*/scRNAseq_factors_in_umap&quot;.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">original_h5ad</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">umap_plot_sc</span>
    <span class="n">umap</span><span class="o">=</span><span class="n">original_h5ad</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="n">umap_tag</span><span class="p">]</span>
    <span class="n">barcode</span><span class="o">=</span><span class="n">original_h5ad</span><span class="o">.</span><span class="n">obs_names</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span>
    <span class="n">barcode</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">barcode</span><span class="p">,(</span><span class="nb">len</span><span class="p">(</span><span class="n">barcode</span><span class="p">),</span><span class="mi">1</span><span class="p">))</span>
    <span class="n">umap_not_order</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">hstack</span><span class="p">((</span><span class="n">barcode</span><span class="p">,</span><span class="n">umap</span><span class="p">))</span>

    <span class="n">cellname</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">annotation_singlecell_barcode_id</span><span class="p">,(</span><span class="nb">len</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">annotation_singlecell_barcode_id</span><span class="p">),</span><span class="mi">1</span><span class="p">)</span> <span class="p">)</span>
    <span class="n">umap_data</span><span class="o">=</span><span class="n">sort_index_in_right_order</span><span class="p">(</span><span class="n">cellname</span><span class="p">,</span><span class="n">umap_not_order</span><span class="p">)</span>

    <span class="c1">#sc_ct_name=np.array([input.singlecell_unique_clusterid,input.singlecell_unique_clustername]))</span>
    <span class="c1">#sc_cluster=np.array([input.annotation_singlecell_barcode_id,input.annotation_singlecell_cluster_id])</span>

    <span class="n">PCA_of_sc_cluster_accordingto_spatial_clusterid</span><span class="p">,</span><span class="n">save_scFactors</span><span class="p">,</span><span class="n">save_spFactors</span><span class="o">=</span><span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">covariation_dir</span><span class="o">+</span><span class="s1">&#39;factors_info.p&#39;</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">))</span>


    <span class="n">t</span><span class="o">=</span><span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">no_of_pc</span><span class="p">):</span>
        <span class="n">t</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">H_sc</span><span class="o">=</span><span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">annotation_singlecell_barcode_id</span><span class="p">)):</span>
        <span class="n">cellid</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">annotation_singlecell_barcode_id</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">cellid</span> <span class="ow">in</span> <span class="n">save_scFactors</span><span class="p">:</span>
            <span class="n">H_sc</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">save_scFactors</span><span class="p">[</span><span class="n">cellid</span><span class="p">])</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">H_sc</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">t</span><span class="p">))</span>
    <span class="n">H_sc</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">H_sc</span><span class="p">)</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">visualize_factors_id</span><span class="p">)</span><span class="o">==</span><span class="mi">1</span><span class="p">:</span>
        <span class="n">fig</span><span class="p">,(</span><span class="n">ax</span><span class="p">)</span><span class="o">=</span><span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">)</span>
        <span class="n">v1</span><span class="o">=</span><span class="n">H_sc</span><span class="p">[:,(</span><span class="n">visualize_factors_id</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">-</span><span class="mi">1</span><span class="p">)]</span>
        <span class="n">CTname</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">singlecell_unique_clustername</span>
        <span class="n">cellsinCT</span><span class="o">=</span><span class="p">{}</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">annotation_singlecell_barcode_id</span><span class="p">)):</span>
            <span class="n">clu_id</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">annotation_singlecell_celltypename</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">clu_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">cellsinCT</span><span class="p">:</span>
                <span class="n">cellsinCT</span><span class="p">[</span><span class="n">clu_id</span><span class="p">]</span><span class="o">=</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">cellsinCT</span><span class="p">[</span><span class="n">clu_id</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>

        <span class="n">plot_all_ct</span><span class="p">(</span><span class="n">CTname</span><span class="p">,</span><span class="n">umap_data</span><span class="p">[:,[</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">]],</span><span class="n">cellsinCT</span><span class="p">,</span><span class="n">ax</span><span class="p">,[</span><span class="n">choose_interacting_celltype_pair</span><span class="p">[</span><span class="mi">0</span><span class="p">]],</span><span class="n">v1</span><span class="p">,</span><span class="n">cmap</span><span class="p">,</span><span class="n">ms</span><span class="p">,</span><span class="n">msna</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_axis_off</span><span class="p">()</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">choose_interacting_celltype_pair</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">+</span><span class="s1">&#39;:Fa&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">visualize_factors_id</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span>



    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">visualize_factors_id</span><span class="p">)</span><span class="o">==</span><span class="mi">2</span><span class="p">:</span>
        <span class="n">fig</span><span class="p">,(</span><span class="n">ax</span><span class="p">)</span><span class="o">=</span><span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">)</span>
        <span class="n">v1</span><span class="o">=</span><span class="n">H_sc</span><span class="p">[:,(</span><span class="n">visualize_factors_id</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">-</span><span class="mi">1</span><span class="p">)]</span>
        <span class="n">v2</span><span class="o">=</span><span class="n">H_sc</span><span class="p">[:,(</span><span class="n">visualize_factors_id</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">-</span><span class="mi">1</span><span class="p">)]</span>

        <span class="n">CTname</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">singlecell_unique_clustername</span>
        <span class="n">cellsinCT</span><span class="o">=</span><span class="p">{}</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">annotation_singlecell_barcode_id</span><span class="p">)):</span>
            <span class="n">clu_id</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">annotation_singlecell_celltypename</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">clu_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">cellsinCT</span><span class="p">:</span>
                <span class="n">cellsinCT</span><span class="p">[</span><span class="n">clu_id</span><span class="p">]</span><span class="o">=</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">cellsinCT</span><span class="p">[</span><span class="n">clu_id</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>

        <span class="n">plot_all_ct</span><span class="p">(</span><span class="n">CTname</span><span class="p">,</span><span class="n">umap_data</span><span class="p">[:,[</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">]],</span><span class="n">cellsinCT</span><span class="p">,</span><span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">],[</span><span class="n">choose_interacting_celltype_pair</span><span class="p">[</span><span class="mi">0</span><span class="p">]],</span><span class="n">v1</span><span class="p">,</span><span class="n">cmap</span><span class="p">,</span><span class="n">ms</span><span class="p">,</span><span class="n">msna</span><span class="p">)</span>
        <span class="n">plot_all_ct</span><span class="p">(</span><span class="n">CTname</span><span class="p">,</span><span class="n">umap_data</span><span class="p">[:,[</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">]],</span><span class="n">cellsinCT</span><span class="p">,</span><span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">],[</span><span class="n">choose_interacting_celltype_pair</span><span class="p">[</span><span class="mi">1</span><span class="p">]],</span><span class="n">v2</span><span class="p">,</span><span class="n">cmap</span><span class="p">,</span><span class="n">ms</span><span class="p">,</span><span class="n">msna</span><span class="p">)</span>
        <span class="c1">#ax1[1].legend(loc=&#39;upper right&#39;,bbox_to_anchor=(1.50, 1),ncol=1, frameon=False,borderaxespad=0.,prop={&quot;size&quot;:10},fancybox=True, shadow=True)</span>
        <span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
        <span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
        <span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
        <span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
        <span class="c1">#plt.gca().axes.get_yaxis().set_visible(False)</span>
        <span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_axis_off</span><span class="p">()</span>
        <span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_axis_off</span><span class="p">()</span>
        <span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">choose_interacting_celltype_pair</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">+</span><span class="s1">&#39;:Fa&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">visualize_factors_id</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span>
        <span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">choose_interacting_celltype_pair</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">+</span><span class="s1">&#39;:Fa&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">visualize_factors_id</span><span class="p">[</span><span class="mi">1</span><span class="p">]))</span>

    <span class="n">fig</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;The figures are saved: &quot;</span><span class="p">,</span> <span class="nb">input</span><span class="o">.</span><span class="n">covariation_dir</span><span class="o">+</span><span class="s1">&#39;scRNAseq_factors_in_umap.&#39;</span><span class="o">+</span><span class="n">saveas</span><span class="p">)</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">covariation_dir</span><span class="o">+</span><span class="s1">&#39;scRNAseq_factors_in_umap.&#39;</span><span class="o">+</span><span class="n">saveas</span><span class="p">,</span><span class="n">bbox_inches</span><span class="o">=</span><span class="s1">&#39;tight&#39;</span><span class="p">,</span><span class="n">transparent</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">showit</span><span class="p">:</span>
        <span class="k">pass</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="s1">&#39;all&#39;</span><span class="p">)</span>
    <span class="k">return</span> <span class="mi">0</span></div>



<div class="viewcode-block" id="plot_all_ct">
<a class="viewcode-back" href="../../nico_covariations.html#nico_covariations.Covariations.plot_all_ct">[docs]</a>
<span class="k">def</span> <span class="nf">plot_all_ct</span><span class="p">(</span><span class="n">CTname</span><span class="p">,</span><span class="n">PP</span><span class="p">,</span><span class="n">cellsinCT</span><span class="p">,</span><span class="n">ax</span><span class="p">,</span><span class="n">mycelltype</span><span class="p">,</span><span class="n">Fa</span><span class="p">,</span><span class="n">cmap</span><span class="p">,</span><span class="n">ms</span><span class="p">,</span><span class="n">msna</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Visualize factor values in UMAP for all cell types.</span>

<span class="sd">    This helper function is used for visualizing factor values in UMAP, showing the distribution of cells across different cell types and highlighting specific cell types of interest.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    -----------</span>
<span class="sd">    CTname : list of str</span>
<span class="sd">        List of cell type names.</span>
<span class="sd">    PP : np.ndarray</span>
<span class="sd">        UMAP embedding coordinates for all cells.</span>
<span class="sd">    cellsinCT : dict</span>
<span class="sd">        Dictionary where keys are cell type names and values are lists of cell indices corresponding to each cell type.</span>
<span class="sd">    ax : matplotlib.axes.Axes</span>
<span class="sd">        Matplotlib Axes object where the UMAP plot will be drawn.</span>
<span class="sd">    mycelltype : list of str</span>
<span class="sd">        List of cell types to highlight.</span>
<span class="sd">    Fa : np.ndarray</span>
<span class="sd">        Array of factor values corresponding to each cell.</span>
<span class="sd">    cmap : str or matplotlib.colors.Colormap</span>
<span class="sd">        Colormap used for plotting the factor values.</span>
<span class="sd">    ms : int</span>
<span class="sd">        Marker size for the highlighted cell types.</span>
<span class="sd">    msna : int</span>
<span class="sd">        Marker size for the non-highlighted (NA) cell types.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1">#cmap=plt.cm.get_cmap(&#39;Spectral&#39;)</span>
    <span class="c1">#cmap=plt.cm.get_cmap(&#39;jet&#39;)</span>
    <span class="n">cumsum</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">CTname</span><span class="p">))</span>

    <span class="n">naindex</span><span class="o">=</span><span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">CTname</span><span class="p">)):</span>
        <span class="n">index</span><span class="o">=</span><span class="n">cellsinCT</span><span class="p">[</span><span class="n">CTname</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span>
        <span class="n">labelname</span><span class="o">=</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;-&#39;</span><span class="o">+</span><span class="n">CTname</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">+</span><span class="s1">&#39;-&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">index</span><span class="p">))</span>
        <span class="k">if</span> <span class="n">CTname</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="ow">in</span> <span class="n">mycelltype</span><span class="p">:</span>
            <span class="c1">#p1=ax.plot(PP[index,0],PP[index,1],&#39;o&#39;,label=labelname,color=rgba,markersize=1)</span>
            <span class="n">p1</span><span class="o">=</span><span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">PP</span><span class="p">[</span><span class="n">index</span><span class="p">,</span><span class="mi">0</span><span class="p">],</span><span class="n">PP</span><span class="p">[</span><span class="n">index</span><span class="p">,</span><span class="mi">1</span><span class="p">],</span><span class="n">s</span><span class="o">=</span><span class="n">ms</span><span class="p">,</span><span class="n">c</span><span class="o">=</span><span class="n">Fa</span><span class="p">[</span><span class="n">index</span><span class="p">],</span><span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">,</span><span class="n">cmap</span><span class="o">=</span><span class="n">cmap</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">naindex</span><span class="o">=</span><span class="n">naindex</span><span class="o">+</span><span class="n">index</span>

    <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">PP</span><span class="p">[</span><span class="n">naindex</span><span class="p">,</span><span class="mi">0</span><span class="p">],</span><span class="n">PP</span><span class="p">[</span><span class="n">naindex</span><span class="p">,</span><span class="mi">1</span><span class="p">],</span><span class="s1">&#39;.&#39;</span><span class="p">,</span><span class="n">color</span><span class="o">=</span><span class="s2">&quot;0.5&quot;</span><span class="p">,</span><span class="n">label</span><span class="o">=</span><span class="s1">&#39;NA&#39;</span><span class="p">,</span><span class="n">markersize</span><span class="o">=</span><span class="n">msna</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">colorbar</span><span class="p">(</span><span class="n">p1</span><span class="p">,</span><span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span><span class="n">shrink</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span></div>





<div class="viewcode-block" id="visualize_factors_in_spatial_umap">
<a class="viewcode-back" href="../../nico_covariations.html#nico_covariations.Covariations.visualize_factors_in_spatial_umap">[docs]</a>
<span class="k">def</span> <span class="nf">visualize_factors_in_spatial_umap</span><span class="p">(</span><span class="nb">input</span><span class="p">,</span>
<span class="n">choose_interacting_celltype_pair</span><span class="p">,</span>
<span class="n">visualize_factors_id</span><span class="p">,</span>
<span class="n">umap_tag</span><span class="o">=</span><span class="s1">&#39;X_umap&#39;</span><span class="p">,</span>
<span class="n">quepath</span><span class="o">=</span><span class="s1">&#39;./inputQuery/&#39;</span><span class="p">,</span><span class="n">msna</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span><span class="n">ms</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span><span class="n">cmap</span><span class="o">=</span><span class="n">matplotlib</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s2">&quot;image.cmap&quot;</span><span class="p">],</span>
<span class="n">saveas</span><span class="o">=</span><span class="s1">&#39;pdf&#39;</span><span class="p">,</span><span class="n">transparent_mode</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">showit</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span><span class="mf">3.5</span><span class="p">)):</span>

<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Visualize factors in spatial UMAP for cell type interactions.</span>

<span class="sd">    This function is used to visualize the factors of interacting cell types in a spatial UMAP embedding. It generates and saves plots showing the distribution and factor values of cells.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    -----------</span>
<span class="sd">    input : str</span>
<span class="sd">        The primary input is the output from gene_covariation_analysis.</span>

<span class="sd">    choose_interacting_celltype_pair : list of str</span>
<span class="sd">        Define the cell type single or in pairs for visualization in the spatial UMAP.</span>
<span class="sd">        Example: choose_interacting_celltype_pair=[&#39;CentralCellType&#39;, &#39;NicheCellType&#39;]</span>

<span class="sd">    visualize_factors_id : list of str</span>
<span class="sd">        Define the factor IDs single or in pairs for visualization in the spatial UMAP.</span>
<span class="sd">        Example: visualize_factors_id=[1, 3]</span>

<span class="sd">    umap_tag : str, optional</span>
<span class="sd">        Slot for UMAP embedding in the AnnData object. Default is &#39;X_umap&#39;.</span>

<span class="sd">    quepath : str, optional</span>
<span class="sd">        Path to the query spatial count matrix in scTransform-like normalization in the common gene space. The filename should be sct_spatial.h5ad.</span>
<span class="sd">        Default is &#39;./inputQuery/&#39;.</span>

<span class="sd">    msna : float, optional</span>
<span class="sd">        Marker size for not selected (NA) cell types. Default is 0.1.</span>

<span class="sd">    ms : float, optional</span>
<span class="sd">        Marker size for selected cell types. Default is 5.</span>

<span class="sd">    cmap : str or matplotlib.colors.Colormap, optional</span>
<span class="sd">        Colormap used for visualizing the factor values. Default is matplotlib.rcParams[&quot;image.cmap&quot;].</span>

<span class="sd">    saveas : str, optional</span>
<span class="sd">        Format to save the figures, either &#39;pdf&#39; or &#39;png&#39;. Default is &#39;pdf&#39;.</span>

<span class="sd">    transparent_mode : bool, optional</span>
<span class="sd">        Background color of the figures. If True, the figures have a transparent background. Default is False.</span>

<span class="sd">    showit : bool, optional</span>
<span class="sd">        If True, the figures will be displayed. Default is True.</span>

<span class="sd">    figsize : tuple of float, optional</span>
<span class="sd">        Dimension of the figure size. Default is (8, 3.5).</span>

<span class="sd">    Output:</span>
<span class="sd">    -------</span>
<span class="sd">    The output figure will be saved in nico_out/covariations_R*_F*/spatial_factors_in_umap*.</span>


<span class="sd">    &quot;&quot;&quot;</span>


    <span class="n">sct_ad_sp</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">umap_plot_sp</span>
    <span class="n">umap</span><span class="o">=</span><span class="n">sct_ad_sp</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="n">umap_tag</span><span class="p">]</span>
    <span class="n">barcode</span><span class="o">=</span><span class="n">sct_ad_sp</span><span class="o">.</span><span class="n">obs_names</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span>
    <span class="n">barcode</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">barcode</span><span class="p">,(</span><span class="nb">len</span><span class="p">(</span><span class="n">barcode</span><span class="p">),</span><span class="mi">1</span><span class="p">))</span>
    <span class="n">umap_not_order</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">hstack</span><span class="p">((</span><span class="n">barcode</span><span class="p">,</span><span class="n">umap</span><span class="p">))</span>

    <span class="n">cellname</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">annotation_spatial_barcode_id</span><span class="p">,(</span><span class="nb">len</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">annotation_spatial_barcode_id</span><span class="p">),</span><span class="mi">1</span><span class="p">)</span> <span class="p">)</span>
    <span class="n">umap_data</span><span class="o">=</span><span class="n">sort_index_in_right_order</span><span class="p">(</span><span class="n">cellname</span><span class="p">,</span><span class="n">umap_not_order</span><span class="p">)</span>
    <span class="n">PCA_of_sc_cluster_accordingto_spatial_clusterid</span><span class="p">,</span><span class="n">save_scFactors</span><span class="p">,</span><span class="n">save_spFactors</span><span class="o">=</span><span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">covariation_dir</span><span class="o">+</span><span class="s1">&#39;factors_info.p&#39;</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">))</span>

    <span class="n">t</span><span class="o">=</span><span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">no_of_pc</span><span class="p">):</span>
        <span class="n">t</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>

    <span class="n">H_sp</span><span class="o">=</span><span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">annotation_spatial_cluster_id</span><span class="p">)):</span>
        <span class="n">cellid</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">annotation_spatial_barcode_id</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">cellid</span> <span class="ow">in</span> <span class="n">save_spFactors</span><span class="p">:</span>
            <span class="n">H_sp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">save_spFactors</span><span class="p">[</span><span class="n">cellid</span><span class="p">])</span>
            <span class="c1">#print(save_scFactors[cellid])</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">H_sp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">t</span><span class="p">))</span>
    <span class="n">H_sp</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">H_sp</span><span class="p">)</span>


    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">visualize_factors_id</span><span class="p">)</span><span class="o">==</span><span class="mi">1</span><span class="p">:</span>
        <span class="n">fig</span><span class="p">,(</span><span class="n">ax</span><span class="p">)</span><span class="o">=</span><span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">)</span>
        <span class="n">v1</span><span class="o">=</span><span class="n">H_sp</span><span class="p">[:,(</span><span class="n">visualize_factors_id</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">-</span><span class="mi">1</span><span class="p">)]</span>
        <span class="n">CTname</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clustername</span>
        <span class="n">cellsinCT</span><span class="o">=</span><span class="p">{}</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">annotation_spatial_cluster_id</span><span class="p">)):</span>
            <span class="n">clu_id</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">annotation_spatial_celltypename</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
            <span class="c1">#cel_id=sc_cluster[i][0]</span>
            <span class="k">if</span> <span class="n">clu_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">cellsinCT</span><span class="p">:</span>
                <span class="n">cellsinCT</span><span class="p">[</span><span class="n">clu_id</span><span class="p">]</span><span class="o">=</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">cellsinCT</span><span class="p">[</span><span class="n">clu_id</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
        <span class="n">plot_all_ct</span><span class="p">(</span><span class="n">CTname</span><span class="p">,</span><span class="n">umap_data</span><span class="p">[:,[</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">]],</span><span class="n">cellsinCT</span><span class="p">,</span><span class="n">ax</span><span class="p">,[</span><span class="n">choose_interacting_celltype_pair</span><span class="p">[</span><span class="mi">0</span><span class="p">]],</span><span class="n">v1</span><span class="p">,</span><span class="n">cmap</span><span class="p">,</span><span class="n">ms</span><span class="p">,</span><span class="n">msna</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_axis_off</span><span class="p">()</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">choose_interacting_celltype_pair</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">+</span><span class="s1">&#39;:Fa&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">visualize_factors_id</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span>



    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">visualize_factors_id</span><span class="p">)</span><span class="o">==</span><span class="mi">2</span><span class="p">:</span>
        <span class="n">fig</span><span class="p">,(</span><span class="n">ax</span><span class="p">)</span><span class="o">=</span><span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">)</span>
        <span class="n">v1</span><span class="o">=</span><span class="n">H_sp</span><span class="p">[:,(</span><span class="n">visualize_factors_id</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">-</span><span class="mi">1</span><span class="p">)]</span>
        <span class="n">v2</span><span class="o">=</span><span class="n">H_sp</span><span class="p">[:,(</span><span class="n">visualize_factors_id</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">-</span><span class="mi">1</span><span class="p">)]</span>

        <span class="n">CTname</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clustername</span>
        <span class="n">cellsinCT</span><span class="o">=</span><span class="p">{}</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">annotation_spatial_cluster_id</span><span class="p">)):</span>
            <span class="n">clu_id</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">annotation_spatial_celltypename</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
            <span class="c1">#cel_id=sc_cluster[i][0]</span>
            <span class="k">if</span> <span class="n">clu_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">cellsinCT</span><span class="p">:</span>
                <span class="n">cellsinCT</span><span class="p">[</span><span class="n">clu_id</span><span class="p">]</span><span class="o">=</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">cellsinCT</span><span class="p">[</span><span class="n">clu_id</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>

        <span class="n">plot_all_ct</span><span class="p">(</span><span class="n">CTname</span><span class="p">,</span><span class="n">umap_data</span><span class="p">[:,[</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">]],</span><span class="n">cellsinCT</span><span class="p">,</span><span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">],[</span><span class="n">choose_interacting_celltype_pair</span><span class="p">[</span><span class="mi">0</span><span class="p">]],</span><span class="n">v1</span><span class="p">,</span><span class="n">cmap</span><span class="p">,</span><span class="n">ms</span><span class="p">,</span><span class="n">msna</span><span class="p">)</span>
        <span class="n">plot_all_ct</span><span class="p">(</span><span class="n">CTname</span><span class="p">,</span><span class="n">umap_data</span><span class="p">[:,[</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">]],</span><span class="n">cellsinCT</span><span class="p">,</span><span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">],[</span><span class="n">choose_interacting_celltype_pair</span><span class="p">[</span><span class="mi">1</span><span class="p">]],</span><span class="n">v2</span><span class="p">,</span><span class="n">cmap</span><span class="p">,</span><span class="n">ms</span><span class="p">,</span><span class="n">msna</span><span class="p">)</span>
        <span class="c1">#ax1[1].legend(loc=&#39;upper right&#39;,bbox_to_anchor=(1.50, 1),ncol=1, frameon=False,borderaxespad=0.,prop={&quot;size&quot;:10},fancybox=True, shadow=True)</span>
        <span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
        <span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
        <span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
        <span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
        <span class="c1">#plt.gca().axes.get_yaxis().set_visible(False)</span>
        <span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_axis_off</span><span class="p">()</span>
        <span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_axis_off</span><span class="p">()</span>
        <span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">choose_interacting_celltype_pair</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">+</span><span class="s1">&#39;:Fa&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">visualize_factors_id</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span>
        <span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">choose_interacting_celltype_pair</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">+</span><span class="s1">&#39;:Fa&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">visualize_factors_id</span><span class="p">[</span><span class="mi">1</span><span class="p">]))</span>

    <span class="n">fig</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;The figures are saved: &quot;</span><span class="p">,</span> <span class="nb">input</span><span class="o">.</span><span class="n">covariation_dir</span><span class="o">+</span><span class="s1">&#39;spatial_factors_in_umap.&#39;</span><span class="o">+</span><span class="n">saveas</span><span class="p">)</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">covariation_dir</span><span class="o">+</span><span class="s1">&#39;spatial_factors_in_umap.&#39;</span><span class="o">+</span><span class="n">saveas</span><span class="p">,</span><span class="n">bbox_inches</span><span class="o">=</span><span class="s1">&#39;tight&#39;</span><span class="p">,</span><span class="n">transparent</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">showit</span><span class="p">:</span>
        <span class="k">pass</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="s1">&#39;all&#39;</span><span class="p">)</span>
    <span class="k">return</span> <span class="mi">0</span></div>



<div class="viewcode-block" id="read_LigRecDb">
<a class="viewcode-back" href="../../nico_covariations.html#nico_covariations.Covariations.read_LigRecDb">[docs]</a>
<span class="k">def</span> <span class="nf">read_LigRecDb</span><span class="p">(</span><span class="n">contdb</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Reads the ligand-receptor database file.</span>

<span class="sd">    This function processes a database of ligand-receptor pairs, identifying unique ligands, receptors, and elements that act as both. The input should be a list of strings, where each string represents a line from the database file.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    -----------</span>
<span class="sd">    contdb : list of str</span>
<span class="sd">        A list of strings, where each string is a line from the ligand-receptor database file. Each line contains a ligand and a receptor separated by whitespace.</span>


<span class="sd">    Example:</span>
<span class="sd">    --------</span>
<span class="sd">    &gt;&gt;&gt; contdb = [</span>
<span class="sd">            &quot;LIG1 REC1&quot;,</span>
<span class="sd">            &quot;LIG2 REC2&quot;,</span>
<span class="sd">            &quot;REC1 LIG1&quot;,  # REC1 is both a receptor and a ligand</span>
<span class="sd">            &quot;LIG3 REC3&quot;</span>
<span class="sd">        ]</span>
<span class="sd">    &gt;&gt;&gt; totalLRpairs, ligand, receptor, either = read_LigRecDb(contdb)</span>
<span class="sd">    &gt;&gt;&gt; print(totalLRpairs)</span>
<span class="sd">    [[&#39;LIG1&#39;, &#39;REC1&#39;], [&#39;LIG2&#39;, &#39;REC2&#39;], [&#39;REC1&#39;, &#39;LIG1&#39;], [&#39;LIG3&#39;, &#39;REC3&#39;]]</span>
<span class="sd">    &gt;&gt;&gt; print(ligand)</span>
<span class="sd">    {&#39;LIG2&#39;: 1, &#39;LIG3&#39;: 1}</span>
<span class="sd">    &gt;&gt;&gt; print(receptor)</span>
<span class="sd">    {&#39;REC2&#39;: 1, &#39;REC3&#39;: 1}</span>
<span class="sd">    &gt;&gt;&gt; print(either)</span>
<span class="sd">    {&#39;LIG1&#39;: 1, &#39;REC1&#39;: 1}</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1">#f=open(&#39;sort_3_db_L_R_high_confident.dat&#39;,&#39;r&#39;)</span>
    <span class="n">totalLRpairs</span><span class="o">=</span><span class="p">[]</span>
    <span class="n">ligand</span><span class="o">=</span><span class="p">{}</span>
    <span class="n">receptor</span><span class="o">=</span><span class="p">{}</span>
    <span class="n">either</span><span class="o">=</span><span class="p">{}</span>
    <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">contdb</span><span class="p">)):</span>
        <span class="n">l</span><span class="o">=</span><span class="n">contdb</span><span class="p">[</span><span class="n">j</span><span class="p">][</span><span class="mi">0</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
        <span class="n">ligand</span><span class="p">[</span><span class="n">l</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">upper</span><span class="p">()]</span><span class="o">=</span><span class="mi">1</span>
        <span class="n">receptor</span><span class="p">[</span><span class="n">l</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">upper</span><span class="p">()]</span><span class="o">=</span><span class="mi">1</span>
        <span class="k">if</span> <span class="p">[</span><span class="n">l</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">l</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">totalLRpairs</span><span class="p">:</span>
            <span class="n">totalLRpairs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span> <span class="p">[</span><span class="n">l</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">upper</span><span class="p">(),</span> <span class="n">l</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="p">])</span>

    <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">ligand</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">receptor</span><span class="p">:</span>
            <span class="n">either</span><span class="p">[</span><span class="n">key</span><span class="p">]</span><span class="o">=</span><span class="mi">1</span>
    <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">either</span><span class="p">:</span>
        <span class="n">ligand</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="n">key</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
        <span class="n">receptor</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="n">key</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">totalLRpairs</span><span class="p">,</span><span class="n">ligand</span><span class="p">,</span><span class="n">receptor</span><span class="p">,</span><span class="n">either</span></div>


<div class="viewcode-block" id="sort_index_in_right_order">
<a class="viewcode-back" href="../../nico_covariations.html#nico_covariations.Covariations.sort_index_in_right_order">[docs]</a>
<span class="k">def</span> <span class="nf">sort_index_in_right_order</span><span class="p">(</span><span class="n">correct</span><span class="p">,</span><span class="n">wrong</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Sorts the &#39;wrong&#39; array to match the order of the &#39;correct&#39; array based on the first column values.</span>

<span class="sd">    This function reorders the rows of the &#39;wrong&#39; array to match the order of the &#39;correct&#39; array based on the values in the first column. It is a helper function used in visualizing cell type annotations.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    -----------</span>
<span class="sd">    correct : ndarray</span>
<span class="sd">        An array with the correct order of elements. The sorting is based on the values in the first column.</span>

<span class="sd">    wrong : ndarray</span>
<span class="sd">        An array that needs to be reordered to match the &#39;correct&#39; array. The sorting is based on the values in the first column.</span>

<span class="sd">    Returns:</span>
<span class="sd">    --------</span>
<span class="sd">    right : ndarray</span>
<span class="sd">        The &#39;wrong&#39; array reordered to match the order of the &#39;correct&#39; array based on the first column values.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">d</span><span class="o">=</span><span class="p">{}</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">wrong</span><span class="p">)):</span>
        <span class="n">d</span><span class="p">[</span><span class="n">wrong</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="mi">0</span><span class="p">]]</span><span class="o">=</span><span class="n">i</span>
    <span class="n">index</span><span class="o">=</span><span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">correct</span><span class="p">)):</span>
        <span class="n">index</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">d</span><span class="p">[</span><span class="n">correct</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="mi">0</span><span class="p">]])</span>
    <span class="n">right</span><span class="o">=</span><span class="n">wrong</span><span class="p">[</span><span class="n">index</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">right</span></div>



<div class="viewcode-block" id="plot_top_genes_for_a_given_celltype_from_all_three_factors">
<a class="viewcode-back" href="../../nico_covariations.html#nico_covariations.Covariations.plot_top_genes_for_a_given_celltype_from_all_three_factors">[docs]</a>
<span class="k">def</span> <span class="nf">plot_top_genes_for_a_given_celltype_from_all_three_factors</span><span class="p">(</span><span class="nb">input</span><span class="p">,</span><span class="n">choose_celltypes</span><span class="o">=</span><span class="p">[],</span><span class="n">top_NOG</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span><span class="n">rps_rpl_mt_genes_included</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">correlation_with_spearman</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">saveas</span><span class="o">=</span><span class="s1">&#39;pdf&#39;</span><span class="p">,</span><span class="n">transparent_mode</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">showit</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">10</span><span class="p">)):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Visualize top genes associated with given cell types across all three factors.</span>

<span class="sd">    This function generates plots of the top N genes associated with specified cell types from the input data.</span>
<span class="sd">    The associations can be visualized using either Spearman correlation coefficient or cosine similarity.</span>
<span class="sd">    Optionally, the visualization can include rps, rpl, and mt genes.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    -----------</span>
<span class="sd">    input : dict</span>
<span class="sd">        The main input is the output from gene_covariation_analysis.</span>

<span class="sd">    choose_celltypes : list, optional</span>
<span class="sd">        The cell type for which the gene-factor associations should be displayed.</span>
<span class="sd">        If the list is empty, the output will be generated for all the cell types.</span>
<span class="sd">        (default is [])</span>

<span class="sd">    top_NOG : int, optional</span>
<span class="sd">        Number of genes to visualize.</span>
<span class="sd">        (default is 20)</span>

<span class="sd">    rps_rpl_mt_genes_included : bool, optional</span>
<span class="sd">        For pathway analysis, decide whether to include rps, rpl, and mt genes. If True, they are included.</span>
<span class="sd">        (default is True)</span>

<span class="sd">    correlation_with_spearman : bool, optional</span>
<span class="sd">        If True, visualize gene-factor association obtained as Spearman correlation coefficient; otherwise, cosine similarity is displayed.</span>
<span class="sd">        (default is True)</span>

<span class="sd">    saveas : str, optional</span>
<span class="sd">        Save the figures in PDF or PNG format (dpi for PNG format is 300).</span>
<span class="sd">        (default is &#39;pdf&#39;)</span>

<span class="sd">    transparent_mode : bool, optional</span>
<span class="sd">        Background color of the figures.</span>
<span class="sd">        (default is False)</span>

<span class="sd">    showit : bool, optional</span>
<span class="sd">        Whether to display the plot or not.</span>
<span class="sd">        (default is True)</span>

<span class="sd">    figsize : tuple, optional</span>
<span class="sd">        Dimension of the figure size.</span>
<span class="sd">        (default is (12, 10))</span>

<span class="sd">    Outputs:</span>
<span class="sd">    --------</span>
<span class="sd">    The gene visualization figures are saved in ./nico_out/covariations_R*_F*/dotplots/*</span>

<span class="sd">    Example:</span>
<span class="sd">    --------</span>
<span class="sd">    &gt;&gt;&gt; input_data = load_data_from_analysis()  # hypothetical function to load data</span>
<span class="sd">    &gt;&gt;&gt; plot_top_genes_for_a_given_celltype_from_all_three_factors(input_data, choose_celltypes=[&#39;CellType1&#39;], top_NOG=10)</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">savefigdir</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">covariation_dir</span><span class="o">+</span> <span class="s1">&#39;dotplots/&#39;</span>
    <span class="n">create_directory</span><span class="p">(</span><span class="n">savefigdir</span><span class="p">)</span>

    <span class="n">PCA_of_sc_cluster_accordingto_spatial_clusterid</span><span class="p">,</span><span class="n">save_scFactors</span><span class="p">,</span><span class="n">save_spFactors</span><span class="o">=</span><span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">covariation_dir</span><span class="o">+</span><span class="s1">&#39;factors_info.p&#39;</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">))</span>
    <span class="n">n</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clustername</span><span class="p">)</span>
    <span class="n">perform</span><span class="o">=</span><span class="p">[]</span>
    <span class="n">found</span><span class="o">=</span><span class="p">[]</span>
    <span class="k">for</span> <span class="n">fi</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="p">):</span>
        <span class="n">CC_celltype_name</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clustername</span><span class="p">[</span><span class="n">fi</span><span class="p">]</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">choose_celltypes</span><span class="p">)</span><span class="o">==</span><span class="mi">0</span><span class="p">:</span>
            <span class="n">perform</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fi</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">CC_celltype_name</span> <span class="ow">in</span> <span class="n">choose_celltypes</span><span class="p">:</span>
                <span class="n">perform</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fi</span><span class="p">)</span>
                <span class="n">found</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">CC_celltype_name</span><span class="p">)</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">choose_celltypes</span><span class="p">)</span><span class="o">!=</span><span class="mi">0</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;cell types found &quot;</span><span class="p">,</span><span class="n">found</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">fi</span> <span class="ow">in</span> <span class="n">perform</span><span class="p">:</span>
        <span class="n">clid</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clusterid</span><span class="p">[</span><span class="n">fi</span><span class="p">]</span>
        <span class="n">spearman_factors</span><span class="p">,</span><span class="n">CC_PCA</span><span class="p">,</span><span class="n">CC_gene</span><span class="p">,</span><span class="n">CC_meanExpression</span><span class="p">,</span><span class="n">CC_popExpression</span><span class="p">,</span><span class="n">cosine_factors</span><span class="p">,</span><span class="n">alpha</span><span class="o">=</span><span class="n">PCA_of_sc_cluster_accordingto_spatial_clusterid</span><span class="p">[</span><span class="n">clid</span><span class="p">]</span>
        <span class="n">CC_celltype_name</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clustername</span><span class="p">[</span><span class="n">fi</span><span class="p">]</span>
        <span class="n">mu</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">CC_meanExpression</span><span class="p">)</span>
        <span class="n">pop</span><span class="o">=</span><span class="n">CC_popExpression</span>

        <span class="n">nvr1</span><span class="o">=</span><span class="p">[]</span>
        <span class="n">nvr2</span><span class="o">=</span><span class="p">[]</span>
        <span class="n">nvr3</span><span class="o">=</span><span class="p">[]</span>
        <span class="n">comgene</span><span class="o">=</span><span class="p">[]</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">no_of_pc</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">correlation_with_spearman</span><span class="p">:</span>
                <span class="n">source</span><span class="o">=</span><span class="n">spearman_factors</span><span class="p">[:,</span><span class="n">j</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">source</span><span class="o">=</span><span class="n">cosine_factors</span><span class="p">[:,</span><span class="n">j</span><span class="p">]</span>
            <span class="n">ind</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="o">-</span><span class="n">source</span><span class="p">)</span>
            <span class="n">interestofGene</span><span class="o">=</span><span class="p">[]</span>
            <span class="n">value_fact</span><span class="o">=</span><span class="p">[]</span>
            <span class="n">value_pop</span><span class="o">=</span><span class="p">[]</span>
            <span class="n">value_avgexp</span><span class="o">=</span><span class="p">[]</span>
            <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">source</span><span class="p">)):</span>
                <span class="n">temp</span><span class="o">=</span><span class="n">CC_gene</span><span class="p">[</span><span class="n">ind</span><span class="p">[</span><span class="n">k</span><span class="p">]]</span>
                <span class="k">if</span> <span class="n">rps_rpl_mt_genes_included</span><span class="p">:</span>
                    <span class="n">flag</span><span class="o">=</span><span class="mi">1</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">flag</span><span class="o">=</span><span class="mi">1</span>
                    <span class="k">if</span> <span class="n">temp</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">3</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;Rps&#39;</span><span class="p">:</span>
                        <span class="n">flag</span><span class="o">=</span><span class="mi">0</span>
                    <span class="k">if</span> <span class="n">temp</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">3</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;Rpl&#39;</span><span class="p">:</span>
                        <span class="n">flag</span><span class="o">=</span><span class="mi">0</span>
                    <span class="k">if</span> <span class="n">temp</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">3</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;mt-&#39;</span><span class="p">:</span>
                        <span class="n">flag</span><span class="o">=</span><span class="mi">0</span>
                <span class="k">if</span> <span class="n">flag</span><span class="o">==</span><span class="mi">1</span><span class="p">:</span>
                    <span class="n">interestofGene</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">CC_gene</span><span class="p">[</span><span class="n">ind</span><span class="p">[</span><span class="n">k</span><span class="p">]])</span>
                    <span class="n">value_fact</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">source</span><span class="p">[</span><span class="n">ind</span><span class="p">[</span><span class="n">k</span><span class="p">]])</span>
                    <span class="n">value_pop</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">pop</span><span class="p">[</span><span class="n">ind</span><span class="p">[</span><span class="n">k</span><span class="p">]])</span>
                    <span class="n">value_avgexp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mu</span><span class="p">[</span><span class="n">ind</span><span class="p">[</span><span class="n">k</span><span class="p">]])</span>

            <span class="n">value_fact</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">value_fact</span><span class="p">)</span>
            <span class="n">value_pop</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">value_pop</span><span class="p">)</span>
            <span class="n">value_avgexp</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">value_avgexp</span><span class="p">)</span>

            <span class="n">interestofGene</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">interestofGene</span><span class="p">)</span>
            <span class="n">index_pos</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="o">-</span><span class="n">value_fact</span><span class="p">)</span>
            <span class="n">index_neg</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">value_fact</span><span class="p">)</span>

            <span class="n">gp1</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">interestofGene</span><span class="p">[</span><span class="n">index_pos</span><span class="p">])</span>
            <span class="n">gn1</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">interestofGene</span><span class="p">[</span><span class="n">index_neg</span><span class="p">])</span>

            <span class="n">comgene</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gp1</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="n">top_NOG</span><span class="p">])</span>
            <span class="n">comgene</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gn1</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="n">top_NOG</span><span class="p">])</span>

            <span class="n">gex</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">top_NOG</span><span class="p">,</span><span class="mi">1</span><span class="p">),</span><span class="n">dtype</span><span class="o">=</span><span class="nb">float</span><span class="p">)</span>

            <span class="n">vp1</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">value_fact</span><span class="p">[</span><span class="n">index_pos</span><span class="p">])[</span><span class="mi">0</span><span class="p">:</span><span class="n">top_NOG</span><span class="p">]</span>
            <span class="n">vn1</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">value_fact</span><span class="p">[</span><span class="n">index_neg</span><span class="p">])[</span><span class="mi">0</span><span class="p">:</span><span class="n">top_NOG</span><span class="p">]</span>
            <span class="n">pos_pop1</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">value_pop</span><span class="p">[</span><span class="n">index_pos</span><span class="p">])[</span><span class="mi">0</span><span class="p">:</span><span class="n">top_NOG</span><span class="p">]</span>
            <span class="n">neg_pop1</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">value_pop</span><span class="p">[</span><span class="n">index_neg</span><span class="p">])[</span><span class="mi">0</span><span class="p">:</span><span class="n">top_NOG</span><span class="p">]</span>
            <span class="n">pos_avg1</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">value_avgexp</span><span class="p">[</span><span class="n">index_pos</span><span class="p">])[</span><span class="mi">0</span><span class="p">:</span><span class="n">top_NOG</span><span class="p">]</span>
            <span class="n">neg_avg1</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">value_avgexp</span><span class="p">[</span><span class="n">index_neg</span><span class="p">])[</span><span class="mi">0</span><span class="p">:</span><span class="n">top_NOG</span><span class="p">]</span>

            <span class="n">nvr1</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">hstack</span><span class="p">((</span><span class="n">np</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">vp1</span><span class="p">,(</span><span class="nb">len</span><span class="p">(</span><span class="n">vp1</span><span class="p">),</span><span class="mi">1</span><span class="p">)),</span><span class="n">gex</span><span class="p">)))</span>
            <span class="n">nvr1</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">hstack</span><span class="p">((</span><span class="n">np</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">vn1</span><span class="p">,(</span><span class="nb">len</span><span class="p">(</span><span class="n">vn1</span><span class="p">),</span><span class="mi">1</span><span class="p">)),</span><span class="n">gex</span><span class="p">)))</span>
            <span class="n">nvr2</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">hstack</span><span class="p">((</span><span class="n">np</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">pos_pop1</span><span class="p">,(</span><span class="nb">len</span><span class="p">(</span><span class="n">pos_pop1</span><span class="p">),</span><span class="mi">1</span><span class="p">)),</span><span class="n">gex</span><span class="p">)))</span>
            <span class="n">nvr2</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">hstack</span><span class="p">((</span><span class="n">np</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">neg_pop1</span><span class="p">,(</span><span class="nb">len</span><span class="p">(</span><span class="n">neg_pop1</span><span class="p">),</span><span class="mi">1</span><span class="p">)),</span><span class="n">gex</span><span class="p">)))</span>
            <span class="n">nvr3</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">hstack</span><span class="p">((</span><span class="n">np</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">pos_avg1</span><span class="p">,(</span><span class="nb">len</span><span class="p">(</span><span class="n">pos_avg1</span><span class="p">),</span><span class="mi">1</span><span class="p">)),</span><span class="n">gex</span><span class="p">)))</span>
            <span class="n">nvr3</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">hstack</span><span class="p">((</span><span class="n">np</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">neg_avg1</span><span class="p">,(</span><span class="nb">len</span><span class="p">(</span><span class="n">neg_avg1</span><span class="p">),</span><span class="mi">1</span><span class="p">)),</span><span class="n">gex</span><span class="p">)))</span>


        <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="mi">6</span><span class="p">,</span><span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">)</span>
        <span class="n">title</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Pos Fa1&#39;</span><span class="p">,</span><span class="s1">&#39;Neg Fa1&#39;</span><span class="p">,</span><span class="s1">&#39;Pos Fa2&#39;</span><span class="p">,</span><span class="s1">&#39;Neg Fa2&#39;</span><span class="p">,</span><span class="s1">&#39;Pos Fa3&#39;</span><span class="p">,</span><span class="s1">&#39;Neg Fa3&#39;</span><span class="p">]</span>

        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">6</span><span class="p">):</span>
                <span class="n">x</span><span class="p">,</span><span class="n">y</span><span class="p">,</span><span class="n">z</span><span class="p">,</span><span class="n">bigs</span><span class="o">=</span><span class="n">findXYZC</span><span class="p">(</span><span class="n">nvr1</span><span class="p">[</span><span class="n">i</span><span class="p">],</span><span class="n">nvr2</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
                <span class="n">p0</span><span class="o">=</span><span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">x</span><span class="p">,</span><span class="n">y</span><span class="p">,</span><span class="n">s</span><span class="o">=</span><span class="n">bigs</span><span class="p">,</span><span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">,</span><span class="n">c</span><span class="o">=</span><span class="n">z</span><span class="p">,</span><span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;RdBu_r&#39;</span><span class="p">)</span> <span class="c1">#&#39;cm.cmap_name</span>
                <span class="n">x</span><span class="p">,</span><span class="n">y</span><span class="p">,</span><span class="n">z</span><span class="p">,</span><span class="n">bigs</span><span class="o">=</span><span class="n">findXYZC</span><span class="p">(</span><span class="n">nvr3</span><span class="p">[</span><span class="n">i</span><span class="p">],</span><span class="n">nvr2</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
                <span class="n">p1</span><span class="o">=</span><span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">x</span><span class="p">,</span><span class="n">y</span><span class="p">,</span><span class="n">s</span><span class="o">=</span><span class="n">bigs</span><span class="p">,</span><span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">,</span><span class="n">c</span><span class="o">=</span><span class="n">z</span><span class="p">,</span><span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;RdBu_r&#39;</span><span class="p">)</span>
                <span class="n">plt</span><span class="o">.</span><span class="n">colorbar</span><span class="p">(</span><span class="n">p0</span><span class="p">,</span><span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="n">i</span><span class="p">],</span><span class="n">shrink</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
                <span class="n">plt</span><span class="o">.</span><span class="n">colorbar</span><span class="p">(</span><span class="n">p1</span><span class="p">,</span><span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="n">i</span><span class="p">],</span><span class="n">shrink</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
                <span class="n">kw</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">prop</span><span class="o">=</span><span class="s2">&quot;sizes&quot;</span><span class="p">,</span> <span class="n">num</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.6</span><span class="p">,</span> <span class="n">fmt</span><span class="o">=</span><span class="s2">&quot;% </span><span class="si">{x:.0f}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="n">legend2</span> <span class="o">=</span> <span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="o">*</span><span class="n">p1</span><span class="o">.</span><span class="n">legend_elements</span><span class="p">(</span><span class="o">**</span><span class="n">kw</span><span class="p">),</span><span class="n">loc</span><span class="o">=</span><span class="s2">&quot;upper right&quot;</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s2">&quot;Fraction of </span><span class="se">\n</span><span class="s2">cells </span><span class="se">\n</span><span class="s2">expressed&quot;</span><span class="p">,</span><span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mf">1.0</span><span class="p">,</span> <span class="mi">0</span><span class="p">),</span><span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">6</span><span class="p">):</span>
            <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">2</span><span class="p">):</span>
                <span class="n">ax</span><span class="p">[</span><span class="n">j</span><span class="p">,</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">nvr1</span><span class="p">[</span><span class="n">i</span><span class="p">])))</span>
                <span class="n">ax</span><span class="p">[</span><span class="n">j</span><span class="p">,</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticklabels</span><span class="p">(</span><span class="n">comgene</span><span class="p">[</span><span class="n">i</span><span class="p">],</span><span class="n">style</span><span class="o">=</span><span class="s1">&#39;italic&#39;</span><span class="p">)</span>
                <span class="n">ax</span><span class="p">[</span><span class="n">j</span><span class="p">,</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span><span class="c1">#range(1))</span>
                <span class="n">ax</span><span class="p">[</span><span class="n">j</span><span class="p">,</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">set_xticklabels</span><span class="p">([])</span><span class="c1">#xlabels[i],rotation=30)</span>
                <span class="n">ax</span><span class="p">[</span><span class="n">j</span><span class="p">,</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">([</span><span class="o">-</span><span class="mf">0.5</span><span class="p">,</span><span class="mf">0.5</span><span class="p">])</span>
                <span class="n">ax</span><span class="p">[</span><span class="n">j</span><span class="p">,</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">([</span><span class="o">-</span><span class="mf">0.5</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">nvr1</span><span class="p">[</span><span class="n">i</span><span class="p">])</span><span class="o">+</span><span class="mf">0.5</span><span class="p">])</span>
                <span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">title</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>

        <span class="n">grid</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">GridSpec</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">6</span><span class="p">)</span>
        <span class="n">create_subtitle</span><span class="p">(</span><span class="n">fig</span><span class="p">,</span> <span class="n">grid</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">::],</span> <span class="n">CC_celltype_name</span><span class="o">+</span><span class="s1">&#39; Spearman correlation&#39;</span><span class="p">)</span>
        <span class="n">create_subtitle</span><span class="p">(</span><span class="n">fig</span><span class="p">,</span> <span class="n">grid</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">::],</span>  <span class="n">CC_celltype_name</span><span class="o">+</span><span class="s1">&#39; log(avg expression)&#39;</span><span class="p">)</span>

        <span class="n">fig</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;The figures are saved: &quot;</span><span class="p">,</span> <span class="n">savefigdir</span><span class="o">+</span><span class="n">remove_extra_character_from_name</span><span class="p">(</span><span class="n">CC_celltype_name</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;.&#39;</span><span class="o">+</span><span class="n">saveas</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">savefigdir</span><span class="o">+</span><span class="n">remove_extra_character_from_name</span><span class="p">(</span><span class="n">CC_celltype_name</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;.&#39;</span><span class="o">+</span><span class="n">saveas</span><span class="p">,</span><span class="n">bbox_inches</span><span class="o">=</span><span class="s1">&#39;tight&#39;</span><span class="p">,</span><span class="n">transparent</span><span class="o">=</span><span class="n">transparent_mode</span><span class="p">,</span><span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">showit</span><span class="p">:</span>
            <span class="k">pass</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="s1">&#39;all&#39;</span><span class="p">)</span></div>




<div class="viewcode-block" id="plot_top_genes_for_pair_of_celltypes_from_two_chosen_factors">
<a class="viewcode-back" href="../../nico_covariations.html#nico_covariations.Covariations.plot_top_genes_for_pair_of_celltypes_from_two_chosen_factors">[docs]</a>
<span class="k">def</span> <span class="nf">plot_top_genes_for_pair_of_celltypes_from_two_chosen_factors</span><span class="p">(</span><span class="nb">input</span><span class="p">,</span>
<span class="n">choose_interacting_celltype_pair</span><span class="p">,</span>
<span class="n">visualize_factors_id</span><span class="p">,</span>
<span class="n">top_NOG</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span>
<span class="n">rps_rpl_mt_genes_included</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">correlation_with_spearman</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="n">saveas</span><span class="o">=</span><span class="s1">&#39;pdf&#39;</span><span class="p">,</span><span class="n">transparent_mode</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">showit</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">8</span><span class="p">)):</span>


<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Visualize top genes associated with a pair of cell types from chosen factors.</span>

<span class="sd">    This function generates plots of the top 20 genes in the factors associated with specified cell types</span>
<span class="sd">    from the input data, using either Spearman correlation coefficient or cosine similarity.</span>
<span class="sd">    The visualizations include comparisons between the chosen factors for each cell type.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    -----------</span>
<span class="sd">    input : object</span>
<span class="sd">        The main input is the output from gene_covariation_analysis, which includes factor and expression data.</span>

<span class="sd">    choose_interacting_celltype_pair : list</span>
<span class="sd">        Define the cell type pairs for visualization. The first entry is the central cell type, and the second is the niche cell type.</span>

<span class="sd">    visualize_factors_id : list</span>
<span class="sd">        Define the factor IDs for visualization. The first entry is the factor ID of the central cell type, and the second is the factor ID of the niche cell type.</span>

<span class="sd">    top_NOG : int, optional</span>
<span class="sd">        Number of genes to visualize.</span>
<span class="sd">        (default is 20)</span>

<span class="sd">    rps_rpl_mt_genes_included : bool, optional</span>
<span class="sd">        For pathway analysis, decide whether to include rps, rpl, and mt genes. If True, they are included.</span>
<span class="sd">        (default is True)</span>

<span class="sd">    correlation_with_spearman : bool, optional</span>
<span class="sd">        If True, visualize gene-factor association obtained as Spearman correlation coefficient; otherwise, cosine similarity is displayed.</span>
<span class="sd">        (default is True)</span>

<span class="sd">    saveas : str, optional</span>
<span class="sd">        Save the figures in PDF or PNG format (dpi for PNG format is 300).</span>
<span class="sd">        (default is &#39;pdf&#39;)</span>

<span class="sd">    transparent_mode : bool, optional</span>
<span class="sd">        Background color of the figures.</span>
<span class="sd">        (default is False)</span>

<span class="sd">    showit : bool, optional</span>
<span class="sd">        Whether to display the plot or not.</span>
<span class="sd">        (default is True)</span>

<span class="sd">    figsize : tuple, optional</span>
<span class="sd">        Dimension of the figure size.</span>
<span class="sd">        (default is (5, 8))</span>

<span class="sd">    Outputs:</span>
<span class="sd">    --------</span>
<span class="sd">    The gene visualization figures are saved in ./nico_out/covariations_R*_F*/dotplots/*</span>

<span class="sd">    Example:</span>
<span class="sd">    --------</span>
<span class="sd">    &gt;&gt;&gt; scov.plot_top_genes_for_pair_of_celltypes_from_two_chosen_factors(cov_out,</span>
<span class="sd">    choose_interacting_celltype_pair=[&#39;Stem/TA&#39;,&#39;Paneth&#39;],</span>
<span class="sd">    visualize_factors_id=[1,1],</span>
<span class="sd">    top_NOG=20,saveas=saveas,transparent_mode=transparent_mode)</span>
<span class="sd">    &quot;&quot;&quot;</span>


    <span class="n">savefigdir</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">covariation_dir</span><span class="o">+</span> <span class="s1">&#39;dotplots/&#39;</span>
    <span class="n">create_directory</span><span class="p">(</span><span class="n">savefigdir</span><span class="p">)</span>

    <span class="n">PCA_of_sc_cluster_accordingto_spatial_clusterid</span><span class="p">,</span><span class="n">save_scFactors</span><span class="p">,</span><span class="n">save_spFactors</span><span class="o">=</span><span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">covariation_dir</span><span class="o">+</span><span class="s1">&#39;factors_info.p&#39;</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">))</span>
    <span class="n">n</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clustername</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">fi</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="p">):</span>
        <span class="n">clid</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clusterid</span><span class="p">[</span><span class="n">fi</span><span class="p">]</span>
        <span class="n">spearman_factors</span><span class="p">,</span><span class="n">CC_PCA</span><span class="p">,</span><span class="n">CC_gene</span><span class="p">,</span><span class="n">CC_meanExpression</span><span class="p">,</span><span class="n">CC_popExpression</span><span class="p">,</span><span class="n">cosine_factors</span><span class="p">,</span><span class="n">alpha</span><span class="o">=</span><span class="n">PCA_of_sc_cluster_accordingto_spatial_clusterid</span><span class="p">[</span><span class="n">clid</span><span class="p">]</span>
        <span class="n">CC_celltype_name</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clustername</span><span class="p">[</span><span class="n">fi</span><span class="p">]</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">CC_celltype_name</span><span class="o">==</span><span class="n">choose_interacting_celltype_pair</span><span class="p">[</span><span class="mi">0</span><span class="p">]):</span>
            <span class="n">mu1</span><span class="o">=</span><span class="n">CC_meanExpression</span>
            <span class="n">pop1</span><span class="o">=</span><span class="n">CC_popExpression</span>
            <span class="k">if</span> <span class="n">correlation_with_spearman</span><span class="p">:</span>
                <span class="n">z1</span><span class="o">=</span><span class="n">spearman_factors</span><span class="p">[:,</span><span class="n">visualize_factors_id</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">z1</span><span class="o">=</span><span class="n">cosine_factors</span><span class="p">[:,</span><span class="n">visualize_factors_id</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
            <span class="n">ga1</span><span class="p">,</span><span class="n">va1</span><span class="o">=</span><span class="n">find_interest_of_genes</span><span class="p">(</span><span class="n">z1</span><span class="p">,</span><span class="n">pop1</span><span class="p">,</span><span class="n">mu1</span><span class="p">,</span><span class="n">rps_rpl_mt_genes_included</span><span class="p">,</span><span class="n">CC_gene</span><span class="p">,</span><span class="n">top_NOG</span><span class="p">)</span>


    <span class="k">for</span> <span class="n">fi</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="p">):</span>
        <span class="n">clid</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clusterid</span><span class="p">[</span><span class="n">fi</span><span class="p">]</span>
        <span class="n">spearman_factors</span><span class="p">,</span><span class="n">CC_PCA</span><span class="p">,</span><span class="n">CC_gene</span><span class="p">,</span><span class="n">CC_meanExpression</span><span class="p">,</span><span class="n">CC_popExpression</span><span class="p">,</span><span class="n">cosine_factors</span><span class="p">,</span><span class="n">alpha</span><span class="o">=</span><span class="n">PCA_of_sc_cluster_accordingto_spatial_clusterid</span><span class="p">[</span><span class="n">clid</span><span class="p">]</span>
        <span class="n">CC_celltype_name</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">spatialcell_unique_clustername</span><span class="p">[</span><span class="n">fi</span><span class="p">]</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">CC_celltype_name</span><span class="o">==</span><span class="n">choose_interacting_celltype_pair</span><span class="p">[</span><span class="mi">1</span><span class="p">]):</span>
            <span class="n">mu2</span><span class="o">=</span><span class="n">CC_meanExpression</span>
            <span class="n">pop2</span><span class="o">=</span><span class="n">CC_popExpression</span>
            <span class="k">if</span> <span class="n">correlation_with_spearman</span><span class="p">:</span>
                <span class="n">z2</span><span class="o">=</span><span class="n">spearman_factors</span><span class="p">[:,</span><span class="n">visualize_factors_id</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="c1">#np.hstack((z,spearman_factors))</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">z2</span><span class="o">=</span><span class="n">cosine_factors</span><span class="p">[:,</span><span class="n">visualize_factors_id</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="c1">#np.hstack((z,cosine_factors))</span>
            <span class="n">ga2</span><span class="p">,</span><span class="n">va2</span><span class="o">=</span><span class="n">find_interest_of_genes</span><span class="p">(</span><span class="n">z2</span><span class="p">,</span><span class="n">pop2</span><span class="p">,</span><span class="n">mu2</span><span class="p">,</span><span class="n">rps_rpl_mt_genes_included</span><span class="p">,</span><span class="n">CC_gene</span><span class="p">,</span><span class="n">top_NOG</span><span class="p">)</span>


    <span class="n">comgenes</span><span class="o">=</span><span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">ga1</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">i</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">comgenes</span><span class="p">:</span>
            <span class="n">comgenes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">ga2</span><span class="p">:</span>
        <span class="c1">#if i not in comgenes:</span>
            <span class="n">comgenes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
    <span class="n">n</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">comgenes</span><span class="p">)</span>
    <span class="n">gex</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">n</span><span class="p">,</span><span class="mi">2</span><span class="p">),</span><span class="n">dtype</span><span class="o">=</span><span class="nb">float</span><span class="p">)</span>
    <span class="n">gpop</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">n</span><span class="p">,</span><span class="mi">2</span><span class="p">),</span><span class="n">dtype</span><span class="o">=</span><span class="nb">float</span><span class="p">)</span>
    <span class="n">gfact</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">n</span><span class="p">,</span><span class="mi">2</span><span class="p">),</span><span class="n">dtype</span><span class="o">=</span><span class="nb">float</span><span class="p">)</span>
    <span class="n">max1</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="nb">max</span><span class="p">(</span><span class="n">mu1</span><span class="p">))</span>
    <span class="n">max2</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="nb">max</span><span class="p">(</span><span class="n">mu2</span><span class="p">))</span>

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">comgenes</span><span class="p">)):</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">CC_gene</span><span class="p">)):</span>
            <span class="k">if</span> <span class="n">comgenes</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">==</span><span class="n">CC_gene</span><span class="p">[</span><span class="n">j</span><span class="p">]:</span>
                    <span class="n">gex</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="mi">0</span><span class="p">]</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">mu1</span><span class="p">[</span><span class="n">j</span><span class="p">])</span><span class="c1">#np.log(mu1[j])/max1 #</span>
                    <span class="n">gpop</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="mi">0</span><span class="p">]</span><span class="o">=</span><span class="n">pop1</span><span class="p">[</span><span class="n">j</span><span class="p">]</span>
                    <span class="n">gex</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="mi">1</span><span class="p">]</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">mu2</span><span class="p">[</span><span class="n">j</span><span class="p">])</span><span class="c1">#np.log(mu2[j])/max2 #</span>
                    <span class="n">gpop</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="mi">1</span><span class="p">]</span><span class="o">=</span><span class="n">pop2</span><span class="p">[</span><span class="n">j</span><span class="p">]</span>
                    <span class="n">gfact</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="mi">0</span><span class="p">]</span><span class="o">=</span><span class="n">z1</span><span class="p">[</span><span class="n">j</span><span class="p">]</span>
                    <span class="n">gfact</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="mi">1</span><span class="p">]</span><span class="o">=</span><span class="n">z2</span><span class="p">[</span><span class="n">j</span><span class="p">]</span>

    <span class="n">xlabels11</span><span class="o">=</span><span class="p">[</span><span class="n">choose_interacting_celltype_pair</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">,</span><span class="s1">&#39; &#39;</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;:avg exp&#39;</span><span class="p">,</span><span class="n">choose_interacting_celltype_pair</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">,</span><span class="s1">&#39; &#39;</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;:Fa&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">visualize_factors_id</span><span class="p">[</span><span class="mi">0</span><span class="p">]),</span>
                 <span class="n">choose_interacting_celltype_pair</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">,</span><span class="s1">&#39; &#39;</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;:avg exp&#39;</span><span class="p">,</span><span class="n">choose_interacting_celltype_pair</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">,</span><span class="s1">&#39; &#39;</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;:Fa&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">visualize_factors_id</span><span class="p">[</span><span class="mi">1</span><span class="p">])]</span>
    <span class="n">xlabels</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">xlabels</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">=</span><span class="p">[</span><span class="n">choose_interacting_celltype_pair</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">,</span><span class="s1">&#39; &#39;</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;:Fa&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">visualize_factors_id</span><span class="p">[</span><span class="mi">0</span><span class="p">]),</span><span class="n">choose_interacting_celltype_pair</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">,</span><span class="s1">&#39; &#39;</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;:Fa&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">visualize_factors_id</span><span class="p">[</span><span class="mi">1</span><span class="p">])]</span>
    <span class="n">xlabels</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">=</span><span class="p">[</span><span class="n">choose_interacting_celltype_pair</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">,</span><span class="s1">&#39; &#39;</span><span class="p">),</span><span class="n">choose_interacting_celltype_pair</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">,</span><span class="s1">&#39; &#39;</span><span class="p">)]</span>



    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">)</span>
    <span class="n">x</span><span class="p">,</span><span class="n">y</span><span class="p">,</span><span class="n">z</span><span class="p">,</span><span class="n">bigs</span><span class="o">=</span><span class="n">findXYZC</span><span class="p">(</span><span class="n">gfact</span><span class="p">,</span><span class="n">gpop</span><span class="p">)</span>
    <span class="n">p0</span><span class="o">=</span><span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">x</span><span class="p">,</span><span class="n">y</span><span class="p">,</span><span class="n">s</span><span class="o">=</span><span class="n">bigs</span><span class="p">,</span><span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">,</span><span class="n">c</span><span class="o">=</span><span class="n">z</span><span class="p">,</span><span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;RdBu_r&#39;</span><span class="p">)</span> <span class="c1">#&#39;cm.cmap_name</span>
    <span class="n">x</span><span class="p">,</span><span class="n">y</span><span class="p">,</span><span class="n">z</span><span class="p">,</span><span class="n">bigs</span><span class="o">=</span><span class="n">findXYZC</span><span class="p">(</span><span class="n">gex</span><span class="p">,</span><span class="n">gpop</span><span class="p">)</span>
    <span class="n">p1</span><span class="o">=</span><span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">x</span><span class="p">,</span><span class="n">y</span><span class="p">,</span><span class="n">s</span><span class="o">=</span><span class="n">bigs</span><span class="p">,</span><span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">,</span><span class="n">c</span><span class="o">=</span><span class="n">z</span><span class="p">,</span><span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;RdBu_r&#39;</span><span class="p">)</span>

    <span class="n">kw</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">prop</span><span class="o">=</span><span class="s2">&quot;sizes&quot;</span><span class="p">,</span> <span class="n">num</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.6</span><span class="p">,</span> <span class="n">fmt</span><span class="o">=</span><span class="s2">&quot;% </span><span class="si">{x:.0f}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">legend2</span> <span class="o">=</span> <span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="o">*</span><span class="n">p1</span><span class="o">.</span><span class="n">legend_elements</span><span class="p">(</span><span class="o">**</span><span class="n">kw</span><span class="p">),</span><span class="n">loc</span><span class="o">=</span><span class="s2">&quot;upper right&quot;</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s2">&quot;Fraction of </span><span class="se">\n</span><span class="s2">cells </span><span class="se">\n</span><span class="s2">expressed&quot;</span><span class="p">,</span><span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mf">1.</span><span class="p">))</span>


    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">2</span><span class="p">):</span>
        <span class="n">ax</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">gfact</span><span class="p">)))</span>
        <span class="k">if</span> <span class="n">i</span><span class="o">==</span><span class="mi">0</span><span class="p">:</span>
            <span class="n">ax</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticklabels</span><span class="p">(</span><span class="n">comgenes</span><span class="p">,</span><span class="n">style</span><span class="o">=</span><span class="s1">&#39;italic&#39;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">ax</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticklabels</span><span class="p">([])</span>
        <span class="n">ax</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">2</span><span class="p">))</span>
        <span class="n">ax</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">set_xticklabels</span><span class="p">(</span><span class="n">xlabels</span><span class="p">[</span><span class="n">i</span><span class="p">],</span><span class="n">rotation</span><span class="o">=</span><span class="mi">30</span><span class="p">)</span>
        <span class="n">ax</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">([</span><span class="o">-</span><span class="mf">0.5</span><span class="p">,</span><span class="mf">1.5</span><span class="p">])</span>
        <span class="n">ax</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">([</span><span class="o">-</span><span class="mf">0.5</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">gfact</span><span class="p">)</span><span class="o">+</span><span class="mf">0.5</span><span class="p">])</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">colorbar</span><span class="p">(</span><span class="n">p0</span><span class="p">,</span><span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="n">shrink</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">colorbar</span><span class="p">(</span><span class="n">p1</span><span class="p">,</span><span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span><span class="n">shrink</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
    <span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;spearman correlation&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;log(avg expression)&#39;</span><span class="p">)</span>
    <span class="n">pos</span><span class="o">=</span><span class="n">top_NOG</span><span class="o">-</span><span class="mf">0.5</span>
    <span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span><span class="mi">3</span><span class="p">],[</span><span class="n">pos</span><span class="p">,</span><span class="n">pos</span><span class="p">],</span><span class="s1">&#39;k--&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span><span class="mi">3</span><span class="p">],[</span><span class="n">pos</span><span class="p">,</span><span class="n">pos</span><span class="p">],</span><span class="s1">&#39;k--&#39;</span><span class="p">)</span>


    <span class="n">fig</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>

    <span class="n">savename</span> <span class="o">=</span> <span class="n">remove_extra_character_from_name</span><span class="p">(</span><span class="n">choose_interacting_celltype_pair</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span><span class="o">+</span><span class="s1">&#39;_&#39;</span><span class="o">+</span><span class="n">remove_extra_character_from_name</span><span class="p">(</span><span class="n">choose_interacting_celltype_pair</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;The figures are saved: &quot;</span><span class="p">,</span> <span class="n">savefigdir</span><span class="o">+</span><span class="s1">&#39;combined_&#39;</span><span class="o">+</span><span class="n">savename</span><span class="o">+</span><span class="s1">&#39;.&#39;</span><span class="o">+</span><span class="n">saveas</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">savefigdir</span><span class="o">+</span><span class="s1">&#39;combined_&#39;</span><span class="o">+</span><span class="n">savename</span><span class="o">+</span><span class="s1">&#39;.&#39;</span><span class="o">+</span><span class="n">saveas</span><span class="p">,</span><span class="n">bbox_inches</span><span class="o">=</span><span class="s1">&#39;tight&#39;</span><span class="p">,</span><span class="n">transparent</span><span class="o">=</span><span class="n">transparent_mode</span><span class="p">,</span><span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">showit</span><span class="p">:</span>
        <span class="k">pass</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="s1">&#39;all&#39;</span><span class="p">)</span></div>







<div class="viewcode-block" id="find_interest_of_genes">
<a class="viewcode-back" href="../../nico_covariations.html#nico_covariations.Covariations.find_interest_of_genes">[docs]</a>
<span class="k">def</span> <span class="nf">find_interest_of_genes</span><span class="p">(</span><span class="n">source</span><span class="p">,</span><span class="n">pop</span><span class="p">,</span><span class="n">mu</span><span class="p">,</span><span class="n">rps_rpl_mt_genes_included</span><span class="p">,</span><span class="n">CC_gene</span><span class="p">,</span><span class="n">top_NOG</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Find genes of interest based on factor values, population expression, and average expression.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    -----------</span>
<span class="sd">    source : np.array</span>
<span class="sd">        Factor values associated with genes.</span>

<span class="sd">    pop : np.array</span>
<span class="sd">        Population expression values.</span>

<span class="sd">    mu : np.array</span>
<span class="sd">        Mean expression values.</span>

<span class="sd">    rps_rpl_mt_genes_included : bool</span>
<span class="sd">        Include Rps, Rpl, and mt genes if True.</span>

<span class="sd">    CC_gene : np.array</span>
<span class="sd">        List of gene names.</span>

<span class="sd">    top_NOG : int</span>
<span class="sd">        Number of genes to select.</span>

<span class="sd">    Returns:</span>
<span class="sd">    --------</span>
<span class="sd">    gp1 : list</span>
<span class="sd">        Genes of interest based on factor values.</span>

<span class="sd">    vp1 : list</span>
<span class="sd">        Corresponding factor values of the selected genes.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">ind</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="o">-</span><span class="n">source</span><span class="p">)</span>
    <span class="n">interestofGene</span><span class="o">=</span><span class="p">[]</span>
    <span class="n">value_fact</span><span class="o">=</span><span class="p">[]</span>
    <span class="n">value_pop</span><span class="o">=</span><span class="p">[]</span>
    <span class="n">value_avgexp</span><span class="o">=</span><span class="p">[]</span>
    <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">source</span><span class="p">)):</span>
        <span class="n">temp</span><span class="o">=</span><span class="n">CC_gene</span><span class="p">[</span><span class="n">ind</span><span class="p">[</span><span class="n">k</span><span class="p">]]</span>
        <span class="k">if</span> <span class="n">rps_rpl_mt_genes_included</span><span class="p">:</span>
            <span class="n">flag</span><span class="o">=</span><span class="mi">1</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">flag</span><span class="o">=</span><span class="mi">1</span>
            <span class="k">if</span> <span class="n">temp</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">3</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;Rps&#39;</span><span class="p">:</span>
                <span class="n">flag</span><span class="o">=</span><span class="mi">0</span>
            <span class="k">if</span> <span class="n">temp</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">3</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;Rpl&#39;</span><span class="p">:</span>
                <span class="n">flag</span><span class="o">=</span><span class="mi">0</span>
            <span class="k">if</span> <span class="n">temp</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">3</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;mt-&#39;</span><span class="p">:</span>
                <span class="n">flag</span><span class="o">=</span><span class="mi">0</span>
        <span class="k">if</span> <span class="n">flag</span><span class="o">==</span><span class="mi">1</span><span class="p">:</span>
            <span class="n">interestofGene</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">CC_gene</span><span class="p">[</span><span class="n">ind</span><span class="p">[</span><span class="n">k</span><span class="p">]])</span>
            <span class="n">value_fact</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">source</span><span class="p">[</span><span class="n">ind</span><span class="p">[</span><span class="n">k</span><span class="p">]])</span>
            <span class="n">value_pop</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">pop</span><span class="p">[</span><span class="n">ind</span><span class="p">[</span><span class="n">k</span><span class="p">]])</span>
            <span class="n">value_avgexp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mu</span><span class="p">[</span><span class="n">ind</span><span class="p">[</span><span class="n">k</span><span class="p">]])</span>

    <span class="n">value_fact</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">value_fact</span><span class="p">)</span>
    <span class="n">value_pop</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">value_pop</span><span class="p">)</span>
    <span class="n">value_avgexp</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">value_avgexp</span><span class="p">)</span>
    <span class="n">interestofGene</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">interestofGene</span><span class="p">)</span>

    <span class="n">index_pos</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="o">-</span><span class="n">value_fact</span><span class="p">)</span>
    <span class="c1">#index_neg=np.argsort(value_fact)</span>

    <span class="c1">#print(interestofGene[index_pos])</span>
    <span class="n">gp1</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">interestofGene</span><span class="p">[</span><span class="n">index_pos</span><span class="p">])[</span><span class="mi">0</span><span class="p">:</span><span class="n">top_NOG</span><span class="p">]</span>
    <span class="n">vp1</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">value_fact</span><span class="p">[</span><span class="n">index_pos</span><span class="p">])[</span><span class="mi">0</span><span class="p">:</span><span class="n">top_NOG</span><span class="p">]</span>
    <span class="n">pos_pop1</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">value_pop</span><span class="p">[</span><span class="n">index_pos</span><span class="p">])[</span><span class="mi">0</span><span class="p">:</span><span class="n">top_NOG</span><span class="p">]</span>
    <span class="n">pos_avg1</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">value_avgexp</span><span class="p">[</span><span class="n">index_pos</span><span class="p">])[</span><span class="mi">0</span><span class="p">:</span><span class="n">top_NOG</span><span class="p">]</span>

    <span class="c1">#gn1=list(interestofGene[index_neg])</span>

    <span class="c1">#comgene.append(gn1[0:top_NOG])</span>
    <span class="c1">#gex=np.zeros((top_NOG,1),dtype=float)</span>
    <span class="c1">#vn1=list(value_fact[index_neg])[0:top_NOG]</span>
    <span class="c1">#neg_pop1=list(value_pop[index_neg])[0:top_NOG]</span>
    <span class="c1">#neg_avg1=list(value_avgexp[index_neg])[0:top_NOG]</span>
    <span class="k">return</span> <span class="n">gp1</span><span class="p">,</span> <span class="n">vp1</span></div>

</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Grn lab.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>