<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>nico_covariations package &mdash; NiCo 1.0.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=19f00094" />
      <link rel="stylesheet" type="text/css" href="_static/copybutton.css?v=76b2166b" />

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="_static/jquery.js?v=5d32c60e"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="_static/documentation_options.js?v=8d563738"></script>
        <script src="_static/doctools.js?v=888ff710"></script>
        <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
        <script src="_static/clipboard.min.js?v=a7894cd8"></script>
        <script src="_static/copybutton.js?v=f281be69"></script>
        <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="nico_covariations.utils package" href="nico_covariations.utils.html" />
    <link rel="prev" title="nico_annotations.utils package" href="nico_annotations.utils.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html">
            
              <img src="_static/nico-logo.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples.html">Examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples.html#tutorials-with-visualization">Tutorials with visualization</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="modules.html">NiCo_Modules</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="nico_annotations.html">nico_annotations package</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">nico_covariations package</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#subpackages">Subpackages</a></li>
<li class="toctree-l3"><a class="reference internal" href="#submodules">Submodules</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-nico_covariations.Covariations">nico_covariations.Covariations module</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-nico_covariations">Module contents</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="nico_interactions.html">nico_interactions package</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html">utils package</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="api_reference.html">API Reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="faq.html">Frequently Asked Questions (FAQ)</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">NiCo</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="modules.html">NiCo_Modules</a></li>
      <li class="breadcrumb-item active">nico_covariations package</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/nico_covariations.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="nico-covariations-package">
<h1>nico_covariations package<a class="headerlink" href="#nico-covariations-package" title="Link to this heading"></a></h1>
<section id="subpackages">
<h2>Subpackages<a class="headerlink" href="#subpackages" title="Link to this heading"></a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="nico_covariations.utils.html">nico_covariations.utils package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="nico_covariations.utils.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="nico_covariations.utils.html#module-nico_covariations.utils.pyliger_utilities">nico_covariations.utils.pyliger_utilities module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="nico_covariations.utils.html#nico_covariations.utils.pyliger_utilities.NMF_obj_eval"><code class="docutils literal notranslate"><span class="pre">NMF_obj_eval()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="nico_covariations.utils.html#nico_covariations.utils.pyliger_utilities.column_group_sub"><code class="docutils literal notranslate"><span class="pre">column_group_sub()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="nico_covariations.utils.html#nico_covariations.utils.pyliger_utilities.iNMF"><code class="docutils literal notranslate"><span class="pre">iNMF()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="nico_covariations.utils.html#nico_covariations.utils.pyliger_utilities.nnlsm_blockpivot"><code class="docutils literal notranslate"><span class="pre">nnlsm_blockpivot()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="nico_covariations.utils.html#nico_covariations.utils.pyliger_utilities.nonneg"><code class="docutils literal notranslate"><span class="pre">nonneg()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="nico_covariations.utils.html#nico_covariations.utils.pyliger_utilities.normal_eq_comb"><code class="docutils literal notranslate"><span class="pre">normal_eq_comb()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="nico_covariations.utils.html#module-nico_covariations.utils">Module contents</a></li>
</ul>
</li>
</ul>
</div>
</section>
<section id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Link to this heading"></a></h2>
</section>
<section id="module-nico_covariations.Covariations">
<span id="nico-covariations-covariations-module"></span><h2>nico_covariations.Covariations module<a class="headerlink" href="#module-nico_covariations.Covariations" title="Link to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="nico_covariations.Covariations.alignment_score">
<span class="sig-prename descclassname"><span class="pre">nico_covariations.Covariations.</span></span><span class="sig-name descname"><span class="pre">alignment_score</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">H</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spH</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ind_H</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ind_spH</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/nico_covariations/Covariations.html#alignment_score"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#nico_covariations.Covariations.alignment_score" title="Link to this definition"></a></dt>
<dd><p>Calculate the alignment score between factors from two different modalities during integrated NMF.</p>
<p>This helper function is used in <cite>find_PC_of_invidualCluster_in_SC</cite> to evaluate the alignment
score between factors from scRNAseq data and spatial data.</p>
<section id="parameters">
<h3>Parameters<a class="headerlink" href="#parameters" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>H<span class="classifier">numpy.ndarray</span></dt><dd><p>The matrix representing the factors from the scRNAseq data.
Each row corresponds to a sample, and each column corresponds to a factor.</p>
</dd>
<dt>spH<span class="classifier">numpy.ndarray</span></dt><dd><p>The matrix representing the factors from the spatial data.
Each row corresponds to a sample, and each column corresponds to a factor.</p>
</dd>
<dt>ind_H<span class="classifier">numpy.ndarray or list</span></dt><dd><p>Indices of the common genes in the scRNAseq data.</p>
</dd>
<dt>ind_spH<span class="classifier">numpy.ndarray or list</span></dt><dd><p>Indices of the common genes in the spatial data.</p>
</dd>
</dl>
</section>
<section id="returns">
<h3>Returns<a class="headerlink" href="#returns" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>float</dt><dd><p>The alignment score between the factors from the scRNAseq and spatial data.</p>
</dd>
</dl>
</section>
<section id="notes">
<h3>Notes<a class="headerlink" href="#notes" title="Link to this heading"></a></h3>
<p>The alignment score is calculated by computing the cosine similarity between the factors of
the common genes in the scRNAseq and spatial data. A higher score indicates better alignment
between the factors from the two modalities.</p>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="nico_covariations.Covariations.compute_PC_space">
<span class="sig-prename descclassname"><span class="pre">nico_covariations.Covariations.</span></span><span class="sig-name descname"><span class="pre">compute_PC_space</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sct_ad_sc_full</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/nico_covariations/Covariations.html#compute_PC_space"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#nico_covariations.Covariations.compute_PC_space" title="Link to this definition"></a></dt>
<dd><p>Helper function in gene_covariation_analysis to find the weighted neighborhood
average of cell types from the spatial factors.</p>
<p>This function computes the weighted neighborhood average of principal components (PCs) for each cell type from spatial transcriptomics data. The weights are based on the inverse of the distances between neighboring cells.</p>
<section id="id1">
<h3>Parameters<a class="headerlink" href="#id1" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>input<span class="classifier">object</span></dt><dd><p>An object containing the following attributes:</p>
<ul class="simple">
<li><p>spatialcell_unique_clusterid: list of unique spatial cell cluster IDs.</p></li>
<li><p>neighbors: list of neighbors for each cell.</p></li>
<li><p>neigh_distances: list of distances to neighbors for each cell.</p></li>
<li><p>annotation_spatial_barcode_id: list of spatial barcode IDs for each cell.</p></li>
<li><p>annotation_spatial_cluster_id: list of spatial cluster IDs for each cell.</p></li>
<li><p>pc_of_sp_clusterid: matrix of principal components for each spatial cluster ID.</p></li>
<li><p>no_of_pc: int, number of principal components.</p></li>
<li><p>outputname: str, the name of the output file to save the results.</p></li>
</ul>
</dd>
</dl>
</section>
<section id="id2">
<h3>Returns<a class="headerlink" href="#id2" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>None</dt><dd><p>This function saves the weighted neighborhood of factors in a niche to a .npz file specified by input.outputname.</p>
</dd>
</dl>
</section>
<section id="id3">
<h3>Notes<a class="headerlink" href="#id3" title="Link to this heading"></a></h3>
<ul class="simple">
<li><p>This function calculates the weighted average of the principal components (PCs) for each cell’s neighborhood, using the inverse of the distances to its neighbors as weights.</p></li>
<li><p>The result is a matrix where each row represents a cell, and each column represents the weighted average PC values for each cluster in the cell’s neighborhood.</p></li>
<li><p>The weighted neighborhood feature matrix is saved to a file in .npz format.</p></li>
</ul>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="nico_covariations.Covariations.create_directory">
<span class="sig-prename descclassname"><span class="pre">nico_covariations.Covariations.</span></span><span class="sig-name descname"><span class="pre">create_directory</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">outputFolder</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/nico_covariations/Covariations.html#create_directory"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#nico_covariations.Covariations.create_directory" title="Link to this definition"></a></dt>
<dd><p>Create an empty directory.</p>
<p>This function checks if a specified directory exists, and if not, it creates the directory.</p>
<section id="id4">
<h3>Parameters<a class="headerlink" href="#id4" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>outputFolder<span class="classifier">str</span></dt><dd><p>The path of the directory to be created.</p>
</dd>
</dl>
</section>
<section id="raises">
<h3>Raises<a class="headerlink" href="#raises" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>OSError</dt><dd><p>If the directory cannot be created due to permission issues or other OS-related errors.</p>
</dd>
</dl>
</section>
<section id="id5">
<h3>Notes<a class="headerlink" href="#id5" title="Link to this heading"></a></h3>
<ul class="simple">
<li><p>If the directory already exists, no action is taken.</p></li>
<li><p>This function ensures that the directory path is available for subsequent file operations.</p></li>
</ul>
</section>
<section id="example">
<h3>Example<a class="headerlink" href="#example" title="Link to this heading"></a></h3>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">create_directory</span><span class="p">(</span><span class="s1">&#39;./new_out/&#39;</span><span class="p">)</span>
</pre></div>
</div>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="nico_covariations.Covariations.create_subtitle">
<span class="sig-prename descclassname"><span class="pre">nico_covariations.Covariations.</span></span><span class="sig-name descname"><span class="pre">create_subtitle</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fig</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Figure</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">grid</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">SubplotSpec</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/nico_covariations/Covariations.html#create_subtitle"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#nico_covariations.Covariations.create_subtitle" title="Link to this definition"></a></dt>
<dd><p>Add a title to a specific set of subplots within a figure.</p>
<p>This helper function is used to create a title for a subset of plots within
a Matplotlib figure. The title is added with a specific formatting and the
subplot is hidden from view (no axes or frames).</p>
<section id="id6">
<h3>Parameters:<a class="headerlink" href="#id6" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>fig<span class="classifier">matplotlib.figure.Figure</span></dt><dd><p>The figure object to which the subplot belongs.</p>
</dd>
<dt>grid<span class="classifier">matplotlib.gridspec.SubplotSpec</span></dt><dd><p>The subplot specification that defines the location and size of the subplot within the figure.</p>
</dd>
<dt>title<span class="classifier">str</span></dt><dd><p>The title text to be displayed above the subplot.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="nico_covariations.Covariations.extract_and_plot_top_genes_from_chosen_factor_in_celltype">
<span class="sig-prename descclassname"><span class="pre">nico_covariations.Covariations.</span></span><span class="sig-name descname"><span class="pre">extract_and_plot_top_genes_from_chosen_factor_in_celltype</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">choose_celltype</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">choose_factor_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">top_NOG</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rps_rpl_mt_genes_included</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">correlation_with_spearman</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">positively_correlated</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">saveas</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'pdf'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'RdBu_r'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">transparent_mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">showit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(5,</span> <span class="pre">6)</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/nico_covariations/Covariations.html#extract_and_plot_top_genes_from_chosen_factor_in_celltype"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#nico_covariations.Covariations.extract_and_plot_top_genes_from_chosen_factor_in_celltype" title="Link to this definition"></a></dt>
<dd><p>Extract and plot top genes associated with a chosen factor in a specified cell type.</p>
<p>This function uses the output from gene_covariation_analysis to identify and visualize the top genes associated with a chosen factor in a specified cell type. The genes can be filtered and visualized based on their correlation with the factor, with options to include or exclude specific gene types.</p>
<section id="id7">
<h3>Parameters<a class="headerlink" href="#id7" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>input<span class="classifier">object</span></dt><dd><p>The main input is the output from gene_covariation_analysis.</p>
</dd>
<dt>choose_celltype<span class="classifier">str</span></dt><dd><p>Define the cell type to include in the analysis.</p>
</dd>
<dt>choose_factor_id<span class="classifier">int</span></dt><dd><p>Define the factor ID of the cell type to be analyzed.</p>
</dd>
<dt>top_NOG<span class="classifier">int, optional</span></dt><dd><p>Number of top genes to visualize.
(default is 30)</p>
</dd>
<dt>rps_rpl_mt_genes_included<span class="classifier">bool, optional</span></dt><dd><p>Decide whether to include rps, rpl, and mt genes in the pathway analysis. If True, they are included.
(default is True)</p>
</dd>
<dt>correlation_with_spearman<span class="classifier">bool, optional</span></dt><dd><p>If True, visualize gene-factor association using the Spearman correlation coefficient; otherwise, use cosine similarity.
(default is True)</p>
</dd>
<dt>positively_correlated<span class="classifier">bool, optional</span></dt><dd><p>If the gene-factor association is selected as Spearman correlation, choose whether the associated genes should be positively correlated (True) or negatively correlated (False).
(default is True)</p>
</dd>
<dt>saveas<span class="classifier">str, optional</span></dt><dd><p>Save the figures in PDF or PNG format (dpi for PNG format is 300).
(default is ‘pdf’)</p>
</dd>
<dt>cmap<span class="classifier">str, optional</span></dt><dd><p>Define the colormap for visualizing factors.
(default is ‘RdBu_r’)</p>
</dd>
<dt>transparent_mode<span class="classifier">bool, optional</span></dt><dd><p>Define the background color of the figures. If True, figures have a transparent background.
(default is False)</p>
</dd>
<dt>showit<span class="classifier">bool, optional</span></dt><dd><p>If True, the generated figures will be displayed.
(default is True)</p>
</dd>
<dt>figsize<span class="classifier">tuple, optional</span></dt><dd><p>Dimension of the figure size.
(default is (5, 6))</p>
</dd>
</dl>
</section>
<section id="outputs">
<h3>Outputs<a class="headerlink" href="#outputs" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>pd.DataFrame</dt><dd><p>Returns a DataFrame containing the gene, factor, average expression, and proportion of the population expressing that gene.</p>
</dd>
</dl>
</section>
<section id="id8">
<h3>Notes<a class="headerlink" href="#id8" title="Link to this heading"></a></h3>
<ul class="simple">
<li><p>The function saves the figures in the directory “nico_out/covariations_R*_F*/dotplots/Factors*”.</p></li>
<li><p>The DataFrame returned includes detailed information about the top genes associated with the chosen factor.</p></li>
</ul>
</section>
<section id="id9">
<h3>Example<a class="headerlink" href="#id9" title="Link to this heading"></a></h3>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">extract_and_plot_top_genes_from_chosen_factor_in_celltype</span><span class="p">(</span><span class="n">input_data</span><span class="p">,</span> <span class="s1">&#39;CellTypeA&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">top_NOG</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">saveas</span><span class="o">=</span><span class="s1">&#39;png&#39;</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
</pre></div>
</div>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="nico_covariations.Covariations.findXYZC">
<span class="sig-prename descclassname"><span class="pre">nico_covariations.Covariations.</span></span><span class="sig-name descname"><span class="pre">findXYZC</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">c</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">s</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/nico_covariations/Covariations.html#findXYZC"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#nico_covariations.Covariations.findXYZC" title="Link to this definition"></a></dt>
<dd><p>Helper function used in plot_top_selected_genes_as_dotplot.</p>
<p>This function extracts and transforms data from the given matrices c and s, creating four lists:
x-coordinates, y-coordinates, values (z), and sizes (bigs).</p>
<section id="id10">
<h3>Parameters<a class="headerlink" href="#id10" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>c<span class="classifier">array-like</span></dt><dd><p>A 2D array (matrix) where each element represents a value at a specific (i, j) coordinate.</p>
</dd>
<dt>s<span class="classifier">array-like</span></dt><dd><p>A 2D array (matrix) of the same shape as c, where each element represents a size multiplier for the corresponding
element in c.</p>
</dd>
</dl>
</section>
<section id="id11">
<h3>Returns<a class="headerlink" href="#id11" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>x<span class="classifier">list</span></dt><dd><p>List of x-coordinates for each element in c.</p>
</dd>
<dt>y<span class="classifier">list</span></dt><dd><p>List of y-coordinates for each element in c.</p>
</dd>
<dt>z<span class="classifier">list</span></dt><dd><p>List of values from c corresponding to each (x, y) coordinate.</p>
</dd>
<dt>bigs<span class="classifier">list</span></dt><dd><p>List of sizes, where each size is calculated as 100 times the corresponding element in s.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="nico_covariations.Covariations.find_LR_interactions_in_interacting_cell_types">
<span class="sig-prename descclassname"><span class="pre">nico_covariations.Covariations.</span></span><span class="sig-name descname"><span class="pre">find_LR_interactions_in_interacting_cell_types</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">choose_interacting_celltype_pair</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">choose_factors_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pvalueCutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">correlation_with_spearman</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">LR_plot_NMF_Fa_thres</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">LR_plot_Exp_thres</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">saveas</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'pdf'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">transparent_mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">showit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(12,</span> <span class="pre">10)</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/nico_covariations/Covariations.html#find_LR_interactions_in_interacting_cell_types"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#nico_covariations.Covariations.find_LR_interactions_in_interacting_cell_types" title="Link to this definition"></a></dt>
<dd><p>Find ligand-receptor (LR) interactions in interacting cell types and visualize them.</p>
<p>This function processes the output from gene_covariation_analysis to identify significant LR interactions between specified cell type pairs and visualizes the results.</p>
<section id="id12">
<h3>Parameters<a class="headerlink" href="#id12" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>input<span class="classifier">object</span></dt><dd><p>The main input is the output from gene_covariation_analysis.</p>
</dd>
<dt>choose_interacting_celltype_pair<span class="classifier">list, optional</span></dt><dd><p>Define the cell type pairs for which information on LR communication should be returned. The first element of the list is the central cell type (CC), and the second element is the niche cell type (NC).
If the list is empty, LR interactions will be returned for all significant interacting cell types.
Default is [].</p>
</dd>
<dt>choose_factors_id<span class="classifier">list, optional</span></dt><dd><p>Define factor IDs for which LR interactions are visualized. The first element of the list is the factor ID of the central cell type, and the second element is the factor ID of the niche cell type.
If the list is empty, LR plots will be saved for all significant niche cell type factor interactions.
Default is [].</p>
</dd>
<dt>pvalueCutoff<span class="classifier">float, optional</span></dt><dd><p>The p-value cutoff used to find the significant central cell type factor and niche cell type factor interactions.
Default is 0.05.</p>
</dd>
<dt>correlation_with_spearman<span class="classifier">bool, optional</span></dt><dd><p>If True, compute gene-factor correlation as Spearman correlation coefficient; otherwise, compute as cosine similarity.
Default is True.</p>
</dd>
<dt>LR_plot_NMF_Fa_thres<span class="classifier">float, optional</span></dt><dd><p>Only ligands or receptors that exhibit a correlation to the respective factors higher than this cutoff are retained.
Default is 0.2.</p>
</dd>
<dt>LR_plot_Exp_thres<span class="classifier">float, optional</span></dt><dd><p>Only ligands or receptors that are expressed in a fraction of cells of the respective cell types exceeding this cutoff are retained.
Default is 0.2.</p>
</dd>
<dt>saveas<span class="classifier">str, optional</span></dt><dd><p>Save the figures in PDF or PNG format (dpi for PNG format is 300).
Default is ‘pdf’.</p>
</dd>
<dt>transparent_mode<span class="classifier">bool, optional</span></dt><dd><p>Background color in the figures.
Default is False.</p>
</dd>
<dt>showit<span class="classifier">bool, optional</span></dt><dd><p>If True, the figures are shown interactively.
Default is True.</p>
</dd>
<dt>figsize<span class="classifier">tuple, optional</span></dt><dd><p>Dimension of the figure size. The figure size on the X-axis direction is the (number of genes) multiplied by factor 12/34.
The figure size on the Y-axis direction is the (number of genes) multiplied by factor 10/44.
All generated figure size are scaled according to the above factors.
Initital figure size is (12, 10).</p>
</dd>
</dl>
</section>
<section id="id13">
<h3>Outputs<a class="headerlink" href="#id13" title="Link to this heading"></a></h3>
<ul class="simple">
<li><p>The LR interaction figures are saved in “./nico_out/covariations_R*_F*/Plot_ligand_receptor_in_niche*”.</p></li>
</ul>
</section>
<section id="id14">
<h3>Notes<a class="headerlink" href="#id14" title="Link to this heading"></a></h3>
<ul class="simple">
<li><p>Our analysis accounts for bidirectional cellular crosstalk interactions of ligands and receptors in cell types A and B.</p></li>
<li><p>The ligand can be expressed on cell type A and signal to the receptor detected on cell type B, or vice versa.</p></li>
<li><p>Both ligand-receptor plots and Excel sheets profile bidirectional cellular crosstalk of ligand and receptor in cell types A and B.</p></li>
</ul>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="nico_covariations.Covariations.find_PC_of_invidualCluster_in_SC">
<span class="sig-prename descclassname"><span class="pre">nico_covariations.Covariations.</span></span><span class="sig-name descname"><span class="pre">find_PC_of_invidualCluster_in_SC</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seed</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spatial_integration_modality</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scbarcode</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">iNMFmode</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scadata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">no_of_pc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spbarcode</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spadata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sct_ad_sc_full</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">celltype_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cutoff_to_count_exp_cell_population</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/nico_covariations/Covariations.html#find_PC_of_invidualCluster_in_SC"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#nico_covariations.Covariations.find_PC_of_invidualCluster_in_SC" title="Link to this definition"></a></dt>
<dd><p>Helper function used in compute_PC_space to find principal components (PCs) for individual clusters in single-cell RNA sequencing (scRNA-seq) data and spatial transcriptomics data.</p>
<p>This function integrates scRNA-seq and spatial transcriptomics data using non-negative matrix factorization (NMF) or integrative NMF (iNMF), and computes the alignment score, correlation, and other metrics for the identified principal components.</p>
<section id="id15">
<h3>Parameters<a class="headerlink" href="#id15" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>seed<span class="classifier">int</span></dt><dd><p>Random seed for reproducibility.</p>
</dd>
<dt>spatial_integration_modality<span class="classifier">str</span></dt><dd><p>Modality for spatial integration, either ‘single’ or ‘double’.</p>
</dd>
<dt>scbarcode<span class="classifier">list</span></dt><dd><p>List of single-cell barcodes.</p>
</dd>
<dt>iNMFmode<span class="classifier">bool</span></dt><dd><p>Flag indicating whether to use iNMF (True) or not (False).</p>
</dd>
<dt>scadata<span class="classifier">AnnData</span></dt><dd><p>Single-cell RNA-seq data in AnnData format.</p>
</dd>
<dt>no_of_pc<span class="classifier">int</span></dt><dd><p>Number of principal components to compute.</p>
</dd>
<dt>spbarcode<span class="classifier">list</span></dt><dd><p>List of spatial transcriptomics barcodes.</p>
</dd>
<dt>spadata<span class="classifier">AnnData</span></dt><dd><p>Spatial transcriptomics data in AnnData format.</p>
</dd>
<dt>sct_ad_sc_full<span class="classifier">AnnData</span></dt><dd><p>Full single-cell RNA-seq data in AnnData format.</p>
</dd>
<dt>celltype_name<span class="classifier">str</span></dt><dd><p>Name of the cell type being analyzed.</p>
</dd>
<dt>cutoff_to_count_exp_cell_population<span class="classifier">float</span></dt><dd><p>Expression cutoff to count the proportion of cell population expressing a gene.</p>
</dd>
</dl>
</section>
<section id="id16">
<h3>Returns<a class="headerlink" href="#id16" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>transfer_sp_com<span class="classifier">ndarray</span></dt><dd><p>Transformed spatial component matrix.</p>
</dd>
<dt>transfer_sc_com<span class="classifier">list</span></dt><dd><p>Transformed single-cell component matrix (currently not populated).</p>
</dd>
<dt>sc_spearman<span class="classifier">ndarray</span></dt><dd><p>Spearman correlation between genes and principal components in single-cell data.</p>
</dd>
<dt>sc_cosine<span class="classifier">ndarray</span></dt><dd><p>Cosine similarity between genes and principal components in single-cell data.</p>
</dd>
<dt>sc_genenames<span class="classifier">ndarray</span></dt><dd><p>Array of gene names.</p>
</dd>
<dt>H<span class="classifier">ndarray</span></dt><dd><p>Principal component matrix for single-cell data.</p>
</dd>
<dt>spH<span class="classifier">ndarray</span></dt><dd><p>Principal component matrix for spatial data.</p>
</dd>
<dt>sc_cluster_mean_exp<span class="classifier">ndarray</span></dt><dd><p>Mean expression of genes across single-cell clusters.</p>
</dd>
<dt>sc_cluster_exp_more_than_threshold<span class="classifier">ndarray</span></dt><dd><p>Proportion of single-cell clusters expressing genes above the cutoff.</p>
</dd>
<dt>alpha<span class="classifier">int</span></dt><dd><p>Optimal alpha value used in iNMF.</p>
</dd>
</dl>
</section>
<section id="id17">
<h3>Notes<a class="headerlink" href="#id17" title="Link to this heading"></a></h3>
<ul class="simple">
<li><p>This function normalizes gene expression data and computes principal components using either NMF or iNMF.</p></li>
<li><p>It calculates the alignment score for spatial and single-cell data integration.</p></li>
<li><p>Spearman correlation and cosine similarity between genes and PCs are computed.</p></li>
<li><p>The results include the transformed spatial component matrix, gene correlations, and other metrics for downstream analysis.</p></li>
</ul>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="nico_covariations.Covariations.find_correlation_bw_genes_and_PC_component_in_singlecell">
<span class="sig-prename descclassname"><span class="pre">nico_covariations.Covariations.</span></span><span class="sig-name descname"><span class="pre">find_correlation_bw_genes_and_PC_component_in_singlecell</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">KcomponentCluster</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">clusterExpression</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/nico_covariations/Covariations.html#find_correlation_bw_genes_and_PC_component_in_singlecell"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#nico_covariations.Covariations.find_correlation_bw_genes_and_PC_component_in_singlecell" title="Link to this definition"></a></dt>
<dd><p>Calculate Spearman correlation between genes and principal components in single-cell data.</p>
<p>This helper function is used within the <cite>find_PC_of_invidualCluster_in_SC</cite> function to determine
the Spearman correlation between common gene scRNAseq factors (principal components) and scRNAseq
gene expression data.</p>
<section id="id18">
<h3>Parameters<a class="headerlink" href="#id18" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>KcomponentCluster<span class="classifier">numpy.ndarray or pandas.DataFrame</span></dt><dd><p>The matrix representing the principal components (factors) from scRNAseq data. Each column
corresponds to a principal component.</p>
</dd>
<dt>clusterExpression<span class="classifier">numpy.ndarray or pandas.DataFrame</span></dt><dd><p>The matrix representing the gene expression data from scRNAseq. Each row corresponds to a gene
and each column corresponds to a cell.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="nico_covariations.Covariations.find_correlation_bw_genes_and_PC_component_in_singlecell_cosine">
<span class="sig-prename descclassname"><span class="pre">nico_covariations.Covariations.</span></span><span class="sig-name descname"><span class="pre">find_correlation_bw_genes_and_PC_component_in_singlecell_cosine</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">KcomponentCluster</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">clusterExpression</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/nico_covariations/Covariations.html#find_correlation_bw_genes_and_PC_component_in_singlecell_cosine"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#nico_covariations.Covariations.find_correlation_bw_genes_and_PC_component_in_singlecell_cosine" title="Link to this definition"></a></dt>
<dd><p>Calculate cosine similarity between common gene scRNAseq factors and scRNAseq count data.</p>
<p>This helper function is used within the <cite>find_PC_of_invidualCluster_in_SC</cite> function to determine
the cosine similarity between common gene scRNAseq factors (principal components) and scRNAseq
gene expression data.</p>
<section id="id19">
<h3>Parameters<a class="headerlink" href="#id19" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>KcomponentCluster<span class="classifier">numpy.ndarray or pandas.DataFrame</span></dt><dd><p>The matrix representing the principal components (factors) from scRNAseq data. Each column
corresponds to a principal component.</p>
</dd>
<dt>clusterExpression<span class="classifier">numpy.ndarray or pandas.DataFrame</span></dt><dd><p>The matrix representing the gene expression data from scRNAseq. Each row corresponds to a gene
and each column corresponds to a cell.</p>
</dd>
</dl>
</section>
<section id="id20">
<h3>Returns<a class="headerlink" href="#id20" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>numpy.ndarray</dt><dd><p>A matrix containing the cosine similarity scores between each gene and each principal component.
Each row corresponds to a gene, and each column corresponds to a principal component.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="nico_covariations.Covariations.find_fold_change">
<span class="sig-prename descclassname"><span class="pre">nico_covariations.Covariations.</span></span><span class="sig-name descname"><span class="pre">find_fold_change</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">PCA</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">NH_PCA</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">CCPC</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">NCPC</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">totalLRpairs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">LRcutoff</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">CC_meanExpression</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">NC_meanExpression</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">CC_popExpression</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">NC_popExpression</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">number_of_top_genes_to_print</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/nico_covariations/Covariations.html#find_fold_change"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#nico_covariations.Covariations.find_fold_change" title="Link to this definition"></a></dt>
<dd><p>Identify ligand-receptor genes for cell type interaction analysis.</p>
<p>This helper function is used in <cite>find_LR_interactions_in_interacting_cell_types</cite> to find ligand-receptor (LR) genes based on principal component analysis (PCA) data. It identifies the top genes and checks for LR interactions between specific cell types.</p>
<section id="id21">
<h3>Parameters:<a class="headerlink" href="#id21" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>PCA<span class="classifier">numpy.ndarray</span></dt><dd><p>PCA data for the cell type of interest.</p>
</dd>
<dt>NH_PCA<span class="classifier">numpy.ndarray</span></dt><dd><p>PCA data for non-host cell types.</p>
</dd>
<dt>gene<span class="classifier">list of str</span></dt><dd><p>List of gene names corresponding to the PCA data.</p>
</dd>
<dt>CCPC<span class="classifier">int</span></dt><dd><p>Principal component index for the cell type of interest.</p>
</dd>
<dt>NCPC<span class="classifier">int</span></dt><dd><p>Principal component index for the non-host cell type.</p>
</dd>
<dt>totalLRpairs<span class="classifier">list of tuples</span></dt><dd><p>List of tuples representing all possible ligand-receptor pairs.</p>
</dd>
<dt>LRcutoff<span class="classifier">float</span></dt><dd><p>Threshold for selecting significant ligand-receptor interactions.</p>
</dd>
<dt>CC_meanExpression<span class="classifier">numpy.ndarray</span></dt><dd><p>Mean expression values for the central cell type.</p>
</dd>
<dt>NC_meanExpression<span class="classifier">numpy.ndarray</span></dt><dd><p>Mean expression values for the niche cell type.</p>
</dd>
<dt>CC_popExpression<span class="classifier">numpy.ndarray</span></dt><dd><p>Population expression values for the central cell type.</p>
</dd>
<dt>NC_popExpression<span class="classifier">numpy.ndarray</span></dt><dd><p>Population expression values for the niche cell type.</p>
</dd>
<dt>number_of_top_genes_to_print<span class="classifier">int</span></dt><dd><p>Number of top genes to include in the output.</p>
</dd>
</dl>
</section>
<section id="id22">
<h3>Returns:<a class="headerlink" href="#id22" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>cc_genes<span class="classifier">list of str</span></dt><dd><p>List of significant genes for the cell type of interest.</p>
</dd>
<dt>nc_genes<span class="classifier">list of str</span></dt><dd><p>List of significant genes for the non-host cell type.</p>
</dd>
<dt>cc_genes5<span class="classifier">list of list</span></dt><dd><p>Top genes for the cell type of interest with their PCA scores.</p>
</dd>
<dt>nc_genes5<span class="classifier">list of list</span></dt><dd><p>Top genes for the non-host cell type with their PCA scores.</p>
</dd>
<dt>Found1<span class="classifier">list of list</span></dt><dd><p>Ligand-receptor pairs with ligands in the cell type of interest and receptors in the non-host cell type.</p>
</dd>
<dt>Found2<span class="classifier">list of list</span></dt><dd><p>Ligand-receptor pairs with ligands in the non-host cell type and receptors in the cell type of interest.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="nico_covariations.Covariations.find_index">
<span class="sig-prename descclassname"><span class="pre">nico_covariations.Covariations.</span></span><span class="sig-name descname"><span class="pre">find_index</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sp_genename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sc_genename</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/nico_covariations/Covariations.html#find_index"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#nico_covariations.Covariations.find_index" title="Link to this definition"></a></dt>
<dd><p>Find the common gene space submatrix between two modalities.</p>
<p>This helper function is used within the <cite>gene_covariation_analysis</cite> function to identify the indices of common genes
between two lists of gene names corresponding to spatial and scRNAseq modalities.</p>
<section id="id23">
<h3>Parameters<a class="headerlink" href="#id23" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>sp_genename<span class="classifier">list</span></dt><dd><p>A list of gene names from the spatial modality.</p>
</dd>
<dt>sc_genename<span class="classifier">list</span></dt><dd><p>A list of gene names from the scRNAseq modality.</p>
</dd>
</dl>
</section>
<section id="id24">
<h3>Returns<a class="headerlink" href="#id24" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>list</dt><dd><p>A list of indices corresponding to the common genes found in both <cite>sp_genename</cite> and <cite>sc_genename</cite>.</p>
</dd>
</dl>
</section>
<section id="id25">
<h3>Example<a class="headerlink" href="#id25" title="Link to this heading"></a></h3>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sp_genes</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;gene1&#39;</span><span class="p">,</span> <span class="s1">&#39;gene2&#39;</span><span class="p">,</span> <span class="s1">&#39;gene3&#39;</span><span class="p">,</span> <span class="s1">&#39;gene4&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sc_genes</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;gene3&#39;</span><span class="p">,</span> <span class="s1">&#39;gene4&#39;</span><span class="p">,</span> <span class="s1">&#39;gene5&#39;</span><span class="p">,</span> <span class="s1">&#39;gene6&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">index_sp</span><span class="p">,</span><span class="n">index_sc</span> <span class="o">=</span> <span class="n">find_index</span><span class="p">(</span><span class="n">sp_genes</span><span class="p">,</span> <span class="n">sc_genes</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">index_sp</span><span class="p">)</span>
<span class="go">[2, 3]</span>
</pre></div>
</div>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="nico_covariations.Covariations.find_interest_of_genes">
<span class="sig-prename descclassname"><span class="pre">nico_covariations.Covariations.</span></span><span class="sig-name descname"><span class="pre">find_interest_of_genes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">source</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pop</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mu</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rps_rpl_mt_genes_included</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">CC_gene</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">top_NOG</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/nico_covariations/Covariations.html#find_interest_of_genes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#nico_covariations.Covariations.find_interest_of_genes" title="Link to this definition"></a></dt>
<dd><p>Find genes of interest based on factor values, population expression, and average expression.</p>
<section id="id26">
<h3>Parameters:<a class="headerlink" href="#id26" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>source<span class="classifier">np.array</span></dt><dd><p>Factor values associated with genes.</p>
</dd>
<dt>pop<span class="classifier">np.array</span></dt><dd><p>Population expression values.</p>
</dd>
<dt>mu<span class="classifier">np.array</span></dt><dd><p>Mean expression values.</p>
</dd>
<dt>rps_rpl_mt_genes_included<span class="classifier">bool</span></dt><dd><p>Include Rps, Rpl, and mt genes if True.</p>
</dd>
<dt>CC_gene<span class="classifier">np.array</span></dt><dd><p>List of gene names.</p>
</dd>
<dt>top_NOG<span class="classifier">int</span></dt><dd><p>Number of genes to select.</p>
</dd>
</dl>
</section>
<section id="id27">
<h3>Returns:<a class="headerlink" href="#id27" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>gp1<span class="classifier">list</span></dt><dd><p>Genes of interest based on factor values.</p>
</dd>
<dt>vp1<span class="classifier">list</span></dt><dd><p>Corresponding factor values of the selected genes.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="nico_covariations.Covariations.find_logistic_regression_interacting_score">
<span class="sig-prename descclassname"><span class="pre">nico_covariations.Covariations.</span></span><span class="sig-name descname"><span class="pre">find_logistic_regression_interacting_score</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmn</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">coef</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">CTFeatures</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nameOfCellType</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">logistic_coef_cutoff</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/nico_covariations/Covariations.html#find_logistic_regression_interacting_score"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#nico_covariations.Covariations.find_logistic_regression_interacting_score" title="Link to this definition"></a></dt>
<dd><p>Helper function used in gene_covariation_analysis to find niche interaction scores from logistic regression classifier.</p>
<p>This function identifies the interacting cell types by analyzing the coefficients of a logistic regression classifier.
It normalizes the coefficients, sorts them, and identifies the significant interactions based on a specified cutoff value.</p>
<section id="id28">
<h3>Parameters<a class="headerlink" href="#id28" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>cmn<span class="classifier">array</span></dt><dd><p>Confusion matrix or similar matrix representing the performance of the logistic regression classifier.</p>
</dd>
<dt>coef<span class="classifier">array</span></dt><dd><p>Coefficients of the logistic regression model.</p>
</dd>
<dt>CTFeatures<span class="classifier">list</span></dt><dd><p>List of cell type features used in the logistic regression model.</p>
</dd>
<dt>nameOfCellType<span class="classifier">list</span></dt><dd><p>List of names corresponding to cell types.</p>
</dd>
<dt>logistic_coef_cutoff<span class="classifier">float</span></dt><dd><p>The cutoff value to consider a coefficient as significant for interaction.</p>
</dd>
</dl>
</section>
<section id="id29">
<h3>Returns<a class="headerlink" href="#id29" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>logistic_predicted_interactions<span class="classifier">dict</span></dt><dd><p>A dictionary where keys are cell types and values are lists of interacting cell types with their interaction scores.</p>
</dd>
</dl>
</section>
<section id="id30">
<h3>Notes<a class="headerlink" href="#id30" title="Link to this heading"></a></h3>
<ul class="simple">
<li><p>The function normalizes the logistic regression coefficients.</p></li>
<li><p>It identifies the most significant interactions based on the absolute value of the coefficients.</p></li>
<li><p>Interactions with coefficients above the cutoff value are considered significant and are included in the output.</p></li>
</ul>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="nico_covariations.Covariations.gene_covariation_analysis">
<span class="sig-prename descclassname"><span class="pre">nico_covariations.Covariations.</span></span><span class="sig-name descname"><span class="pre">gene_covariation_analysis</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">Radius</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_niche_prediction_dir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">refpath</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'./inputRef/'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quepath</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'./inputQuery/'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_cluster_tag</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'cluster'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_original_counts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Original_counts.h5ad'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">LRdbFilename</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'./utils/NiCoLRdb.txt'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">iNMFmode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">no_of_factors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">shap_analysis</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">shap_cluster_cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cutoff_to_count_exp_cell_population</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">541</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spatial_integration_modality</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'double'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">anndata_object_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'nico_celltype_annotation.h5ad'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lambda_c</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[0.0009765625,</span> <span class="pre">0.001953125,</span> <span class="pre">0.00390625,</span> <span class="pre">0.0078125,</span> <span class="pre">0.015625,</span> <span class="pre">0.03125,</span> <span class="pre">0.0625,</span> <span class="pre">0.125,</span> <span class="pre">0.25,</span> <span class="pre">0.5,</span> <span class="pre">1.0,</span> <span class="pre">2.0,</span> <span class="pre">4.0,</span> <span class="pre">8.0,</span> <span class="pre">16.0,</span> <span class="pre">32.0,</span> <span class="pre">64.0,</span> <span class="pre">128.0,</span> <span class="pre">256.0,</span> <span class="pre">512.0]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">coeff_cutoff_for_rid_reg</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">logistic_coef_cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/nico_covariations/Covariations.html#gene_covariation_analysis"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#nico_covariations.Covariations.gene_covariation_analysis" title="Link to this definition"></a></dt>
<dd><p>Perform gene covariation analysis within the niche.</p>
<p>This is the primary function called by the user to perform gene covariation analysis within the niche.
Before calling this function, the user must call the <cite>spatial_neighborhood_analysis</cite> function from the interaction module.</p>
<section id="id31">
<h3>Parameters<a class="headerlink" href="#id31" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>Radius<span class="classifier">int, optional</span></dt><dd><p>This radius parameter should be the same as used in spatial neighborhood analysis to find the niche interactions.
Default is 0.</p>
</dd>
<dt>output_niche_prediction_dir<span class="classifier">str, optional</span></dt><dd><p>The output directory location from the previous niche interaction runs generated by the function <cite>spatial_neighborhood_analysis</cite>.
Default is ‘./nico_out/’.</p>
</dd>
<dt>refpath<span class="classifier">str, optional</span></dt><dd><p>Path to the reference scRNAseq count matrix in scTransform-like normalization. The filename must be <cite>sct_singleCell.h5ad</cite>.
Default is ‘./inputRef/’.</p>
</dd>
<dt>quepath<span class="classifier">str, optional</span></dt><dd><p>Path to the query spatial count matrix in scTransform-like normalization. The filename must be <cite>sct_spatial.h5ad</cite>.
Default is ‘./inputQuery/’.</p>
</dd>
<dt>ref_cluster_tag<span class="classifier">str, optional</span></dt><dd><p>The slot in the reference anndata object file where cell type information is stored.
Default is ‘cluster’.</p>
</dd>
<dt>ref_original_counts<span class="classifier">str, optional</span></dt><dd><p>Path to the original count data of scRNAseq in anndata object. Must have the cluster information in <cite>.obs</cite> and the umap information in <cite>.obsm[‘X_umap’]</cite>.
anndata.raw layer should have count matrix data. It will used to find the Spearman correlation and cosine similarity.
Default is ‘Original_counts.h5ad’.</p>
</dd>
<dt>LRdbFilename<span class="classifier">str, optional</span></dt><dd><p>Filename of the ligand-receptor database. The first column should be Ligand, the second column Receptor, and the third column the resource list.
Default is ‘./utils/NiCoLRdb.txt’.</p>
</dd>
<dt>iNMFmode<span class="classifier">bool, optional</span></dt><dd><p>If True, uses an integrated NMF approach to learn a gene-by-factor submatrix from both modalities.
If False, uses an ordinary NMF approach to learn a gene by factor submatrix only from scRNAseq data and transfers these factors to the spatial modality for learning the gene weights.
Default is True.</p>
</dd>
<dt>no_of_factors<span class="classifier">int, optional</span></dt><dd><p>Number of factors used in NMF for finding the common gene latent dimension space.
Default is 3.</p>
</dd>
<dt>lambda_c<span class="classifier">list, optional</span></dt><dd><p>Initial range of regularization parameters used in the ridge regression step to find the optimal parameter.
Default is <cite>list(np.power(2.0, np.arange(-10, 10)))</cite>.</p>
</dd>
<dt>shap_analysis<span class="classifier">bool, optional</span></dt><dd><p>Flag to perform SHAP analysis.
Default is False.</p>
</dd>
<dt>shap_cluster_cutoff<span class="classifier">float, optional</span></dt><dd><p>SHAP analysis cutoff parameter.
Default is 0.5.</p>
</dd>
<dt>coeff_cutoff_for_rid_reg<span class="classifier">float, optional</span></dt><dd><p>Cutoff used to create the list of significant celltype_factor-celltype_factor niche covariations with an absolute regression coefficient greater than this.
Default is 0.</p>
</dd>
<dt>cutoff_to_count_exp_cell_population<span class="classifier">int, optional</span></dt><dd><p>Parameter to find the percentage of the cell population that express a given gene in a given cell type. Value 0 is acceptable with count data.
Default is 0.</p>
</dd>
<dt>seed<span class="classifier">int, optional</span></dt><dd><p>Random seed used in RepeatedStratifiedKFold.
Default is 541.</p>
</dd>
<dt>spatial_integration_modality<span class="classifier">str, optional</span></dt><dd><p>Modality for spatial integration if both scRNAseq and Spatial data is available.
Default is ‘double’.
For only spatial data this value must be ‘single’</p>
</dd>
<dt>anndata_object_name<span class="classifier">str, optional</span></dt><dd><p>Name of the spatial anndata object name.
Default is ‘nico_celltype_annotation.h5ad’.</p>
</dd>
<dt>logistic_coef_cutoff<span class="classifier">float, optional</span></dt><dd><p>Cutoff to retrieve the positive niche interactions (cell type - cell type). For values &gt;0, cell type pairs are likely to interact.
Default is 0.</p>
</dd>
</dl>
</section>
<section id="id32">
<h3>Outputs<a class="headerlink" href="#id32" title="Link to this heading"></a></h3>
<p>The output is saved in the directory specified by <cite>output_niche_prediction_dir</cite>, with default location being ‘./nico_out/covariations_R*_F*’.</p>
</section>
<section id="id33">
<h3>Notes<a class="headerlink" href="#id33" title="Link to this heading"></a></h3>
<ul class="simple">
<li><p>Please provide <cite>Original_counts.h5ad</cite>, <cite>sct_singleCell.h5ad</cite> files from scRNAseq data.</p></li>
<li><p>Provide  <cite>sct_spatial.h5ad</cite> files for the spatial transcriptomics data.</p></li>
<li><p><cite>Original_counts.h5ad</cite> object should also have the cluster information in <cite>.obs</cite> and the umap information in <cite>.obsm</cite> and .raw layer has count data.</p></li>
</ul>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="nico_covariations.Covariations.makePCneighboorhoodFeatureMatrix">
<span class="sig-prename descclassname"><span class="pre">nico_covariations.Covariations.</span></span><span class="sig-name descname"><span class="pre">makePCneighboorhoodFeatureMatrix</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/nico_covariations/Covariations.html#makePCneighboorhoodFeatureMatrix"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#nico_covariations.Covariations.makePCneighboorhoodFeatureMatrix" title="Link to this definition"></a></dt>
<dd><p>Helper function in gene_covariation_analysis to find the weighted neighborhood
average of cell types from the spatial factors.</p>
<p>This function computes a matrix where each row corresponds to a cell and each column
corresponds to a weighted average of principal components (PCs) from neighboring cells.</p>
<section id="id34">
<h3>Parameters:<a class="headerlink" href="#id34" title="Link to this heading"></a></h3>
<dl>
<dt>input<span class="classifier">object</span></dt><dd><p>An object containing various attributes required for computation, such as:</p>
<blockquote>
<div><ul class="simple">
<li><p>spatialcell_unique_clusterid: Unique cluster IDs for spatial cells.</p></li>
<li><p>neighbors: List of neighboring cells for each cell.</p></li>
<li><p>neigh_distances: Distances to neighbors.</p></li>
<li><p>annotation_spatial_barcode_id: Barcode IDs for spatial annotations.</p></li>
<li><p>annotation_spatial_cluster_id: Cluster IDs for spatial annotations.</p></li>
<li><p>pc_of_sp_clusterid: Principal components for spatial cluster IDs.</p></li>
<li><p>no_of_pc: Number of principal components.</p></li>
<li><p>outputname: Name of the output file.</p></li>
</ul>
</div></blockquote>
</dd>
</dl>
</section>
<section id="id35">
<h3>Outputs:<a class="headerlink" href="#id35" title="Link to this heading"></a></h3>
<p>A .npz file containing the matrix of weighted neighborhood principal components.</p>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="nico_covariations.Covariations.make_excel_sheet_for_gene_correlation">
<span class="sig-prename descclassname"><span class="pre">nico_covariations.Covariations.</span></span><span class="sig-name descname"><span class="pre">make_excel_sheet_for_gene_correlation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/nico_covariations/Covariations.html#make_excel_sheet_for_gene_correlation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#nico_covariations.Covariations.make_excel_sheet_for_gene_correlation" title="Link to this definition"></a></dt>
<dd><p>Create an Excel sheet compiling gene correlations with factors across different cell types.</p>
<p>This function generates an Excel sheet that provides a structured and accessible representation of gene factors associated with cell types. It includes various types of information, such as average gene expression, Spearman correlation values, and cosine similarity values for both scRNASeq and spatial data.</p>
<section id="id36">
<h3>Parameters<a class="headerlink" href="#id36" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>input<span class="classifier">object</span></dt><dd><p>The main input is the output from gene_covariation_analysis.</p>
</dd>
</dl>
</section>
<section id="id37">
<h3>Outputs<a class="headerlink" href="#id37" title="Link to this heading"></a></h3>
<ul class="simple">
<li><dl class="simple">
<dt>Excel sheets categorized into different types of information:</dt><dd><ul>
<li><p>‘avg gene exp’: Average gene expression.</p></li>
<li><p>‘spearman scRNAseq Fa(i)’: Spearman correlation values for different factors within scRNASeq data.</p></li>
<li><p>‘cosine scRNAseq Fa(i)’: Cosine similarity values within scRNASeq data.</p></li>
<li><p>‘spearman spatial Fa(i)’: Spearman correlation values for common genes in the spatial data.</p></li>
<li><p>‘cosine spatial Fa(i)’: Cosine similarity values for common genes in the spatial data.</p></li>
</ul>
</dd>
</dl>
</li>
</ul>
</section>
<section id="id38">
<h3>Notes<a class="headerlink" href="#id38" title="Link to this heading"></a></h3>
<ul class="simple">
<li><p>In the sheet names, ‘i’ corresponds to the factor ID.</p></li>
<li><p>Columns include factors representing all cell types.</p></li>
<li><p>For each factor, genes are sorted based on their association with the factor ID corresponding to the respective sheet.</p></li>
<li><dl class="simple">
<dt>A color-coding scheme is used to distinguish genes:</dt><dd><ul>
<li><p>Ligands are depicted in blue.</p></li>
<li><p>Receptors are depicted in red.</p></li>
<li><p>Genes with both ligand and receptor functions are depicted in magenta.</p></li>
</ul>
</dd>
</dl>
</li>
</ul>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="nico_covariations.Covariations.model_linear_regression">
<span class="sig-prename descclassname"><span class="pre">nico_covariations.Covariations.</span></span><span class="sig-name descname"><span class="pre">model_linear_regression</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">logistic_predicted_interactions</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/nico_covariations/Covariations.html#model_linear_regression"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#nico_covariations.Covariations.model_linear_regression" title="Link to this definition"></a></dt>
<dd><p>Helper function for gene_covariation_analysis to prepare data Y (central cell factors)
and X (neighborhood average spatial cell factors) for each cell type to perform regression.</p>
<p>This function loads the precomputed neighborhood feature matrix and prepares the data for linear
regression analysis. It then performs ridge regression for each cell type to find the relationship
between the central cell factors (Y) and the neighborhood average spatial cell factors (X).</p>
<section id="id39">
<h3>Parameters<a class="headerlink" href="#id39" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>input<span class="classifier">object</span></dt><dd><p>An object containing the following attributes:</p>
<ul class="simple">
<li><p>shap_cluster_cutoff : float, cutoff value for SHAP clustering.</p></li>
<li><p>outputname : str, the name of the input file containing precomputed neighborhood features.</p></li>
<li><p>no_of_pc : int, number of principal components.</p></li>
<li><p>spatialcell_unique_clusterid : list, unique cluster IDs of spatial cells.</p></li>
<li><p>annotation_spatial_cluster_id : list, cluster IDs for each spatial cell.</p></li>
<li><p>spatialcell_unique_clustername : list, unique cluster names of spatial cells.</p></li>
<li><p>seed : int, seed value for regression.</p></li>
<li><p>lambda_c : float, regularization parameter for ridge regression.</p></li>
<li><p>K_fold : int, number of folds for cross-validation.</p></li>
<li><p>n_repeats : int, number of repeats for cross-validation.</p></li>
</ul>
</dd>
<dt>logistic_predicted_interactions<span class="classifier">dict</span></dt><dd><p>A dictionary where keys are cell type names and values are lists of tuples. Each tuple contains
a cell type name and a score representing the predicted interaction strength with the key cell type.</p>
</dd>
</dl>
</section>
<section id="id40">
<h3>Returns<a class="headerlink" href="#id40" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>save_coef<span class="classifier">dict</span></dt><dd><p>A dictionary where keys are unique cluster IDs and values are lists containing the following elements:</p>
<ul class="simple">
<li><p>coef : array, coefficients of the ridge regression model.</p></li>
<li><p>intercept : array, intercepts of the ridge regression model.</p></li>
<li><p>alpha : float, regularization parameter of the ridge regression model.</p></li>
<li><p>xlabel : array, names of the features.</p></li>
<li><p>score : array, scores of the features.</p></li>
<li><p>target : array, target values (central cell factors).</p></li>
<li><p>neighborhoodClass : array, neighborhood average spatial cell factors.</p></li>
<li><p>pv : array, p-values of the regression coefficients.</p></li>
<li><p>percent_variance_explained : array, percentage of variance explained by the model.</p></li>
<li><p>residual_variance_explained : array, residual variance explained by the model.</p></li>
</ul>
</dd>
</dl>
</section>
<section id="id41">
<h3>Notes<a class="headerlink" href="#id41" title="Link to this heading"></a></h3>
<ul class="simple">
<li><p>The function uses ridge regression to model the relationship between central cell factors and neighborhood factors.</p></li>
<li><p>The precomputed neighborhood feature matrix is loaded from a file and NaN values are replaced with zeros.</p></li>
<li><p>The function selects relevant features based on logistic_predicted_interactions and performs ridge regression.</p></li>
<li><p>The results are stored in a dictionary and returned for further analysis.</p></li>
</ul>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="nico_covariations.Covariations.multiplicative_method">
<span class="sig-prename descclassname"><span class="pre">nico_covariations.Covariations.</span></span><span class="sig-name descname"><span class="pre">multiplicative_method</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">W</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">H</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">A</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_iter</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/nico_covariations/Covariations.html#multiplicative_method"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#nico_covariations.Covariations.multiplicative_method" title="Link to this definition"></a></dt>
<dd><blockquote>
<div><p>Perform conventional Non-negative Matrix Factorization (NMF) using a multiplicative update rule.</p>
<p>This helper function is used in <cite>find_PC_of_invidualCluster_in_SC</cite> to decompose matrix <cite>A</cite> into
two non-negative matrices <cite>W</cite> and <cite>H</cite> such that <cite>A ≈ W &#64; H</cite>.</p>
<dl class="simple">
<dt>W<span class="classifier">ndarray</span></dt><dd><p>Initial matrix representing the basis vectors. Shape(n_samples, n_components).</p>
</dd>
<dt>H<span class="classifier">ndarray</span></dt><dd><p>Initial matrix representing the coefficients. Shape(n_components, n_features).</p>
</dd>
<dt>A<span class="classifier">ndarray</span></dt><dd><p>The input data matrix to be factorized. Shape(n_samples, n_features).</p>
</dd>
<dt>max_iter<span class="classifier">int</span></dt><dd><p>The maximum number of iterations for the multiplicative update algorithm.</p>
</dd>
</dl>
<dl class="simple">
<dt>W<span class="classifier">ndarray</span></dt><dd><p>Updated basis matrix after NMF. Shape: (n_samples, n_components).</p>
</dd>
<dt>H<span class="classifier">ndarray</span></dt><dd><p>Updated coefficient matrix after NMF. Shape: (n_components, n_features).</p>
</dd>
<dt>norms<span class="classifier">list</span></dt><dd><p>List of Frobenius norms of the difference between <cite>A</cite> and <cite>W &#64; H</cite> for each iteration.</p>
</dd>
</dl>
<p>The update rules for <cite>W</cite> and <cite>H</cite> are based on minimizing the Frobenius norm of the difference between <cite>A</cite> and <cite>W &#64; H</cite>. The update for <cite>H</cite> is performed as:</p>
<div class="math notranslate nohighlight">
\[H_{ij} = H_{ij}  \]</div>
</div></blockquote>
<p>rac{(W^T A)_{ij}}{(W^T W H)_{ij} + epsilon}</p>
<blockquote>
<div><p>where <cite>ε</cite> is a small constant to prevent division by zero.</p>
<p>The update for <cite>W</cite> has been commented out but follows a similar form. Uncomment the lines under “Update W” to perform updates for <cite>W</cite> as well.</p>
<div class="math notranslate nohighlight">
\[W_{ij} = W_{ij}  \]</div>
</div></blockquote>
<p>rac{(A H^T)_{ij}}{(W H H^T)_{ij} + epsilon}</p>
<blockquote>
<div><p>This method is sensitive to initializations of <cite>W</cite> and <cite>H</cite>, and the results may vary across runs.</p>
</div></blockquote>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="nico_covariations.Covariations.pathway_analysis">
<span class="sig-prename descclassname"><span class="pre">nico_covariations.Covariations.</span></span><span class="sig-name descname"><span class="pre">pathway_analysis</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">NOG_pathway</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">choose_factors_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">correlation_with_spearman</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">saveas</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'pdf'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">savefigure</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">positively_correlated</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rps_rpl_mt_genes_included</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">choose_celltypes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">circlesize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">12</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pathwayCutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pathwayorganism</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Mouse'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">database</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['GO_Biological_Process_2021',</span> <span class="pre">'BioPlanet_2019',</span> <span class="pre">'Reactome_2016']</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/nico_covariations/Covariations.html#pathway_analysis"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#nico_covariations.Covariations.pathway_analysis" title="Link to this definition"></a></dt>
<dd><p>Perform pathway analysis for gene covariations within niches.</p>
<p>This function utilizes the gene covariation data from the gene_covariation_analysis to perform pathway enrichment analysis using the GSEApy package. It identifies and visualizes the pathways that are significantly enriched among the top genes associated with NMF factors.</p>
<section id="id42">
<h3>Parameters<a class="headerlink" href="#id42" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>input<span class="classifier">object</span></dt><dd><p>The main input is the output from gene_covariation_analysis.</p>
</dd>
<dt>NOG_pathway<span class="classifier">int, optional</span></dt><dd><p>Number of top genes associated with NMF factors to search in the pathway database. If no pathways are observed, increase the number of genes or try different databases.
(default is 50)</p>
</dd>
<dt>choose_factors_id<span class="classifier">list, optional</span></dt><dd><p>Define the factor IDs for which pathway enrichments should be visualized. If empty, pathways will be saved for all factors.
(default is [])</p>
</dd>
<dt>correlation_with_spearman<span class="classifier">bool, optional</span></dt><dd><p>If True, visualize gene-factor correlations obtained by Spearman correlation coefficient; otherwise, use cosine similarities.
(default is True)</p>
</dd>
<dt>positively_correlated<span class="classifier">bool, optional</span></dt><dd><p>If True and correlation_with_spearman is selected, select positively correlated genes; otherwise, select negatively correlated genes.
(default is True)</p>
</dd>
<dt>rps_rpl_mt_genes_included<span class="classifier">bool, optional</span></dt><dd><p>If True, include rps, rpl, and mt genes in the pathway analysis; if False, filter these genes out.
(default is True)</p>
</dd>
<dt>pathwayCutoff<span class="classifier">float, optional</span></dt><dd><p>The cutoff parameter for finding pathway-enriched libraries from the top genes for each factor of a given cell type using GSEApy.
(default is 0.5)</p>
</dd>
<dt>pathwayorganism<span class="classifier">str, optional</span></dt><dd><p>The organism used in the GSEApy package.
(default is ‘Mouse’)</p>
</dd>
<dt>database<span class="classifier">list, optional</span></dt><dd><p>The databases used in the GSEApy package for pathway analysis. The default includes ‘GO_Biological_Process_2021’, ‘BioPlanet_2019’, and ‘Reactome_2016’.
(default is [‘GO_Biological_Process_2021’, ‘BioPlanet_2019’, ‘Reactome_2016’])</p>
</dd>
<dt>choose_celltypes<span class="classifier">list, optional</span></dt><dd><p>Define the cell types for which pathway enrichments should be returned from the Enrichr library. If empty, the output will be generated for all cell types.
(default is [])</p>
</dd>
<dt>saveas<span class="classifier">str, optional</span></dt><dd><p>Save the figures in PDF or PNG format (dpi for PNG format is 300).
(default is ‘pdf’)</p>
</dd>
<dt>circlesize<span class="classifier">int, optional</span></dt><dd><p>The size of the points in the pathway figure. Increase this value if the dots are too small.
(default is 12)</p>
</dd>
<dt>savefigure<span class="classifier">bool, optional</span></dt><dd><p>If True, the generated figures will be saved.
(default is False)</p>
</dd>
<dt>showit<span class="classifier">bool, optional</span></dt><dd><p>If True, the generated figures will be displayed.
(default is True)</p>
</dd>
<dt>figsize<span class="classifier">tuple, optional</span></dt><dd><p>The dimension of the figure size.
(default is (12, 10))</p>
</dd>
</dl>
</section>
<section id="id43">
<h3>Outputs<a class="headerlink" href="#id43" title="Link to this heading"></a></h3>
<p>The pathway figures are saved in “./nico_out/covariations_R*_F*/Pathway_figures/”.</p>
</section>
<section id="id44">
<h3>Notes<a class="headerlink" href="#id44" title="Link to this heading"></a></h3>
<ul class="simple">
<li><p>In the sheet names, ‘i’ corresponds to the factor ID.</p></li>
<li><p>Columns include factors representing all cell types.</p></li>
<li><p>For each factor, genes are sorted based on their association with the factor ID corresponding to the respective sheet.</p></li>
<li><dl class="simple">
<dt>A color-coding scheme is used to distinguish genes:</dt><dd><ul>
<li><p>Ligands are depicted in blue.</p></li>
<li><p>Receptors are depicted in red.</p></li>
<li><p>Genes with both ligand and receptor functions are depicted in magenta.</p></li>
</ul>
</dd>
</dl>
</li>
</ul>
</section>
<section id="id45">
<h3>Example<a class="headerlink" href="#id45" title="Link to this heading"></a></h3>
<p>For other available databases, check for species ‘Human,’ ‘Mouse,’ ‘Yeast,’ ‘Fly,’ ‘Fish,’ and ‘Worm’ in the following way:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">gseapy</span> <span class="k">as</span> <span class="nn">gp</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mouse</span> <span class="o">=</span> <span class="n">gp</span><span class="o">.</span><span class="n">get_library_name</span><span class="p">(</span><span class="n">organism</span><span class="o">=</span><span class="s1">&#39;Mouse&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">human</span> <span class="o">=</span> <span class="n">gp</span><span class="o">.</span><span class="n">get_library_name</span><span class="p">(</span><span class="n">organism</span><span class="o">=</span><span class="s1">&#39;Human&#39;</span><span class="p">)</span>
</pre></div>
</div>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="nico_covariations.Covariations.plot_all_ct">
<span class="sig-prename descclassname"><span class="pre">nico_covariations.Covariations.</span></span><span class="sig-name descname"><span class="pre">plot_all_ct</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">CTname</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PP</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cellsinCT</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mycelltype</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Fa</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmap</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">msna</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/nico_covariations/Covariations.html#plot_all_ct"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#nico_covariations.Covariations.plot_all_ct" title="Link to this definition"></a></dt>
<dd><p>Visualize factor values in UMAP for all cell types.</p>
<p>This helper function is used for visualizing factor values in UMAP, showing the distribution of cells across different cell types and highlighting specific cell types of interest.</p>
<section id="id46">
<h3>Parameters:<a class="headerlink" href="#id46" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>CTname<span class="classifier">list of str</span></dt><dd><p>List of cell type names.</p>
</dd>
<dt>PP<span class="classifier">np.ndarray</span></dt><dd><p>UMAP embedding coordinates for all cells.</p>
</dd>
<dt>cellsinCT<span class="classifier">dict</span></dt><dd><p>Dictionary where keys are cell type names and values are lists of cell indices corresponding to each cell type.</p>
</dd>
<dt>ax<span class="classifier">matplotlib.axes.Axes</span></dt><dd><p>Matplotlib Axes object where the UMAP plot will be drawn.</p>
</dd>
<dt>mycelltype<span class="classifier">list of str</span></dt><dd><p>List of cell types to highlight.</p>
</dd>
<dt>Fa<span class="classifier">np.ndarray</span></dt><dd><p>Array of factor values corresponding to each cell.</p>
</dd>
<dt>cmap<span class="classifier">str or matplotlib.colors.Colormap</span></dt><dd><p>Colormap used for plotting the factor values.</p>
</dd>
<dt>ms<span class="classifier">int</span></dt><dd><p>Marker size for the highlighted cell types.</p>
</dd>
<dt>msna<span class="classifier">int</span></dt><dd><p>Marker size for the non-highlighted (NA) cell types.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="nico_covariations.Covariations.plot_cosine_and_spearman_correlation_to_factors">
<span class="sig-prename descclassname"><span class="pre">nico_covariations.Covariations.</span></span><span class="sig-name descname"><span class="pre">plot_cosine_and_spearman_correlation_to_factors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">choose_celltypes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">NOG_Fa</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">saveas</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'pdf'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">transparent_mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">showit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(15,</span> <span class="pre">10)</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/nico_covariations/Covariations.html#plot_cosine_and_spearman_correlation_to_factors"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#nico_covariations.Covariations.plot_cosine_and_spearman_correlation_to_factors" title="Link to this definition"></a></dt>
<dd><p>Plots cosine and Spearman correlation to factors for given cell types.</p>
<section id="id47">
<h3>Parameters:<a class="headerlink" href="#id47" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>input<span class="classifier">object</span></dt><dd><p>The main input object containing the output from gene_covariation_analysis.</p>
</dd>
<dt>choose_celltypes<span class="classifier">list, optional</span></dt><dd><p>The cell types for which you want to inspect the covariation pattern.
If the list is empty, the output will be generated for all cell types.
Default is [].</p>
</dd>
<dt>NOG_Fa<span class="classifier">int, optional</span></dt><dd><p>Number of genes to visualize for each factor.
Default is 30.</p>
</dd>
<dt>saveas<span class="classifier">str, optional</span></dt><dd><p>Format to save the figures. Options are ‘pdf’ or ‘png’. If ‘png’ is chosen, dpi is set to 300.
Default is ‘pdf’.</p>
</dd>
<dt>transparent_mode<span class="classifier">bool, optional</span></dt><dd><p>Whether to save the figure with a transparent background.
Default is False.</p>
</dd>
<dt>showit<span class="classifier">bool, optional</span></dt><dd><p>Whether to display the plot interactively.
Default is True.</p>
</dd>
<dt>figsize<span class="classifier">tuple, optional</span></dt><dd><p>Dimensions of the figure size.
Default is (15, 10).</p>
</dd>
</dl>
</section>
<section id="id48">
<h3>Outputs:<a class="headerlink" href="#id48" title="Link to this heading"></a></h3>
<p>The output NMF plots are saved in ./&lt;output_nico_dir&gt;/covariations_R*_F*/NMF_output.</p>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="nico_covariations.Covariations.plot_feature_matrices">
<span class="sig-prename descclassname"><span class="pre">nico_covariations.Covariations.</span></span><span class="sig-name descname"><span class="pre">plot_feature_matrices</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">showit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">saveas</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'pdf'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">transparent_mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(10,</span> <span class="pre">10)</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/nico_covariations/Covariations.html#plot_feature_matrices"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#nico_covariations.Covariations.plot_feature_matrices" title="Link to this definition"></a></dt>
<dd><p>Plots feature vectors of the spatial factors from the regression step.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="nico_covariations.Covariations.plot_ligand_receptor_in_interacting_celltypes">
<span class="sig-prename descclassname"><span class="pre">nico_covariations.Covariations.</span></span><span class="sig-name descname"><span class="pre">plot_ligand_receptor_in_interacting_celltypes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">CC_celltype_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">NC_celltype_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">logRegScore</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pc1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pc2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ridgeRegScore</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pvalue</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Found1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Found2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">saveLRplots</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">LR_plot_Exp_thres</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">saveas</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">transparent_mode</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">showit</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flag</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/nico_covariations/Covariations.html#plot_ligand_receptor_in_interacting_celltypes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#nico_covariations.Covariations.plot_ligand_receptor_in_interacting_celltypes" title="Link to this definition"></a></dt>
<dd><p>Plot ligand-receptor interactions for interacting cell types.</p>
<p>This helper function is used in <cite>find_LR_interactions_in_interacting_cell_types</cite> to plot
rectangle p-value figures representing ligand-receptor interactions between two cell types.</p>
<p>The Y-axis shows the central cell type factors, and the X-axis shows the colocalized neighborhood cell type factors.
The circle size denotes the p-values, and the circle size scales with significance.</p>
<section id="id49">
<h3>Parameters:<a class="headerlink" href="#id49" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>CC_celltype_name<span class="classifier">str</span></dt><dd><p>Name of the central cell type.</p>
</dd>
<dt>NC_celltype_name<span class="classifier">str</span></dt><dd><p>Name of the neighborhood cell type.</p>
</dd>
<dt>logRegScore<span class="classifier">float</span></dt><dd><p>Logistic regression score.</p>
</dd>
<dt>pc1<span class="classifier">int</span></dt><dd><p>Principal component for the central cell type.</p>
</dd>
<dt>pc2<span class="classifier">int</span></dt><dd><p>Principal component for the neighborhood cell type.</p>
</dd>
<dt>ridgeRegScore<span class="classifier">float</span></dt><dd><p>Ridge regression score.</p>
</dd>
<dt>pvalue<span class="classifier">float</span></dt><dd><p>P-value for the interaction.</p>
</dd>
<dt>Found1<span class="classifier">list</span></dt><dd><p>List of found ligand-receptor interactions where the ligand is in the central cell type.</p>
</dd>
<dt>Found2<span class="classifier">list</span></dt><dd><p>List of found ligand-receptor interactions where the ligand is in the neighborhood cell type.</p>
</dd>
<dt>saveLRplots<span class="classifier">str</span></dt><dd><p>Directory to save the ligand-receptor plots.</p>
</dd>
<dt>LR_plot_Exp_thres<span class="classifier">float</span></dt><dd><p>Expression threshold for plotting.</p>
</dd>
<dt>saveas<span class="classifier">str</span></dt><dd><p>File format to save the plots (e.g., ‘png’, ‘pdf’).</p>
</dd>
<dt>transparent_mode<span class="classifier">bool</span></dt><dd><p>Whether to save the plots with a transparent background.</p>
</dd>
<dt>showit<span class="classifier">bool</span></dt><dd><p>Whether to display the plots.</p>
</dd>
<dt>figsize<span class="classifier">tuple</span></dt><dd><p>Size of the figure.</p>
</dd>
<dt>flag<span class="classifier">str</span></dt><dd><p>Flag indicating which interactions to plot (‘First’, ‘Second’, ‘Both’).</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="nico_covariations.Covariations.plot_significant_regression_covariations_as_circleplot">
<span class="sig-prename descclassname"><span class="pre">nico_covariations.Covariations.</span></span><span class="sig-name descname"><span class="pre">plot_significant_regression_covariations_as_circleplot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">choose_celltypes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">saveas</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'pdf'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pvalue_cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mention_pvalue</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">transparent_mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">showit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(6,</span> <span class="pre">1.25)</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/nico_covariations/Covariations.html#plot_significant_regression_covariations_as_circleplot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#nico_covariations.Covariations.plot_significant_regression_covariations_as_circleplot" title="Link to this definition"></a></dt>
<dd><p>Plot significant regression covariations as a circle plot.</p>
<p>This function visualizes the significant regression covariations identified in the gene covariation analysis.</p>
<section id="id50">
<h3>Parameters<a class="headerlink" href="#id50" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>input<span class="classifier">object</span></dt><dd><p>The main input is the output from gene_covariation_analysis.</p>
</dd>
<dt>choose_celltypes<span class="classifier">list, optional</span></dt><dd><p>The cell type(s) for which you want to inspect the covariation pattern.
If the list is empty, the output will be generated for all cell types.
Default is [].</p>
</dd>
<dt>saveas<span class="classifier">str, optional</span></dt><dd><p>Format to save the figures in, either ‘pdf’ or ‘png’ (dpi for PNG format is 300).
Default is ‘pdf’.</p>
</dd>
<dt>pvalue_cutoff<span class="classifier">float, optional</span></dt><dd><p>The p-value cutoff used to print the -log10(pvalue) on top of the circle.
Default is 0.05.</p>
</dd>
<dt>mention_pvalue<span class="classifier">bool, optional</span></dt><dd><p>Whether to highlight the p-value on the circle plot. If False, it will not be shown.
Default is True.</p>
</dd>
<dt>transparent_mode<span class="classifier">bool, optional</span></dt><dd><p>Background color in the figures. If True, the background will be transparent.
Default is False.</p>
</dd>
<dt>showit<span class="classifier">bool, optional</span></dt><dd><p>Whether to display the plot. If False, the plot will be saved but not shown.
Default is True.</p>
</dd>
<dt>figsize<span class="classifier">tuple, optional</span></dt><dd><p>Dimension of the figure size.
Default is (6, 1.25).</p>
</dd>
</dl>
</section>
<section id="id51">
<h3>Outputs<a class="headerlink" href="#id51" title="Link to this heading"></a></h3>
<p>The regression figures are saved in ‘./nico_out/covariations_R*_F*/Regression_outputs/’.</p>
</section>
<section id="id52">
<h3>Notes<a class="headerlink" href="#id52" title="Link to this heading"></a></h3>
<ul class="simple">
<li><p>The main input is the output from gene_covariation_analysis.</p></li>
<li><p>The output directory for saving the figures should exist prior to running this function.</p></li>
</ul>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="nico_covariations.Covariations.plot_significant_regression_covariations_as_heatmap">
<span class="sig-prename descclassname"><span class="pre">nico_covariations.Covariations.</span></span><span class="sig-name descname"><span class="pre">plot_significant_regression_covariations_as_heatmap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">choose_celltypes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">saveas</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'pdf'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">transparent_mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">showit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(6,</span> <span class="pre">10)</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/nico_covariations/Covariations.html#plot_significant_regression_covariations_as_heatmap"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#nico_covariations.Covariations.plot_significant_regression_covariations_as_heatmap" title="Link to this definition"></a></dt>
<dd><p>Plot significant regression covariations as a heatmap.</p>
<p>This function visualizes the significant regression covariations from the gene covariation analysis as a heatmap.</p>
<section id="id53">
<h3>Parameters<a class="headerlink" href="#id53" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>input<span class="classifier">object</span></dt><dd><p>The main input is the output from the gene_covariation_analysis.</p>
</dd>
<dt>choose_celltypes<span class="classifier">list, optional</span></dt><dd><p>The cell types for which you want to inspect the covariation regression pattern.
If the list is empty, the output will be generated for all cell types.
Default is an empty list [].</p>
</dd>
<dt>saveas<span class="classifier">str, optional</span></dt><dd><p>Format to save the figures in, either ‘pdf’ or ‘png’ (dpi for PNG format is 300).
Default is ‘pdf’.</p>
</dd>
<dt>transparent_mode<span class="classifier">bool, optional</span></dt><dd><p>Background color in the figures. If True, the background will be transparent.
Default is False.</p>
</dd>
<dt>showit<span class="classifier">bool, optional</span></dt><dd><p>Whether to display the plot. If False, the plot will be saved but not shown.
Default is True.</p>
</dd>
<dt>figsize<span class="classifier">tuple, optional</span></dt><dd><p>Dimension of the figure size.
Default is (6, 10).</p>
</dd>
</dl>
</section>
<section id="id54">
<h3>Outputs<a class="headerlink" href="#id54" title="Link to this heading"></a></h3>
<p>The regression heatmap figures are saved in the specified format and location as defined in the function implementation.
Default save location is ./nico_out/covariations_R*_F*/Regression_outputs/</p>
</section>
<section id="id55">
<h3>Notes<a class="headerlink" href="#id55" title="Link to this heading"></a></h3>
<ul class="simple">
<li><p>Ensure the input contains the necessary data from the gene covariation analysis.</p></li>
<li><p>Ensure the output directory exists and is writable before running this function.</p></li>
</ul>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="nico_covariations.Covariations.plot_top_genes_for_a_given_celltype_from_all_three_factors">
<span class="sig-prename descclassname"><span class="pre">nico_covariations.Covariations.</span></span><span class="sig-name descname"><span class="pre">plot_top_genes_for_a_given_celltype_from_all_three_factors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">choose_celltypes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">top_NOG</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rps_rpl_mt_genes_included</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">correlation_with_spearman</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">saveas</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'pdf'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">transparent_mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">showit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(12,</span> <span class="pre">10)</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/nico_covariations/Covariations.html#plot_top_genes_for_a_given_celltype_from_all_three_factors"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#nico_covariations.Covariations.plot_top_genes_for_a_given_celltype_from_all_three_factors" title="Link to this definition"></a></dt>
<dd><p>Visualize top genes associated with given cell types across all three factors.</p>
<p>This function generates plots of the top N genes associated with specified cell types from the input data.
The associations can be visualized using either Spearman correlation coefficient or cosine similarity.
Optionally, the visualization can include rps, rpl, and mt genes.</p>
<section id="id56">
<h3>Parameters:<a class="headerlink" href="#id56" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>input<span class="classifier">dict</span></dt><dd><p>The main input is the output from gene_covariation_analysis.</p>
</dd>
<dt>choose_celltypes<span class="classifier">list, optional</span></dt><dd><p>The cell type for which the gene-factor associations should be displayed.
If the list is empty, the output will be generated for all the cell types.
(default is [])</p>
</dd>
<dt>top_NOG<span class="classifier">int, optional</span></dt><dd><p>Number of genes to visualize.
(default is 20)</p>
</dd>
<dt>rps_rpl_mt_genes_included<span class="classifier">bool, optional</span></dt><dd><p>For pathway analysis, decide whether to include rps, rpl, and mt genes. If True, they are included.
(default is True)</p>
</dd>
<dt>correlation_with_spearman<span class="classifier">bool, optional</span></dt><dd><p>If True, visualize gene-factor association obtained as Spearman correlation coefficient; otherwise, cosine similarity is displayed.
(default is True)</p>
</dd>
<dt>saveas<span class="classifier">str, optional</span></dt><dd><p>Save the figures in PDF or PNG format (dpi for PNG format is 300).
(default is ‘pdf’)</p>
</dd>
<dt>transparent_mode<span class="classifier">bool, optional</span></dt><dd><p>Background color of the figures.
(default is False)</p>
</dd>
<dt>showit<span class="classifier">bool, optional</span></dt><dd><p>Whether to display the plot or not.
(default is True)</p>
</dd>
<dt>figsize<span class="classifier">tuple, optional</span></dt><dd><p>Dimension of the figure size.
(default is (12, 10))</p>
</dd>
</dl>
</section>
<section id="id57">
<h3>Outputs:<a class="headerlink" href="#id57" title="Link to this heading"></a></h3>
<p>The gene visualization figures are saved in ./nico_out/covariations_R*_F*/dotplots/*</p>
</section>
<section id="id58">
<h3>Example:<a class="headerlink" href="#id58" title="Link to this heading"></a></h3>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">input_data</span> <span class="o">=</span> <span class="n">load_data_from_analysis</span><span class="p">()</span>  <span class="c1"># hypothetical function to load data</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">plot_top_genes_for_a_given_celltype_from_all_three_factors</span><span class="p">(</span><span class="n">input_data</span><span class="p">,</span> <span class="n">choose_celltypes</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;CellType1&#39;</span><span class="p">],</span> <span class="n">top_NOG</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="nico_covariations.Covariations.plot_top_genes_for_pair_of_celltypes_from_two_chosen_factors">
<span class="sig-prename descclassname"><span class="pre">nico_covariations.Covariations.</span></span><span class="sig-name descname"><span class="pre">plot_top_genes_for_pair_of_celltypes_from_two_chosen_factors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">choose_interacting_celltype_pair</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">visualize_factors_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">top_NOG</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rps_rpl_mt_genes_included</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">correlation_with_spearman</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">saveas</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'pdf'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">transparent_mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">showit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(5,</span> <span class="pre">8)</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/nico_covariations/Covariations.html#plot_top_genes_for_pair_of_celltypes_from_two_chosen_factors"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#nico_covariations.Covariations.plot_top_genes_for_pair_of_celltypes_from_two_chosen_factors" title="Link to this definition"></a></dt>
<dd><p>Visualize top genes associated with a pair of cell types from chosen factors.</p>
<p>This function generates plots of the top 20 genes in the factors associated with specified cell types
from the input data, using either Spearman correlation coefficient or cosine similarity.
The visualizations include comparisons between the chosen factors for each cell type.</p>
<section id="id59">
<h3>Parameters:<a class="headerlink" href="#id59" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>input<span class="classifier">object</span></dt><dd><p>The main input is the output from gene_covariation_analysis, which includes factor and expression data.</p>
</dd>
<dt>choose_interacting_celltype_pair<span class="classifier">list</span></dt><dd><p>Define the cell type pairs for visualization. The first entry is the central cell type, and the second is the niche cell type.</p>
</dd>
<dt>visualize_factors_id<span class="classifier">list</span></dt><dd><p>Define the factor IDs for visualization. The first entry is the factor ID of the central cell type, and the second is the factor ID of the niche cell type.</p>
</dd>
<dt>top_NOG<span class="classifier">int, optional</span></dt><dd><p>Number of genes to visualize.
(default is 20)</p>
</dd>
<dt>rps_rpl_mt_genes_included<span class="classifier">bool, optional</span></dt><dd><p>For pathway analysis, decide whether to include rps, rpl, and mt genes. If True, they are included.
(default is True)</p>
</dd>
<dt>correlation_with_spearman<span class="classifier">bool, optional</span></dt><dd><p>If True, visualize gene-factor association obtained as Spearman correlation coefficient; otherwise, cosine similarity is displayed.
(default is True)</p>
</dd>
<dt>saveas<span class="classifier">str, optional</span></dt><dd><p>Save the figures in PDF or PNG format (dpi for PNG format is 300).
(default is ‘pdf’)</p>
</dd>
<dt>transparent_mode<span class="classifier">bool, optional</span></dt><dd><p>Background color of the figures.
(default is False)</p>
</dd>
<dt>showit<span class="classifier">bool, optional</span></dt><dd><p>Whether to display the plot or not.
(default is True)</p>
</dd>
<dt>figsize<span class="classifier">tuple, optional</span></dt><dd><p>Dimension of the figure size.
(default is (5, 8))</p>
</dd>
</dl>
</section>
<section id="id60">
<h3>Outputs:<a class="headerlink" href="#id60" title="Link to this heading"></a></h3>
<p>The gene visualization figures are saved in ./nico_out/covariations_R*_F*/dotplots/*</p>
</section>
<section id="id61">
<h3>Example:<a class="headerlink" href="#id61" title="Link to this heading"></a></h3>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">scov</span><span class="o">.</span><span class="n">plot_top_genes_for_pair_of_celltypes_from_two_chosen_factors</span><span class="p">(</span><span class="n">cov_out</span><span class="p">,</span>
<span class="go">choose_interacting_celltype_pair=[&#39;Stem/TA&#39;,&#39;Paneth&#39;],</span>
<span class="go">visualize_factors_id=[1,1],</span>
<span class="go">top_NOG=20,saveas=saveas,transparent_mode=transparent_mode)</span>
</pre></div>
</div>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="nico_covariations.Covariations.read_LigRecDb">
<span class="sig-prename descclassname"><span class="pre">nico_covariations.Covariations.</span></span><span class="sig-name descname"><span class="pre">read_LigRecDb</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">contdb</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/nico_covariations/Covariations.html#read_LigRecDb"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#nico_covariations.Covariations.read_LigRecDb" title="Link to this definition"></a></dt>
<dd><p>Reads the ligand-receptor database file.</p>
<p>This function processes a database of ligand-receptor pairs, identifying unique ligands, receptors, and elements that act as both. The input should be a list of strings, where each string represents a line from the database file.</p>
<section id="id62">
<h3>Parameters:<a class="headerlink" href="#id62" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>contdb<span class="classifier">list of str</span></dt><dd><p>A list of strings, where each string is a line from the ligand-receptor database file. Each line contains a ligand and a receptor separated by whitespace.</p>
</dd>
</dl>
</section>
<section id="id63">
<h3>Example:<a class="headerlink" href="#id63" title="Link to this heading"></a></h3>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">contdb</span> <span class="o">=</span> <span class="p">[</span>
<span class="go">        &quot;LIG1 REC1&quot;,</span>
<span class="go">        &quot;LIG2 REC2&quot;,</span>
<span class="go">        &quot;REC1 LIG1&quot;,  # REC1 is both a receptor and a ligand</span>
<span class="go">        &quot;LIG3 REC3&quot;</span>
<span class="go">    ]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">totalLRpairs</span><span class="p">,</span> <span class="n">ligand</span><span class="p">,</span> <span class="n">receptor</span><span class="p">,</span> <span class="n">either</span> <span class="o">=</span> <span class="n">read_LigRecDb</span><span class="p">(</span><span class="n">contdb</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">totalLRpairs</span><span class="p">)</span>
<span class="go">[[&#39;LIG1&#39;, &#39;REC1&#39;], [&#39;LIG2&#39;, &#39;REC2&#39;], [&#39;REC1&#39;, &#39;LIG1&#39;], [&#39;LIG3&#39;, &#39;REC3&#39;]]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">ligand</span><span class="p">)</span>
<span class="go">{&#39;LIG2&#39;: 1, &#39;LIG3&#39;: 1}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">receptor</span><span class="p">)</span>
<span class="go">{&#39;REC2&#39;: 1, &#39;REC3&#39;: 1}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">either</span><span class="p">)</span>
<span class="go">{&#39;LIG1&#39;: 1, &#39;REC1&#39;: 1}</span>
</pre></div>
</div>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="nico_covariations.Covariations.read_spatial_data">
<span class="sig-prename descclassname"><span class="pre">nico_covariations.Covariations.</span></span><span class="sig-name descname"><span class="pre">read_spatial_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">clusterFilename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">celltypeFilename</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/nico_covariations/Covariations.html#read_spatial_data"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#nico_covariations.Covariations.read_spatial_data" title="Link to this definition"></a></dt>
<dd><p>Read the cluster information for spatial data.</p>
<p>This helper function is used within the <cite>gene_covariation_analysis</cite> function to read the cluster
and cell type information from the specified files.</p>
<section id="id64">
<h3>Parameters<a class="headerlink" href="#id64" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>clusterFilename<span class="classifier">str</span></dt><dd><p>The file path of the cluster information file.</p>
</dd>
<dt>celltypeFilename<span class="classifier">str</span></dt><dd><p>The file path of the cell type information file.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="nico_covariations.Covariations.remove_extra_character_from_name">
<span class="sig-prename descclassname"><span class="pre">nico_covariations.Covariations.</span></span><span class="sig-name descname"><span class="pre">remove_extra_character_from_name</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/nico_covariations/Covariations.html#remove_extra_character_from_name"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#nico_covariations.Covariations.remove_extra_character_from_name" title="Link to this definition"></a></dt>
<dd><p>Remove special characters from cell type names to avoid errors while saving figures.</p>
<p>This function replaces certain special characters in the input <cite>name</cite> with
underscores or other appropriate characters to ensure the name is safe for use
as a filename.</p>
<section id="id65">
<h3>Parameters<a class="headerlink" href="#id65" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>name<span class="classifier">str</span></dt><dd><p>The original cell type name that may contain special characters.</p>
</dd>
</dl>
</section>
<section id="id66">
<h3>Returns<a class="headerlink" href="#id66" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>str</dt><dd><p>The modified cell type name with special characters removed or replaced.</p>
</dd>
</dl>
</section>
<section id="id67">
<h3>Example<a class="headerlink" href="#id67" title="Link to this heading"></a></h3>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">name</span> <span class="o">=</span> <span class="s1">&#39;T-cell (CD4+)/CD8+&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">clean_name</span> <span class="o">=</span> <span class="n">remove_extra_character_from_name</span><span class="p">(</span><span class="n">name</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">clean_name</span><span class="p">)</span>
<span class="go">&#39;T-cell_CD4p_CD8p&#39;</span>
</pre></div>
</div>
</section>
<section id="id68">
<h3>Notes<a class="headerlink" href="#id68" title="Link to this heading"></a></h3>
<p>The following replacements are made:</p>
<blockquote>
<div><ul class="simple">
<li><p>‘/’ is replaced with ‘_’</p></li>
<li><p>‘ ‘ (space) is replaced with ‘_’</p></li>
<li><p>‘”’ (double quote) is removed</p></li>
<li><p>“’” (single quote) is removed</p></li>
<li><p>‘)’ is removed</p></li>
<li><p>‘(’ is removed</p></li>
<li><p>‘+’ is replaced with ‘p’</p></li>
<li><p>‘-’ is replaced with ‘n’</p></li>
<li><p>‘.’ (dot) is removed</p></li>
</ul>
</div></blockquote>
<p>These substitutions help in creating filenames that do not contain characters
that might be problematic for file systems or software.</p>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="nico_covariations.Covariations.run_ridge_regression">
<span class="sig-prename descclassname"><span class="pre">nico_covariations.Covariations.</span></span><span class="sig-name descname"><span class="pre">run_ridge_regression</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">saveoutname</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ylabelname</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">neighborhoodClass</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">shap_cluster_cutoff</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/nico_covariations/Covariations.html#run_ridge_regression"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#nico_covariations.Covariations.run_ridge_regression" title="Link to this definition"></a></dt>
<dd><p>Helper function for model_linear_regression to perform ridge regression per cell type.</p>
<p>This function performs ridge regression for each target variable (central cell factors) using
the neighborhood average spatial cell factors as predictors. It normalizes the data, fits the
regression model, and computes various statistics including p-values and explained variance.</p>
<section id="id69">
<h3>Parameters<a class="headerlink" href="#id69" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>input<span class="classifier">object</span></dt><dd><p>An object containing the following attributes:
- shap_analysis : bool, whether to perform SHAP analysis.
- regression_outdir : str, directory to save regression outputs.
- lambda_c : list, list of regularization parameters for RidgeCV.
- no_of_pc : int, number of principal components.</p>
</dd>
<dt>saveoutname<span class="classifier">str</span></dt><dd><p>The name to save the output of the regression results.</p>
</dd>
<dt>ylabelname<span class="classifier">list</span></dt><dd><p>List of feature names for the predictors.</p>
</dd>
<dt>target<span class="classifier">array</span></dt><dd><p>The target values (central cell factors).</p>
</dd>
<dt>neighborhoodClass<span class="classifier">array</span></dt><dd><p>The neighborhood average spatial cell factors.</p>
</dd>
<dt>shap_cluster_cutoff<span class="classifier">float</span></dt><dd><p>The cutoff value for clustering in SHAP analysis.</p>
</dd>
</dl>
</section>
<section id="id70">
<h3>Returns<a class="headerlink" href="#id70" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>coef<span class="classifier">array</span></dt><dd><p>Coefficients of the ridge regression model.</p>
</dd>
<dt>intercept<span class="classifier">array</span></dt><dd><p>Intercepts of the ridge regression model.</p>
</dd>
<dt>lambda_c<span class="classifier">list</span></dt><dd><p>List of regularization parameters used in the ridge regression model.</p>
</dd>
<dt>percent_variance_explained<span class="classifier">list</span></dt><dd><p>Percentage of variance explained by the model.</p>
</dd>
<dt>residual_variance_explained<span class="classifier">float</span></dt><dd><p>Residual variance explained by the model (currently set to 0).</p>
</dd>
<dt>pv<span class="classifier">array</span></dt><dd><p>P-values of the regression coefficients.</p>
</dd>
</dl>
</section>
<section id="id71">
<h3>Notes<a class="headerlink" href="#id71" title="Link to this heading"></a></h3>
<ul class="simple">
<li><p>The function normalizes the predictors and target variables.</p></li>
<li><p>It fits a ridge regression model for each target variable and computes various statistics.</p></li>
<li><p>If SHAP analysis is enabled, it performs SHAP analysis and saves the plots.</p></li>
</ul>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="nico_covariations.Covariations.save_LR_interactions_in_excelsheet_and_regression_summary_in_textfile_for_interacting_cell_types">
<span class="sig-prename descclassname"><span class="pre">nico_covariations.Covariations.</span></span><span class="sig-name descname"><span class="pre">save_LR_interactions_in_excelsheet_and_regression_summary_in_textfile_for_interacting_cell_types</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pvalueCutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">correlation_with_spearman</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">LR_plot_NMF_Fa_thres</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">LR_plot_Exp_thres</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">number_of_top_genes_to_print</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">20</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/nico_covariations/Covariations.html#save_LR_interactions_in_excelsheet_and_regression_summary_in_textfile_for_interacting_cell_types"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#nico_covariations.Covariations.save_LR_interactions_in_excelsheet_and_regression_summary_in_textfile_for_interacting_cell_types" title="Link to this definition"></a></dt>
<dd><p>Save ligand-receptor (LR) interactions in an Excel sheet and regression summary in a text file for interacting cell types.</p>
<p>This function processes the output from gene_covariation_analysis to identify significant LR interactions and saves the results in an Excel sheet and a text file.</p>
<section id="id72">
<h3>Parameters:<a class="headerlink" href="#id72" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>input<span class="classifier">object</span></dt><dd><p>The main input is the output from gene_covariation_analysis.</p>
</dd>
<dt>pvalueCutoff<span class="classifier">float, optional</span></dt><dd><p>The cutoff used to select the significant central cell type factor and niche cell type factor covariations.
Default is 0.05.</p>
</dd>
<dt>correlation_with_spearman<span class="classifier">bool, optional</span></dt><dd><p>If True, genes factor correlations are computed as Spearman correlation coefficient; otherwise, cosine similarities are computed.
Default is True.</p>
</dd>
<dt>LR_plot_NMF_Fa_thres<span class="classifier">float, optional</span></dt><dd><p>Only ligands or receptors are retained that exhibit a correlation to the respective factors higher than this cutoff.
Default is 0.2.</p>
</dd>
<dt>LR_plot_Exp_thres<span class="classifier">float, optional</span></dt><dd><p>Only ligands or receptors are retained that are expressed in a fraction of cells of the respective cell types exceeding this cutoff.
Default is 0.2.</p>
</dd>
<dt>number_of_top_genes_to_print<span class="classifier">int, optional</span></dt><dd><p>The number of top correlating genes to print in the regression summary text file.
Default is 20.</p>
</dd>
</dl>
</section>
<section id="id73">
<h3>Outputs:<a class="headerlink" href="#id73" title="Link to this heading"></a></h3>
<blockquote>
<div><ul>
<li><p>An Excel sheet with ligand-receptor interaction information for easy access.
The columns are structured as follows in the sheets:</p>
<blockquote>
<div><ul class="simple">
<li><ol class="upperalpha simple">
<li><p>ID of the cell type-cell type interaction</p></li>
</ol>
</li>
<li><p>BC. Interacting cell types A and B</p></li>
<li><ol class="upperalpha simple" start="4">
<li><p>Normalized interaction scores from the logistic regression classifier</p></li>
</ol>
</li>
<li><p>EF. NMF factor IDs (metagenes) in cell types A and B</p></li>
<li><ol class="upperalpha simple" start="7">
<li><p>Ridge regression coefficient indicating the factors’ covariation</p></li>
</ol>
</li>
<li><ol class="upperalpha simple" start="8">
<li><p>Ligand in cell type A</p></li>
</ol>
</li>
<li><ol class="upperroman simple">
<li><p>Receptor in cell type B</p></li>
</ol>
</li>
<li><p>JK. Pearson correlation of ligand and receptor genes in cell types A and B with the corresponding factors</p></li>
<li><p>LM. Average expression of ligands and receptors in cell types A and B</p></li>
<li><p>NO. Fraction of cells expressing these genes with counts greater than zero in cell types A and B</p></li>
</ul>
</div></blockquote>
</li>
<li><p>A regression summary text file with the following structure:</p></li>
<li><p>First row: CC-Fa(i), CC (cell type), niche_score (from classifier), NC-Fa*, NC (cell type), RegCoeff (covariation score), p-value on normal scale, p-value on -log10 scale</p></li>
<li><p>Second row: Top 20 (number_of_top_genes_to_print) genes correlated to Fa(i) of central cell type, with genes and their factor ID indicated in the pair</p></li>
<li><p>Third row: Top 20 (number_of_top_genes_to_print) genes correlated to Fa(j) of niche cell type, with genes and their factor ID indicated in the pair</p></li>
</ul>
</div></blockquote>
</section>
<section id="id74">
<h3>Notes<a class="headerlink" href="#id74" title="Link to this heading"></a></h3>
<ul class="simple">
<li><p>Our analysis accounts for bidirectional cellular crosstalk interactions of ligands and receptors in cell types A and B.</p></li>
<li><p>The ligand can be expressed on cell type A and signal to the receptor detected on cell type B, or vice versa.</p></li>
<li><p>Both ligand-receptor plots and Excel sheets profile bidirectional cellular crosstalk of ligand and receptor in cell types A and B.</p></li>
<li><p>Each central cell type is represented in a separate Excel sheet, while the LR enrichment sheet aggregates all interactions across central cell types.</p></li>
</ul>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="nico_covariations.Covariations.sort_index_in_right_order">
<span class="sig-prename descclassname"><span class="pre">nico_covariations.Covariations.</span></span><span class="sig-name descname"><span class="pre">sort_index_in_right_order</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">correct</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">wrong</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/nico_covariations/Covariations.html#sort_index_in_right_order"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#nico_covariations.Covariations.sort_index_in_right_order" title="Link to this definition"></a></dt>
<dd><p>Sorts the ‘wrong’ array to match the order of the ‘correct’ array based on the first column values.</p>
<p>This function reorders the rows of the ‘wrong’ array to match the order of the ‘correct’ array based on the values in the first column. It is a helper function used in visualizing cell type annotations.</p>
<section id="id75">
<h3>Parameters:<a class="headerlink" href="#id75" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>correct<span class="classifier">ndarray</span></dt><dd><p>An array with the correct order of elements. The sorting is based on the values in the first column.</p>
</dd>
<dt>wrong<span class="classifier">ndarray</span></dt><dd><p>An array that needs to be reordered to match the ‘correct’ array. The sorting is based on the values in the first column.</p>
</dd>
</dl>
</section>
<section id="id76">
<h3>Returns:<a class="headerlink" href="#id76" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>right<span class="classifier">ndarray</span></dt><dd><p>The ‘wrong’ array reordered to match the order of the ‘correct’ array based on the first column values.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="nico_covariations.Covariations.sorting_of_factors_for_showing_the_value_in_excelsheet">
<span class="sig-prename descclassname"><span class="pre">nico_covariations.Covariations.</span></span><span class="sig-name descname"><span class="pre">sorting_of_factors_for_showing_the_value_in_excelsheet</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">CC_corr</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">no_of_pc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genenames</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/nico_covariations/Covariations.html#sorting_of_factors_for_showing_the_value_in_excelsheet"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#nico_covariations.Covariations.sorting_of_factors_for_showing_the_value_in_excelsheet" title="Link to this definition"></a></dt>
<dd><p>Sort factor values for displaying in an Excel sheet.</p>
<p>This helper function is used in <cite>make_excel_sheet_for_gene_correlation</cite> to sort the factor values from correlation data. It separates the sorted values into all genes and a subset of common genes.</p>
<section id="id77">
<h3>Parameters:<a class="headerlink" href="#id77" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>CC_corr<span class="classifier">numpy.ndarray</span></dt><dd><p>Array of correlation values, where rows represent genes and columns represent principal components.</p>
</dd>
<dt>no_of_pc<span class="classifier">int</span></dt><dd><p>Number of principal components.</p>
</dd>
<dt>gene<span class="classifier">list of str</span></dt><dd><p>List of gene names corresponding to the rows in <cite>CC_corr</cite>.</p>
</dd>
<dt>genenames<span class="classifier">list of str</span></dt><dd><p>List of gene names to be included in the common subset.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="nico_covariations.Covariations.top_genes_in_correlation_list_without">
<span class="sig-prename descclassname"><span class="pre">nico_covariations.Covariations.</span></span><span class="sig-name descname"><span class="pre">top_genes_in_correlation_list_without</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">corr_NMFfactors_genes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_top_words</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/nico_covariations/Covariations.html#top_genes_in_correlation_list_without"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#nico_covariations.Covariations.top_genes_in_correlation_list_without" title="Link to this definition"></a></dt>
<dd><p>Identify top genes associated with NMF factors, excluding duplicates.</p>
<p>This helper function sorts the factor values and selects the top genes
associated with each factor. It is used in <cite>plot_cosine_and_spearman_correlation_to_factors</cite>.</p>
<section id="id78">
<h3>Parameters<a class="headerlink" href="#id78" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>genename<span class="classifier">numpy.ndarray or pandas.Series</span></dt><dd><p>Array or Series containing gene names.</p>
</dd>
<dt>corr_NMFfactors_genes<span class="classifier">numpy.ndarray or pandas.DataFrame</span></dt><dd><p>The matrix representing the correlation values between genes and NMF factors.
Each row corresponds to a gene, and each column corresponds to an NMF factor.</p>
</dd>
<dt>n_top_words<span class="classifier">int</span></dt><dd><p>The number of top genes to retrieve for each NMF factor.</p>
</dd>
</dl>
</section>
<section id="id79">
<h3>Returns<a class="headerlink" href="#id79" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>gname<span class="classifier">numpy.ndarray</span></dt><dd><p>Array containing the names of the top genes associated with the NMF factors.</p>
</dd>
<dt>mat<span class="classifier">numpy.ndarray</span></dt><dd><p>Matrix containing the correlation values of the top genes associated with the NMF factors.
Each row corresponds to a selected top gene, and each column corresponds to an NMF factor.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="nico_covariations.Covariations.triangulation_for_triheatmap">
<span class="sig-prename descclassname"><span class="pre">nico_covariations.Covariations.</span></span><span class="sig-name descname"><span class="pre">triangulation_for_triheatmap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">M</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">N</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/nico_covariations/Covariations.html#triangulation_for_triheatmap"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#nico_covariations.Covariations.triangulation_for_triheatmap" title="Link to this definition"></a></dt>
<dd><p>Create triangulation for plotting a ligand-receptor map.</p>
<p>This helper function generates the triangulation needed for plotting a rectangle four faces in
the <cite>plot_ligand_receptor_in_interacting_celltypes</cite> function. It constructs the vertices
and triangles required for visualizing the ligand-receptor interactions on a heatmap.</p>
<section id="id80">
<h3>Parameters:<a class="headerlink" href="#id80" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>M<span class="classifier">int</span></dt><dd><p>Number of columns in the heatmap.</p>
</dd>
<dt>N<span class="classifier">int</span></dt><dd><p>Number of rows in the heatmap.</p>
</dd>
</dl>
</section>
<section id="id81">
<h3>Returns:<a class="headerlink" href="#id81" title="Link to this heading"></a></h3>
<p>list of matplotlib.tri.Triangulation</p>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="nico_covariations.Covariations.visualize_factors_in_scRNAseq_umap">
<span class="sig-prename descclassname"><span class="pre">nico_covariations.Covariations.</span></span><span class="sig-name descname"><span class="pre">visualize_factors_in_scRNAseq_umap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">choose_interacting_celltype_pair</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">visualize_factors_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">umap_tag</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'X_umap'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">msna</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ms</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'viridis'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">saveas</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'pdf'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">transparent_mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">showit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(8,</span> <span class="pre">3.5)</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/nico_covariations/Covariations.html#visualize_factors_in_scRNAseq_umap"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#nico_covariations.Covariations.visualize_factors_in_scRNAseq_umap" title="Link to this definition"></a></dt>
<dd><p>Visualize factors in scRNAseq UMAP embedding.</p>
<p>This function visualizes the factors in single-cell RNA sequencing (scRNAseq) UMAP embeddings. It highlights the interactions between specified cell type pairs and their corresponding factor IDs.</p>
<section id="id82">
<h3>Parameters:<a class="headerlink" href="#id82" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>input<span class="classifier">str</span></dt><dd><p>The primary input is the output from <cite>gene_covariation_analysis</cite>.</p>
</dd>
<dt>choose_interacting_celltype_pair<span class="classifier">list</span></dt><dd><p>List defining the cell type single or in pairs to visualize. At least one cell type need to put by the user.</p>
</dd>
<dt>visualize_factors_id<span class="classifier">list</span></dt><dd><p>List defining the factor IDs single or in pairs to visualize in the UMAP. The chosen factors analogous to defined cell types.</p>
</dd>
<dt>umap_tag<span class="classifier">str, optional</span></dt><dd><p>The UMAP embedding tag in the .obsm field of the AnnData object (default is ‘X_umap’).</p>
</dd>
<dt>msna<span class="classifier">float, optional</span></dt><dd><p>The marker size for non selected (NA) cell types (default is 0.1).</p>
</dd>
<dt>ms<span class="classifier">int, optional</span></dt><dd><p>The marker size for selected cell types (default is 5).</p>
</dd>
<dt>cmap<span class="classifier">str, optional</span></dt><dd><p>Colormap for visualizing factors (default is <cite>matplotlib.rcParams[“image.cmap”]</cite>).</p>
</dd>
<dt>saveas<span class="classifier">str, optional</span></dt><dd><p>Format to save the figures (‘pdf’ or ‘png’) (default is ‘pdf’).</p>
</dd>
<dt>transparent_mode<span class="classifier">bool, optional</span></dt><dd><p>Whether to save the figures with a transparent background (default is False).</p>
</dd>
<dt>showit<span class="classifier">bool, optional</span></dt><dd><p>Whether to display the figures (default is True).</p>
</dd>
<dt>figsize<span class="classifier">tuple, optional</span></dt><dd><p>Size of the figure (default is (8, 3.5)).</p>
</dd>
</dl>
</section>
<section id="id83">
<h3>Outputs:<a class="headerlink" href="#id83" title="Link to this heading"></a></h3>
<p>The factor visualization in scRNAseq embedding is saved in “./nico_out/covariations_R*_F*/scRNAseq_factors_in_umap”.</p>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="nico_covariations.Covariations.visualize_factors_in_spatial_umap">
<span class="sig-prename descclassname"><span class="pre">nico_covariations.Covariations.</span></span><span class="sig-name descname"><span class="pre">visualize_factors_in_spatial_umap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">choose_interacting_celltype_pair</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">visualize_factors_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">umap_tag</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'X_umap'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quepath</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'./inputQuery/'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">msna</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ms</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'viridis'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">saveas</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'pdf'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">transparent_mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">showit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(8,</span> <span class="pre">3.5)</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/nico_covariations/Covariations.html#visualize_factors_in_spatial_umap"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#nico_covariations.Covariations.visualize_factors_in_spatial_umap" title="Link to this definition"></a></dt>
<dd><p>Visualize factors in spatial UMAP for cell type interactions.</p>
<p>This function is used to visualize the factors of interacting cell types in a spatial UMAP embedding. It generates and saves plots showing the distribution and factor values of cells.</p>
<section id="id84">
<h3>Parameters:<a class="headerlink" href="#id84" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>input<span class="classifier">str</span></dt><dd><p>The primary input is the output from gene_covariation_analysis.</p>
</dd>
<dt>choose_interacting_celltype_pair<span class="classifier">list of str</span></dt><dd><p>Define the cell type single or in pairs for visualization in the spatial UMAP.
Example: choose_interacting_celltype_pair=[‘CentralCellType’, ‘NicheCellType’]</p>
</dd>
<dt>visualize_factors_id<span class="classifier">list of str</span></dt><dd><p>Define the factor IDs single or in pairs for visualization in the spatial UMAP.
Example: visualize_factors_id=[1, 3]</p>
</dd>
<dt>umap_tag<span class="classifier">str, optional</span></dt><dd><p>Slot for UMAP embedding in the AnnData object. Default is ‘X_umap’.</p>
</dd>
<dt>quepath<span class="classifier">str, optional</span></dt><dd><p>Path to the query spatial count matrix in scTransform-like normalization in the common gene space. The filename should be sct_spatial.h5ad.
Default is ‘./inputQuery/’.</p>
</dd>
<dt>msna<span class="classifier">float, optional</span></dt><dd><p>Marker size for not selected (NA) cell types. Default is 0.1.</p>
</dd>
<dt>ms<span class="classifier">float, optional</span></dt><dd><p>Marker size for selected cell types. Default is 5.</p>
</dd>
<dt>cmap<span class="classifier">str or matplotlib.colors.Colormap, optional</span></dt><dd><p>Colormap used for visualizing the factor values. Default is matplotlib.rcParams[“image.cmap”].</p>
</dd>
<dt>saveas<span class="classifier">str, optional</span></dt><dd><p>Format to save the figures, either ‘pdf’ or ‘png’. Default is ‘pdf’.</p>
</dd>
<dt>transparent_mode<span class="classifier">bool, optional</span></dt><dd><p>Background color of the figures. If True, the figures have a transparent background. Default is False.</p>
</dd>
<dt>showit<span class="classifier">bool, optional</span></dt><dd><p>If True, the figures will be displayed. Default is True.</p>
</dd>
<dt>figsize<span class="classifier">tuple of float, optional</span></dt><dd><p>Dimension of the figure size. Default is (8, 3.5).</p>
</dd>
</dl>
</section>
<section id="output">
<h3>Output:<a class="headerlink" href="#output" title="Link to this heading"></a></h3>
<p>The output figure will be saved in nico_out/covariations_R*_F*/spatial_factors_in_umap*.</p>
</section>
</dd></dl>

</section>
<section id="module-nico_covariations">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-nico_covariations" title="Link to this heading"></a></h2>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="nico_annotations.utils.html" class="btn btn-neutral float-left" title="nico_annotations.utils package" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="nico_covariations.utils.html" class="btn btn-neutral float-right" title="nico_covariations.utils package" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Grün lab.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>