Introduction
============

Welcome to the Nico Documentation!

NiCo ( **Ni** che **Co** variation) is a tool designed for the comprehensive analysis of imaging- and sequencing-based spatial transcriptomics data.
This documentation provides detailed instructions on how to use the tool, including installation, usage examples, and API references.

Key Features:
-------------

- **Cell Type Annotation:** Accurate annotation of cell types in the spatial modality by label transfer.
- **Visualization of Cell Types in Space:** Generation of UMAP plots and spatial maps to visualize cell types.
- **Spatial Neighborhood Analysis:** Identification and analysis of cell type-cell type interactions within spatial niche neighborhoods.
- **Inference of Cell State Covariation:** Analysis of molecular crosstalk of cell types co-localized in niches using latent variable regression.
- **Visualization of Ligand-Receptor Pairs:** NiCo visualizes ligand-receptor pairs associated with covarying latent variables of interacting cell types.
- **Visualization of Pathways:** NiCo visualizes enriched pathways in the top correlating genes associated with latent variables.

Getting Started:
----------------

To get started with NiCo, follow the installation instructions and refer to the examples and tutorials for detailed examples.

Contributing:
-------------

We welcome contributions to the project. If you would like to contribute, please refer to the contributing guidelines.

Support:
--------

For support, please check the FAQ section.
